
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Impact of filtering PTMs &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'gallery_output/Overview_of_Spliced_PTMs/plot_filter_impact';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../index.html">Gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_ksea.html">Kinase substrate enrichment analysis (KSEA)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_protein_interactions.html">Protein interactions network</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/gallery_output/Overview_of_Spliced_PTMs/plot_filter_impact.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Impact of filtering PTMs</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="sphx-glr-download-link-note admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#sphx-glr-download-gallery-output-overview-of-spliced-ptms-plot-filter-impact-py"><span class="std std-ref">Go to the end</span></a>
to download the full example code.</p>
</div>
<section class="sphx-glr-example-title" id="impact-of-filtering-ptms">
<span id="sphx-glr-gallery-output-overview-of-spliced-ptms-plot-filter-impact-py"></span><h1>Impact of filtering PTMs<a class="headerlink" href="#impact-of-filtering-ptms" title="Link to this heading">#</a></h1>
<p>In some cases, you may wish to filter out PTMs with weaker evidence (only identified in small number of experiments), splice events with small changes, or PTMs that have not previously been observed to be impacted by splicing. This can be done using the <cite>filter_ptms</cite> function in the <cite>helpers</cite> module. We can also visualize the impact of filtering using the <cite>analyze.filter</cite> module.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose</span><span class="w"> </span><span class="kn">import</span> <span class="n">helpers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose.analyze</span><span class="w"> </span><span class="kn">import</span> <span class="nb">filter</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>


<span class="c1">#load example differential inclusion and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>There are several ways to filter PTMs based on the parameter values below. Let’s set these parameters so that we focus on PTMs that have been observed to impacted by splicing (remove_novel = True) and have been observed in at least 2 mass spectrometry experiments (min_experiments = 2). By default, the filter ptms will also remove any insignificant splice events with a dPSI value less than 0.1 (min_dpsi = 0.1, alpha = 0.05), but let’s restrict that a little more to focus on events with changes of at least 20%.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">remove_novel</span> <span class="o">=</span> <span class="kc">True</span>
<span class="n">min_MS_observations</span> <span class="o">=</span> <span class="mi">5</span>
<span class="n">min_dpsi</span> <span class="o">=</span> <span class="mf">0.2</span>

<span class="c1">#other paramter values we could set to filter PTMs</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span> <span class="c1"># p-value threshold for significance</span>
<span class="n">min_studies</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># minimum number of literature publications that support the PTM (high and low throughput)</span>
</pre></div>
</div>
<p>We can now assess the impact of these parameters on the number of PTMs, and type of PTMs, that are present in the dataset after filtering using the <cite>filter.plot_filter_impact()</cite> function</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="nb">filter</span><span class="o">.</span><span class="n">plot_filter_impact</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">output_type</span> <span class="o">=</span> <span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">remove_novel</span> <span class="o">=</span> <span class="n">remove_novel</span><span class="p">,</span> <span class="n">min_MS_observations</span> <span class="o">=</span> <span class="n">min_MS_observations</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">report_removed</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_filter_impact_001.png" srcset="../../_images/sphx_glr_plot_filter_impact_001.png" alt="plot filter impact" class = "sphx-glr-single-img"/><p>Rather than the total number of PTMs, we can also assess how it impacts the proportion of PTMs that are present in the dataset. This is useful to see how the filtering parameters impact the type of PTMs that are present in the dataset, as more restrictive parameters may skew the dataset towards a particular type of PTM (e.g. phosphorylation).</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="nb">filter</span><span class="o">.</span><span class="n">plot_filter_impact</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">output_type</span> <span class="o">=</span> <span class="s1">&#39;fraction&#39;</span><span class="p">,</span> <span class="n">remove_novel</span> <span class="o">=</span> <span class="n">remove_novel</span><span class="p">,</span> <span class="n">min_MS_observations</span> <span class="o">=</span> <span class="n">min_MS_observations</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">report_removed</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_filter_impact_002.png" srcset="../../_images/sphx_glr_plot_filter_impact_002.png" alt="plot filter impact" class = "sphx-glr-single-img"/><p>As you can see, this a pretty restrictive filter, and we are left with only a small number of PTMs (mostly phosphorylation). That may be useful for some analyses, but we may want to relax the parameters a little to include more PTMs. We could also choose to only filter out phosphoryation sites based on evidence, and keep all other PTMs regardless of the evidence (e.g. acetylation, methylation, etc.). This is not perfect, but as phosphorylation is the most commonly studied/measured PTM, it may be a good compromise to keep other PTMs that are not as well studied.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">phospho_only_evidence_filter</span> <span class="o">=</span> <span class="kc">True</span>

<span class="nb">filter</span><span class="o">.</span><span class="n">plot_filter_impact</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">output_type</span> <span class="o">=</span> <span class="s1">&#39;count&#39;</span><span class="p">,</span> <span class="n">remove_novel</span> <span class="o">=</span> <span class="n">remove_novel</span><span class="p">,</span> <span class="n">min_MS_observations</span> <span class="o">=</span> <span class="n">min_MS_observations</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">report_removed</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">phospho_only_evidence_filter</span> <span class="o">=</span> <span class="n">phospho_only_evidence_filter</span><span class="p">)</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_filter_impact_003.png" srcset="../../_images/sphx_glr_plot_filter_impact_003.png" alt="plot filter impact" class = "sphx-glr-single-img"/><p>Now you can see that there is the same number of phosphorylation sites, but now there are other modification types still present.</p>
<p class="sphx-glr-timing"><strong>Total running time of the script:</strong> (0 minutes 0.427 seconds)</p>
<div class="sphx-glr-footer sphx-glr-footer-example docutils container" id="sphx-glr-download-gallery-output-overview-of-spliced-ptms-plot-filter-impact-py">
<div class="sphx-glr-download sphx-glr-download-jupyter docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/a1602bfa6a6b8b26d28729b9606f2a8b/plot_filter_impact.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_filter_impact.ipynb</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-python docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/78c097b1a1833d4b77a3f4163107a97e/plot_filter_impact.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_filter_impact.py</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-zip docutils container">
<p><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">zipped:</span> <span class="pre">plot_filter_impact.zip</span></code></p>
</div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.github.io">Gallery generated by Sphinx-Gallery</a></p>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>