
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>PTM-POSE Reference &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'PTM_POSE';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="&lt;no title&gt;" href="Running_PTM-POSE/index.html" />
    <link rel="prev" title="Dependencies" href="Dependencies.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.3.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="Dependencies.html">Dependencies</a></li>

<li class="toctree-l1 current active"><a class="current reference internal" href="#">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="gallery_output/index.html">Gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="gallery_output/A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="gallery_output/FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="gallery_output/Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/Networks/plot_protein_interactions.html">Protein interactions network</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="gallery_output/flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="gallery_output/flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/PTM_POSE.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>PTM-POSE Reference</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.pose_config">Configuration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.pose_config.download_translator"><code class="docutils literal notranslate"><span class="pre">download_translator()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.project">PTM Projection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.find_ptms_in_region"><code class="docutils literal notranslate"><span class="pre">find_ptms_in_region()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_MATS"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_MATS()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_SpliceSeq"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_SpliceSeq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_splice_events"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_splice_events()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.flanking_sequences">Flanking Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_rMATS"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_rMATS()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_splice_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_splicegraph()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.annotate">Annotating PTMs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_ELM_interactions"><code class="docutils literal notranslate"><span class="pre">add_ELM_interactions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_ELM_matched_motifs"><code class="docutils literal notranslate"><span class="pre">add_ELM_matched_motifs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_custom_annotation"><code class="docutils literal notranslate"><span class="pre">add_custom_annotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_omnipath_data"><code class="docutils literal notranslate"><span class="pre">add_omnipath_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.annotate_ptms"><code class="docutils literal notranslate"><span class="pre">annotate_ptms()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.append_from_gmt"><code class="docutils literal notranslate"><span class="pre">append_from_gmt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.check_file"><code class="docutils literal notranslate"><span class="pre">check_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.check_gmt_file"><code class="docutils literal notranslate"><span class="pre">check_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.combine_enzyme_data"><code class="docutils literal notranslate"><span class="pre">combine_enzyme_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.combine_interaction_data"><code class="docutils literal notranslate"><span class="pre">combine_interaction_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_DEPOD_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_DEPOD_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMInt_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_PTMInt_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMcode_interprotein_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_PTMcode_interprotein_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMsigDB_gmt_files"><code class="docutils literal notranslate"><span class="pre">construct_PTMsigDB_gmt_files()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PhosphoSitePlus_gmt_files"><code class="docutils literal notranslate"><span class="pre">construct_PhosphoSitePlus_gmt_files()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_RegPhos_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_RegPhos_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_annotation_dict_from_gmt"><code class="docutils literal notranslate"><span class="pre">construct_annotation_dict_from_gmt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_custom_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_custom_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_gmt_df"><code class="docutils literal notranslate"><span class="pre">construct_gmt_df()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_omnipath_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_omnipath_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.convert_PSP_label_to_UniProt"><code class="docutils literal notranslate"><span class="pre">convert_PSP_label_to_UniProt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_ids_PTMcode"><code class="docutils literal notranslate"><span class="pre">extract_ids_PTMcode()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_interaction_details"><code class="docutils literal notranslate"><span class="pre">extract_interaction_details()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_positions_from_DEPOD"><code class="docutils literal notranslate"><span class="pre">extract_positions_from_DEPOD()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.get_available_gmt_annotations"><code class="docutils literal notranslate"><span class="pre">get_available_gmt_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.load_gmt_file"><code class="docutils literal notranslate"><span class="pre">load_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.process_database_annotations"><code class="docutils literal notranslate"><span class="pre">process_database_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.simplify_annotation"><code class="docutils literal notranslate"><span class="pre">simplify_annotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.unify_interaction_data"><code class="docutils literal notranslate"><span class="pre">unify_interaction_data()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-modules">Analyze Modules</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.summarize">Summaries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.combine_outputs"><code class="docutils literal notranslate"><span class="pre">combine_outputs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.get_modification_class_data"><code class="docutils literal notranslate"><span class="pre">get_modification_class_data()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.get_modification_counts"><code class="docutils literal notranslate"><span class="pre">get_modification_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.plot_modification_breakdown"><code class="docutils literal notranslate"><span class="pre">plot_modification_breakdown()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.filter">Filtering PTMs and Events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.filter.plot_filter_impact"><code class="docutils literal notranslate"><span class="pre">plot_filter_impact()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.annotations">Annotations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.annotation_enrichment"><code class="docutils literal notranslate"><span class="pre">annotation_enrichment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.draw_pie"><code class="docutils literal notranslate"><span class="pre">draw_pie()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.gene_set_enrichment"><code class="docutils literal notranslate"><span class="pre">gene_set_enrichment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_available_annotations"><code class="docutils literal notranslate"><span class="pre">get_available_annotations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_background_annotation_counts"><code class="docutils literal notranslate"><span class="pre">get_background_annotation_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_enrichment_inputs"><code class="docutils literal notranslate"><span class="pre">get_enrichment_inputs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_ptm_annotations"><code class="docutils literal notranslate"><span class="pre">get_ptm_annotations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_EnrichR_pies"><code class="docutils literal notranslate"><span class="pre">plot_EnrichR_pies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_annotation_counts"><code class="docutils literal notranslate"><span class="pre">plot_annotation_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_available_annotations"><code class="docutils literal notranslate"><span class="pre">plot_available_annotations()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.interactions">Protein Interactions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.get_edge_colors"><code class="docutils literal notranslate"><span class="pre">get_edge_colors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.get_interaction_stats"><code class="docutils literal notranslate"><span class="pre">get_interaction_stats()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.plot_interaction_network"><code class="docutils literal notranslate"><span class="pre">plot_interaction_network()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.plot_network_centrality"><code class="docutils literal notranslate"><span class="pre">plot_network_centrality()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions"><code class="docutils literal notranslate"><span class="pre">protein_interactions</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.compare_to_nease"><code class="docutils literal notranslate"><span class="pre">protein_interactions.compare_to_nease()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_interaction_stats()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_protein_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_nease_comparison()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_network_centrality()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.summarize_protein_network()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.enzyme">Enzyme Regulation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.enzyme.compare_KL_for_sequence"><code class="docutils literal notranslate"><span class="pre">compare_KL_for_sequence()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.enzyme.get_all_KL_scores"><code class="docutils literal notranslate"><span class="pre">get_all_KL_scores()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Flanking Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.compare_inclusion_motifs"><code class="docutils literal notranslate"><span class="pre">compare_inclusion_motifs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.findAlteredPositions"><code class="docutils literal notranslate"><span class="pre">findAlteredPositions()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.find_motifs"><code class="docutils literal notranslate"><span class="pre">find_motifs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.getSequenceIdentity"><code class="docutils literal notranslate"><span class="pre">getSequenceIdentity()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_alterations_matrix"><code class="docutils literal notranslate"><span class="pre">plot_alterations_matrix()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_location_of_altered_flanking_residues"><code class="docutils literal notranslate"><span class="pre">plot_location_of_altered_flanking_residues()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_sequence_differences"><code class="docutils literal notranslate"><span class="pre">plot_sequence_differences()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="ptm-pose-reference">
<h1>PTM-POSE Reference<a class="headerlink" href="#ptm-pose-reference" title="Link to this heading">#</a></h1>
<section id="module-ptm_pose.pose_config">
<span id="configuration"></span><h2>Configuration<a class="headerlink" href="#module-ptm_pose.pose_config" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.pose_config.download_translator">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.pose_config.</span></span><span class="sig-name descname"><span class="pre">download_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/pose_config.html#download_translator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.pose_config.download_translator" title="Link to this definition">#</a></dt>
<dd><p>Using rest API from UniProt, download mapping information between UniProt IDs, Gene names, and Ensembl Gene IDs. This information is used to convert between different gene identifiers and UniProt IDs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>save</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to save the translator file locally. The default is False.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.project">
<span id="ptm-projection"></span><h2>PTM Projection<a class="headerlink" href="#module-ptm_pose.project" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.find_ptms_in_region">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">find_ptms_in_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#find_ptms_in_region"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.find_ptms_in_region" title="Link to this definition">#</a></dt>
<dd><p>Given an genomic region in either hg38 or hg19 coordinates (such as the region encoding an exon of interest), identify PTMs that are mapped to that region. If so, return the exon number. If none are found, return np.nan.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>chromosome: str</strong></dt><dd><p>chromosome where region is located</p>
</dd>
<dt><strong>strand: int</strong></dt><dd><p>DNA strand for region is found on (1 for forward, -1 for reverse)</p>
</dd>
<dt><strong>start: int</strong></dt><dd><p>start position of region on the chromosome/strand (should always be less than end)</p>
</dd>
<dt><strong>end: int</strong></dt><dd><p>end position of region on the chromosome/strand (should always be greater than start)</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ptms_in_region: pandas.DataFrame</dt><dd><p>dataframe containing all PTMs found in the region. If no PTMs are found, returns np.nan.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.project_ptms_onto_MATS">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">project_ptms_onto_MATS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">SE_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A5SS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A3SS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RI_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MXE_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identify_flanking_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'meanDeltaPSI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FDR'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_modification_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PROCESSES</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#project_ptms_onto_MATS"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.project_ptms_onto_MATS" title="Link to this definition">#</a></dt>
<dd><p>Given splice quantification from the MATS algorithm, annotate with PTMs that are found in the differentially included regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_coordinates: pandas.DataFrame</strong></dt><dd><p>dataframe containing PTM information, including chromosome, strand, and genomic location of PTMs</p>
</dd>
<dt><strong>SE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing skipped exon event information from MATS</p>
</dd>
<dt><strong>A5SS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 5’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>A3SS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 3’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>RI_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing retained intron event information from MATS</p>
</dd>
<dt><strong>MXE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing mutually exclusive exon event information from MATS</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
<dt><strong>dPSI_col: str</strong></dt><dd><p>Column name indicating delta PSI value. Default is ‘meanDeltaPSI’.</p>
</dd>
<dt><strong>sig_col: str</strong></dt><dd><p>Column name indicating significance of the event. Default is ‘FDR’.</p>
</dd>
<dt><strong>extra_cols: list</strong></dt><dd><p>List of column names for additional information to add to the results. Default is None.</p>
</dd>
<dt><strong>separate_modification_types: bool</strong></dt><dd><p>Indicate whether residues with multiple modifications (i.e. phosphorylation and acetylation) should be treated as separate PTMs and be placed in unique rows of the output dataframe. Default is False.</p>
</dd>
<dt><strong>PROCESSES: int</strong></dt><dd><p>Number of processes to use for multiprocessing. Default is 1.</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the find_ptms_in_many_regions function, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.project_ptms_onto_SpliceSeq">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">project_ptms_onto_SpliceSeq</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psi_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicegraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'symbol'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_modification_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identify_flanking_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PROCESSES</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#project_ptms_onto_SpliceSeq"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.project_ptms_onto_SpliceSeq" title="Link to this definition">#</a></dt>
<dd><p>Given splice event quantification from SpliceSeq (such as what can be downloaded from TCGASpliceSeq), annotate with PTMs that are found in the differentially included regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>psi_data: pandas.DataFrame</strong></dt><dd><p>dataframe containing splice event quantification from SpliceSeq. Must contain the following columns: ‘symbol’, ‘exons’, ‘splice_type’.</p>
</dd>
<dt><strong>splicegraph: pandas.DataFrame</strong></dt><dd><p>dataframe containing exon information from the splicegraph used during splice event quantification. Must contain the following columns: ‘Symbol’, ‘Exon’, ‘Chr_Start’, ‘Chr_Stop’.</p>
</dd>
<dt><strong>gene_col: str</strong></dt><dd><p>column name in psi_data that contains the gene name. Default is ‘symbol’.</p>
</dd>
<dt><strong>dPSI_col: str</strong></dt><dd><p>column name in psi_data that contains the delta PSI value for the splice event. Default is None, which will not include this information in the output.</p>
</dd>
<dt><strong>sig_col: str</strong></dt><dd><p>column name in psi_data that contains the significance value for the splice event. Default is None, which will not include this information in the output.</p>
</dd>
<dt><strong>extra_cols: list</strong></dt><dd><p>list of additional columns to include in the output dataframe. Default is None, which will not include any additional columns.</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg19’.</p>
</dd>
<dt><strong>separate_modification_types: bool</strong></dt><dd><p>Indicate whether to store PTM sites with multiple modification types as multiple rows. For example, if a site at K100 was both an acetylation and methylation site, these will be separated into unique rows with the same site number but different modification types. Default is True.</p>
</dd>
<dt><strong>identify_flanking_sequences: bool</strong></dt><dd><p>Indicate whether to identify and return the flanking sequences for the splice events. Default is False.</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>Size of the flanking sequence to extract from the splice event. Default is 5, which will extract 5 bases upstream and downstream of the splice event. Only relevant if identify_flanking_sequences is True.</p>
</dd>
<dt><strong>PROCESSES: int</strong></dt><dd><p>Number of processes to use for multiprocessing. Default is 1 (single processing).</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the find_ptms_in_many_regions function, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.project_ptms_onto_splice_events">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">project_ptms_onto_splice_events</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">splice_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_original_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'chr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strand'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exonStart_0base'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exonEnd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_modification_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_coordinate_system</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1-based'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_coordinate_system</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1-based'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taskbar_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PROCESSES</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#project_ptms_onto_splice_events"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.project_ptms_onto_splice_events" title="Link to this definition">#</a></dt>
<dd><p>Given splice event quantification data, project PTMs onto the regions impacted by the splice events. Assumes that the splice event data will have chromosome, strand, and genomic start/end positions for the regions of interest, and each row of the splice_event_data corresponds to a unique region.</p>
<p>Important note: PTM-POSE relies on Ensembl based coordinates (1-based), so if the coordinates are 0-based, make sure to indicate using the start_coordinate_system and end_coordinate_system parameters. For example, rMATS uses 0-based for the start coordinates, but 1-based for the end coordinates. In this case, set start_coordinate_system = ‘0-based’ and end_coordinate_system = ‘1-based’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>splice_data: pandas.DataFrame</strong></dt><dd><p>dataframe containing splice event information, including chromosome, strand, and genomic location of regions of interest</p>
</dd>
<dt><strong>ptm_coordinates: pandas.DataFrame</strong></dt><dd><p>dataframe containing PTM information, including chromosome, strand, and genomic location of PTMs. If none, it will pull from the config file.</p>
</dd>
<dt><strong>chromosome_col: str</strong></dt><dd><p>column name in splice_data that contains chromosome information. Default is ‘chr’. Expects it to be a str with only the chromosome number: ‘Y’, ‘1’, ‘2’, etc.</p>
</dd>
<dt><strong>strand_col: str</strong></dt><dd><p>column name in splice_data that contains strand information. Default is ‘strand’. Expects it to be a str with ‘+’ or ‘-’, or integers as 1 or -1. Will convert to integers automatically if string format is provided.</p>
</dd>
<dt><strong>region_start_col: str</strong></dt><dd><p>column name in splice_data that contains the start position of the region of interest. Default is ‘exonStart_0base’.</p>
</dd>
<dt><strong>region_end_col: str</strong></dt><dd><p>column name in splice_data that contains the end position of the region of interest. Default is ‘exonEnd’.</p>
</dd>
<dt><strong>event_id_col: str</strong></dt><dd><p>column name in splice_data that contains the unique identifier for the splice event. If provided, will be used to annotate the ptm information with the specific splice event ID. Default is None.</p>
</dd>
<dt><strong>gene_col: str</strong></dt><dd><p>column name in splice_data that contains the gene name. If provided, will be used to make sure the projected PTMs stem from the same gene (some cases where genomic coordiantes overlap between distinct genes). Default is None.</p>
</dd>
<dt><strong>dPSI_col: str</strong></dt><dd><p>column name in splice_data that contains the delta PSI value for the splice event. Default is None, which will not include this information in the output</p>
</dd>
<dt><strong>sig_col: str</strong></dt><dd><p>column name in splice_data that contains the significance value for the splice event. Default is None, which will not include this information in the output.</p>
</dd>
<dt><strong>extra_cols: list</strong></dt><dd><p>list of additional columns to include in the output dataframe. Default is None, which will not include any additional columns.</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
<dt><strong>start_coordinate_system: str</strong></dt><dd><p>indicates the coordinate system used for the start position. Either ‘0-based’ or ‘1-based’. Default is ‘1-based’.</p>
</dd>
<dt><strong>end_coordinate_system: str</strong></dt><dd><p>indicates the coordinate system used for the end position. Either ‘0-based’ or ‘1-based’. Default is ‘1-based’.</p>
</dd>
<dt><strong>separate_modification_types: bool</strong></dt><dd><p>Indicate whether to store PTM sites with  multiple modification types as multiple rows. For example, if a site at K100 was both an acetylation and methylation site, these will be separated into unique rows with the same site number but different modification types. Default is True.</p>
</dd>
<dt><strong>taskbar_label: str</strong></dt><dd><p>Label to display in the tqdm progress bar. Default is None, which will automatically state “Projecting PTMs onto regions using —– coordinates”.</p>
</dd>
<dt><strong>PROCESSES: int</strong></dt><dd><p>Number of processes to use for multiprocessing. Default is 1 (single processing)</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the find_ptms_in_many_regions function, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>spliced_ptm_info: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events</p>
</dd>
<dt>splice_data: pandas.DataFrame</dt><dd><p>dataframe containing the original splice data with an additional column ‘PTMs’ that contains the PTMs found in the region of interest, in the format of ‘SiteNumber(ModificationType)’. If no PTMs are found, the value will be np.nan.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.flanking_sequences">
<span id="flanking-sequences"></span><h2>Flanking Sequences<a class="headerlink" href="#module-ptm_pose.flanking_sequences" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Coordinates'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes" title="Link to this definition">#</a></dt>
<dd><p>Given flanking and spliced regions associated with a splice event, identify PTMs that have potential to have an altered flanking sequence depending on whether spliced region is included or excluded (if PTM is close to splice boundary). For these PTMs, extract the flanking sequences associated with the inclusion and exclusion cases and translate into amino acid sequences. If the PTM is not associated with a codon that codes for the expected amino acid, the PTM will be excluded from the results.</p>
<p>Important note: It is assumed that all region coordinates are based on a 1-based coordinate system, not 0-based, consistent with Ensembl. If using a 0-based system, please adjust the coordinates accordingly prior to running this function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>chromosome</strong><span class="classifier">str</span></dt><dd><p>Chromosome associated with the splice event</p>
</dd>
<dt><strong>strand</strong><span class="classifier">int</span></dt><dd><p>Strand associated with the splice event (1 for forward, -1 for negative)</p>
</dd>
<dt><strong>first_flank_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the first flanking region (first is currently defined based on location the genome not coding sequence)</p>
</dd>
<dt><strong>spliced_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the spliced region</p>
</dd>
<dt><strong>second_flank_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the second flanking region (second is currently defined based on location the genome not coding sequence)</p>
</dd>
<dt><strong>event_id</strong><span class="classifier">str, optional</span></dt><dd><p>Event ID associated with the splice event, by default None</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 7</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg38’. Other options is hg19.</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM in returned flanking sequences, by default True</p>
</dd>
<dt><strong>order_by</strong><span class="classifier">str, optional</span></dt><dd><p>Whether the first, spliced and second regions are defined by their genomic coordinates (first has smallest coordinate, spliced next, then second), or if they are defined by their translation (first the first when translated, etc.)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd><p>DataFrame containing the PTMs associated with the flanking regions and the amino acid sequences of the flanking regions in the inclusion and exclusion cases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes_from_rMATS">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes_from_rMATS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">SE_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A5SS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A3SS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RI_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'meanDeltaPSI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FDR'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes_from_rMATS"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_rMATS" title="Link to this definition">#</a></dt>
<dd><p>Given splice events identified rMATS extract quantified PTMs that are nearby the splice boundary (potential for flanking sequence to be altered). Coordinate information of individual exons should be found in splicegraph. You can also provide columns with specific psi or significance information. Extra cols not in these categories can be provided with extra_cols parameter.</p>
<p>Only use this function if you do not care about differentially included sites, otherwise you can use the project module set <cite>identify_flanking_sequences = True</cite> (project.project_ptms_onto_MATS(identify_flanking_sequences = True))</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_coordinates: pandas.DataFrame</strong></dt><dd><p>dataframe containing PTM information, including chromosome, strand, and genomic location of PTMs. If none, will use the PTM coordinates from the pose_config file.</p>
</dd>
<dt><strong>SE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing skipped exon event information from MATS</p>
</dd>
<dt><strong>A5SS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 5’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>A3SS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 3’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>RI_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing retained intron event information from MATS</p>
</dd>
<dt><strong>MXE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing mutually exclusive exon event information from MATS</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
<dt><strong>dPSI_col: str</strong></dt><dd><p>Column name indicating delta PSI value. Default is ‘meanDeltaPSI’.</p>
</dd>
<dt><strong>sig_col: str</strong></dt><dd><p>Column name indicating significance of the event. Default is ‘FDR’.</p>
</dd>
<dt><strong>extra_cols: list</strong></dt><dd><p>List of column names for additional information to add to the results. Default is None.</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes_from_splice_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">splice_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start_coordinate_system</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1-based'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_coordinate_system</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'1-based'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes_from_splice_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data" title="Link to this definition">#</a></dt>
<dd><p>Given a DataFrame containing information about splice events, extract the flanking sequences associated with the PTMs in the flanking regions if there is potential for this to be altered. The DataFrame should contain columns for the chromosome, strand, start and stop locations of the first flanking region, spliced region, and second flanking region. The DataFrame should also contain a column for the event ID associated with the splice event. If the DataFrame does not contain the necessary columns, the function will raise an error.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>splice_data</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about splice events</p>
</dd>
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>chromosome_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating chromosome, by default None</p>
</dd>
<dt><strong>strand_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating strand, by default None</p>
</dd>
<dt><strong>first_flank_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the first flanking region, by default None</p>
</dd>
<dt><strong>first_flank_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the first flanking region, by default None</p>
</dd>
<dt><strong>spliced_region_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the spliced region, by default None</p>
</dd>
<dt><strong>spliced_region_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the spliced region, by default None</p>
</dd>
<dt><strong>second_flank_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the second flanking region, by default None</p>
</dd>
<dt><strong>second_flank_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the second flanking region, by default None</p>
</dd>
<dt><strong>event_id_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating event ID, by default None</p>
</dd>
<dt><strong>gene_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating gene name, by default None</p>
</dd>
<dt><strong>extra_cols</strong><span class="classifier">list, optional</span></dt><dd><p>List of additional columns to include in the output DataFrame, by default None</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 7</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg38’. Other options is hg19.</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM in returned flanking sequences, by default True</p>
</dd>
<dt><strong>start_coordinate_system</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the start locations of the regions, by default ‘1-based’. Other option is ‘0-based’.</p>
</dd>
<dt><strong>end_coordinate_system</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the end locations of the regions, by default ‘1-based’. Other option is ‘0-based’.</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">keyword arguments, optional</span></dt><dd><p>Additional keyword arguments to pass to the find_ptms_in_many_regions function, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>list</dt><dd><p>List containing DataFrames with the PTMs associated with the flanking regions and the amino acid sequences of the flanking regions in the inclusion and exclusion cases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes_from_splicegraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psi_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicegraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'symbol'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes_from_splicegraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph" title="Link to this definition">#</a></dt>
<dd><p>Given a DataFrame containing information about splice events  obtained from SpliceSeq and the corresponding splicegraph, extract the flanking sequences of PTMs that are nearby the splice boundary (potential for flanking sequence to be altered). Coordinate information of individual exons should be found in splicegraph. You can also provide columns with specific psi or significance information. Extra cols not in these categories can be provided with extra_cols parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>psi_data</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about splice events obtained from SpliceSeq</p>
</dd>
<dt><strong>splicegraph</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about individual exons and their coordinates</p>
</dd>
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>dPSI_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating delta PSI value, by default None</p>
</dd>
<dt><strong>sig_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating significance of the event, by default None</p>
</dd>
<dt><strong>event_id_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating event ID, by default None</p>
</dd>
<dt><strong>extra_cols</strong><span class="classifier">list, optional</span></dt><dd><p>List of column names for additional information to add to the results, by default None</p>
</dd>
<dt><strong>gene_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating gene symbol of spliced gene, by default ‘symbol’</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 5</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg19’. Other options is hg38.</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>altered_flanks</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing the PTMs associated with the flanking regions that are altered, and the flanking sequences that arise depending on whether the flanking sequence is included or not</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.annotate">
<span id="annotating-ptms"></span><h2>Annotating PTMs<a class="headerlink" href="#module-ptm_pose.annotate" title="Link to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_ELM_interactions">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_ELM_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_ELM_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_ELM_interactions" title="Link to this definition">#</a></dt>
<dd><p>Given a spliced ptms or altered flanks dataframe from the project module, add ELM interaction data to the dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pandas.DataFrame</strong></dt><dd><p>Contains the PTMs identified across the different splice events, either differentially included events, or altered flanking sequences</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>Path to the ELM data file. If not provided, the data will be downloaded directly from the ELM website</p>
</dd>
<dt><strong>report_success: bool</strong></dt><dd><p>If True, will print out the number of PTMs identified in the dataset that have ELM interaction information</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with additional columns for ELM interaction data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_ELM_matched_motifs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_ELM_matched_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_ELM_matched_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_ELM_matched_motifs" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptms or altered flanks dataframes, compare the canonical flanking sequences of each PTM to motifs recorded in the ELM database. If a match is found, the ELM motif will be recorded in the ELM:Motif Matches column</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pandas.DataFrame</strong></dt><dd><p>Contains the PTMs identified across the different splice events, either differentially included events, or altered flanking sequences</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>Number of residues to include on either side of the PTM for the motif search. Default is 7</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>Path to the ELM data file. If not provided, the data will be downloaded directly from the ELM website</p>
</dd>
<dt><strong>report_success: bool</strong></dt><dd><p>If True, will print out the number of PTMs identified in the dataset that have ELM motif data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_custom_annotation">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_custom_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accession_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UniProtKB</span> <span class="pre">Accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Residue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PTM</span> <span class="pre">Position</span> <span class="pre">in</span> <span class="pre">Isoform'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_custom_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_custom_annotation" title="Link to this definition">#</a></dt>
<dd><p>Add custom annotation data to ptms or altered flanking sequence dataframes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_data: pandas.DataFrame</strong></dt><dd><p>Dataframe containing the annotation data to be added to the ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added</p>
</dd>
<dt><strong>source_name: str</strong></dt><dd><p>Name of the source of the annotation data, will be used to label the columns in the ptms dataframe</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation data being added, will be used to label the columns in the ptms dataframe</p>
</dd>
<dt><strong>annotation_col: str</strong></dt><dd><p>Column name in the annotation data that contains the annotation data to be added to the ptms dataframe</p>
</dd>
<dt><strong>accession_col: str</strong></dt><dd><p>Column name in the annotation data that contains the UniProtKB Accession information. Default is ‘UniProtKB Accession’</p>
</dd>
<dt><strong>residue_col: str</strong></dt><dd><p>Column name in the annotation data that contains the residue information</p>
</dd>
<dt><strong>position_col: str</strong></dt><dd><p>Column name in the annotation data that contains the PTM position information</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with an additional column for the custom annotation data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_omnipath_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_omnipath_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_to_gene_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace_old_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_omnipath_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_omnipath_data" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptms or altered flanks dataframe, append enzyme-substrate interactions recorded in OmniPath database. These will be split between ‘Writer’ enzymes, or enzymes that add the modification (OmniPath:Writer Enzyme), and ‘Eraser’ enzymes, or enzymes that remove the modification (OmniPath:Eraser Enzyme). Note, we do not consider the ‘post translational modification’ or ‘cleavage’ entries for this purpose.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>Spliced PTMs or altered flanks dataframe.</p>
</dd>
<dt><strong>min_sources</strong><span class="classifier">int</span></dt><dd><p>Minimum number of sources (i.e. database) for enzyme-substrate interaction. Default is 1, or all entries.</p>
</dd>
<dt><strong>min_references</strong><span class="classifier">int</span></dt><dd><p>Minimum number of references (i.e. publications) for enzyme-substrate interaction. Default is 1, or all entries.</p>
</dd>
<dt><strong>convert_to_gene_name</strong><span class="classifier">bool</span></dt><dd><p>Whether to convert enzyme names from UniProt IDs to gene names using pose_config.uniprot_to_genename. Default is True.</p>
</dd>
<dt><strong>report_success</strong><span class="classifier">bool</span></dt><dd><p>Whether to report success message. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.annotate_ptms">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">annotate_ptms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phosphositeplus</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptmsigdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptmcode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptmint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">omnipath</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regphos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactions_to_combine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzymes_to_combine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'All'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#annotate_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.annotate_ptms" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptm data, add annotations from various databases. The annotations that can be added are the following:</p>
<p>PhosphoSitePlus: regulatory site data (file must be provided), kinase-substrate data (file must be provided), and disease association data (file must be provided)
ELM: interaction data (can be downloaded automatically or provided as a file), motif matches (elm class data can be downloaded automatically or provided as a file)
PTMInt: interaction data (will be downloaded automatically)
PTMcode: intraprotein interactions (can be downloaded automatically or provided as a file), interprotein interactions (can be downloaded automatically or provided as a file)
DEPOD: phosphatase-substrate data (will be downloaded automatically)
RegPhos: kinase-substrate data (will be downloaded automatically)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Spliced PTM data from project module</p>
</dd>
<dt><strong>psp_regulatory</strong><span class="classifier">bool</span></dt><dd></dd>
<dt><strong>interactions_to_combine: list</strong></dt><dd><p>List of databases to combine interaction data from. Default is [‘PTMcode’, ‘PhosphoSitePlus’, ‘RegPhos’, ‘PTMInt’]</p>
</dd>
<dt><strong>kinases_to_combine: list</strong></dt><dd><p>List of databases to combine kinase-substrate data from. Default is [‘PhosphoSitePlus’, ‘RegPhos’]</p>
</dd>
<dt><strong>combine_similar: bool</strong></dt><dd><p>Whether to combine annotations of similar information (kinase, interactions, etc) from multiple databases into another column labeled as ‘Combined’. Default is True</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.append_from_gmt">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">append_from_gmt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gmt_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#append_from_gmt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.append_from_gmt" title="Link to this definition">#</a></dt>
<dd><p>Given a gmt annotation file format, add the annotations to the ptms dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>dataframe containing ptm information, which can be the spliced_ptms or altered_flanks dataframe generated during projection</p>
</dd>
<dt><strong>database</strong><span class="classifier">str</span></dt><dd><p>Name of the database for the annotation. Used to identify proper annotation if gmt_df not provided</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>Type of annotation to append to the ptms dataframe</p>
</dd>
<dt><strong>gmt_df</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>If using custom gmt file, provide the dataframe loaded from the GMT file. This will override the database and annot_type parameters if provided.</p>
</dd>
<dt><strong>column_name</strong><span class="classifier">str or None</span></dt><dd><p>Name of the column to use for the annotations in the ptms dataframe. If None, will use a default name based on the database and annot_type. Default is None.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">additional keyword arguments</span></dt><dd><p>Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.check_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">check_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expected_extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.tsv'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#check_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.check_file" title="Link to this definition">#</a></dt>
<dd><p>Given a file name, check if the file exists and has the expected extension. If the file does not exist or has the wrong extension, raise an error.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fname: str</strong></dt><dd><p>File name to check</p>
</dd>
<dt><strong>expected_extension: str</strong></dt><dd><p>Expected file extension. Default is ‘.tsv’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.check_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">check_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gmt_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatic_download</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#check_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.check_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Given a gmt file path, check to make sure it exists. If it doesn’t, either raise error or download and save a gmt file in the provided directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gmt_file</strong><span class="classifier">str</span></dt><dd><p>file path to gmt file</p>
</dd>
<dt><strong>database</strong><span class="classifier">str</span></dt><dd><p>name of database associated with gmt file</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>type of annotation to check for. This is used to provide more specific error messages</p>
</dd>
<dt><strong>automatic download: bool</strong></dt><dd><p>whether to automatically download data and process into gmt file if it does not exist and can be done. Default is false</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str or None</span></dt><dd><p>location to save annotations, if automatic download is true</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">additional keyword arguments</span></dt><dd><p>Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.combine_enzyme_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">combine_enzyme_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">enzyme_databases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'RegPhos',</span> <span class="pre">'OmniPath',</span> <span class="pre">'DEPOD']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regphos_conversion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'ABL1(ABL)':</span> <span class="pre">'ABL1',</span> <span class="pre">'CDC2':</span> <span class="pre">'CDK1',</span> <span class="pre">'CK2A1':</span> <span class="pre">'CSNK2A1',</span> <span class="pre">'ERK1(MAPK3)':</span> <span class="pre">'MAPK3',</span> <span class="pre">'ERK2(MAPK1)':</span> <span class="pre">'MAPK1',</span> <span class="pre">'JNK2(MAPK9)':</span> <span class="pre">'MAPK9',</span> <span class="pre">'PKACA':</span> <span class="pre">'PRKACA'}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#combine_enzyme_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.combine_enzyme_data" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptm information, combine enzyme-substrate data from multiple databases (currently support PhosphoSitePlus, RegPhos, OmniPath, DEPOD, and iKiP downloaded from PTMsigDB), assuming that the enzyme data from these resources has already been added to the spliced ptm data. The combined kinase data will be added as a new column labeled ‘Combined:Writer Enzyme’ and ‘Combined:Eraser Enzyme’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Spliced PTM data from project module</p>
</dd>
<dt><strong>enzyme_databases: list</strong></dt><dd><p>List of databases to combine enzyme data from. Currently support PhosphoSitePlus, RegPhos, OmniPath, and DEPOD</p>
</dd>
<dt><strong>regphos_conversion: dict</strong></dt><dd><p>Allows conversion of RegPhos names to matching names in PhosphoSitePlus.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ptms: pd.DataFrame</dt><dd><p>PTM data with combined kinase data added</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.combine_interaction_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">combine_interaction_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_databases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'PTMcode',</span> <span class="pre">'PTMInt',</span> <span class="pre">'RegPhos',</span> <span class="pre">'DEPOD']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_enzyme_interactions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#combine_interaction_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.combine_interaction_data" title="Link to this definition">#</a></dt>
<dd><p>Given annotated spliced ptm data, extract interaction data from various databases and combine into a single dataframe. This will include the interacting protein, the type of interaction, and the source of the interaction data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe containing PTM data and associated interaction annotations from various databases</p>
</dd>
<dt><strong>interaction_databases: list</strong></dt><dd><p>List of databases to extract interaction data from. Options include ‘PhosphoSitePlus’, ‘PTMcode’, ‘PTMInt’, ‘RegPhos’, ‘DEPOD’. These should already have annotation columns in the ptms dataframe, otherwise they will be ignored. For kinase-substrate interactions, if combined column is present, will use that instead of individual databases</p>
</dd>
<dt><strong>include_enzyme_interactions: bool</strong></dt><dd><p>If True, will include kinase-substrate and phosphatase interactions in the output dataframe</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>interact_data: list</dt><dd><p>List of dataframes containing PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES), and the source of the interaction data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_DEPOD_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_DEPOD_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_DEPOD_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_DEPOD_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Download and process DEPOD data to create a GMT file for PTM-POSE. DEPOD contains information on dephosphorylation sites and their corresponding substrates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>odir</strong><span class="classifier">str, optional</span></dt><dd><p>Output directory for the GMT file. If not provided, it will default to the ‘Resource_Files/Annotations/DEPOD’ directory within the PTM-POSE package directory.</p>
</dd>
<dt><strong>overwrite</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, will overwrite any existing GMT file in the output directory. If False and the GMT file already exists, the function will skip processing and print a message.</p>
</dd>
<dt><strong>max_retries</strong><span class="classifier">int, optional</span></dt><dd><p>Number of times to try downloading data from DEPOD</p>
</dd>
<dt><strong>delay</strong><span class="classifier">int, optional</span></dt><dd><p>Delay in seconds between download attempts. Default is 10 seconds.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_PTMInt_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_PTMInt_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_PTMInt_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_PTMInt_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Download and process PTMInt interaction data to create gmt files for PTM-POSE</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str, optional</span></dt><dd><p>Path to the PTMInt data file. If not provided, the data will be downloaded directly from the PTMInt website. Default is None.</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str, optional</span></dt><dd><p>Output directory for the gmt file. If not provided, will default to the PTM-POSE resource directory for annotations. Default is None.</p>
</dd>
<dt><strong>overwrite</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, will overwrite any existing gmt files in the output directory. If False, will skip the creation of the gmt file if it already exists. Default is False.</p>
</dd>
<dt><strong>max_retries</strong><span class="classifier">int, optional</span></dt><dd><p>Number of times to retry downloading the PTMInt data if the download fails. Default is 5.</p>
</dd>
<dt><strong>delay</strong><span class="classifier">int, optional</span></dt><dd><p>Amount of time to wait (in seconds) before retrying the download if it fails. Default is 10 seconds.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_PTMcode_interprotein_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_PTMcode_interprotein_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_PTMcode_interprotein_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_PTMcode_interprotein_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Given the PTMcode interprotein interaction data, convert to readily usable format with PTM-POSE in gmt file format</p>
<dl class="simple">
<dt>file: str</dt><dd><p>Path to the PTMcode interprotein interaction data file. If not provided, the data will be downloaded directly from the PTMcode website</p>
</dd>
<dt>odir<span class="classifier">str</span></dt><dd><p>Output directory for the gmt file. If not provided, will default to the PTM-POSE resource directory for annotations</p>
</dd>
<dt>overwrite<span class="classifier">bool, optional</span></dt><dd><p>If True, will overwrite any existing gmt files in the output directory. If False, will skip the creation of the gmt file if it already exists. Default is False.</p>
</dd>
<dt>max_retries<span class="classifier">int, optional</span></dt><dd><p>Number of times to retry downloading the PTMcode data if the initial attempt fails. Default is 5.</p>
</dd>
<dt>delay<span class="classifier">int, optional</span></dt><dd><p>Number of seconds to wait between retries if the download fails. Default is 10 seconds.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_PTMsigDB_gmt_files">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_PTMsigDB_gmt_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">process_PSP_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_PTMsigDB_gmt_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_PTMsigDB_gmt_files" title="Link to this definition">#</a></dt>
<dd><p>Given the PTMsigDB xlsx file, convert to readily usable format with PTM-POSE in gmt file format. This will also process the PhosphoSitePlus data in PTMsigDB if requested.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file</strong><span class="classifier">str</span></dt><dd><p>PTMsigDB excel file path. This file can be downloaded from the PTMsigDB website.</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str</span></dt><dd><p>Output directory for the gmt files. If None, will default to the PTM-POSE resource directory.</p>
</dd>
<dt><strong>overwrite</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, will overwrite any existing gmt files in the output directory. Default is False.</p>
</dd>
<dt><strong>process_PSP_data</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, will process the PhosphoSitePlus data included in the PTMsigDB file, but only if not already found in odir. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_PhosphoSitePlus_gmt_files">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_PhosphoSitePlus_gmt_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regulatory_site_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinase_substrate_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disease_association_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_PhosphoSitePlus_gmt_files"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_PhosphoSitePlus_gmt_files" title="Link to this definition">#</a></dt>
<dd><p>Given three PhosphoSitePlus annotation files, convert to readily usable format with PTM-POSE in gmt file format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>regulatory_site_file: str or None</strong></dt><dd><p>Path to the PhosphoSitePlus regulatory site file (gzipped). If None, will skip creating function annotations.</p>
</dd>
<dt><strong>kinase_substrate_file: str or None</strong></dt><dd><p>Path to the PhosphoSitePlus kinase-substrate file (gzipped). If None, will skip creating kinase-substrate annotations.</p>
</dd>
<dt><strong>disease_association_file: str or None</strong></dt><dd><p>Path to the PhosphoSitePlus disease association file (gzipped). If None, will skip creating disease association annotations.</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str or None</span></dt><dd><p>Path to the output directory where the GMT files will be saved. If None, will save to the default resource directory for PhosphoSitePlus annotations.</p>
</dd>
<dt><strong>overwrite</strong><span class="classifier">bool</span></dt><dd><p>If True, will overwrite existing GMT files if they already exist. If False, will skip creating the GMT files if they already exist. Default is False.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_RegPhos_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_RegPhos_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_RegPhos_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_RegPhos_gmt_file" title="Link to this definition">#</a></dt>
<dd><dl class="simple">
<dt>file<span class="classifier">str</span></dt><dd><p>RegPhos text file path. This file can be downloaded from the RegPhos website. If None, the function will raise an error.</p>
</dd>
<dt>odir<span class="classifier">str</span></dt><dd><p>Output directory for the gmt files. If None, will default to the PTM-POSE resource directory.</p>
</dd>
<dt>overwrite<span class="classifier">bool, optional</span></dt><dd><p>If True, will overwrite any existing gmt files in the output directory. Default is False.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_annotation_dict_from_gmt">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_annotation_dict_from_gmt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gmt_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'annotation'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_annotation_dict_from_gmt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_annotation_dict_from_gmt" title="Link to this definition">#</a></dt>
<dd><p>Given a gmt annotation file format, construct a dictionary mapping each item to its annotations, with either the annotation as key or PTM as the key</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_custom_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_custom_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accession_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UniProtKB</span> <span class="pre">Accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Residue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PTM</span> <span class="pre">Position</span> <span class="pre">in</span> <span class="pre">Isofrom'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_custom_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_custom_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Function for constructing a gmt file for annotations not currently provided by PTM-POSE. Ideally, these annotations should be partially processed to have the same format as PTM-POSE annotations. For example, they should have columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_df: pandas.DataFrame</strong></dt><dd><p>Dataframe containing the annotation data to be added to the ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</p>
</dd>
<dt><strong>database</strong><span class="classifier">str</span></dt><dd><p>Name of the database for the annotation. This will be used to create the output directory and file name.</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>Type of annotation data being added. This will be used to create the output file name and description.</p>
</dd>
<dt><strong>annot_col</strong><span class="classifier">str</span></dt><dd><p>Column name in the annotation data that contains the annotation data to be added to the ptms dataframe. This will be used as the annotation column in the output GMT file.</p>
</dd>
<dt><strong>accession_col</strong><span class="classifier">str</span></dt><dd><p>Column name in the annotation data that contains the UniProtKB Accession information. Default is ‘UniProtKB Accession’.</p>
</dd>
<dt><strong>residue_col</strong><span class="classifier">str</span></dt><dd><p>Column name in the annotation data that contains the residue information. Default is ‘Residue’.</p>
</dd>
<dt><strong>position_col</strong><span class="classifier">str</span></dt><dd><p>Column name in the annotation data that contains the PTM position information. Default is ‘PTM Position in Isoform’.</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str or None</span></dt><dd><p>Path to the output directory where the GMT file will be saved. If None, will save to the default resource directory for annotations. Default is None.</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">additional keyword arguments</span></dt><dd><p>additional keywords to pass to construct_gmt_df function. This can include parameters such as annotation_separator, description, and compressed.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_gmt_df">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_gmt_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_separator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compressed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_gmt_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_gmt_df" title="Link to this definition">#</a></dt>
<dd><p>Given annotation data, construct a dataframe in the gmt file format. Save if odir and fname are provided</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>df</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Dataframe containing the annotation data to be converted to GMT format. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</p>
</dd>
<dt><strong>annotation_col</strong><span class="classifier">str</span></dt><dd><p>Column name in the dataframe that contains the annotation data to be added to the GMT file. This will be used as the annotation column in the output GMT file.</p>
</dd>
<dt><strong>description</strong><span class="classifier">str or np.nan</span></dt><dd><p>description to add to description column</p>
</dd>
<dt><strong>annotation_separator</strong><span class="classifier">str or None</span></dt><dd><p>what separator to use for splitting annotations in the annotation_col. If None, will not split annotations. Default is None.</p>
</dd>
<dt><strong>odir</strong><span class="classifier">str or None</span></dt><dd><p>file path to output directory where the GMT file will be saved. If None, will not save. Default is None.</p>
</dd>
<dt><strong>fname</strong><span class="classifier">str or None:</span></dt><dd><p>name of output file. If None, will use the annotation_col as the file name. Default is None.</p>
</dd>
<dt><strong>compressed</strong><span class="classifier">bool</span></dt><dd><p>whether to save gmt file in gzip format. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.construct_omnipath_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">construct_omnipath_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_sources</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_references</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_to_gene_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">odir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#construct_omnipath_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.construct_omnipath_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Download enzyme-substrate interactions from the OmniPath database. The data will be filtered based on the number of sources and references specified. The resulting data will be split into two categories: ‘Writer’ enzymes, which add the modification, and ‘Eraser’ enzymes, which remove the modification. The output will be saved as GMT files in resource files directory.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.convert_PSP_label_to_UniProt">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">convert_PSP_label_to_UniProt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#convert_PSP_label_to_UniProt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.convert_PSP_label_to_UniProt" title="Link to this definition">#</a></dt>
<dd><p>Given a label for an interacting protein from PhosphoSitePlus, convert to UniProtKB accession. Required as PhosphoSitePlus interactions are recorded in various ways that aren’t necessarily consistent with other databases (i.e. not always gene name)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>label: str</strong></dt><dd><p>Label for interacting protein from PhosphoSitePlus</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.extract_ids_PTMcode">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">extract_ids_PTMcode</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'##</span> <span class="pre">Protein1'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#extract_ids_PTMcode"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.extract_ids_PTMcode" title="Link to this definition">#</a></dt>
<dd><p>Many different ID forms are used in PTMcode, but we rely on UniProt IDs. This function is used to convert between Ensembl Gene IDs to UniProt IDs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.extract_interaction_details">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">extract_interaction_details</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interaction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus:Interactions'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#extract_interaction_details"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.extract_interaction_details" title="Link to this definition">#</a></dt>
<dd><p>Given an interaction string from a specific database, extract the type of interaction and the interacting protein. This is required as different databases format their interaction strings differently.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.extract_positions_from_DEPOD">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">extract_positions_from_DEPOD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#extract_positions_from_DEPOD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.extract_positions_from_DEPOD" title="Link to this definition">#</a></dt>
<dd><p>Given string object consisting of multiple modifications in the form of ‘Residue-Position’ separated by ‘, ‘, extract the residue and position. Ignore any excess details in the string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>x</strong><span class="classifier">str</span></dt><dd><p>dephosphosite entry from DEPOD data</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>new_x :str</dt><dd><p>ptm residue and position in format that PTM-POSE recognizes</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.get_available_gmt_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">get_available_gmt_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dict'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#get_available_gmt_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.get_available_gmt_annotations" title="Link to this definition">#</a></dt>
<dd><p>Get the annotations available in resource files in GMT format. Can be outputted as either a dictionary or pandas DataFrame</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>format: str</strong></dt><dd><p>Format to output the available annotations. Options are ‘dict’ or ‘dataframe’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.load_gmt_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">load_gmt_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gmt_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#load_gmt_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.load_gmt_file" title="Link to this definition">#</a></dt>
<dd><p>Load a GMT file into a pandas DataFrame</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.process_database_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">process_database_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">key_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'annotation'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapsed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">resource_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">automatic_download</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#process_database_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.process_database_annotations" title="Link to this definition">#</a></dt>
<dd><p>Given a database and annotation type, find and process the annotations into a dictionary mapping each PTM to its annotations, or vice versa</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>database: str</strong></dt><dd><p>source of annotation</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>type of annotation to retrieve</p>
</dd>
<dt><strong>key_type</strong><span class="classifier">str</span></dt><dd><p>whether the annotation or ptm should be the key of the output dictionary. Default is annotation</p>
</dd>
<dt><strong>collapsed</strong><span class="classifier">bool</span></dt><dd><p>whether to combine annotations for similar types into a single annotation. For example, ‘cell growth, induced’ and ‘cell growth, inhibited’ would be simplified to ‘cell growth’. Default is False.</p>
</dd>
<dt><strong>resource_dir</strong><span class="classifier">str or None</span></dt><dd><p>location of annotations. By default, this will look for annotations in PTM-POSE resource directory</p>
</dd>
<dt><strong>automatic_download: bool</strong></dt><dd><p>Whether to automatically download annotations that are not yet present in resource files directory</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">additional keyword arguments</span></dt><dd><p>Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.simplify_annotation">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">simplify_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#simplify_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.simplify_annotation" title="Link to this definition">#</a></dt>
<dd><p>Given an annotation, remove additional information such as whether or not a function is increasing or decreasing. For example, ‘cell growth, induced’ would be simplified to ‘cell growth’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation: str</strong></dt><dd><p>Annotation to simplify</p>
</dd>
<dt><strong>sep: str</strong></dt><dd><p>Separator that splits the core annotation from additional detail. Default is ‘,’. Assumes the first element is the core annotation.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annotation: str</dt><dd><p>Simplified annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.unify_interaction_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">unify_interaction_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#unify_interaction_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.unify_interaction_data" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptm data and a column containing interaction data, extract the interacting protein, type of interaction, and convert to UniProtKB accession. This will be added as a new column labeled ‘Interacting ID’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe containing PTM data</p>
</dd>
<dt><strong>interaction_col: str</strong></dt><dd><p>column containing interaction information from a specific database</p>
</dd>
<dt><strong>name_dict: dict</strong></dt><dd><p>dictionary to convert names within given database to UniProt IDs. For cases when name is not necessarily one of the gene names listed in UniProt</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>interact: pd.DataFrame</dt><dd><p>Contains PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="analyze-modules">
<h2>Analyze Modules<a class="headerlink" href="#analyze-modules" title="Link to this heading">#</a></h2>
<section id="module-ptm_pose.analyze.summarize">
<span id="summaries"></span><h3>Summaries<a class="headerlink" href="#module-ptm_pose.analyze.summarize" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.summarize.combine_outputs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.summarize.</span></span><span class="sig-name descname"><span class="pre">combine_outputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_removed_annotations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop_codon_introduction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_conflicting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/summarize.html#combine_outputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.summarize.combine_outputs" title="Link to this definition">#</a></dt>
<dd><p>Given the spliced_ptms (differentially included) and altered_flanks (altered flanking sequences) dataframes obtained from project and flanking_sequences modules, combine the two into a single dataframe that categorizes each PTM by the impact on the PTM site</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases</p>
</dd>
<dt><strong>include_stop_codon_introduction: bool</strong></dt><dd><p>Whether to include PTMs that introduce stop codons in the altered flanks. Default is False.</p>
</dd>
<dt><strong>remove_conflicting: bool</strong></dt><dd><p>Whether to remove PTMs that are both included and excluded across different splicing events. Default is True.</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Additional keyword arguments to pass to the function, will be passed to <cite>helpers.filter_ptms</cite> if filtering is desired. Will automatically filter out insignificant events if not provided</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.summarize.get_modification_class_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.summarize.</span></span><span class="sig-name descname"><span class="pre">get_modification_class_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/summarize.html#get_modification_class_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.summarize.get_modification_class_data" title="Link to this definition">#</a></dt>
<dd><p>Given ptm dataframe and a specific modification class, return a dataframe with only the PTMs of that class</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Dataframe with ptm information, such as the spliced_ptms or altered_flanks dataframe obtained during projection</p>
</dd>
<dt><strong>mod_class</strong><span class="classifier">str</span></dt><dd><p>The modification class to filter by, e.g. ‘Phosphorylation’, ‘Acetylation’, etc.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.summarize.get_modification_counts">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.summarize.</span></span><span class="sig-name descname"><span class="pre">get_modification_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/summarize.html#get_modification_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.summarize.get_modification_counts" title="Link to this definition">#</a></dt>
<dd><p>Given PTM data (either spliced ptms, altered flanks, or combined data), return the counts of each modification class</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events or with altered flanking sequences</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>modification_counts: pd.Series</dt><dd><p>Series with the counts of each modification class</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.summarize.plot_modification_breakdown">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.summarize.</span></span><span class="sig-name descname"><span class="pre">plot_modification_breakdown</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[(0.00392156862745098,</span> <span class="pre">0.45098039215686275,</span> <span class="pre">0.6980392156862745),</span> <span class="pre">(0.8705882352941177,</span> <span class="pre">0.5607843137254902,</span> <span class="pre">0.0196078431372549),</span> <span class="pre">(0.00784313725490196,</span> <span class="pre">0.6196078431372549,</span> <span class="pre">0.45098039215686275),</span> <span class="pre">(0.8352941176470589,</span> <span class="pre">0.3686274509803922,</span> <span class="pre">0.0),</span> <span class="pre">(0.8,</span> <span class="pre">0.47058823529411764,</span> <span class="pre">0.7372549019607844),</span> <span class="pre">(0.792156862745098,</span> <span class="pre">0.5686274509803921,</span> <span class="pre">0.3803921568627451),</span> <span class="pre">(0.984313725490196,</span> <span class="pre">0.6862745098039216,</span> <span class="pre">0.8941176470588236),</span> <span class="pre">(0.5803921568627451,</span> <span class="pre">0.5803921568627451,</span> <span class="pre">0.5803921568627451),</span> <span class="pre">(0.9254901960784314,</span> <span class="pre">0.8823529411764706,</span> <span class="pre">0.2),</span> <span class="pre">(0.33725490196078434,</span> <span class="pre">0.7058823529411765,</span> <span class="pre">0.9137254901960784)]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/summarize.html#plot_modification_breakdown"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.summarize.plot_modification_breakdown" title="Link to this definition">#</a></dt>
<dd><p>Plot the number of PTMs that are differentially included or have altered flanking sequences, separated by PTM type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs that are differentially included</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs that have altered flanking sequences</p>
</dd>
<dt><strong>colors: list</strong></dt><dd><p>List of colors to use for the bar plot (first two will be used). Default is seaborn colorblind palette.</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
<dt><strong>kwargs: dict</strong></dt><dd><p>Additional keyword arguments to pass to the function, will be passed to <cite>helpers.filter_ptms</cite> if filtering is desired. Will automatically filter out insignificant events by min_dpsi and significance if the columns are present</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.analyze.filter">
<span id="filtering-ptms-and-events"></span><h3>Filtering PTMs and Events<a class="headerlink" href="#module-ptm_pose.analyze.filter" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.filter.plot_filter_impact">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.filter.</span></span><span class="sig-name descname"><span class="pre">plot_filter_impact</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'count'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">topn</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/filter.html#plot_filter_impact"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.filter.plot_filter_impact" title="Link to this definition">#</a></dt>
<dd><p>Given a dataframe of PTMs and a set of filter arguments to be passed to helpers.filter_ptms, this function will plot the number or fraction of PTMs that are retained after filtering for each modification type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Dataframe containing PTM data with a column ‘Modification Class’ that contains the type of modification (e.g. phosphorylation, acetylation, etc.)</p>
</dd>
<dt><strong>output_type</strong><span class="classifier">str, optional</span></dt><dd><p>Type of output to plot, either ‘count’ or ‘fraction’. The default is ‘count’.</p>
</dd>
<dt><strong>topn</strong><span class="classifier">int, optional</span></dt><dd><p>The number of top modification classes to plot. The default is 10.</p>
</dd>
<dt><strong>ax</strong><span class="classifier">matplotlib.axes.Axes, optional</span></dt><dd><p>The axes to plot on. If None, a new figure and axes will be created. The default is None.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">keyword arguments</span></dt><dd><p>Additional keyword arguments to be passed to the filter_ptms function (e.g. min_studies, min_compendia, etc.). These will be extracted and checked for validity.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.analyze.annotations">
<span id="annotations"></span><h3>Annotations<a class="headerlink" href="#module-ptm_pose.analyze.annotations" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.annotation_enrichment">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">annotation_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#annotation_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.annotation_enrichment" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptm information (differential inclusion, altered flanking sequences, or both), calculate the enrichment of specific annotations in the dataset using a hypergeometric test. Background data can be provided/constructed in a few ways:</p>
<ol class="arabic simple">
<li><p>Use annotations from the entire phosphoproteome (background_type = ‘all’)</p></li>
<li><p>Use the alpha and min_dpsi parameter to construct a foreground that only includes significantly spliced PTMs, and use the entire provided spliced_ptms dataframe as the background. This will allow you to compare the enrichment of specific annotations in the significantly spliced PTMs compared to the entire dataset. Will do this automatically if alpha or min_dpsi is provided.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, ‘PTMInt’, ‘PTMcode’, ‘DEPOD’, ‘RegPhos’, ‘PTMsigDB’. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>annot_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>background_type: str</strong></dt><dd><p>how to construct the background data. Options include ‘pregenerated’ (default) and ‘significance’. If ‘significance’ is selected, the alpha and min_dpsi parameters must be provided. Otherwise, will use whole proteome in the ptm_coordinates dataframe as the background.</p>
</dd>
<dt><strong>collapse_on_similar: bool</strong></dt><dd><p>Whether to collapse similar annotations (for example, increasing and decreasing functions) into a single category. Default is False.</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset, if any</p>
</dd>
<dt><strong>alpha: float</strong></dt><dd><p>significance threshold to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>min_dpsi: float</strong></dt><dd><p>minimum delta PSI value to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.draw_pie">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">draw_pie</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dist</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xpos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ypos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'donut'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#draw_pie"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.draw_pie" title="Link to this definition">#</a></dt>
<dd><p>Draws pies individually, as if points on a scatter plot. This function was taken from this stack overflow post: <a class="reference external" href="https://stackoverflow.com/questions/56337732/how-to-plot-scatter-pie-chart-using-matplotlib">https://stackoverflow.com/questions/56337732/how-to-plot-scatter-pie-chart-using-matplotlib</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>dist: list</strong></dt><dd><p>list of values to be represented as pie slices for a single point</p>
</dd>
<dt><strong>xpos: float</strong></dt><dd><p>x position of pie chart in the scatter plot</p>
</dd>
<dt><strong>ypos: float</strong></dt><dd><p>y position of pie chart in the scatter plot</p>
</dd>
<dt><strong>size: float</strong></dt><dd><p>size of pie chart</p>
</dd>
<dt><strong>colors: list</strong></dt><dd><p>list of colors to use for pie slices</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>axis to plot on, if None, will create new figure</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.gene_set_enrichment">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">gene_set_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Significance'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpsi_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'dPSI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['GO_Biological_Process_2023',</span> <span class="pre">'Reactome_2022']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_sig_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#gene_set_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.gene_set_enrichment" title="Link to this definition">#</a></dt>
<dd><p>Given spliced_ptms and/or altered_flanks dataframes (or the dataframes combined from combine_outputs()), perform gene set enrichment analysis using the enrichr API</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with differentially included PTMs projected onto splicing events and with annotations appended from various databases. Default is None (will not be considered in analysis). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases. Default is None (will not be considered). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>combined: pd.DataFrame</strong></dt><dd><p>Combined dataframe with spliced_ptms and altered_flanks dataframes. Default is None. If provided, spliced_ptms and altered_flanks dataframes will be ignored.</p>
</dd>
<dt><strong>gene_sets: list</strong></dt><dd><p>List of gene sets to use in enrichment analysis. Default is [‘KEGG_2021_Human’, ‘GO_Biological_Process_2023’, ‘GO_Cellular_Component_2023’, ‘GO_Molecular_Function_2023’,’Reactome_2022’]. Look at gseapy and enrichr documentation for other available gene sets</p>
</dd>
<dt><strong>background: list</strong></dt><dd><p>List of genes to use as background in enrichment analysis. Default is None (all genes in the gene set database will be used).</p>
</dd>
<dt><strong>return_sig_only: bool</strong></dt><dd><p>Whether to return only significantly enriched gene sets. Default is True.</p>
</dd>
<dt><strong>max_retries: int</strong></dt><dd><p>Number of times to retry downloading gene set enrichment data from enrichr API. Default is 5.</p>
</dd>
<dt><strong>delay: int</strong></dt><dd><p>Number of seconds to wait between retries. Default is 10.</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the combine_outputs() function from the summarize module. These will be used to filter the spliced_ptms and altered_flanks dataframes before performing gene set enrichment analysis. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>combine_outputs()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>results: pd.DataFrame</dt><dd><p>Dataframe with gene set enrichment results from enrichr API</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.get_available_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">get_available_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#get_available_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.get_available_annotations" title="Link to this definition">#</a></dt>
<dd><p>Given a PTM dataframe, indicate the annotations that are available for analysis and indicate whether they have already been appended to the PTM dataset</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>contains PTM information and may have annotations already appended, such as spliced_ptms and altered_flanks dataframes generated during projection</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>available_annots</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>DataFrame indicating the available annotation types and their sources, as well as whether they have been appended to the PTM data.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.get_background_annotation_counts">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">get_background_annotation_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#get_background_annotation_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.get_background_annotation_counts" title="Link to this definition">#</a></dt>
<dd><p>Given a database and annotation type, retrieve the counts of PTMs associated with the requested annotation type across all PTMs in the ptm_coordinates dataframe used for projection</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>database</strong><span class="classifier">str</span></dt><dd><p>Source of annotation. Default is PhosphoSitePlus</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>Type of annotation that can be found in indicated database. Default is ‘Function’. Other options include ‘Process’, ‘Disease’, ‘Enzyme’, ‘Interactions’, etc.</p>
</dd>
<dt><strong>kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.get_enrichment_inputs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">get_enrichment_inputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#get_enrichment_inputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.get_enrichment_inputs" title="Link to this definition">#</a></dt>
<dd><p>Given the spliced ptms, altered_flanks, or combined PTMs dataframe, identify the number of PTMs corresponding to specific annotations in the foreground (PTMs impacted by splicing) and the background (all PTMs in the proteome or all PTMs in dataset not impacted by splicing). This information can be used to calculate the enrichment of specific annotations among PTMs impacted by splicing. Several options are provided for constructing the background data: all (based on entire proteome in the ptm_coordinates dataframe) or significance (foreground PTMs are extracted from provided spliced PTMs based on significance and minimum delta PSI)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases. This can be either the spliced_ptms, altered_flanks, or combined dataframe.</p>
</dd>
<dt><strong>annot_type</strong><span class="classifier">str</span></dt><dd><p>type of annotation to pull the annotations from. Default is ‘Function’.</p>
</dd>
<dt><strong>database</strong><span class="classifier">str</span></dt><dd><p>source of annotations. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>background_type</strong><span class="classifier">str</span></dt><dd><p>Type of background to construct. Options are either ‘all’ (all PTMs in proteome) or ‘significance’ (only PTMs in dataset). Note that significance option assumes that PTMs have not already been filtered for significance.</p>
</dd>
<dt><strong>collapse_on_similar</strong><span class="classifier">bool</span></dt><dd><p>Whether to collapse similar annotations (for example, “cell growth, increased” and “cell growth, decreased”) into a single category. Default is False.</p>
</dd>
<dt><strong>mod_class</strong><span class="classifier">str</span></dt><dd><p>Type of modification to perform enrichment for</p>
</dd>
<dt><strong>min_dpsi: float</strong></dt><dd><p>Minimum change in PSI required to return a PTM as associated with the annotation. Default is 0.1. This can be used to filter out PTMs that are not significantly spliced.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float</span></dt><dd><p>Significance threshold to use to filter PTMs based on their significance. Default is 0.05. This can be used to filter out PTMs that are not significantly spliced.</p>
</dd>
<dt><strong>kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.get_ptm_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">get_ptm_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#get_ptm_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.get_ptm_annotations" title="Link to this definition">#</a></dt>
<dd><p>Given spliced ptm information obtained from project and annotate modules, grab PTMs in spliced ptms associated with specific PTM modules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>annot_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, or ‘PTMInt’. ELM and PTMInt data will automatically be downloaded, but due to download restrictions, PhosphoSitePlus data must be manually downloaded and annotated in the spliced_ptms data using functions from the annotate module. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>collapse_on_similar</strong><span class="classifier">bool</span></dt><dd><p>Whether to collapse similar annotations (for example, “cell growth, increased” and “cell growth, decreased”) into a single category. Default is False.</p>
</dd>
<dt><strong>min_dpsi: float</strong></dt><dd><p>Minimum change in PSI required to return a PTM as associated with the annotation. Default is 0.1. This can be used to filter out PTMs that are not significantly spliced.</p>
</dd>
<dt><strong>alpha</strong><span class="classifier">float</span></dt><dd><p>Significance threshold to use to filter PTMs based on their significance. Default is 0.05. This can be used to filter out PTMs that are not significantly spliced.</p>
</dd>
<dt><strong>kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments to pass to the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>annotations</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Individual PTM information of PTMs that have been associated with the requested annotation type.</p>
</dd>
<dt><strong>annotation_counts</strong><span class="classifier">pd.Series or pd.DataFrame</span></dt><dd><p>Number of PTMs associated with each annotation of the requested annotation type. If dPSI col is provided or impact col is present, will output annotation counts for each type of impact (‘Included’, ‘Excluded’, ‘Altered Flank’) separately.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.plot_EnrichR_pies">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">plot_EnrichR_pies</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enrichr_results</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_terms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">terms_to_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">edgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">row_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'circle'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#plot_EnrichR_pies"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.plot_EnrichR_pies" title="Link to this definition">#</a></dt>
<dd><p>Given PTM-specific EnrichR results, plot EnrichR score for the provided terms, with each self point represented as a pie chart indicating the fraction of genes in the group with PTMs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_results: pd.selfFrame</strong></dt><dd><p>selfFrame containing PTM-specific results from EnrichR analysis</p>
</dd>
<dt><strong>num_to_plot: int</strong></dt><dd><p>number of terms to plot, if None, will plot all terms. Ignored if specific terms are provided in terms to plot list</p>
</dd>
<dt><strong>terms_to_plot: list</strong></dt><dd><p>list of terms to plot</p>
</dd>
<dt><strong>colors: list</strong></dt><dd><p>list of colors to use for pie slices. Default is None, which will use seaborn colorblind palette</p>
</dd>
<dt><strong>edgecolor: str</strong></dt><dd><p>color to use for edge of pie slices. Default is None, which will use the same color as the pie slice</p>
</dd>
<dt><strong>row_height: float</strong></dt><dd><p>height of each row in the plot. Default is 0.3.</p>
</dd>
<dt><strong>type: str</strong></dt><dd><p>type of pie chart to plot. Default is ‘circle’. Options include ‘circle’ and ‘donut’ (hole in center).</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>axis to plot on, if None, will create new figure</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.plot_annotation_counts">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">plot_annotation_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annot_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">colors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_terms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_loc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1.05,</span> <span class="pre">0.5)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'database'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#plot_annotation_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.plot_annotation_counts" title="Link to this definition">#</a></dt>
<dd><p>Given a dataframe with PTM annotations added, plot the top annotations associated with the PTMs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with differentially included PTMs</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>Database to use for annotations. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>annot_type: str</strong></dt><dd><p>Type of annotation to plot. Default is ‘Function’.</p>
</dd>
<dt><strong>collapse_on_similar: bool</strong></dt><dd><p>Whether to collapse similar annotations into a single category. Default is True.</p>
</dd>
<dt><strong>colors: list</strong></dt><dd><p>List of colors to use for the bar plot. Default is None.</p>
</dd>
<dt><strong>top_terms: int</strong></dt><dd><p>Number of top terms to plot. Default is 5.</p>
</dd>
<dt><strong>legend: bool</strong></dt><dd><p>Whether to show the legend. Default is True.</p>
</dd>
<dt><strong>legend_loc: tuple</strong></dt><dd><p>Location of the legend. Default is None, which will place the legend in the upper right corner.</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
<dt><strong>title_type: str</strong></dt><dd><p>Type of title to use for the plot. Default is ‘database’. Options include ‘database’ and ‘detailed’.</p>
</dd>
<dt><strong>title: str</strong></dt><dd><p>Title to use for the plot. Default is None.</p>
</dd>
<dt><strong>fontsize: int</strong></dt><dd><p>Font size for the plot. Default is 10.</p>
</dd>
<dt><strong>legend_loc: tuple</strong></dt><dd><p>Location of the legend. Default is None, which will place the legend to the right of the plot.</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotations.plot_available_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.annotations.</span></span><span class="sig-name descname"><span class="pre">plot_available_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_annotations_in_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_all_ptm_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/annotations.html#plot_available_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotations.plot_available_annotations" title="Link to this definition">#</a></dt>
<dd><p>Given a dataframe with ptm annotations added, show the number of PTMs associated with each annotation type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs and annotations added</p>
</dd>
<dt><strong>only_annotations_in_data</strong><span class="classifier">bool</span></dt><dd><p>Only plot annotations that are already appended to the dataset</p>
</dd>
<dt><strong>show_all_ptm_count: bool</strong></dt><dd><p>Whether to show the total number of PTMs in the dataset. Default is True.</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.analyze.interactions">
<span id="protein-interactions"></span><h3>Protein Interactions<a class="headerlink" href="#module-ptm_pose.analyze.interactions" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.get_edge_colors">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.interactions.</span></span><span class="sig-name descname"><span class="pre">get_edge_colors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interaction_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defaultedgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_edges_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Database'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database_color_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'Multiple':</span> <span class="pre">'purple',</span> <span class="pre">'PSP/RegPhos':</span> <span class="pre">'red',</span> <span class="pre">'PTMInt':</span> <span class="pre">'gold',</span> <span class="pre">'PTMcode':</span> <span class="pre">'blue',</span> <span class="pre">'PhosphoSitePlus':</span> <span class="pre">'green'}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#get_edge_colors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.get_edge_colors" title="Link to this definition">#</a></dt>
<dd><p>Get the edge colors to use for a provided networkx graph, either plotting all edges the same color or coloring them based on the database they are from.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>interaction_graph: nx.Graph</strong></dt><dd><p>networkx graph containing the interaction network</p>
</dd>
<dt><strong>network_data: pd.DataFrame</strong></dt><dd><p>specific network edge data that contains information on which database the interaction is from and any other relevant information (such as regulation change)</p>
</dd>
<dt><strong>defaultedgecolor</strong><span class="classifier">‘str’</span></dt><dd><p>Default color to use for edges if no specific database color is found. Default is ‘gray’.</p>
</dd>
<dt><strong>color_edges_by</strong><span class="classifier">str</span></dt><dd><p>How to color the edges. Options are ‘Database’ to color by the database they are from, or ‘Same’ to color all edges the same color. Default is ‘Database’.</p>
</dd>
<dt><strong>database_color_dict</strong><span class="classifier">dict</span></dt><dd><p>Colors to use for specific databases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.get_interaction_stats">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.interactions.</span></span><span class="sig-name descname"><span class="pre">get_interaction_stats</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interaction_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#get_interaction_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.get_interaction_stats" title="Link to this definition">#</a></dt>
<dd><p>Given the networkx interaction graph, calculate various network centrality measures to identify the most relevant PTMs or genes in the network</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.plot_interaction_network">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.interactions.</span></span><span class="sig-name descname"><span class="pre">plot_interaction_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">interaction_graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_stats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modified_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modified_node_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lightblue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_node_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defaultedgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_edges_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Same'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend_fontsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proteins_to_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelcolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#plot_interaction_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.plot_interaction_network" title="Link to this definition">#</a></dt>
<dd><p>Given the interaction graph and network data outputted from analyze.protein_interactions, plot the interaction network, signifying which proteins or ptms are altered by splicing and the specific regulation change that occurs. by default, will only label proteins</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>interaction_graph: nx.Graph</strong></dt><dd><p>NetworkX graph object representing the interaction network, created from analyze.get_interaction_network</p>
</dd>
<dt><strong>network_data: pd.DataFrame</strong></dt><dd><p>Dataframe containing details about specifici protein interactions (including which protein contains the spliced PTMs)</p>
</dd>
<dt><strong>network_stats: pd.DataFrame</strong></dt><dd><p>Dataframe containing network statistics for each protein in the interaction network, obtained from analyze.get_interaction_stats(). Default is None, which will not label any proteins in the network.</p>
</dd>
<dt><strong>modified_color: str</strong></dt><dd><p>Color to use for proteins that are spliced. Default is ‘red’.</p>
</dd>
<dt><strong>modified_node_size: int</strong></dt><dd><p>Size of nodes that are spliced. Default is 10.</p>
</dd>
<dt><strong>interacting_color: str</strong></dt><dd><p>Color to use for proteins that are not spliced. Default is ‘lightblue’.</p>
</dd>
<dt><strong>interacting_node_size: int</strong></dt><dd><p>Size of nodes that are not spliced. Default is 1.</p>
</dd>
<dt><strong>edgecolor: str</strong></dt><dd><p>Color to use for edges in the network. Default is ‘gray’.</p>
</dd>
<dt><strong>seed: int</strong></dt><dd><p>Seed to use for spring layout of network. Default is 200.</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
<dt><strong>proteins_to_label: list, int, or str</strong></dt><dd><p>Specific proteins to label in the network. If list, will label all proteins in the list. If int, will label the top N proteins by degree centrality. If str, will label the specific protein. Default is None, which will not label any proteins in the network.</p>
</dd>
<dt><strong>labelcolor: str</strong></dt><dd><p>Color to use for labels. Default is ‘black’.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.plot_network_centrality">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.interactions.</span></span><span class="sig-name descname"><span class="pre">plot_network_centrality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_stats</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_data</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">centrality_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Degree'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modified_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#plot_network_centrality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.plot_network_centrality" title="Link to this definition">#</a></dt>
<dd><p>Given the network statistics data obtained from analyze.get_interaction_stats(), plot the top N proteins in the protein interaction network based on centrality measure (Degree, Betweenness, or Closeness)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>network_stats: pd.DataFrame</strong></dt><dd><p>Dataframe containing network statistics for each protein in the interaction network, obtained from analyze.get_interaction_stats()</p>
</dd>
<dt><strong>network_data: pd.DataFrame</strong></dt><dd><p>Dataframe containing information on which proteins are spliced and how they are altered. Default is None, which will plot all proteins the same color (interacting_color)</p>
</dd>
<dt><strong>centrality_measure: str</strong></dt><dd><p>Centrality measure to use for plotting. Default is ‘Degree’. Options include ‘Degree’, ‘Degree Centrality’, ‘Betweenness Centrality’, ‘Closeness Centrality’.</p>
</dd>
<dt><strong>top_N: int</strong></dt><dd><p>Number of top proteins to plot. Default is 10.</p>
</dd>
<dt><strong>modified_color: str</strong></dt><dd><p>Color to use for proteins that are spliced. Default is ‘red’.</p>
</dd>
<dt><strong>interacting_color: str</strong></dt><dd><p>Color to use for proteins that are not spliced. Default is ‘black’.</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.interactions.</span></span><span class="sig-name descname"><span class="pre">protein_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_enzyme_interactions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_databases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'PTMcode',</span> <span class="pre">'PTMInt',</span> <span class="pre">'RegPhos',</span> <span class="pre">'DEPOD',</span> <span class="pre">'OmniPath']</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions" title="Link to this definition">#</a></dt>
<dd><p>Class to assess interactions facilitated by PTMs in splicing network</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>include_enzyme_interactions: bool</strong></dt><dd><p>Whether to include interactions with enzymes in the network, such as kinase-substrate interactions. Default is True</p>
</dd>
<dt><strong>interaction_databases: list</strong></dt><dd><p>List of databases whose information to include in the network. Default is [‘PhosphoSitePlus’, ‘PTMcode’, ‘PTMInt’, ‘RegPhos’, ‘DEPOD’, ‘ELM’]</p>
</dd>
<dt><strong>**kwargs: additional keyword arguments</strong></dt><dd><p>Additional keyword arguments, which will be fed into the <cite>filter_ptms()</cite> function from the helper module. These will be used to filter ptms with lower evidence. For example, if you want to filter PTMs based on the number of MS observations, you can add ‘min_MS_observations = 2’ to the kwargs. This will filter out any PTMs that have less than 2 MS observations. See the <cite>filter_ptms()</cite> function for more options.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>interaction_graph: nx.Graph</strong></dt><dd><p>NetworkX graph object representing the interaction network, created from analyze.get_interaction_network</p>
</dd>
<dt><strong>network_data: pd.DataFrame</strong></dt><dd><p>Dataframe containing details about specifici protein interactions (including which protein contains the spliced PTMs)</p>
</dd>
<dt><strong>network_stats: pd.DataFrame</strong></dt><dd><p>Dataframe containing network statistics for each protein in the interaction network, obtained from analyze.get_interaction_stats(). Default is None, which will not label any proteins in the network.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<div class="pst-scrollable-table-container"><table class="autosummary longtable table autosummary">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.compare_to_nease" title="ptm_pose.analyze.interactions.protein_interactions.compare_to_nease"><code class="xref py py-obj docutils literal notranslate"><span class="pre">compare_to_nease</span></code></a>(nease_edges)</p></td>
<td><p>Given the network edges generated by NEASE, compare the network edges generated by NEASE to the network edges generated by the PTM-driven interactions</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_network" title="ptm_pose.analyze.interactions.protein_interactions.get_interaction_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_interaction_network</span></code></a>([node_type])</p></td>
<td><p>Given the spliced PTM data, extract interaction information and construct a dataframe containing all possible interactions driven by PTMs, either centered around specific PTMs or specific genes.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats" title="ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_interaction_stats</span></code></a>()</p></td>
<td><p>Given the networkx interaction graph, calculate various network centrality measures to identify the most relevant PTMs or genes in the network</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network" title="ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_protein_interaction_network</span></code></a>(protein)</p></td>
<td><p>Given a specific protein, return the network data for that protein</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network" title="ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_interaction_network</span></code></a>([modified_color, ...])</p></td>
<td><p>Given the interactiong graph and network data outputted from analyze.get_interaction_network, plot the interaction network, signifying which proteins or ptms are altered by splicing and the specific regulation change that occurs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison" title="ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_nease_comparison</span></code></a>([ax, nease_edges])</p></td>
<td><p>Given the comparison data generated by compare_to_nease, plot the number of edges identified by NEASE and PTM-POSE</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality" title="ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_network_centrality</span></code></a>([...])</p></td>
<td><p>Plot the centrality measure for the top N proteins in the interaction network based on a specified centrality measure.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network" title="ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network"><code class="xref py py-obj docutils literal notranslate"><span class="pre">summarize_protein_network</span></code></a>(protein)</p></td>
<td><p>Given a protein of interest, summarize the network data for that protein</p></td>
</tr>
</tbody>
</table>
</div>
<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.compare_to_nease">
<span class="sig-name descname"><span class="pre">compare_to_nease</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nease_edges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.compare_to_nease"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.compare_to_nease" title="Link to this definition">#</a></dt>
<dd><p>Given the network edges generated by NEASE, compare the network edges generated by NEASE to the network edges generated by the PTM-driven interactions</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>nease_edges</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Interactions found by NEASE, which is the output of nease.get_edges() function after running NEASE (see nease_runner module for example)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>nease_comparison</strong><span class="classifier">pd.DataFrame</span></dt><dd><p>Dataframe containing the all edges found by NEASE and PTM-POSE. This will include edges that are unique to NEASE, unique to PTM-POSE, and common between the two.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.get_interaction_network">
<span class="sig-name descname"><span class="pre">get_interaction_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Gene'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.get_interaction_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_network" title="Link to this definition">#</a></dt>
<dd><p>Given the spliced PTM data, extract interaction information and construct a dataframe containing all possible interactions driven by PTMs, either centered around specific PTMs or specific genes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>node_type: str</strong></dt><dd><p>What to define interactions by. Can either be by ‘PTM’, which will consider each PTM as a separate node, or by ‘Gene’, which will aggregate information across all PTMs of a single gene into a single node. Default is ‘Gene’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats">
<span class="sig-name descname"><span class="pre">get_interaction_stats</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.get_interaction_stats"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats" title="Link to this definition">#</a></dt>
<dd><p>Given the networkx interaction graph, calculate various network centrality measures to identify the most relevant PTMs or genes in the network</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network">
<span class="sig-name descname"><span class="pre">get_protein_interaction_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.get_protein_interaction_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network" title="Link to this definition">#</a></dt>
<dd><p>Given a specific protein, return the network data for that protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>protein: str</strong></dt><dd><p>Gene name of the protein of interest</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>protein_network: pd.DataFrame</dt><dd><p>Dataframe containing network data for the protein of interest</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network">
<span class="sig-name descname"><span class="pre">plot_interaction_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">modified_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modified_node_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lightblue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_node_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">defaultedgecolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gray'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">color_edges_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Same'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">proteins_to_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labelcolor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">legend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_nease_comparison</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.plot_interaction_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network" title="Link to this definition">#</a></dt>
<dd><p>Given the interactiong graph and network data outputted from analyze.get_interaction_network, plot the interaction network, signifying which proteins or ptms are altered by splicing and the specific regulation change that occurs. by default, will only label proteins</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>modified_color: str</strong></dt><dd><p>Color to use for nodes that are modified by splicing. Default is ‘red’</p>
</dd>
<dt><strong>modified_node_size: int</strong></dt><dd><p>Size of nodes that are modified by splicing. Default is 10</p>
</dd>
<dt><strong>interacting_color: str</strong></dt><dd><p>Color to use for nodes that interact with modified nodes. Default is ‘lightblue’</p>
</dd>
<dt><strong>interacting_node_size: int</strong></dt><dd><p>Size of nodes that interact with modified nodes. Default is 5</p>
</dd>
<dt><strong>defaultedgecolor: str</strong></dt><dd><p>Color to use for edges in the network. Default is ‘gray’. Can choose to color by database by providing a dictionary with database names as keys and colors as values or by specifying ‘database’ as color</p>
</dd>
<dt><strong>color_edges_by: str</strong></dt><dd><p>How to color edges in the network. Default is ‘Same’, which will color all edges the same color. Can also specify ‘Database’ to color edges based on the database they are from. If using ‘Database’, please provide a dictionary with database names as keys and colors as values in defaultedgecolor parameter.</p>
</dd>
<dt><strong>seed: int</strong></dt><dd><p>Seed to use for random number generator. Default is 200</p>
</dd>
<dt><strong>ax: matplotlib.pyplot.Axes</strong></dt><dd><p>Axes object to plot the network. Default is None</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison">
<span class="sig-name descname"><span class="pre">plot_nease_comparison</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nease_edges</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.plot_nease_comparison"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison" title="Link to this definition">#</a></dt>
<dd><p>Given the comparison data generated by compare_to_nease, plot the number of edges identified by NEASE and PTM-POSE</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ax</strong><span class="classifier">matplotlib.pyplot.Axes</span></dt><dd><p>axes to plot on</p>
</dd>
<dt><strong>nease_edges</strong><span class="classifier">pd.DataFrame, optional</span></dt><dd><p>Interactions found by NEASE, which is the output of nease.get_edges() function after running NEASE (see nease_runner module for example). Only needed if you have not run comparison_to_nease() previously</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality">
<span class="sig-name descname"><span class="pre">plot_network_centrality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">centrality_measure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Degree'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_N</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modified_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'red'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interacting_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'black'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.plot_network_centrality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality" title="Link to this definition">#</a></dt>
<dd><p>Plot the centrality measure for the top N proteins in the interaction network based on a specified centrality measure. This will help identify the most relevant PTMs/genes in the network.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>centrality_measure: str</strong></dt><dd><p>How to calculate centrality. Options include ‘Degree’, ‘Degree Centrality’, ‘Betweenness Centrality’, ‘Closeness Centrality’, and ‘Eigenvector Centrality’. Default is ‘Degree’.</p>
</dd>
<dt><strong>top_N: int</strong></dt><dd><p>Number of top proteins to plot based on the centrality measure. Default is 10.</p>
</dd>
<dt><strong>modified_color: str</strong></dt><dd><p>Color to use for proteins that are spliced. Default is ‘red’.</p>
</dd>
<dt><strong>interacting_color: str</strong></dt><dd><p>Color to use for proteins that are not spliced. Default is ‘black’.</p>
</dd>
<dt><strong>ax: matplotlib.pyplot.Axes</strong></dt><dd><p>Axes object to plot the centrality bar plot. If None, a new figure will be created. Default is None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network">
<span class="sig-name descname"><span class="pre">summarize_protein_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/interactions.html#protein_interactions.summarize_protein_network"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network" title="Link to this definition">#</a></dt>
<dd><p>Given a protein of interest, summarize the network data for that protein</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>protein: str</strong></dt><dd><p>Gene name of the protein of interest</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-ptm_pose.analyze.enzyme">
<span id="enzyme-regulation"></span><h3>Enzyme Regulation<a class="headerlink" href="#module-ptm_pose.analyze.enzyme" title="Link to this heading">#</a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.enzyme.compare_KL_for_sequence">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.enzyme.</span></span><span class="sig-name descname"><span class="pre">compare_KL_for_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inclusion_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclusion_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comparison_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'percentile'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/enzyme.html#compare_KL_for_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.enzyme.compare_KL_for_sequence" title="Link to this definition">#</a></dt>
<dd><p>Given two sequences, compare the kinase library scores, percentiles, or ranks for each sequence. Optionally, provide a dPSI value to calculate the relative change in preference for each kinase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>inclusion_seq</strong><span class="classifier">str</span></dt><dd><p>sequence to score for inclusion preference, with modification lowercased</p>
</dd>
<dt><strong>exclusion_seq</strong><span class="classifier">str</span></dt><dd><p>sequence to score for exclusion preference, with modification lowercased</p>
</dd>
<dt><strong>dpsi</strong><span class="classifier">float</span></dt><dd><p>dPSI value for the PTM event, which will be used to calculate the relative change in preference for each kinase (score difference * dPSI). Default is None.</p>
</dd>
<dt><strong>comparison_type</strong><span class="classifier">str</span></dt><dd><p>type of comparison to perform. Can be ‘percentile’, ‘score’, or ‘rank’. Default is ‘percentile’.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.enzyme.get_all_KL_scores">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.enzyme.</span></span><span class="sig-name descname"><span class="pre">get_all_KL_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kin_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['ser_thr',</span> <span class="pre">'tyrosine']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'percentiles'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/enzyme.html#get_all_KL_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.enzyme.get_all_KL_scores" title="Link to this definition">#</a></dt>
<dd><p>Given a dataset with flanking sequences, score each flanking sequence</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq_data</strong><span class="classifier">pandas dataframe</span></dt><dd><p>processed dataframe containing flanking sequences to score</p>
</dd>
<dt><strong>seq_col</strong><span class="classifier">str</span></dt><dd><p>column in seq_data containing the flanking sequences</p>
</dd>
<dt><strong>kin_type</strong><span class="classifier">list</span></dt><dd><p>list of kinase types to score. Can be ‘ser_thr’ or ‘ST’ for serine/threonine kinases, or ‘tyrosine’ or ‘Y’ for tyrosine kinases. Default is [‘ser_thr’, ‘tyrosine’].</p>
</dd>
<dt><strong>score_type</strong><span class="classifier">str</span></dt><dd><p>type of score to calculate. Can be ‘percentile’, ‘score’, or ‘rank’. Default is ‘percentile’.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>merged_data</strong><span class="classifier">dict</span></dt><dd><p>dictionary containing the merged dataframes for each kinase type with KinaseLibrary scores</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="id1">
<h3>Flanking Sequences<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<span class="target" id="module-ptm_pose.analyze.flanking_sequences"></span><dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.compare_inclusion_motifs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">compare_inclusion_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flanking_sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#compare_inclusion_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.compare_inclusion_motifs" title="Link to this definition">#</a></dt>
<dd><p>Given a DataFrame containing flanking sequences with changes and a DataFrame containing ELM class information, identify motifs that are found in the inclusion and exclusion events, identifying motifs unique to each case. This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>flanking_sequences: pandas.DataFrame</strong></dt><dd><p>DataFrame containing flanking sequences with changes, obtained from get_flanking_changes_from_splice_data()</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>). Recommended to download this file and input it manually, but will download from ELM otherwise</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>flanking_sequences: pandas.DataFrame</dt><dd><p>DataFrame containing flanking sequences with changes and motifs found in the inclusion and exclusion events</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.findAlteredPositions">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">findAlteredPositions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#findAlteredPositions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.findAlteredPositions" title="Link to this definition">#</a></dt>
<dd><p>Given two sequences, identify the location of positions that have changed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>sequences to compare (order does not matter)</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>size of the flanking sequences (default is 5). This is used to make sure the provided sequences are the correct length</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>altered_positions: list</dt><dd><p>list of positions that have changed</p>
</dd>
<dt>residue_change: list</dt><dd><p>list of residues that have changed associated with that position</p>
</dd>
<dt>flank_side: str</dt><dd><p>indicates which side of the flanking sequence the change has occurred (N-term, C-term, or Both)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.find_motifs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">find_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#find_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.find_motifs" title="Link to this definition">#</a></dt>
<dd><p>Given a sequence and a dataframe containinn ELM class information, identify motifs that can be found in the provided sequence using the RegEx expression provided by ELM (PTMs not considered). This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq: str</strong></dt><dd><p>sequence to search for motifs</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.getSequenceIdentity">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">getSequenceIdentity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#getSequenceIdentity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.getSequenceIdentity" title="Link to this definition">#</a></dt>
<dd><p>Given two flanking sequences, calculate the sequence identity between them using Biopython and parameters definded by Pillman et al. BMC Bioinformatics 2011</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>flanking sequence</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>normalized_score: float</dt><dd><p>normalized score of sequence similarity between flanking sequences (calculated similarity/max possible similarity)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.plot_alterations_matrix">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">plot_alterations_matrix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modification_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">title</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ax</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#plot_alterations_matrix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.plot_alterations_matrix" title="Link to this definition">#</a></dt>
<dd><p>Given the altered flanking sequences dataframe, plot a matrix showing the positions of altered residues for specific proteins, as well as the specific change</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with altered flanking sequences, and annotated information added with analyze.compare_flanking_sequences</p>
</dd>
<dt><strong>modification_class: str</strong></dt><dd><p>Specific modification class to plot. Default is None, which will plot all modification classes.</p>
</dd>
<dt><strong>residue: str</strong></dt><dd><p>Specific residue to plot. Default is None, which will plot all residues.</p>
</dd>
<dt><strong>title: str</strong></dt><dd><p>Title of the plot. Default is ‘’ (no title).</p>
</dd>
<dt><strong>ax: matplotlib.Axes</strong></dt><dd><p>Axis to plot on. If None, will create new figure. Default is None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.plot_location_of_altered_flanking_residues">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">plot_location_of_altered_flanking_residues</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">modification_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#plot_location_of_altered_flanking_residues"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.plot_location_of_altered_flanking_residues" title="Link to this definition">#</a></dt>
<dd><p>Plot the number of PTMs with altered residues as specific positions relative to the PTM site. This includes the specific position of the residue (-5 to +5 from PTM site) and the specific side of the PTM site that is altered (N-term or C-term)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with altered flanking sequences, and annotated information added with analyze.compare_flanking_sequences</p>
</dd>
<dt><strong>figsize: tuple</strong></dt><dd><p>Size of the figure. Default is (4,3).</p>
</dd>
<dt><strong>modification_class: str</strong></dt><dd><p>Specific modification class to plot. Default is None, which will plot all modification classes.</p>
</dd>
<dt><strong>residue: str</strong></dt><dd><p>Specific residue to plot. Default is None, which will plot all residues.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.flanking_sequences.plot_sequence_differences">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">plot_sequence_differences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inclusion_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclusion_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dpsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">figsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(3,</span> <span class="pre">1)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze/flanking_sequences.html#plot_sequence_differences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.flanking_sequences.plot_sequence_differences" title="Link to this definition">#</a></dt>
<dd><p>Given the flanking sequences for a PTM resulting from a specific splice event, plot the differences between the two sequences, coloring the changing residues in red. If dPSI is also provided, will add an arrow to the plot indicating the direction of change</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>inclusion_seq: str</strong></dt><dd><p>Sequence of the inclusion isoform (with spliced region included)</p>
</dd>
<dt><strong>exclusion_seq: str</strong></dt><dd><p>Sequence of the exclusion isoform (with spliced region excluded)</p>
</dd>
<dt><strong>dpsi: float</strong></dt><dd><p>Change in PSI for the specific splice event. Default is None, which will not add an arrow to the plot.</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>Size of flanking region to plot. Default is 5. This must be less than half the length of the shortest sequence.</p>
</dd>
<dt><strong>figsize: tuple</strong></dt><dd><p>Size of the figure. Default is (3,1).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Dependencies.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Dependencies</p>
      </div>
    </a>
    <a class="right-next"
       href="Running_PTM-POSE/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">&lt;no title&gt;</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.pose_config">Configuration</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.pose_config.download_translator"><code class="docutils literal notranslate"><span class="pre">download_translator()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.project">PTM Projection</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.find_ptms_in_region"><code class="docutils literal notranslate"><span class="pre">find_ptms_in_region()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_MATS"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_MATS()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_SpliceSeq"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_SpliceSeq()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.project.project_ptms_onto_splice_events"><code class="docutils literal notranslate"><span class="pre">project_ptms_onto_splice_events()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.flanking_sequences">Flanking Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_rMATS"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_rMATS()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_splice_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph"><code class="docutils literal notranslate"><span class="pre">get_flanking_changes_from_splicegraph()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.annotate">Annotating PTMs</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_ELM_interactions"><code class="docutils literal notranslate"><span class="pre">add_ELM_interactions()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_ELM_matched_motifs"><code class="docutils literal notranslate"><span class="pre">add_ELM_matched_motifs()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_custom_annotation"><code class="docutils literal notranslate"><span class="pre">add_custom_annotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.add_omnipath_data"><code class="docutils literal notranslate"><span class="pre">add_omnipath_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.annotate_ptms"><code class="docutils literal notranslate"><span class="pre">annotate_ptms()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.append_from_gmt"><code class="docutils literal notranslate"><span class="pre">append_from_gmt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.check_file"><code class="docutils literal notranslate"><span class="pre">check_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.check_gmt_file"><code class="docutils literal notranslate"><span class="pre">check_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.combine_enzyme_data"><code class="docutils literal notranslate"><span class="pre">combine_enzyme_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.combine_interaction_data"><code class="docutils literal notranslate"><span class="pre">combine_interaction_data()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_DEPOD_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_DEPOD_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMInt_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_PTMInt_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMcode_interprotein_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_PTMcode_interprotein_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PTMsigDB_gmt_files"><code class="docutils literal notranslate"><span class="pre">construct_PTMsigDB_gmt_files()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_PhosphoSitePlus_gmt_files"><code class="docutils literal notranslate"><span class="pre">construct_PhosphoSitePlus_gmt_files()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_RegPhos_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_RegPhos_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_annotation_dict_from_gmt"><code class="docutils literal notranslate"><span class="pre">construct_annotation_dict_from_gmt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_custom_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_custom_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_gmt_df"><code class="docutils literal notranslate"><span class="pre">construct_gmt_df()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.construct_omnipath_gmt_file"><code class="docutils literal notranslate"><span class="pre">construct_omnipath_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.convert_PSP_label_to_UniProt"><code class="docutils literal notranslate"><span class="pre">convert_PSP_label_to_UniProt()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_ids_PTMcode"><code class="docutils literal notranslate"><span class="pre">extract_ids_PTMcode()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_interaction_details"><code class="docutils literal notranslate"><span class="pre">extract_interaction_details()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.extract_positions_from_DEPOD"><code class="docutils literal notranslate"><span class="pre">extract_positions_from_DEPOD()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.get_available_gmt_annotations"><code class="docutils literal notranslate"><span class="pre">get_available_gmt_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.load_gmt_file"><code class="docutils literal notranslate"><span class="pre">load_gmt_file()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.process_database_annotations"><code class="docutils literal notranslate"><span class="pre">process_database_annotations()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.simplify_annotation"><code class="docutils literal notranslate"><span class="pre">simplify_annotation()</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.annotate.unify_interaction_data"><code class="docutils literal notranslate"><span class="pre">unify_interaction_data()</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#analyze-modules">Analyze Modules</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.summarize">Summaries</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.combine_outputs"><code class="docutils literal notranslate"><span class="pre">combine_outputs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.get_modification_class_data"><code class="docutils literal notranslate"><span class="pre">get_modification_class_data()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.get_modification_counts"><code class="docutils literal notranslate"><span class="pre">get_modification_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.summarize.plot_modification_breakdown"><code class="docutils literal notranslate"><span class="pre">plot_modification_breakdown()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.filter">Filtering PTMs and Events</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.filter.plot_filter_impact"><code class="docutils literal notranslate"><span class="pre">plot_filter_impact()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.annotations">Annotations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.annotation_enrichment"><code class="docutils literal notranslate"><span class="pre">annotation_enrichment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.draw_pie"><code class="docutils literal notranslate"><span class="pre">draw_pie()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.gene_set_enrichment"><code class="docutils literal notranslate"><span class="pre">gene_set_enrichment()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_available_annotations"><code class="docutils literal notranslate"><span class="pre">get_available_annotations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_background_annotation_counts"><code class="docutils literal notranslate"><span class="pre">get_background_annotation_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_enrichment_inputs"><code class="docutils literal notranslate"><span class="pre">get_enrichment_inputs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.get_ptm_annotations"><code class="docutils literal notranslate"><span class="pre">get_ptm_annotations()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_EnrichR_pies"><code class="docutils literal notranslate"><span class="pre">plot_EnrichR_pies()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_annotation_counts"><code class="docutils literal notranslate"><span class="pre">plot_annotation_counts()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.annotations.plot_available_annotations"><code class="docutils literal notranslate"><span class="pre">plot_available_annotations()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.interactions">Protein Interactions</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.get_edge_colors"><code class="docutils literal notranslate"><span class="pre">get_edge_colors()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.get_interaction_stats"><code class="docutils literal notranslate"><span class="pre">get_interaction_stats()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.plot_interaction_network"><code class="docutils literal notranslate"><span class="pre">plot_interaction_network()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.plot_network_centrality"><code class="docutils literal notranslate"><span class="pre">plot_network_centrality()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions"><code class="docutils literal notranslate"><span class="pre">protein_interactions</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.compare_to_nease"><code class="docutils literal notranslate"><span class="pre">protein_interactions.compare_to_nease()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_interaction_stats"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_interaction_stats()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.get_protein_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.get_protein_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_interaction_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_interaction_network()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_nease_comparison"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_nease_comparison()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.plot_network_centrality"><code class="docutils literal notranslate"><span class="pre">protein_interactions.plot_network_centrality()</span></code></a></li>
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.interactions.protein_interactions.summarize_protein_network"><code class="docutils literal notranslate"><span class="pre">protein_interactions.summarize_protein_network()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze.enzyme">Enzyme Regulation</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.enzyme.compare_KL_for_sequence"><code class="docutils literal notranslate"><span class="pre">compare_KL_for_sequence()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.enzyme.get_all_KL_scores"><code class="docutils literal notranslate"><span class="pre">get_all_KL_scores()</span></code></a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Flanking Sequences</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.compare_inclusion_motifs"><code class="docutils literal notranslate"><span class="pre">compare_inclusion_motifs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.findAlteredPositions"><code class="docutils literal notranslate"><span class="pre">findAlteredPositions()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.find_motifs"><code class="docutils literal notranslate"><span class="pre">find_motifs()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.getSequenceIdentity"><code class="docutils literal notranslate"><span class="pre">getSequenceIdentity()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_alterations_matrix"><code class="docutils literal notranslate"><span class="pre">plot_alterations_matrix()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_location_of_altered_flanking_residues"><code class="docutils literal notranslate"><span class="pre">plot_location_of_altered_flanking_residues()</span></code></a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#ptm_pose.analyze.flanking_sequences.plot_sequence_differences"><code class="docutils literal notranslate"><span class="pre">plot_sequence_differences()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>