

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>PTM-POSE Reference &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-dataframe.css" />
    <link rel="stylesheet" type="text/css" href="_static/sg_gallery-rendered-html.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'PTM_POSE';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Running PTM-POSE" href="General_Instructions.html" />
    <link rel="prev" title="Installation" href="quickstart.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="Dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Installation</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">PTM-POSE Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="General_Instructions.html">Running PTM-POSE</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="gallery_output/index.html">Types of Analysis Performed with PTM-POSE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="gallery_output/plot_protein_interactions.html">Identify protein interactions that may be impacted by splicing of PTMs</a></li>
<li class="toctree-l2"><a class="reference internal" href="gallery_output/plot_num_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l2"><a class="reference internal" href="gallery_output/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="gallery_output/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="Examples/Examples.html">Full Analysis Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="Examples/ESRP1_knockdown.html">Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/PTM_POSE.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>PTM-POSE Reference</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.pose_config">Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.project">PTM Projection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.flanking_sequences">Flanking Sequences</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.annotate">Annotating PTMs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze">Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-0">Plotting</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="ptm-pose-reference">
<h1>PTM-POSE Reference<a class="headerlink" href="#ptm-pose-reference" title="Permalink to this heading">#</a></h1>
<section id="module-ptm_pose.pose_config">
<span id="configuration"></span><h2>Configuration<a class="headerlink" href="#module-ptm_pose.pose_config" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.pose_config.download_ptm_coordinates">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.pose_config.</span></span><span class="sig-name descname"><span class="pre">download_ptm_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">save</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/pose_config.html#download_ptm_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.pose_config.download_ptm_coordinates" title="Permalink to this definition">#</a></dt>
<dd><p>Download ptm_coordinates dataframe from GitHub Large File Storage (LFS). By default, this will not save the file locally due the larger size (do not want to force users to download but highly encourage), but an option to save the file is provided if desired</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>save</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to save the file locally into Resource Files directory. The default is False.</p>
</dd>
<dt><strong>max_retries</strong><span class="classifier">int, optional</span></dt><dd><p>Number of times to attempt to download the file. The default is 5.</p>
</dd>
<dt><strong>delay</strong><span class="classifier">int, optional</span></dt><dd><p>Time to wait between download attempts. The default is 10.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.project">
<span id="ptm-projection"></span><h2>PTM Projection<a class="headerlink" href="#module-ptm_pose.project" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.project_ptms_onto_MATS">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">project_ptms_onto_MATS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">SE_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fiveASS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threeASS_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RI_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">MXE_events</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identify_flanking_sequences</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'meanDeltaPSI'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'FDR'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_modification_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PROCESSES</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#project_ptms_onto_MATS"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.project_ptms_onto_MATS" title="Permalink to this definition">#</a></dt>
<dd><p>Given splice quantification from the MATS algorithm, annotate with PTMs that are found in the differentially included regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_coordinates: pandas.DataFrame</strong></dt><dd><p>dataframe containing PTM information, including chromosome, strand, and genomic location of PTMs</p>
</dd>
<dt><strong>SE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing skipped exon event information from MATS</p>
</dd>
<dt><strong>fiveASS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 5’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>threeASS_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing 3’ alternative splice site event information from MATS</p>
</dd>
<dt><strong>RI_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing retained intron event information from MATS</p>
</dd>
<dt><strong>MXE_events: pandas.DataFrame</strong></dt><dd><p>dataframe containing mutually exclusive exon event information from MATS</p>
</dd>
<dt><strong>coordinate_type: str</strong></dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
<dt><strong>identify_flanking_sequences: bool</strong></dt><dd><p>Indicate whether to look for altered flanking sequences from spliced events, in addition to those directly in the spliced region. Default is False. (not yet active)</p>
</dd>
<dt><strong>PROCESSES: int</strong></dt><dd><p>Number of processes to use for multiprocessing. Default is 1.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.project.project_ptms_onto_splice_events">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.project.</span></span><span class="sig-name descname"><span class="pre">project_ptms_onto_splice_events</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">splice_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate_original_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'chr'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strand'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exonStart_0base'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'exonEnd'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">separate_modification_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taskbar_label</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PROCESSES</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/project.html#project_ptms_onto_splice_events"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.project.project_ptms_onto_splice_events" title="Permalink to this definition">#</a></dt>
<dd><p>Given splice event quantification data, project PTMs onto the regions impacted by the splice events. Assumes that the splice event data will have chromosome, strand, and genomic start/end positions for the regions of interest, and each row of the splice_event_data corresponds to a unique region.</p>
<p>Parameters</p>
<dl class="simple">
<dt>splice_data: pandas.DataFrame</dt><dd><p>dataframe containing splice event information, including chromosome, strand, and genomic location of regions of interest</p>
</dd>
<dt>ptm_coordinates: pandas.DataFrame</dt><dd><p>dataframe containing PTM information, including chromosome, strand, and genomic location of PTMs. If none, it will pull from the config file.</p>
</dd>
<dt>chromosome_col: str</dt><dd><p>column name in splice_data that contains chromosome information. Default is ‘chr’. Expects it to be a str with only the chromosome number: ‘Y’, ‘1’, ‘2’, etc.</p>
</dd>
<dt>strand_col: str</dt><dd><p>column name in splice_data that contains strand information. Default is ‘strand’. Expects it to be a str with ‘+’ or ‘-’, or integers as 1 or -1. Will convert to integers automatically if string format is provided.</p>
</dd>
<dt>region_start_col: str</dt><dd><p>column name in splice_data that contains the start position of the region of interest. Default is ‘exonStart_0base’.</p>
</dd>
<dt>region_end_col: str</dt><dd><p>column name in splice_data that contains the end position of the region of interest. Default is ‘exonEnd’.</p>
</dd>
<dt>event_id_col: str</dt><dd><p>column name in splice_data that contains the unique identifier for the splice event. If provided, will be used to annotate the ptm information with the specific splice event ID. Default is None.</p>
</dd>
<dt>gene_col: str</dt><dd><p>column name in splice_data that contains the gene name. If provided, will be used to make sure the projected PTMs stem from the same gene (some cases where genomic coordiantes overlap between distinct genes). Default is None.</p>
</dd>
<dt>dPSI_col: str</dt><dd><p>column name in splice_data that contains the delta PSI value for the splice event. Default is None, which will not include this information in the output</p>
</dd>
<dt>sig_col: str</dt><dd><p>column name in splice_data that contains the significance value for the splice event. Default is None, which will not include this information in the output.</p>
</dd>
<dt>extra_cols: list</dt><dd><p>list of additional columns to include in the output dataframe. Default is None, which will not include any additional columns.</p>
</dd>
<dt>coordinate_type: str</dt><dd><p>indicates the coordinate system used for the start and end positions. Either hg38 or hg19. Default is ‘hg38’.</p>
</dd>
<dt>separate_modification_types: bool</dt><dd><p>Indicate whether to store PTM sites with  multiple modification types as multiple rows. For example, if a site at K100 was both an acetylation and methylation site, these will be separated into unique rows with the same site number but different modification types. Default is True.</p>
</dd>
<dt>taskbar_label: str</dt><dd><p>Label to display in the tqdm progress bar. Default is None, which will automatically state “Projecting PTMs onto regions using —– coordinates”.</p>
</dd>
<dt>PROCESSES: int</dt><dd><p>Number of processes to use for multiprocessing. Default is 1 (single processing)</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>spliced_ptm_info: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events</p>
</dd>
<dt>splice_data: pandas.DataFrame</dt><dd><p>dataframe containing the original splice data with an additional column ‘PTMs’ that contains the PTMs found in the region of interest, in the format of ‘SiteNumber(ModificationType)’. If no PTMs are found, the value will be np.nan.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.flanking_sequences">
<span id="flanking-sequences"></span><h2>Flanking Sequences<a class="headerlink" href="#module-ptm_pose.flanking_sequences" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.extract_region_from_splicegraph">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">extract_region_from_splicegraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">splicegraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#extract_region_from_splicegraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.extract_region_from_splicegraph" title="Permalink to this definition">#</a></dt>
<dd><p>Given a region id and the splicegraph from SpliceSeq, extract the chromosome, strand, and start and stop locations of that exon. Start and stop are forced to be in ascending order, which is not necessarily true from the splice graph (i.e. start &gt; stop for negative strand exons). This is done to make the region extraction consistent with the rest of the codebase.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliceseq</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>SpliceSeq splicegraph dataframe, with region_id as index</p>
</dd>
<dt><strong>region_id</strong><span class="classifier">str</span></dt><dd><p>Region ID to extract information from, in the format of ‘GeneName_ExonNumber’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>list</dt><dd><p>List containing the chromosome, strand (1 for forward, -1 for negative), start, and stop locations of the region</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Coordinates'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes" title="Permalink to this definition">#</a></dt>
<dd><p>Currently has been tested with MATS splicing events.</p>
<p>Given flanking and spliced regions associated with a splice event, identify PTMs that have potential to have an altered flanking sequence depending on whether spliced region is included or excluded (if PTM is close to splice boundary). For these PTMs, extract the flanking sequences associated with the inclusion and exclusion cases and translate into amino acid sequences. If the PTM is not associated with a codon that codes for the expected amino acid, the PTM will be excluded from the results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>chromosome</strong><span class="classifier">str</span></dt><dd><p>Chromosome associated with the splice event</p>
</dd>
<dt><strong>strand</strong><span class="classifier">int</span></dt><dd><p>Strand associated with the splice event (1 for forward, -1 for negative)</p>
</dd>
<dt><strong>first_flank_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the first flanking region (first is currently defined based on location the genome not coding sequence)</p>
</dd>
<dt><strong>spliced_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the spliced region</p>
</dd>
<dt><strong>second_flank_region</strong><span class="classifier">list</span></dt><dd><p>List containing the start and stop locations of the second flanking region (second is currently defined based on location the genome not coding sequence)</p>
</dd>
<dt><strong>event_id</strong><span class="classifier">str, optional</span></dt><dd><p>Event ID associated with the splice event, by default None</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 7</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg38’. Other options is hg19.</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM in returned flanking sequences, by default True</p>
</dd>
<dt><strong>order_by</strong><span class="classifier">str, optional</span></dt><dd><p>Whether the first, spliced and second regions are defined by their genomic coordinates (first has smallest coordinate, spliced next, then second), or if they are defined by their translation (first the first when translated, etc.)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>pandas.DataFrame</dt><dd><p>DataFrame containing the PTMs associated with the flanking regions and the amino acid sequences of the flanking regions in the inclusion and exclusion cases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes_from_splice_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">splice_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_region_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_start_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_end_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg38'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes_from_splice_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splice_data" title="Permalink to this definition">#</a></dt>
<dd><p>Given a DataFrame containing information about splice events, extract the flanking sequences associated with the PTMs in the flanking regions if there is potential for this to be altered. The DataFrame should contain columns for the chromosome, strand, start and stop locations of the first flanking region, spliced region, and second flanking region. The DataFrame should also contain a column for the event ID associated with the splice event. If the DataFrame does not contain the necessary columns, the function will raise an error.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>splice_data</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about splice events</p>
</dd>
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>chromosome_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating chromosome, by default None</p>
</dd>
<dt><strong>strand_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating strand, by default None</p>
</dd>
<dt><strong>first_flank_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the first flanking region, by default None</p>
</dd>
<dt><strong>first_flank_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the first flanking region, by default None</p>
</dd>
<dt><strong>spliced_region_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the spliced region, by default None</p>
</dd>
<dt><strong>spliced_region_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the spliced region, by default None</p>
</dd>
<dt><strong>second_flank_start_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating start location of the second flanking region, by default None</p>
</dd>
<dt><strong>second_flank_end_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating end location of the second flanking region, by default None</p>
</dd>
<dt><strong>event_id_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating event ID, by default None</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 7</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg38’. Other options is hg19.</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM in returned flanking sequences, by default True</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>list</dt><dd><p>List containing DataFrames with the PTMs associated with the flanking regions and the amino acid sequences of the flanking regions in the inclusion and exclusion cases</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_changes_from_splicegraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">psi_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicegraph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">event_id_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extra_cols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'symbol'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_changes_from_splicegraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_changes_from_splicegraph" title="Permalink to this definition">#</a></dt>
<dd><p>Given a DataFrame containing information about splice events  obtained from SpliceSeq and the corresponding splicegraph, extract the flanking sequences of PTMs that are nearby the splice boundary (potential for flanking sequence to be altered). Coordinate information of individual exons should be found in splicegraph. You can also provide columns with specific psi or significance information. Extra cols not in these categories can be provided with extra_cols parameter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>psi_data</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about splice events obtained from SpliceSeq</p>
</dd>
<dt><strong>splicegraph</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about individual exons and their coordinates</p>
</dd>
<dt><strong>ptm_coordinates</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing PTM coordinate information for identify PTMs in the flanking regions</p>
</dd>
<dt><strong>dPSI_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating delta PSI value, by default None</p>
</dd>
<dt><strong>sig_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating significance of the event, by default None</p>
</dd>
<dt><strong>event_id_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating event ID, by default None</p>
</dd>
<dt><strong>extra_cols</strong><span class="classifier">list, optional</span></dt><dd><p>List of column names for additional information to add to the results, by default None</p>
</dd>
<dt><strong>gene_col</strong><span class="classifier">str, optional</span></dt><dd><p>Column name indicating gene symbol of spliced gene, by default ‘symbol’</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 5</p>
</dd>
<dt><strong>coordinate_type</strong><span class="classifier">str, optional</span></dt><dd><p>Coordinate system used for the regions, by default ‘hg19’. Other options is hg38.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>altered_flanks</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing the PTMs associated with the flanking regions that are altered, and the flanking sequences that arise depending on whether the flanking sequence is included or not</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_flanking_sequence">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_flanking_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_loc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptm_residue</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_flanking_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_flanking_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_flanking_sequence" title="Permalink to this definition">#</a></dt>
<dd><p>Given a PTM location in a sequence of DNA, extract the flanking sequence around the PTM location and translate into the amino acid sequence. If the sequence is not the correct length, the function will attempt to extract the flanking sequence with spaces to account for missing parts if full_flanking_seq is not True. If the sequence is still not the correct length, the function will raise an error. Any unrecognized codons that are found in the sequence and are not in the standard codon table, including stop codons, will be translated as ‘X’ (unknown) or ‘*’ (stop codon).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_loc</strong><span class="classifier">int</span></dt><dd><p>Location of the first base pair associated with PTM in the DNA sequence</p>
</dd>
<dt><strong>seq</strong><span class="classifier">str</span></dt><dd><p>DNA sequence containing the PTM</p>
</dd>
<dt><strong>ptm_residue</strong><span class="classifier">str</span></dt><dd><p>Amino acid residue associated with the PTM</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 5</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM, by default True</p>
</dd>
<dt><strong>full_flanking_seq</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to require the flanking sequence to be the correct length, by default False</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>str</dt><dd><p>Amino acid sequence of the flanking sequence around the PTM if translation was successful, otherwise np.nan</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_ptm_locs_in_spliced_sequences">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_ptm_locs_in_spliced_sequences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm_loc_in_flank</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_flank_seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_flank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'First'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">order_by</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Coordinates'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_ptm_locs_in_spliced_sequences"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_ptm_locs_in_spliced_sequences" title="Permalink to this definition">#</a></dt>
<dd><p>Given the location of a PTM in a flanking sequence, extract the location of the PTM in the Inclusion Flanking Sequence and the Exclusion Flanking Sequence associated with a given splice event. Inclusion Flanking Sequence will include the skipped exon region, retained intron, or longer alternative splice site depending on event type. The PTM location should be associated with where the PTM is located relative to spliced region (before = ‘First’, after = ‘Second’).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm_loc_in_flank</strong><span class="classifier">int</span></dt><dd><p>Location of the PTM in the flanking sequence it is found (either first or second)</p>
</dd>
<dt><strong>first_flank_seq</strong><span class="classifier">str</span></dt><dd><p>Flanking exon sequence before the spliced region</p>
</dd>
<dt><strong>spliced_seq</strong><span class="classifier">str</span></dt><dd><p>Spliced region sequence</p>
</dd>
<dt><strong>second_flank_seq</strong><span class="classifier">str</span></dt><dd><p>Flanking exon sequence after the spliced region</p>
</dd>
<dt><strong>which_flank</strong><span class="classifier">str, optional</span></dt><dd><p>Which flank the PTM is associated with, by default ‘First’</p>
</dd>
<dt><strong>order_by</strong><span class="classifier">str, optional</span></dt><dd><p>Whether the first, spliced and second regions are defined by their genomic coordinates (first has smallest coordinate, spliced next, then second), or if they are defined by their translation (first the first when translated, etc.)</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>tuple</dt><dd><p>Tuple containing the PTM location in the Inclusion Flanking Sequence and the Exclusion Flanking Sequence</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_ptms_in_splicegraph_flank">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_ptms_in_splicegraph_flank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chromosome</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_region_start</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_region_end</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_flank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'First'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_ptms_in_splicegraph_flank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_ptms_in_splicegraph_flank" title="Permalink to this definition">#</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_spliceseq_event_regions">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_spliceseq_event_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_exon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spliced_exons</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_exon</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">splicegraph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_spliceseq_event_regions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_spliceseq_event_regions" title="Permalink to this definition">#</a></dt>
<dd><p>Given all exons associated with a splicegraph event, obtain the coordinates associated with the flanking exons and the spliced region. The spliced region is defined as the exons that are associated with psi values, while flanking regions include the “from” and “to” exons that indicate the adjacent, unspliced exons.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>gene_name</strong><span class="classifier">str</span></dt><dd><p>Gene name associated with the splice event</p>
</dd>
<dt><strong>from_exon</strong><span class="classifier">int</span></dt><dd><p>Exon number associated with the first flanking exon</p>
</dd>
<dt><strong>spliced_exons</strong><span class="classifier">str</span></dt><dd><p>Exon numbers associated with the spliced region, separated by colons for each unique exon</p>
</dd>
<dt><strong>to_exon</strong><span class="classifier">int</span></dt><dd><p>Exon number associated with the second flanking exon</p>
</dd>
<dt><strong>splicegraph</strong><span class="classifier">pandas.DataFrame</span></dt><dd><p>DataFrame containing information about individual exons and their coordinates</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>tuple</dt><dd><p>Tuple containing the genomic coordinates of the first flanking region, spliced regions, and second flanking region</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.get_spliceseq_flank_loc">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">get_spliceseq_flank_loc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptm</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_region_coords</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_region_coords</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coordinate_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hg19'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#get_spliceseq_flank_loc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.get_spliceseq_flank_loc" title="Permalink to this definition">#</a></dt>
<dd><p>Given ptm information for identifying flanking sequences from splicegraph information, extract the relative location of the ptm in the flanking region (where it is located in translation of the flanking region).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptm</strong><span class="classifier">pandas.Series</span></dt><dd><p>Series containing PTM information</p>
</dd>
<dt><strong>strand</strong><span class="classifier">int</span></dt><dd><p>Strand associated with the splice event (1 for forward, -1 for negative)</p>
</dd>
<dt><strong>from_region_coords</strong><span class="classifier">list</span></dt><dd><p>List containing the chromosome, strand, start, and stop locations of the first flanking region</p>
</dd>
<dt><strong>to_region_coords</strong><span class="classifier">list</span></dt><dd><p>List containing the chromosome, strand, start, and stop locations of the second flanking region</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>int</dt><dd><p>Relative location of the PTM in the flanking region</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.flanking_sequences.translate_flanking_sequence">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.flanking_sequences.</span></span><span class="sig-name descname"><span class="pre">translate_flanking_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_flanking_seq</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lowercase_mod</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_flank_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stop_codon_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'*'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unknown_codon_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'X'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/flanking_sequences.html#translate_flanking_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.flanking_sequences.translate_flanking_sequence" title="Permalink to this definition">#</a></dt>
<dd><p>Given a DNA sequence, translate the sequence into an amino acid sequence. If the sequence is not the correct length, the function will attempt to extract the flanking sequence with spaces to account for missing parts if full_flanking_seq is not True. If the sequence is still not the correct length, the function will raise an error. Any unrecognized codons that are found in the sequence and are not in the standard codon table, including stop codons, will be translated as ‘X’ (unknown) or ‘*’ (stop codon).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq</strong><span class="classifier">str</span></dt><dd><p>DNA sequence to translate</p>
</dd>
<dt><strong>flank_size</strong><span class="classifier">int, optional</span></dt><dd><p>Number of amino acids to include flanking the PTM, by default 7</p>
</dd>
<dt><strong>full_flanking_seq</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to require the flanking sequence to be the correct length, by default True</p>
</dd>
<dt><strong>lowercase_mod</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to lowercase the amino acid associated with the PTM, by default True</p>
</dd>
<dt><strong>first_flank_length</strong><span class="classifier">int, optional</span></dt><dd><p>Length of the flanking sequence in front of the PTM, by default None. If full_flanking_seq is False and sequence is not the correct length, this is required.</p>
</dd>
<dt><strong>stop_codon_symbol</strong><span class="classifier">str, optional</span></dt><dd><p>Symbol to use for stop codons, by default ‘*’</p>
</dd>
<dt><strong>unknown_codon_symbol</strong><span class="classifier">str, optional</span></dt><dd><p>Symbol to use for unknown codons, by default ‘X’</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>str</dt><dd><p>Amino acid sequence of the flanking sequence if translation was successful, otherwise np.nan</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.annotate">
<span id="annotating-ptms"></span><h2>Annotating PTMs<a class="headerlink" href="#module-ptm_pose.annotate" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_ELM_interactions">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_ELM_interactions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_ELM_interactions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_ELM_interactions" title="Permalink to this definition">#</a></dt>
<dd><p>Given a spliced ptms dataframe from the project module, add ELM interaction data to the dataframe</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_PSP_disease_association">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_PSP_disease_association</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Disease-associated_sites.gz'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_PSP_disease_association"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_PSP_disease_association" title="Permalink to this definition">#</a></dt>
<dd><p>Process disease asociation data from PhosphoSitePlus (Disease-associated_sites.gz), and add to spliced_ptms dataframe from project_ptms_onto_splice_events() function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path to the PhosphoSitePlus Kinase_Substrate_Dataset.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>spliced_ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with an additional column indicating the kinases known to phosphorylate that site (not relevant to non-phosphorylation PTMs)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_PSP_kinase_substrate_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_PSP_kinase_substrate_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Kinase_Substrate_Dataset.gz'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_PSP_kinase_substrate_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_PSP_kinase_substrate_data" title="Permalink to this definition">#</a></dt>
<dd><p>Add kinase substrate data from PhosphoSitePlus (Kinase_Substrate_Dataset.gz) to spliced_ptms dataframe from project_ptms_onto_splice_events() function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path to the PhosphoSitePlus Kinase_Substrate_Dataset.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>spliced_ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with an additional column indicating the kinases known to phosphorylate that site (not relevant to non-phosphorylation PTMs)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_PSP_regulatory_site_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_PSP_regulatory_site_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Regulatory_sites.gz'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_PSP_regulatory_site_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_PSP_regulatory_site_data" title="Permalink to this definition">#</a></dt>
<dd><p>Add functional information from PhosphoSitePlus (Regulatory_sites.gz) to spliced_ptms dataframe from project_ptms_onto_splice_events() function</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>file: str</strong></dt><dd><p>Path to the PhosphoSitePlus Regulatory_sites.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>spliced_ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with additional columns for regulatory site information, including domains, biological process, functions, and protein interactions associated with the PTMs</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_PTMInt_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_PTMInt_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_success</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_PTMInt_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_PTMInt_data" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced_ptms data from project module, add PTMInt interaction data, which will include the protein that is being interacted with, whether it enchances or inhibits binding, and the localization of the interaction. This will be added as a new column labeled PTMInt:Interactions and each entry will be formatted like ‘Protein-&gt;Effect|Localization’. If multiple interactions, they will be separated by a semicolon</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_annotation">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_existing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_annotation" title="Permalink to this definition">#</a></dt>
<dd><p>Given a desired database and annotation type, add the corresponding annotation data to the spliced ptm dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe containing PTM data</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>Database to extract annotation data from. Options include ‘PhosphoSitePlus’, ‘PTMcode’, ‘PTMInt’, ‘RegPhos’, ‘DEPOD’</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to extract. Options include ‘Function’, ‘Process’, ‘Interactions’, ‘Disease’, ‘Kinase’, ‘Phosphatase’, but depend on the specific database (run analyze.get_annotation_categories())</p>
</dd>
<dt><strong>file: str</strong></dt><dd><p>File path to annotation data. If None, will download from online source, except for PhosphoSitePlus (due to licensing restrictions)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.add_custom_annotation">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">add_custom_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accession_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UniProtKB</span> <span class="pre">Accession'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">residue_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Residue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PTM</span> <span class="pre">Position</span> <span class="pre">in</span> <span class="pre">Canonical</span> <span class="pre">Isoform'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#add_custom_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.add_custom_annotation" title="Permalink to this definition">#</a></dt>
<dd><p>Add custom annotation data to spliced_ptms or altered flanking sequence dataframes</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation_data: pandas.DataFrame</strong></dt><dd><p>Dataframe containing the annotation data to be added to the spliced_ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Canonical Isoform, and the annotation data to be added</p>
</dd>
<dt><strong>source_name: str</strong></dt><dd><p>Name of the source of the annotation data, will be used to label the columns in the spliced_ptms dataframe</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation data being added, will be used to label the columns in the spliced_ptms dataframe</p>
</dd>
<dt><strong>annotation_col: str</strong></dt><dd><p>Column name in the annotation data that contains the annotation data to be added to the spliced_ptms dataframe</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>spliced_ptms: pandas.DataFrame</dt><dd><p>Contains the PTMs identified across the different splice events with an additional column for the custom annotation data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.annotate_ptms">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">annotate_ptms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psp_regulatory_site_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psp_ks_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">psp_disease_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_interactions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_motifs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PTMint</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PTMcode_interprotein</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DEPOD</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">RegPhos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ptmsigdb_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interactions_to_combine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PTMcode',</span> <span class="pre">'PhosphoSitePlus',</span> <span class="pre">'RegPhos',</span> <span class="pre">'PTMInt']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kinases_to_combine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'RegPhos']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#annotate_ptms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.annotate_ptms" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm data, add annotations from various databases. The annotations that can be added are the following:
- PhosphoSitePlus</p>
<blockquote>
<div><ul class="simple">
<li><p>regulatory site data (file must be provided)</p></li>
<li><p>kinase-substrate data (file must be provided)</p></li>
<li><p>disease association data (file must be provided)</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>ELM </dt><dd><ul>
<li><p>interaction data (can be downloaded automatically or provided as a file)</p></li>
<li><p>motif matches (elm class data can be downloaded automatically or provided as a file)</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>PTMInt</dt><dd><ul>
<li><p>interaction data (will be downloaded automatically)</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>PTMcode</dt><dd><ul>
<li><p>intraprotein interactions (can be downloaded automatically or provided as a file)</p></li>
<li><p>interprotein interactions (can be downloaded automatically or provided as a file)</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>DEPOD</dt><dd><ul>
<li><p>phosphatase-substrate data (will be downloaded automatically)</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>RegPhos</dt><dd><ul>
<li><p>kinase-substrate data (will be downloaded automatically)</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Spliced PTM data from project module</p>
</dd>
<dt><strong>psp_regulatory_site_file: str</strong></dt><dd><p>File path to PhosphoSitePlus regulatory site data</p>
</dd>
<dt><strong>psp_ks_file: str</strong></dt><dd><p>File path to PhosphoSitePlus kinase-substrate data</p>
</dd>
<dt><strong>psp_disease_file: str</strong></dt><dd><p>File path to PhosphoSitePlus disease association data</p>
</dd>
<dt><strong>elm_interactions: bool or str</strong></dt><dd><p>If True, download ELM interaction data automatically. If str, provide file path to ELM interaction data</p>
</dd>
<dt><strong>elm_motifs: bool or str</strong></dt><dd><p>If True, download ELM motif data automatically. If str, provide file path to ELM motif data</p>
</dd>
<dt><strong>PTMint: bool</strong></dt><dd><p>If True, download PTMInt data automatically</p>
</dd>
<dt><strong>PTMcode_intraprotein: bool or str</strong></dt><dd><p>If True, download PTMcode intraprotein data automatically. If str, provide file path to PTMcode intraprotein data</p>
</dd>
<dt><strong>PTMcode_interprotein: bool or str</strong></dt><dd><p>If True, download PTMcode interprotein data automatically. If str, provide file path to PTMcode interprotein data</p>
</dd>
<dt><strong>DEPOD: bool</strong></dt><dd><p>If True, download DEPOD data automatically</p>
</dd>
<dt><strong>RegPhos: bool</strong></dt><dd><p>If True, download RegPhos data automatically</p>
</dd>
<dt><strong>ptmsigdb_file: str</strong></dt><dd><p>File path to PTMsigDB data</p>
</dd>
<dt><strong>interactions_to_combine: list</strong></dt><dd><p>List of databases to combine interaction data from. Default is [‘PTMcode’, ‘PhosphoSitePlus’, ‘RegPhos’, ‘PTMInt’]</p>
</dd>
<dt><strong>kinases_to_combine: list</strong></dt><dd><p>List of databases to combine kinase-substrate data from. Default is [‘PhosphoSitePlus’, ‘RegPhos’]</p>
</dd>
<dt><strong>combine_similar: bool</strong></dt><dd><p>Whether to combine annotations of similar information (kinase, interactions, etc) from multiple databases into another column labeled as ‘Combined’. Default is True</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.check_file">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">check_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expected_extension</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'.tsv'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#check_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.check_file" title="Permalink to this definition">#</a></dt>
<dd><p>Given a file name, check if the file exists and has the expected extension. If the file does not exist or has the wrong extension, raise an error.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>fname: str</strong></dt><dd><p>File name to check</p>
</dd>
<dt><strong>expected_extension: str</strong></dt><dd><p>Expected file extension. Default is ‘.tsv’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.combine_KS_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">combine_KS_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ks_databases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'RegPhos']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regphos_conversion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'ABL1(ABL)':</span> <span class="pre">'ABL1',</span> <span class="pre">'CDC2':</span> <span class="pre">'CDK1',</span> <span class="pre">'CK2A1':</span> <span class="pre">'CSNK2A1',</span> <span class="pre">'ERK1(MAPK3)':</span> <span class="pre">'MAPK3',</span> <span class="pre">'ERK2(MAPK1)':</span> <span class="pre">'MAPK1',</span> <span class="pre">'JNK2(MAPK9)':</span> <span class="pre">'MAPK9',</span> <span class="pre">'PKACA':</span> <span class="pre">'PRKACA'}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#combine_KS_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.combine_KS_data" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm information, combine kinase-substrate data from multiple databases (currently support PhosphoSitePlus and RegPhos), assuming that the kinase data from these resources has already been added to the spliced ptm data. The combined kinase data will be added as a new column labeled ‘Combined:Kinase’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Spliced PTM data from project module</p>
</dd>
<dt><strong>ks_databases: list</strong></dt><dd><p>List of databases to combine kinase data from. Currently support PhosphoSitePlus and RegPhos</p>
</dd>
<dt><strong>regphos_conversion: dict</strong></dt><dd><p>Allows conversion of RegPhos names to matching names in PhosphoSitePlus.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>splicde_ptms: pd.DataFrame</dt><dd><p>Spliced PTM data with combined kinase data added</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.combine_interaction_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">combine_interaction_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_databases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['PhosphoSitePlus',</span> <span class="pre">'PTMcode',</span> <span class="pre">'PTMInt',</span> <span class="pre">'RegPhos',</span> <span class="pre">'DEPOD',</span> <span class="pre">'ELM']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_enzyme_interactions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#combine_interaction_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.combine_interaction_data" title="Permalink to this definition">#</a></dt>
<dd><p>Given annotated spliced ptm data, extract interaction data from various databases and combine into a single dataframe. This will include the interacting protein, the type of interaction, and the source of the interaction data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe containing PTM data and associated interaction annotations from various databases</p>
</dd>
<dt><strong>interaction_databases: list</strong></dt><dd><p>List of databases to extract interaction data from. Options include ‘PhosphoSitePlus’, ‘PTMcode’, ‘PTMInt’, ‘RegPhos’, ‘DEPOD’. These should already have annotation columns in the spliced_ptms dataframe, otherwise they will be ignored. For kinase-substrate interactions, if combined column is present, will use that instead of individual databases</p>
</dd>
<dt><strong>include_enzyme_interactions: bool</strong></dt><dd><p>If True, will include kinase-substrate and phosphatase interactions in the output dataframe</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>interact_data: list</dt><dd><p>List of dataframes containing PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES), and the source of the interaction data</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.convert_PSP_label_to_UniProt">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">convert_PSP_label_to_UniProt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#convert_PSP_label_to_UniProt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.convert_PSP_label_to_UniProt" title="Permalink to this definition">#</a></dt>
<dd><p>Given a label for an interacting protein from PhosphoSitePlus, convert to UniProtKB accession. Required as PhosphoSitePlus interactions are recorded in various ways that aren’t necessarily consistent with other databases (i.e. not always gene name)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>label: str</strong></dt><dd><p>Label for interacting protein from PhosphoSitePlus</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.extract_positions_from_DEPOD">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">extract_positions_from_DEPOD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#extract_positions_from_DEPOD"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.extract_positions_from_DEPOD" title="Permalink to this definition">#</a></dt>
<dd><p>Given string object consisting of multiple modifications in the form of ‘Residue-Position’ separated by ‘, ‘, extract the residue and position. Ignore any excess details in the string.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.annotate.unify_interaction_data">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.annotate.</span></span><span class="sig-name descname"><span class="pre">unify_interaction_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interaction_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/annotate.html#unify_interaction_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.annotate.unify_interaction_data" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm data and a column containing interaction data, extract the interacting protein, type of interaction, and convert to UniProtKB accession. This will be added as a new column labeled ‘Interacting ID’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe containing PTM data</p>
</dd>
<dt><strong>interaction_col: str</strong></dt><dd><p>column containing interaction information from a specific database</p>
</dd>
<dt><strong>name_dict: dict</strong></dt><dd><p>dictionary to convert names within given database to UniProt IDs. For cases when name is not necessarily one of the gene names listed in UniProt</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>interact: pd.DataFrame</dt><dd><p>Contains PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-ptm_pose.analyze">
<span id="analysis"></span><h2>Analysis<a class="headerlink" href="#module-ptm_pose.analyze" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.annotation_enrichment">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">annotation_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pregenerated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#annotation_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.annotation_enrichment" title="Permalink to this definition">#</a></dt>
<dd><p>In progress, needs to be tested</p>
<p>Given spliced ptm information (differential inclusion, altered flanking sequences, or both), calculate the enrichment of specific annotations in the dataset using a hypergeometric test. Background data can be provided/constructed in a few ways:</p>
<ol class="arabic simple">
<li><p>Use preconstructed background data for the annotation of interest, which considers the entire proteome present in the ptm_coordinates dataframe. While this is the default, it may not be the most accurate representation of your data, so you may alternative decide to use the other options which will be more specific to your context.</p></li>
<li><p>Use the alpha and min_dPSI parameter to construct a foreground that only includes significantly spliced PTMs, and use the entire provided spliced_ptms dataframe as the background. This will allow you to compare the enrichment of specific annotations in the significantly spliced PTMs compared to the entire dataset. Will do this automatically if alpha or min_dPSI is provided.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, ‘PTMInt’, ‘PTMcode’, ‘DEPOD’, ‘RegPhos’, ‘PTMsigDB’. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>background_type: str</strong></dt><dd><p>how to construct the background data. Options include ‘pregenerated’ (default) and ‘significance’. If ‘significance’ is selected, the alpha and min_dPSI parameters must be provided. Otherwise, will use whole proteome in the ptm_coordinates dataframe as the background.</p>
</dd>
<dt><strong>collapse_on_similar: bool</strong></dt><dd><p>Whether to collapse similar annotations (for example, increasing and decreasing functions) into a single category. Default is False.</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset, if any</p>
</dd>
<dt><strong>alpha: float</strong></dt><dd><p>significance threshold to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>min_dPSI: float</strong></dt><dd><p>minimum delta PSI value to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>annotation_file: str</strong></dt><dd><p>file to use to annotate custom background data. Default is None.</p>
</dd>
<dt><strong>save_background: bool</strong></dt><dd><p>Whether to save the background data constructed from the ptm_coordinates dataframe into Resource_Files within package. Default is False.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.combine_outputs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">combine_outputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop_codon_introduction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_conflicting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#combine_outputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.combine_outputs" title="Permalink to this definition">#</a></dt>
<dd><p>Given the spliced_ptms (differentially included) and altered_flanks (altered flanking sequences) dataframes obtained from project and flanking_sequences modules, combine the two into a single dataframe that categorizes each PTM by the impact on the PTM site</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset, if any</p>
</dd>
<dt><strong>include_stop_codon_introduction: bool</strong></dt><dd><p>Whether to include PTMs that introduce stop codons in the altered flanks. Default is False.</p>
</dd>
<dt><strong>remove_conflicting: bool</strong></dt><dd><p>Whether to remove PTMs that are both included and excluded across different splicing events. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.compare_inclusion_motifs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_inclusion_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flanking_sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#compare_inclusion_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.compare_inclusion_motifs" title="Permalink to this definition">#</a></dt>
<dd><p>Given a DataFrame containing flanking sequences with changes and a DataFrame containing ELM class information, identify motifs that are found in the inclusion and exclusion events, identifying motifs unique to each case. This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>flanking_sequences: pandas.DataFrame</strong></dt><dd><p>DataFrame containing flanking sequences with changes, obtained from get_flanking_changes_from_splice_data()</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>). Recommended to download this file and input it manually, but will download from ELM otherwise</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>flanking_sequences: pandas.DataFrame</dt><dd><p>DataFrame containing flanking sequences with changes and motifs found in the inclusion and exclusion events</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.edit_sequence_for_kinase_library">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">edit_sequence_for_kinase_library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#edit_sequence_for_kinase_library"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.edit_sequence_for_kinase_library" title="Permalink to this definition">#</a></dt>
<dd><p>Convert flanking sequence to version accepted by kinase library (modified residue denoted by asterick)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.findAlteredPositions">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">findAlteredPositions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#findAlteredPositions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.findAlteredPositions" title="Permalink to this definition">#</a></dt>
<dd><p>Given two sequences, identify the location of positions that have changed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>sequences to compare (order does not matter)</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>size of the flanking sequences (default is 5). This is used to make sure the provided sequences are the correct length</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>altered_positions: list</dt><dd><p>list of positions that have changed</p>
</dd>
<dt>residue_change: list</dt><dd><p>list of residues that have changed associated with that position</p>
</dd>
<dt>flank_side: str</dt><dd><p>indicates which side of the flanking sequence the change has occurred (N-term, C-term, or Both)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.find_motifs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">find_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#find_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.find_motifs" title="Permalink to this definition">#</a></dt>
<dd><p>Given a sequence and a dataframe containinn ELM class information, identify motifs that can be found in the provided sequence using the RegEx expression provided by ELM (PTMs not considered). This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq: str</strong></dt><dd><p>sequence to search for motifs</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.gene_set_enrichment">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">gene_set_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['KEGG_2021_Human',</span> <span class="pre">'GO_Biological_Process_2023',</span> <span class="pre">'GO_Cellular_Component_2023',</span> <span class="pre">'GO_Molecular_Function_2023',</span> <span class="pre">'Reactome_2022']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_sig_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#gene_set_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.gene_set_enrichment" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced_ptms and/or altered_flanks dataframes (or the dataframes combined from combine_outputs()), perform gene set enrichment analysis using the enrichr API</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with differentially included PTMs projected onto splicing events and with annotations appended from various databases. Default is None (will not be considered in analysis). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases. Default is None (will not be considered). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>combined: pd.DataFrame</strong></dt><dd><p>Combined dataframe with spliced_ptms and altered_flanks dataframes. Default is None. If provided, spliced_ptms and altered_flanks dataframes will be ignored.</p>
</dd>
<dt><strong>gene_sets: list</strong></dt><dd><p>List of gene sets to use in enrichment analysis. Default is [‘KEGG_2021_Human’, ‘GO_Biological_Process_2023’, ‘GO_Cellular_Component_2023’, ‘GO_Molecular_Function_2023’,’Reactome_2022’]. Look at gseapy and enrichr documentation for other available gene sets</p>
</dd>
<dt><strong>background: list</strong></dt><dd><p>List of genes to use as background in enrichment analysis. Default is None (all genes in the gene set database will be used).</p>
</dd>
<dt><strong>return_sig_only: bool</strong></dt><dd><p>Whether to return only significantly enriched gene sets. Default is True.</p>
</dd>
<dt><strong>max_retries: int</strong></dt><dd><p>Number of times to retry downloading gene set enrichment data from enrichr API. Default is 5.</p>
</dd>
<dt><strong>delay: int</strong></dt><dd><p>Number of seconds to wait between retries. Default is 10.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>results: pd.DataFrame</dt><dd><p>Dataframe with gene set enrichment results from enrichr API</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.getSequenceIdentity">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">getSequenceIdentity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#getSequenceIdentity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.getSequenceIdentity" title="Permalink to this definition">#</a></dt>
<dd><p>Given two flanking sequences, calculate the sequence identity between them using Biopython and parameters definded by Pillman et al. BMC Bioinformatics 2011</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>flanking sequence</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>normalized_score: float</dt><dd><p>normalized score of sequence similarity between flanking sequences (calculated similarity/max possible similarity)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.get_annotation_categories">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_annotation_categories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_annotation_categories"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.get_annotation_categories" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm information, return the available annotation categories that have been appended to dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annot_categories: pd.DataFrame</dt><dd><p>Dataframe that indicates the available databases, annotations from each database, and column associated with that annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.get_annotation_col">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_annotation_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_annotation_col"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.get_annotation_col" title="Permalink to this definition">#</a></dt>
<dd><p>Given the database of interest and annotation type, return the annotation column that will be found in a annotated spliced_ptm dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTM annotations added from annotate module</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, ‘PTMInt’, ‘PTMcode’, ‘DEPOD’, and ‘RegPhos’. Default is ‘PhosphoSitePlus’.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annotation_col: str</dt><dd><p>Column name in spliced_ptms dataframe that contains the requested annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.get_enrichment_inputs">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_enrichment_inputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pregenerated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_enrichment_inputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.get_enrichment_inputs" title="Permalink to this definition">#</a></dt>
<dd><p>Given the spliced ptms, altered_flanks, or combined PTMs dataframe, identify the number of PTMs corresponding to specific annotations in the foreground (PTMs impacted by splicing) and the background (all PTMs in the proteome or all PTMs in dataset not impacted by splicing). This information can be used to calculate the enrichment of specific annotations among PTMs impacted by splicing. Several options are provided for constructing the background data: pregenerated (based on entire proteome in the ptm_coordinates dataframe) or significance (foreground PTMs are extracted from provided spliced PTMs based on significance and minimum delta PSI)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.get_modification_counts">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_modification_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_modification_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.get_modification_counts" title="Permalink to this definition">#</a></dt>
<dd><p>Given PTM data (either spliced ptms, altered flanks, or combined data), return the counts of each modification class</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events or with altered flanking sequences</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>modification_counts: pd.Series</dt><dd><p>Series with the counts of each modification class</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.get_ptm_annotations">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_ptm_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_ptm_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.get_ptm_annotations" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm information obtained from project and annotate modules, grab PTMs in spliced ptms associated with specific PTM modules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, or ‘PTMInt’. ELM and PTMInt data will automatically be downloaded, but due to download restrictions, PhosphoSitePlus data must be manually downloaded and annotated in the spliced_ptms data using functions from the annotate module. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="ptm_pose.analyze.simplify_annotation">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">simplify_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#simplify_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ptm_pose.analyze.simplify_annotation" title="Permalink to this definition">#</a></dt>
<dd><p>Given an annotation, remove additional information such as whether or not a function is increasing or decreasing. For example, ‘cell growth, induced’ would be simplified to ‘cell growth’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation: str</strong></dt><dd><p>Annotation to simplify</p>
</dd>
<dt><strong>sep: str</strong></dt><dd><p>Separator that splits the core annotation from additional detail. Default is ‘,’. Assumes the first element is the core annotation.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annotation: str</dt><dd><p>Simplified annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-0">
<span id="plotting"></span><h2>Plotting<a class="headerlink" href="#module-0" title="Permalink to this heading">#</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="id0">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">annotation_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pregenerated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#annotation_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id0" title="Permalink to this definition">#</a></dt>
<dd><p>In progress, needs to be tested</p>
<p>Given spliced ptm information (differential inclusion, altered flanking sequences, or both), calculate the enrichment of specific annotations in the dataset using a hypergeometric test. Background data can be provided/constructed in a few ways:</p>
<ol class="arabic simple">
<li><p>Use preconstructed background data for the annotation of interest, which considers the entire proteome present in the ptm_coordinates dataframe. While this is the default, it may not be the most accurate representation of your data, so you may alternative decide to use the other options which will be more specific to your context.</p></li>
<li><p>Use the alpha and min_dPSI parameter to construct a foreground that only includes significantly spliced PTMs, and use the entire provided spliced_ptms dataframe as the background. This will allow you to compare the enrichment of specific annotations in the significantly spliced PTMs compared to the entire dataset. Will do this automatically if alpha or min_dPSI is provided.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, ‘PTMInt’, ‘PTMcode’, ‘DEPOD’, ‘RegPhos’, ‘PTMsigDB’. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>background_type: str</strong></dt><dd><p>how to construct the background data. Options include ‘pregenerated’ (default) and ‘significance’. If ‘significance’ is selected, the alpha and min_dPSI parameters must be provided. Otherwise, will use whole proteome in the ptm_coordinates dataframe as the background.</p>
</dd>
<dt><strong>collapse_on_similar: bool</strong></dt><dd><p>Whether to collapse similar annotations (for example, increasing and decreasing functions) into a single category. Default is False.</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset, if any</p>
</dd>
<dt><strong>alpha: float</strong></dt><dd><p>significance threshold to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>min_dPSI: float</strong></dt><dd><p>minimum delta PSI value to use to subset foreground PTMs. Default is None.</p>
</dd>
<dt><strong>annotation_file: str</strong></dt><dd><p>file to use to annotate custom background data. Default is None.</p>
</dd>
<dt><strong>save_background: bool</strong></dt><dd><p>Whether to save the background data constructed from the ptm_coordinates dataframe into Resource_Files within package. Default is False.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id1">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">combine_outputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_stop_codon_introduction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_conflicting</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#combine_outputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id1" title="Permalink to this definition">#</a></dt>
<dd><p>Given the spliced_ptms (differentially included) and altered_flanks (altered flanking sequences) dataframes obtained from project and flanking_sequences modules, combine the two into a single dataframe that categorizes each PTM by the impact on the PTM site</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset, if any</p>
</dd>
<dt><strong>include_stop_codon_introduction: bool</strong></dt><dd><p>Whether to include PTMs that introduce stop codons in the altered flanks. Default is False.</p>
</dd>
<dt><strong>remove_conflicting: bool</strong></dt><dd><p>Whether to remove PTMs that are both included and excluded across different splicing events. Default is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id2">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">compare_inclusion_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">flanking_sequences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#compare_inclusion_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id2" title="Permalink to this definition">#</a></dt>
<dd><p>Given a DataFrame containing flanking sequences with changes and a DataFrame containing ELM class information, identify motifs that are found in the inclusion and exclusion events, identifying motifs unique to each case. This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>flanking_sequences: pandas.DataFrame</strong></dt><dd><p>DataFrame containing flanking sequences with changes, obtained from get_flanking_changes_from_splice_data()</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>). Recommended to download this file and input it manually, but will download from ELM otherwise</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>flanking_sequences: pandas.DataFrame</dt><dd><p>DataFrame containing flanking sequences with changes and motifs found in the inclusion and exclusion events</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id3">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">edit_sequence_for_kinase_library</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#edit_sequence_for_kinase_library"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id3" title="Permalink to this definition">#</a></dt>
<dd><p>Convert flanking sequence to version accepted by kinase library (modified residue denoted by asterick)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id4">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">findAlteredPositions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flank_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#findAlteredPositions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id4" title="Permalink to this definition">#</a></dt>
<dd><p>Given two sequences, identify the location of positions that have changed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>sequences to compare (order does not matter)</p>
</dd>
<dt><strong>flank_size: int</strong></dt><dd><p>size of the flanking sequences (default is 5). This is used to make sure the provided sequences are the correct length</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>altered_positions: list</dt><dd><p>list of positions that have changed</p>
</dd>
<dt>residue_change: list</dt><dd><p>list of residues that have changed associated with that position</p>
</dd>
<dt>flank_side: str</dt><dd><p>indicates which side of the flanking sequence the change has occurred (N-term, C-term, or Both)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id5">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">find_motifs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">elm_classes</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#find_motifs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id5" title="Permalink to this definition">#</a></dt>
<dd><p>Given a sequence and a dataframe containinn ELM class information, identify motifs that can be found in the provided sequence using the RegEx expression provided by ELM (PTMs not considered). This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq: str</strong></dt><dd><p>sequence to search for motifs</p>
</dd>
<dt><strong>elm_classes: pandas.DataFrame</strong></dt><dd><p>DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (<a class="reference external" href="http://elm.eu.org/elms/elms_index.tsv">http://elm.eu.org/elms/elms_index.tsv</a>)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id6">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">gene_set_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">altered_flanks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combined</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_sets</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['KEGG_2021_Human',</span> <span class="pre">'GO_Biological_Process_2023',</span> <span class="pre">'GO_Cellular_Component_2023',</span> <span class="pre">'GO_Molecular_Function_2023',</span> <span class="pre">'Reactome_2022']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_sig_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_retries</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#gene_set_enrichment"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id6" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced_ptms and/or altered_flanks dataframes (or the dataframes combined from combine_outputs()), perform gene set enrichment analysis using the enrichr API</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with differentially included PTMs projected onto splicing events and with annotations appended from various databases. Default is None (will not be considered in analysis). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>altered_flanks: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases. Default is None (will not be considered). If combined dataframe is provided, this dataframe will be ignored.</p>
</dd>
<dt><strong>combined: pd.DataFrame</strong></dt><dd><p>Combined dataframe with spliced_ptms and altered_flanks dataframes. Default is None. If provided, spliced_ptms and altered_flanks dataframes will be ignored.</p>
</dd>
<dt><strong>gene_sets: list</strong></dt><dd><p>List of gene sets to use in enrichment analysis. Default is [‘KEGG_2021_Human’, ‘GO_Biological_Process_2023’, ‘GO_Cellular_Component_2023’, ‘GO_Molecular_Function_2023’,’Reactome_2022’]. Look at gseapy and enrichr documentation for other available gene sets</p>
</dd>
<dt><strong>background: list</strong></dt><dd><p>List of genes to use as background in enrichment analysis. Default is None (all genes in the gene set database will be used).</p>
</dd>
<dt><strong>return_sig_only: bool</strong></dt><dd><p>Whether to return only significantly enriched gene sets. Default is True.</p>
</dd>
<dt><strong>max_retries: int</strong></dt><dd><p>Number of times to retry downloading gene set enrichment data from enrichr API. Default is 5.</p>
</dd>
<dt><strong>delay: int</strong></dt><dd><p>Number of seconds to wait between retries. Default is 10.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>results: pd.DataFrame</dt><dd><p>Dataframe with gene set enrichment results from enrichr API</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id7">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">getSequenceIdentity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#getSequenceIdentity"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id7" title="Permalink to this definition">#</a></dt>
<dd><p>Given two flanking sequences, calculate the sequence identity between them using Biopython and parameters definded by Pillman et al. BMC Bioinformatics 2011</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq1, seq2: str</strong></dt><dd><p>flanking sequence</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>normalized_score: float</dt><dd><p>normalized score of sequence similarity between flanking sequences (calculated similarity/max possible similarity)</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id8">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_annotation_categories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_annotation_categories"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id8" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm information, return the available annotation categories that have been appended to dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annot_categories: pd.DataFrame</dt><dd><p>Dataframe that indicates the available databases, annotations from each database, and column associated with that annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id9">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_annotation_col</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_annotation_col"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id9" title="Permalink to this definition">#</a></dt>
<dd><p>Given the database of interest and annotation type, return the annotation column that will be found in a annotated spliced_ptm dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTM annotations added from annotate module</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, ‘PTMInt’, ‘PTMcode’, ‘DEPOD’, and ‘RegPhos’. Default is ‘PhosphoSitePlus’.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annotation_col: str</dt><dd><p>Column name in spliced_ptms dataframe that contains the requested annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id10">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_enrichment_inputs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pregenerated'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dPSI</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_background</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_enrichment_inputs"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id10" title="Permalink to this definition">#</a></dt>
<dd><p>Given the spliced ptms, altered_flanks, or combined PTMs dataframe, identify the number of PTMs corresponding to specific annotations in the foreground (PTMs impacted by splicing) and the background (all PTMs in the proteome or all PTMs in dataset not impacted by splicing). This information can be used to calculate the enrichment of specific annotations among PTMs impacted by splicing. Several options are provided for constructing the background data: pregenerated (based on entire proteome in the ptm_coordinates dataframe) or significance (foreground PTMs are extracted from provided spliced PTMs based on significance and minimum delta PSI)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id11">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_modification_counts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ptms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_modification_counts"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id11" title="Permalink to this definition">#</a></dt>
<dd><p>Given PTM data (either spliced ptms, altered flanks, or combined data), return the counts of each modification class</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>ptms: pd.DataFrame</strong></dt><dd><p>Dataframe with PTMs projected onto splicing events or with altered flanking sequences</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>modification_counts: pd.Series</dt><dd><p>Series with the counts of each modification class</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id12">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">get_ptm_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">spliced_ptms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Function'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PhosphoSitePlus'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mod_class</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_on_similar</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dPSI_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sig_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#get_ptm_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id12" title="Permalink to this definition">#</a></dt>
<dd><p>Given spliced ptm information obtained from project and annotate modules, grab PTMs in spliced ptms associated with specific PTM modules</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>spliced_ptms: pd.DataFrame</strong></dt><dd><p>PTMs projected onto splicing events and with annotations appended from various databases</p>
</dd>
<dt><strong>annotation_type: str</strong></dt><dd><p>Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is ‘Function’.</p>
</dd>
<dt><strong>database: str</strong></dt><dd><p>database from which PTMs are pulled. Options include ‘PhosphoSitePlus’, ‘ELM’, or ‘PTMInt’. ELM and PTMInt data will automatically be downloaded, but due to download restrictions, PhosphoSitePlus data must be manually downloaded and annotated in the spliced_ptms data using functions from the annotate module. Default is ‘PhosphoSitePlus’.</p>
</dd>
<dt><strong>mod_class: str</strong></dt><dd><p>modification class to subset</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="id13">
<span class="sig-prename descclassname"><span class="pre">ptm_pose.analyze.</span></span><span class="sig-name descname"><span class="pre">simplify_annotation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">','</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/ptm_pose/analyze.html#simplify_annotation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#id13" title="Permalink to this definition">#</a></dt>
<dd><p>Given an annotation, remove additional information such as whether or not a function is increasing or decreasing. For example, ‘cell growth, induced’ would be simplified to ‘cell growth’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>annotation: str</strong></dt><dd><p>Annotation to simplify</p>
</dd>
<dt><strong>sep: str</strong></dt><dd><p>Separator that splits the core annotation from additional detail. Default is ‘,’. Assumes the first element is the core annotation.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>annotation: str</dt><dd><p>Simplified annotation</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="quickstart.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Installation</p>
      </div>
    </a>
    <a class="right-next"
       href="General_Instructions.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Running PTM-POSE</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.pose_config">Configuration</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.project">PTM Projection</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.flanking_sequences">Flanking Sequences</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.annotate">Annotating PTMs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-ptm_pose.analyze">Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-0">Plotting</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>