

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Gallery Tests &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-dataframe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-rendered-html.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Gallery/gallery_tests';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../PTM_POSE.html">PTM-POSE Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../General_Instructions.html">Running PTM-POSE</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../gallery_output/index.html">Types of Analysis Performed with PTM-POSE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_protein_interactions.html">Identify protein interactions that may be impacted by splicing of PTMs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_num_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../Examples/Examples.html">Full Analysis Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../Examples/ESRP1_knockdown.html">Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Gallery/gallery_tests.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Gallery Tests</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-identify-PTMs">Inspecting identify PTMs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-number-of-PTMs-with-annotation-information-available">Inspecting number of PTMs with annotation information available</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-Specific-Annotation">Inspecting Specific Annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotation-Enrichment-Analysis">Annotation Enrichment Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Gene-Set-Enrichment-Analysis">Gene Set Enrichment Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Protein-Interaction-Networks">Protein Interaction Networks</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#KSTAR-Analysis">KSTAR Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Flanking-Sequence-Analysis">Flanking Sequence Analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Location-of-altered-flanks">Location of altered flanks</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Plot-Event">Plot Event</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Gallery-Tests">
<h1>Gallery Tests<a class="headerlink" href="#Gallery-Tests" title="Permalink to this heading">#</a></h1>
<section id="Inspecting-identify-PTMs">
<h2>Inspecting identify PTMs<a class="headerlink" href="#Inspecting-identify-PTMs" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[70]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pose_plots</span><span class="o">.</span><span class="n">modification_breakdown</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_3_0.png" src="../_images/Gallery_gallery_tests_3_0.png" />
</div>
</div>
</section>
<section id="Inspecting-number-of-PTMs-with-annotation-information-available">
<h2>Inspecting number of PTMs with annotation information available<a class="headerlink" href="#Inspecting-number-of-PTMs-with-annotation-information-available" title="Permalink to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
<span class="n">combined_output</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Some annotations in spliced ptms dataframe not found in altered flanks dataframe. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">analyze</span><span class="o">.</span><span class="n">show_available_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-intense-fg ansi-bold">---------------------------------------------------------------------------</span>
<span class="ansi-red-intense-fg ansi-bold">AttributeError</span>                            Traceback (most recent call last)
Cell <span class="ansi-green-intense-fg ansi-bold">In[3], line 3</span>
<span class="ansi-green-fg">      1</span> <span class="ansi-bold" style="color: rgb(0,135,0)">from</span> <span class="ansi-bold" style="color: rgb(0,0,255)">ptm_pose</span> <span class="ansi-bold" style="color: rgb(0,135,0)">import</span> plots <span class="ansi-bold" style="color: rgb(0,135,0)">as</span> pose_plots
<span class="ansi-green-intense-fg ansi-bold">----&gt; 3</span> <span class="ansi-yellow-bg">analyze</span><span class="ansi-yellow-bg" style="color: rgb(98,98,98)">.</span><span class="ansi-yellow-bg">show_available_annotations</span>()

<span class="ansi-red-intense-fg ansi-bold">AttributeError</span>: module &#39;ptm_pose.analyze&#39; has no attribute &#39;show_available_annotations&#39;
</pre></div></div>
</div>
</section>
<section id="Inspecting-Specific-Annotation">
<h2>Inspecting Specific Annotation<a class="headerlink" href="#Inspecting-Specific-Annotation" title="Permalink to this heading">#</a></h2>
<p>Often, we will want to dig deeper into the specific functions, processes, interactions, etc. associated with the proteins in our dataset. First, we can look at the annotations currently available for analysis, based on annotations that have been appended using the annotate module:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
<span class="n">combined_output</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">)</span>

<span class="n">annot_categories</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_annotation_categories</span><span class="p">(</span><span class="n">combined_output</span><span class="p">)</span>
<span class="n">annot_categories</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Some annotations in spliced ptms dataframe not found in altered flanks dataframe. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>database</th>
      <th>annotation_type</th>
      <th>column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>5</th>
      <td>Combined</td>
      <td>Interactions</td>
      <td>Combined:Interactions</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Combined</td>
      <td>Kinase</td>
      <td>Combined:Kinase</td>
    </tr>
    <tr>
      <th>1</th>
      <td>DEPOD</td>
      <td>Phosphatase</td>
      <td>DEPOD:Phosphatase</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ELM</td>
      <td>Interactions</td>
      <td>ELM:Interactions</td>
    </tr>
    <tr>
      <th>0</th>
      <td>PhosphoSitePlus</td>
      <td>Interactions</td>
      <td>PSP:ON_PROT_INTERACT</td>
    </tr>
    <tr>
      <th>3</th>
      <td>PhosphoSitePlus</td>
      <td>Disease</td>
      <td>PSP:Disease_Association</td>
    </tr>
    <tr>
      <th>4</th>
      <td>PhosphoSitePlus</td>
      <td>Process</td>
      <td>PSP:ON_PROCESS</td>
    </tr>
    <tr>
      <th>6</th>
      <td>PhosphoSitePlus</td>
      <td>Function</td>
      <td>PSP:ON_FUNCTION</td>
    </tr>
    <tr>
      <th>7</th>
      <td>RegPhos</td>
      <td>Kinase</td>
      <td>RegPhos:Kinase</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>This will tell us what database information is available, the types of information from that database, and the column associated with that information. Let’s take a closer look at the biological process information from PhosphoSitePlus:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ptms_with_annotation</span><span class="p">,</span> <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s2">&quot;PhosphoSitePlus&quot;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Specific PTMs with annotation:&#39;</span><span class="p">)</span>
<span class="n">ptms_with_annotation</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Specific PTMs with annotation:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene</th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Modification Class</th>
      <th>PSP:ON_PROCESS</th>
      <th>dPSI</th>
      <th>Significance</th>
      <th>Impact</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>BCAR1</td>
      <td>P56945</td>
      <td>Y</td>
      <td>267.0</td>
      <td>Phosphorylation</td>
      <td>cell growth, induced</td>
      <td>-0.07</td>
      <td>0.0458775672499</td>
      <td>Excluded</td>
    </tr>
    <tr>
      <th>1</th>
      <td>BCAR1</td>
      <td>P56945</td>
      <td>Y</td>
      <td>287.0</td>
      <td>Phosphorylation</td>
      <td>cell growth, induced</td>
      <td>-0.07</td>
      <td>0.0458775672499</td>
      <td>Excluded</td>
    </tr>
    <tr>
      <th>2</th>
      <td>BIN1</td>
      <td>O00499</td>
      <td>T</td>
      <td>348.0</td>
      <td>Phosphorylation</td>
      <td>signaling pathway regulation</td>
      <td>-0.112</td>
      <td>0.0233903490744</td>
      <td>Excluded</td>
    </tr>
    <tr>
      <th>3</th>
      <td>CEACAM1</td>
      <td>P13688</td>
      <td>S</td>
      <td>461.0</td>
      <td>Phosphorylation</td>
      <td>apoptosis, altered</td>
      <td>0.525</td>
      <td>1.73943268451e-09</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>4</th>
      <td>CTTN</td>
      <td>Q14247</td>
      <td>K</td>
      <td>272.0</td>
      <td>Acetylation</td>
      <td>cell motility, inhibited</td>
      <td>0.09</td>
      <td>0.0355211287599</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>5</th>
      <td>CTTN</td>
      <td>Q14247</td>
      <td>S</td>
      <td>298.0</td>
      <td>Phosphorylation</td>
      <td>cell motility, altered; cytoskeletal reorganiz...</td>
      <td>0.09</td>
      <td>0.0355211287599</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>6</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>S</td>
      <td>387.0</td>
      <td>Phosphorylation</td>
      <td>cell motility, altered</td>
      <td>0.253</td>
      <td>0.0129400018182</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>7</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>T</td>
      <td>614.0</td>
      <td>Phosphorylation</td>
      <td>cell motility, altered</td>
      <td>0.253</td>
      <td>0.0129400018182</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TSC2</td>
      <td>P49815</td>
      <td>S</td>
      <td>981.0</td>
      <td>Phosphorylation</td>
      <td>carcinogenesis, inhibited; cell growth, inhibi...</td>
      <td>-0.219</td>
      <td>4.18472157275e-05</td>
      <td>Excluded</td>
    </tr>
    <tr>
      <th>9</th>
      <td>YAP1</td>
      <td>P46937</td>
      <td>K</td>
      <td>342.0</td>
      <td>Ubiquitination</td>
      <td>carcinogenesis, altered</td>
      <td>-0.188;-0.161</td>
      <td>0.000211254197372;4.17884655686e-07</td>
      <td>Excluded</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>From this, we note a total of 9 impacted PTMs from 7 genes that have biological process information available. While we could manually look through to look for common processes, we can also inspect the annotation counts object to see the most common processes, including a breakdown by the type of impact (included [dPSI &gt; 0], excluded [dPSI &lt; 0], or altered flanking sequence):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of PTMs associated with each annotation:&#39;</span><span class="p">)</span>
<span class="n">annotation_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Number of PTMs associated with each annotation:
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>All Impacted</th>
      <th>Included</th>
      <th>Excluded</th>
      <th>Altered Flank</th>
    </tr>
    <tr>
      <th>PSP:ON_PROCESS</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>cell motility, altered</th>
      <td>3</td>
      <td>3.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>cell growth, induced</th>
      <td>2</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>signaling pathway regulation</th>
      <td>2</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>apoptosis, altered</th>
      <td>1</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>cell motility, inhibited</th>
      <td>1</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>cytoskeletal reorganization</th>
      <td>1</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>cell adhesion, inhibited</th>
      <td>1</td>
      <td>1.0</td>
      <td>0.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>carcinogenesis, inhibited</th>
      <td>1</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>cell growth, inhibited</th>
      <td>1</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>autophagy, inhibited</th>
      <td>1</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>carcinogenesis, altered</th>
      <td>1</td>
      <td>0.0</td>
      <td>1.0</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Finally, you may prefer to visualize this information as a figure. Here, we can plot the top 10 most common biological processes for the included, excluded, and altered flanking sequence impacts. Notably, we can plot either the annotations as outputted above (includes directionality of PTM role) or we can collapse this information into similar groups (e.g. “cell motility, altered” and “cell motility, included” would be grouped as “cell motility”). Here, we will plot the full information on the
left and the collapsed information on the right:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">wspace</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_annotations</span><span class="p">(</span><span class="n">combined_output</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">top_terms</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Full Annotation&#39;</span><span class="p">)</span>
<span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_annotations</span><span class="p">(</span><span class="n">combined_output</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">top_terms</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Collapsed Annotation&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0.5, 1.0, &#39;Collapsed Annotation&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_14_1.png" src="../_images/Gallery_gallery_tests_14_1.png" />
</div>
</div>
<p>Of note, you can also choose to only show collapsed annotation information for <code class="docutils literal notranslate"><span class="pre">analyze.get_ptm_annotations()</span></code> by setting <code class="docutils literal notranslate"><span class="pre">collapse_on_similar=True</span></code> in the function call, like we have done for the plot on the right.</p>
</section>
<section id="Annotation-Enrichment-Analysis">
<h2>Annotation Enrichment Analysis<a class="headerlink" href="#Annotation-Enrichment-Analysis" title="Permalink to this heading">#</a></h2>
<p>In some cases, you may want to identify PTM-specific annotations that appear more commonly than might be expected based on how often the annotation appears across the entire proteome. We have provided a function to perform this analysis, <code class="docutils literal notranslate"><span class="pre">analyze.ptm_annotation_enrichment()</span></code>. By default, this function will compare the annotations found in your data to the annotations found in the entire proteome (based on ptm_coordinates dataframe), but you can also choose to perform enrichment analysis by
significance. Here, we will we perform enrichment analysis using the entire proteome as the background. First, let’s look at the available annotations for enrichment analysis:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
<span class="n">combined_output</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">)</span>

<span class="n">annot_categories</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_annotation_categories</span><span class="p">(</span><span class="n">combined_output</span><span class="p">)</span>
<span class="n">annot_categories</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Some annotations in spliced ptms dataframe not found in altered flanks dataframe. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>database</th>
      <th>annotation_type</th>
      <th>column</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>4</th>
      <td>Combined</td>
      <td>Interactions</td>
      <td>Combined:Interactions</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Combined</td>
      <td>Kinase</td>
      <td>Combined:Kinase</td>
    </tr>
    <tr>
      <th>2</th>
      <td>DEPOD</td>
      <td>Phosphatase</td>
      <td>DEPOD:Phosphatase</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ELM</td>
      <td>Interactions</td>
      <td>ELM:Interactions</td>
    </tr>
    <tr>
      <th>0</th>
      <td>PhosphoSitePlus</td>
      <td>Process</td>
      <td>PSP:ON_PROCESS</td>
    </tr>
    <tr>
      <th>1</th>
      <td>PhosphoSitePlus</td>
      <td>Interactions</td>
      <td>PSP:ON_PROT_INTERACT</td>
    </tr>
    <tr>
      <th>6</th>
      <td>PhosphoSitePlus</td>
      <td>Disease</td>
      <td>PSP:Disease_Association</td>
    </tr>
    <tr>
      <th>8</th>
      <td>PhosphoSitePlus</td>
      <td>Function</td>
      <td>PSP:ON_FUNCTION</td>
    </tr>
    <tr>
      <th>7</th>
      <td>RegPhos</td>
      <td>Kinase</td>
      <td>RegPhos:Kinase</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We would like to know if the PTMs have been implicated in any biological processes more than expected by chance. We can perform enrichment analysis on the biological process annotations from PhosphoSitePlus. To maximize the ability of the hypergeometric test to capture these results, we will use the collapsed annotation information (ignores directionality of PTM role):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enrichment</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">annotation_enrichment</span><span class="p">(</span><span class="n">combined_output</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">collapse_on_similar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">enrichment</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Using pregenerated background information on all PTMs in the proteome.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Fraction Impacted</th>
      <th>p-value</th>
      <th>Adjusted p-value</th>
      <th>PTM</th>
    </tr>
    <tr>
      <th>PSP:ON_PROCESS</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>cell motility</th>
      <td>5/1078</td>
      <td>0.052579</td>
      <td>0.420633</td>
      <td>ABI1_S392;CTTN_K272;CTTN_S298;SPHK2_S387;SPHK2...</td>
    </tr>
    <tr>
      <th>cell adhesion</th>
      <td>2/324</td>
      <td>0.122466</td>
      <td>0.489864</td>
      <td>CTTN_S298;MPZL1_Y241</td>
    </tr>
    <tr>
      <th>cell growth</th>
      <td>4/1793</td>
      <td>0.427134</td>
      <td>1.000000</td>
      <td>BCAR1_Y267;BCAR1_Y287;BCAR1_Y306;TSC2_S981</td>
    </tr>
    <tr>
      <th>autophagy</th>
      <td>1/306</td>
      <td>0.434215</td>
      <td>0.868429</td>
      <td>TSC2_S981</td>
    </tr>
    <tr>
      <th>cytoskeletal reorganization</th>
      <td>2/796</td>
      <td>0.435637</td>
      <td>0.868429</td>
      <td>ABI1_S392;CTTN_S298</td>
    </tr>
    <tr>
      <th>apoptosis</th>
      <td>2/1179</td>
      <td>0.644065</td>
      <td>0.868429</td>
      <td>CEACAM1_S461;CEACAM1_T457</td>
    </tr>
    <tr>
      <th>signaling pathway regulation</th>
      <td>2/1206</td>
      <td>0.656208</td>
      <td>0.868429</td>
      <td>BIN1_T348;TSC2_S981</td>
    </tr>
    <tr>
      <th>carcinogenesis</th>
      <td>2/1501</td>
      <td>0.768091</td>
      <td>0.868429</td>
      <td>TSC2_S981;YAP1_K342</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can also plot the annotations and include which annotations are enriched (p-value &lt; 0.05) in the plot:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><br/><br/><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;not yet implemented&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Gene-Set-Enrichment-Analysis">
<h2>Gene Set Enrichment Analysis<a class="headerlink" href="#Gene-Set-Enrichment-Analysis" title="Permalink to this heading">#</a></h2>
<p>In addition to looking at the annotations associated with the PTMs, we can also look at the genes themselves with impacted PTMs. We can perform gene set enrichment analysis using EnrichR module of gseapy to identify if any gene sets are enriched in the PTM dataset, as well as break it down by the type of modication. Here, we will use the <code class="docutils literal notranslate"><span class="pre">analyze.gene_set_enrichment()</span></code> function to perform this analysis. First, let’s look at the available gene sets for enrichment analysis:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
<span class="n">combined_output</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Some annotations in spliced ptms dataframe not found in altered flanks dataframe. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enrichr_results</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">gene_set_enrichment</span><span class="p">(</span><span class="n">combined</span> <span class="o">=</span> <span class="n">combined_output</span><span class="p">,</span> <span class="n">gene_sets</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;GO_Biological_Process_2023&#39;</span><span class="p">,</span> <span class="s1">&#39;Reactome_2022&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">enrichr_results</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_set</th>
      <th>Term</th>
      <th>Overlap</th>
      <th>P-value</th>
      <th>Adjusted P-value</th>
      <th>Old P-value</th>
      <th>Old Adjusted P-value</th>
      <th>Odds Ratio</th>
      <th>Combined Score</th>
      <th>Genes</th>
      <th>Type</th>
      <th>Genes with Differentially Included PTMs only</th>
      <th>Genes with PTM with Altered Flanking Sequence only</th>
      <th>Genes with Both</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>GO_Biological_Process_2023</td>
      <td>Regulation Of Neurogenesis (GO:0050767)</td>
      <td>5/67</td>
      <td>0.000018</td>
      <td>0.011675</td>
      <td>0</td>
      <td>0</td>
      <td>17.392181</td>
      <td>189.619722</td>
      <td>YAP1;APLP2;DOCK7;NUMB;NF2</td>
      <td>Differentially Included + Altered Flanking Seq...</td>
      <td>YAP1;APLP2</td>
      <td>NF2</td>
      <td>DOCK7;NUMB</td>
    </tr>
    <tr>
      <th>1</th>
      <td>GO_Biological_Process_2023</td>
      <td>Enzyme-Linked Receptor Protein Signaling Pathw...</td>
      <td>6/124</td>
      <td>0.000031</td>
      <td>0.011675</td>
      <td>0</td>
      <td>0</td>
      <td>11.055131</td>
      <td>114.642865</td>
      <td>CSF1;FGFR3;FGFR2;PTPRF;BCAR1;MPZL1</td>
      <td>Differentially Included + Altered Flanking Seq...</td>
      <td>FGFR2;CSF1;FGFR3</td>
      <td></td>
      <td>MPZL1;BCAR1;PTPRF</td>
    </tr>
    <tr>
      <th>2</th>
      <td>GO_Biological_Process_2023</td>
      <td>Protein Localization To Cell-Cell Junction (GO...</td>
      <td>3/15</td>
      <td>0.000048</td>
      <td>0.011675</td>
      <td>0</td>
      <td>0</td>
      <td>52.901596</td>
      <td>525.813416</td>
      <td>TJP1;LSR;SCRIB</td>
      <td>Differentially Included + Altered Flanking Seq...</td>
      <td></td>
      <td>LSR</td>
      <td>SCRIB;TJP1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>GO_Biological_Process_2023</td>
      <td>Regulation Of Cell Migration (GO:0030334)</td>
      <td>10/434</td>
      <td>0.000049</td>
      <td>0.011675</td>
      <td>0</td>
      <td>0</td>
      <td>5.280579</td>
      <td>52.425684</td>
      <td>TJP1;CEACAM1;CSF1;ADAM15;LIMCH1;APLP2;NUMB;ITG...</td>
      <td>Differentially Included + Altered Flanking Seq...</td>
      <td>APLP2;CSF1;ITGA6</td>
      <td>NF2</td>
      <td>ADAM15;NUMB;LIMCH1;BCAR1;TJP1;CEACAM1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>GO_Biological_Process_2023</td>
      <td>Integrin-Mediated Signaling Pathway (GO:0007229)</td>
      <td>5/85</td>
      <td>0.000058</td>
      <td>0.011675</td>
      <td>0</td>
      <td>0</td>
      <td>13.466712</td>
      <td>131.282293</td>
      <td>CEACAM1;ADAM15;ITGA6;CD47;BCAR1</td>
      <td>Differentially Included + Altered Flanking Seq...</td>
      <td>ITGA6;CD47</td>
      <td></td>
      <td>ADAM15;CEACAM1;BCAR1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The result is the standard output of gseapy, with the specific genes in the gene set with differentially include or altered flanking sequence PTM sites listed. We can also plot the output of the gene set enrichment analysis:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>

<span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_EnrichR_pies</span><span class="p">(</span><span class="n">enrichr_results</span><span class="p">,</span> <span class="n">top_terms</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_27_0.png" src="../_images/Gallery_gallery_tests_27_0.png" />
</div>
</div>
</section>
<section id="Protein-Interaction-Networks">
<h2>Protein Interaction Networks<a class="headerlink" href="#Protein-Interaction-Networks" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">interaction_graph</span><span class="p">,</span> <span class="n">network_data</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_interaction_network</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">node_type</span> <span class="o">=</span> <span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
<span class="n">network_stats</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_interaction_stats</span><span class="p">(</span><span class="n">interaction_graph</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PhosphoSitePlus regulatory site data found and added
Combined kinase-substrate data found and added
PTMInt data found and added
ELM data found and added
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">network_data</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Modified Gene</th>
      <th>Interacting Gene</th>
      <th>Residue</th>
      <th>Type</th>
      <th>Source</th>
      <th>dPSI</th>
      <th>Regulation Change</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ADAM15</td>
      <td>HCK</td>
      <td>Y735;Y715</td>
      <td>REGULATES</td>
      <td>PSP/RegPhos</td>
      <td>0.181;-0.052</td>
      <td>+;-</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ADAM15</td>
      <td>LCK</td>
      <td>Y715</td>
      <td>REGULATES</td>
      <td>PSP/RegPhos</td>
      <td>0.181;-0.052</td>
      <td>+;-</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ADAM15</td>
      <td>SRC</td>
      <td>Y735;Y715</td>
      <td>REGULATES</td>
      <td>PSP/RegPhos</td>
      <td>0.181;-0.052</td>
      <td>+;-</td>
    </tr>
    <tr>
      <th>3</th>
      <td>BCAR1</td>
      <td>SRC</td>
      <td>Y267;Y287</td>
      <td>REGULATES</td>
      <td>PSP/RegPhos</td>
      <td>-0.07</td>
      <td>-</td>
    </tr>
    <tr>
      <th>4</th>
      <td>BIN1</td>
      <td>MAPT</td>
      <td>T348</td>
      <td>INDUCES</td>
      <td>PhosphoSitePlus;PTMInt</td>
      <td>-0.112</td>
      <td>-</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">importlib</span>
<span class="n">importlib</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">analyze</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;module &#39;ptm_pose.analyze&#39; from &#39;C:\\Users\\Sam\\OneDrive\\Documents\\GradSchool\\Research\\Splicing\\PTM_POSE\\ptm_pose\\analyze.py&#39;&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">analyze</span><span class="o">.</span><span class="n">summarize_protein_network</span><span class="p">(</span><span class="n">protein</span> <span class="o">=</span> <span class="s1">&#39;TSC2&#39;</span><span class="p">,</span> <span class="n">interaction_graph</span> <span class="o">=</span> <span class="n">interaction_graph</span><span class="p">,</span> <span class="n">network_data</span> <span class="o">=</span> <span class="n">network_data</span><span class="p">,</span> <span class="n">network_stats</span> <span class="o">=</span> <span class="n">network_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Decreased interaction likelihoods: AKT1, YWHAE, YWHAZ
Number of interactions: 3 (Rank: 2)
Centrality measures -    Degree = 0.2 (Rank: 2)
                         Betweenness = 0.028571428571428574 (Rank: 3)
                         Closeness = 0.2 (Rank: 3)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_interaction_network</span><span class="p">(</span><span class="n">interaction_graph</span><span class="p">,</span> <span class="n">network_data</span><span class="p">,</span> <span class="n">network_stats</span> <span class="o">=</span> <span class="n">network_stats</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_34_0.png" src="../_images/Gallery_gallery_tests_34_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>

<span class="n">network_stats</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_interaction_stats</span><span class="p">(</span><span class="n">interaction_graph</span><span class="p">)</span>
<span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_network_centrality</span><span class="p">(</span><span class="n">network_stats</span><span class="p">,</span> <span class="n">network_data</span><span class="p">,</span> <span class="n">top_N</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">modified_color</span> <span class="o">=</span> <span class="s1">&#39;coral&#39;</span><span class="p">,</span> <span class="n">interacting_color</span> <span class="o">=</span> <span class="s1">&#39;grey&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_35_0.png" src="../_images/Gallery_gallery_tests_35_0.png" />
</div>
</div>
</section>
<section id="KSTAR-Analysis">
<h2>KSTAR Analysis<a class="headerlink" href="#KSTAR-Analysis" title="Permalink to this heading">#</a></h2>
<p>While we provide functions for performing enrichment of known kinase substrates from databases like PhosphoSitePlus, RegPhos, and PTMsigDB, these resources are limited by the overall number of validated substrates (&lt;5%). For this purpose, we have adapted a previously developed algorithm called KSTAR (Kinase Substrate to Activity Relationships) for use with spliced PTM data, which harnesses kinase-substrate predictions to expand the overall number of phosphorylation sites that can be used as
evidence. This particularly important as you may find many of the spliced PTMs in your dataset are less well studied and may not have any annotated kinases.</p>
<p>In order to perform KSTAR analysis, you will first need to download KSTAR networks from the following <a class="reference external" href="https://figshare.com/articles/dataset/NETWORKS/14944305?file=28768155">figshare</a>.</p>
<p>Once you have downloaded the networks, all you need is your PTM data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load spliced ptm and altered flank data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;spliced_ptms.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kstar_enrichment</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">kstar_enrichment</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">network_dir</span> <span class="o">=</span> <span class="s1">&#39;../../../../Database_Information/NETWORKS/NetworKIN/&#39;</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">)</span>
<span class="n">kstar_enrichment</span><span class="o">.</span><span class="n">run_kstar_enrichment</span><span class="p">()</span>
<span class="n">kstar_enrichment</span><span class="o">.</span><span class="n">return_enriched_kinases</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>You can also run the same analysis for serine/threonine kinases:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kstar_enrichment</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">kstar_enrichment</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">network_dir</span> <span class="o">=</span> <span class="s1">&#39;../../../../Database_Information/NETWORKS/NetworKIN/&#39;</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;ST&#39;</span><span class="p">)</span>
<span class="n">kstar_enrichment</span><span class="o">.</span><span class="n">run_kstar_enrichment</span><span class="p">()</span>
<span class="n">kstar_enrichment</span><span class="o">.</span><span class="n">return_enriched_kinases</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([&#39;PRKG2&#39;, &#39;MAPK14&#39;, &#39;PRKCH&#39;, &#39;PRKCG&#39;, &#39;PRKD1&#39;, &#39;PRKCE&#39;, &#39;ROCK1&#39;,
       &#39;TTK&#39;], dtype=object)
</pre></div></div>
</div>
</section>
<section id="Flanking-Sequence-Analysis">
<h2>Flanking Sequence Analysis<a class="headerlink" href="#Flanking-Sequence-Analysis" title="Permalink to this heading">#</a></h2>
<section id="Location-of-altered-flanks">
<h3>Location of altered flanks<a class="headerlink" href="#Location-of-altered-flanks" title="Permalink to this heading">#</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">flanking_sequences</span> <span class="k">as</span> <span class="n">fs</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1"># Load altered flank data</span>
<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;altered_flanks.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">fs</span><span class="o">.</span><span class="n">compare_flanking_sequences</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">)</span>
<span class="n">altered_flanks</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;Inclusion Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Exclusion Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequence Identity&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Positions&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue Change&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Flank Side&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[73]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Modification Class</th>
      <th>Inclusion Sequence</th>
      <th>Exclusion Sequence</th>
      <th>Sequence Identity</th>
      <th>Altered Positions</th>
      <th>Residue Change</th>
      <th>Altered Flank Side</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>P01116</td>
      <td>T</td>
      <td>148</td>
      <td>Phosphorylation</td>
      <td>ETSAKtRQESG</td>
      <td>ETSAKtRQGC*</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>P01116</td>
      <td>K</td>
      <td>147</td>
      <td>Acetylation</td>
      <td>IETSAkTRQES</td>
      <td>IETSAkTRQGC</td>
      <td>0.818182</td>
      <td>[4.0, 5.0]</td>
      <td>[E-&gt;G, S-&gt;C]</td>
      <td>C-term only</td>
    </tr>
    <tr>
      <th>2</th>
      <td>P01116</td>
      <td>K</td>
      <td>147</td>
      <td>Ubiquitination</td>
      <td>IETSAkTRQES</td>
      <td>IETSAkTRQGC</td>
      <td>0.818182</td>
      <td>[4.0, 5.0]</td>
      <td>[E-&gt;G, S-&gt;C]</td>
      <td>C-term only</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Q9UPQ0</td>
      <td>S</td>
      <td>746</td>
      <td>Phosphorylation</td>
      <td>LPNLNsQGVAW</td>
      <td>LPNLNsQGGFS</td>
      <td>0.727273</td>
      <td>[3.0, 4.0, 5.0]</td>
      <td>[V-&gt;G, A-&gt;F, W-&gt;S]</td>
      <td>C-term only</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Q9UPQ0</td>
      <td>S</td>
      <td>750</td>
      <td>Phosphorylation</td>
      <td>PSQVDsPSSEK</td>
      <td>ILKVDsPSSEK</td>
      <td>0.727273</td>
      <td>[-5.0, -4.0, -3.0]</td>
      <td>[P-&gt;I, S-&gt;L, Q-&gt;K]</td>
      <td>N-term only</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Note, we only calculate these metrics for cases where altered flanking sequences do not cause a stop codon to be introduced, as this is harder to interpret (such as for the first PTM in the list). The above table will indicate the positions in the flanking sequence that are altered, how similar the altered flanking sequence is to the original flanking sequence, and the specific residue change that takes place. We can also plot some of this information to get a better sense of the distribution of
altered flanking sequences:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">importlib</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">pose_plots</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;module &#39;ptm_pose.plots&#39; from &#39;C:\\Users\\Sam\\OneDrive\\Documents\\GradSchool\\Research\\Splicing\\PTM_POSE\\ptm_pose\\plots.py&#39;&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>

<span class="n">pose_plots</span><span class="o">.</span><span class="n">location_of_altered_flanking_residues</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\Sam\OneDrive\Documents\GradSchool\Research\Splicing\PTM_POSE\ptm_pose\plots.py:391: UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.
  ax[0].set_xticklabels([&#39;N-term\nonly&#39;, &#39;C-term\nonly&#39;])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_47_1.png" src="../_images/Gallery_gallery_tests_47_1.png" />
</div>
</div>
<p>We can even create the same plot for specific modification types or residues, as well as label the specific residue changes that occur:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[76]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pose_plots</span><span class="o">.</span><span class="n">location_of_altered_flanking_residues</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">modification_class</span><span class="o">=</span><span class="s1">&#39;Acetylation&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
C:\Users\Sam\OneDrive\Documents\GradSchool\Research\Splicing\PTM_POSE\ptm_pose\plots.py:437: UserWarning: set_ticklabels() should only be used with a fixed number of ticks, i.e. after set_ticks() or using a FixedLocator.
  ax[0].set_xticklabels([&#39;N-term\nonly&#39;, &#39;C-term\nonly&#39;])
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_49_1.png" src="../_images/Gallery_gallery_tests_49_1.png" />
</div>
</div>
<p>If we want to dig deeper, we can look at the specific changes that occurring, although this is only recommended with a selected subset of PTMs, such as those that may have a functional impact:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[102]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pose_plots</span><span class="o">.</span><span class="n">alterations_matrix</span><span class="p">(</span><span class="n">altered_flanks</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_51_0.png" src="../_images/Gallery_gallery_tests_51_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[119]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">importlib</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">analyze</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[119]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;module &#39;ptm_pose.analyze&#39; from &#39;C:\\Users\\Sam\\OneDrive\\Documents\\GradSchool\\Research\\Splicing\\PTM_POSE\\ptm_pose\\analyze.py&#39;&gt;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[120]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">compare_inclusion_motifs</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[126]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sh2_motif_changes</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">identify_change_to_specific_motif</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">elm_motif_name</span> <span class="o">=</span> <span class="s1">&#39;14-3-3&#39;</span><span class="p">,</span> <span class="n">modification_class</span> <span class="o">=</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="n">residues</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;S&#39;</span><span class="p">,</span><span class="s1">&#39;T&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[127]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sh2_motif_changes</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[127]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene</th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Modification Class</th>
      <th>Inclusion Sequence</th>
      <th>Exclusion Sequence</th>
      <th>Motif only in Inclusion</th>
      <th>Motif only in Exclusion</th>
      <th>Altered Positions</th>
      <th>Residue Change</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>22</th>
      <td>MLPH</td>
      <td>Q9BV36</td>
      <td>S</td>
      <td>337</td>
      <td>Phosphorylation</td>
      <td>RGRASsESQDL</td>
      <td>RGRASsESQGS</td>
      <td>LIG_14-3-3_CanoR_1</td>
      <td>NaN</td>
      <td>[4.0, 5.0]</td>
      <td>[D-&gt;G, L-&gt;S]</td>
    </tr>
    <tr>
      <th>23</th>
      <td>MLPH</td>
      <td>Q9BV36</td>
      <td>S</td>
      <td>339</td>
      <td>Phosphorylation</td>
      <td>RASSEsQDL*A</td>
      <td>RASSEsQGSRC</td>
      <td>LIG_14-3-3_CanoR_1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>50</th>
      <td>CEACAM1</td>
      <td>P13688</td>
      <td>T</td>
      <td>457</td>
      <td>Phosphorylation</td>
      <td>LHFGKtGRGKR</td>
      <td>LHFGKtGRLRT</td>
      <td>NaN</td>
      <td>LIG_14-3-3_CterR_2</td>
      <td>[3.0, 4.0, 5.0]</td>
      <td>[G-&gt;L, K-&gt;R, R-&gt;T]</td>
    </tr>
    <tr>
      <th>67</th>
      <td>ENAH</td>
      <td>Q8N8S7</td>
      <td>S</td>
      <td>512</td>
      <td>Phosphorylation</td>
      <td>KSPVIsRTGFS</td>
      <td>KSPVIsRTKIH</td>
      <td>LIG_14-3-3_CterR_2</td>
      <td>NaN</td>
      <td>[3.0, 4.0, 5.0]</td>
      <td>[G-&gt;K, F-&gt;I, S-&gt;H]</td>
    </tr>
    <tr>
      <th>93</th>
      <td>LMO7</td>
      <td>Q8WWI1-3</td>
      <td>S</td>
      <td>356</td>
      <td>Phosphorylation</td>
      <td>ADGTFsRTLSK</td>
      <td>ADGTFsRE*VH</td>
      <td>LIG_14-3-3_CterR_2</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>129</th>
      <td>MAP3K7</td>
      <td>O43318</td>
      <td>T</td>
      <td>403</td>
      <td>Phosphorylation</td>
      <td>RIAATtGLFQA</td>
      <td>RIAATtGQRTA</td>
      <td>LIG_14-3-3_CanoR_1</td>
      <td>NaN</td>
      <td>[2.0, 3.0, 4.0]</td>
      <td>[L-&gt;Q, F-&gt;R, Q-&gt;T]</td>
    </tr>
    <tr>
      <th>141</th>
      <td>LMO7</td>
      <td>Q8WWI1-3</td>
      <td>T</td>
      <td>354</td>
      <td>Phosphorylation</td>
      <td>TEADGtFSR*S</td>
      <td>TEADGtFSRE*</td>
      <td>LIG_14-3-3_CterR_2</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[128]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pose_plots</span><span class="o">.</span><span class="n">alterations_matrix</span><span class="p">(</span><span class="n">sh2_motif_changes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Gallery_gallery_tests_56_0.png" src="../_images/Gallery_gallery_tests_56_0.png" />
</div>
</div>
</section>
</section>
<section id="Plot-Event">
<h2>Plot Event<a class="headerlink" href="#Plot-Event" title="Permalink to this heading">#</a></h2>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-identify-PTMs">Inspecting identify PTMs</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-number-of-PTMs-with-annotation-information-available">Inspecting number of PTMs with annotation information available</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Inspecting-Specific-Annotation">Inspecting Specific Annotation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Annotation-Enrichment-Analysis">Annotation Enrichment Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Gene-Set-Enrichment-Analysis">Gene Set Enrichment Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Protein-Interaction-Networks">Protein Interaction Networks</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#KSTAR-Analysis">KSTAR Analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Flanking-Sequence-Analysis">Flanking Sequence Analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Location-of-altered-flanks">Location of altered flanks</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Plot-Event">Plot Event</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>