
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Assessing enriched PTM functions &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'gallery_output/FunctionEnrichment/plot_specific_annotation';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Impact on protein interaction and regulatory networks" href="../Networks/index.html" />
    <link rel="prev" title="Identify enriched gene sets associated with differentially spliced PTMs" href="plot_geneset_enrichment.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../index.html">Gallery</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 has-children"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 current active has-children"><a class="reference internal" href="index.html">Functional Impact of Spliced PTMs</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_ksea.html">Kinase substrate enrichment analysis (KSEA)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_protein_interactions.html">Protein interactions network</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
<li class="toctree-l3"><a class="reference internal" href="../flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/gallery_output/FunctionEnrichment/plot_specific_annotation.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Assessing enriched PTM functions</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="sphx-glr-download-link-note admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#sphx-glr-download-gallery-output-functionenrichment-plot-specific-annotation-py"><span class="std std-ref">Go to the end</span></a>
to download the full example code.</p>
</div>
<section class="sphx-glr-example-title" id="assessing-enriched-ptm-functions">
<span id="sphx-glr-gallery-output-functionenrichment-plot-specific-annotation-py"></span><h1>Assessing enriched PTM functions<a class="headerlink" href="#assessing-enriched-ptm-functions" title="Link to this heading">#</a></h1>
<p>As described in Running PTM-POSE section, PTM-POSE provides various options for annotating functional information for PTMs, coming from various databases. Often, we will want to dig deeper into the specific functions, processes, interactions, etc. associated with the proteins in our dataset. First, we can look at the annotations currently available for analysis, based on annotations that have been appended using the annotate module:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose</span><span class="w"> </span><span class="kn">import</span> <span class="n">helpers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose.analyze</span><span class="w"> </span><span class="kn">import</span> <span class="n">annotations</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>


<span class="n">annot_categories</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">get_available_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
<span class="n">annot_categories</span>
</pre></div>
</div>
<div class="output_subarea output_html rendered_html output_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Database</th>
      <th>Annotation Type</th>
      <th>Appended to PTM data?</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>iKiP</td>
      <td>Enzyme</td>
      <td>No</td>
    </tr>
    <tr>
      <th>1</th>
      <td>PhosphoSitePlus</td>
      <td>Enzyme</td>
      <td>No</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PhosphoSitePlus</td>
      <td>Disease</td>
      <td>No</td>
    </tr>
    <tr>
      <th>3</th>
      <td>PhosphoSitePlus</td>
      <td>Interactions</td>
      <td>No</td>
    </tr>
    <tr>
      <th>4</th>
      <td>PhosphoSitePlus</td>
      <td>Function</td>
      <td>No</td>
    </tr>
    <tr>
      <th>5</th>
      <td>PhosphoSitePlus</td>
      <td>Process</td>
      <td>No</td>
    </tr>
    <tr>
      <th>6</th>
      <td>PhosphoSitePlus</td>
      <td>Perturbation</td>
      <td>No</td>
    </tr>
    <tr>
      <th>7</th>
      <td>PTMsigDB</td>
      <td>Perturbation-DIA</td>
      <td>No</td>
    </tr>
    <tr>
      <th>8</th>
      <td>PTMsigDB</td>
      <td>Perturbation-DIA2</td>
      <td>No</td>
    </tr>
    <tr>
      <th>9</th>
      <td>PTMsigDB</td>
      <td>Pathway-NetPath</td>
      <td>No</td>
    </tr>
    <tr>
      <th>10</th>
      <td>PTMsigDB</td>
      <td>Pathway-WikiPathways</td>
      <td>No</td>
    </tr>
    <tr>
      <th>11</th>
      <td>PTMsigDB</td>
      <td>Perturbation-PRM</td>
      <td>No</td>
    </tr>
    <tr>
      <th>12</th>
      <td>PTMsigDB</td>
      <td>Pathway-BI</td>
      <td>No</td>
    </tr>
    <tr>
      <th>13</th>
      <td>DEPOD</td>
      <td>Enzyme</td>
      <td>No</td>
    </tr>
    <tr>
      <th>14</th>
      <td>RegPhos</td>
      <td>Enzyme</td>
      <td>No</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<br />
<br /><p>This will tell us what database information is available and the types of information from that database. Let’s take a closer look at the biological process information from PhosphoSitePlus:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">ptms_with_annotation</span><span class="p">,</span> <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s2">&quot;PhosphoSitePlus&quot;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Specific PTMs with annotation:&#39;</span><span class="p">)</span>
<span class="n">ptms_with_annotation</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>3 PTMs removed due to insignificant splice event (p &lt; 0.05, dpsi &gt;= 0.1): (33.33%)
Final number of PTMs to be assessed: 6
Specific PTMs with annotation:
</pre></div>
</div>
<div class="output_subarea output_html rendered_html output_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene</th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Isoform</th>
      <th>Modification Class</th>
      <th>PhosphoSitePlus:Function</th>
      <th>dPSI</th>
      <th>Significance</th>
      <th>Impact</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>CEACAM1</td>
      <td>P13688</td>
      <td>S</td>
      <td>461.0</td>
      <td>Phosphorylation</td>
      <td>activity, inhibited</td>
      <td>0.525</td>
      <td>1.73943268451e-09</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>S</td>
      <td>419.0</td>
      <td>Phosphorylation</td>
      <td>intracellular localization</td>
      <td>0.253</td>
      <td>0.0129400018182</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>S</td>
      <td>421.0</td>
      <td>Phosphorylation</td>
      <td>intracellular localization</td>
      <td>0.253</td>
      <td>0.0129400018182</td>
      <td>Included</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TSC2</td>
      <td>P49815</td>
      <td>S</td>
      <td>981.0</td>
      <td>Phosphorylation</td>
      <td>activity, inhibited;intracellular localization...</td>
      <td>-0.219</td>
      <td>4.18472157275e-05</td>
      <td>Excluded</td>
    </tr>
    <tr>
      <th>4</th>
      <td>YAP1</td>
      <td>P46937</td>
      <td>K</td>
      <td>342.0</td>
      <td>Ubiquitination</td>
      <td>protein degradation</td>
      <td>-0.188;-0.161</td>
      <td>0.000211254197372;4.17884655686e-07</td>
      <td>Excluded</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<br />
<br /><p>From this, we note a total of 9 impacted PTMs from 7 genes that have biological process information available. While we could manually look through to look for common processes, we can also inspect the annotation counts object to see the most common processes, including a breakdown by the type of impact (included [dPSI &gt; 0], excluded [dPSI &lt; 0], or altered flanking sequence). We can also look at the number of PTMs associated with each annotation:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Number of PTMs associated with each annotation:&#39;</span><span class="p">)</span>
<span class="n">annotation_counts</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>Number of PTMs associated with each annotation:
</pre></div>
</div>
<div class="output_subarea output_html rendered_html output_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>All Impacted</th>
      <th>Included</th>
      <th>Excluded</th>
    </tr>
    <tr>
      <th>PhosphoSitePlus:Function</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>intracellular localization</th>
      <td>3</td>
      <td>2.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>activity, inhibited</th>
      <td>2</td>
      <td>1.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>molecular association, regulation</th>
      <td>1</td>
      <td>0.0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>protein degradation</th>
      <td>1</td>
      <td>0.0</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<br />
<br /><p>To better visualize the enriched functions, we can also plot the number of PTMs associated with each function:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="c1">#plot annotatinos when not collapsing them</span>
<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">subplots_adjust</span><span class="p">(</span><span class="n">wspace</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">annotations</span><span class="o">.</span><span class="n">plot_annotation_counts</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">top_terms</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Full Annotation&#39;</span><span class="p">)</span>
<span class="c1">#plot annotations when collapsing them</span>
<span class="n">annotations</span><span class="o">.</span><span class="n">plot_annotation_counts</span><span class="p">(</span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">top_terms</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Collapsed Annotation&#39;</span><span class="p">)</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_specific_annotation_001.png" srcset="../../_images/sphx_glr_plot_specific_annotation_001.png" alt="Full Annotation, Collapsed Annotation" class = "sphx-glr-single-img"/><div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;Collapsed Annotation&#39;)
</pre></div>
</div>
<p>We can also perform enrichment analysis to identify if any of the functions are more likely to be present then expected, using collapsed annotations to increase likelihood of finding enriched terms:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">enrichment</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">annotation_enrichment</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">collapse_on_similar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">enrichment</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
<div class="output_subarea output_html rendered_html output_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Fraction Impacted</th>
      <th>p-value</th>
      <th>Adjusted p-value</th>
      <th>PTM</th>
    </tr>
    <tr>
      <th>PhosphoSitePlus:Function</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>intracellular localization</th>
      <td>3/3006</td>
      <td>0.394168</td>
      <td>1.0</td>
      <td>SPHK2_S419;SPHK2_S421;TSC2_S981</td>
    </tr>
    <tr>
      <th>protein degradation</th>
      <td>1/1645</td>
      <td>0.710863</td>
      <td>1.0</td>
      <td>YAP1_K342</td>
    </tr>
    <tr>
      <th>activity</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>CEACAM1_S461;TSC2_S981</td>
    </tr>
    <tr>
      <th>molecular association</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>TSC2_S981</td>
    </tr>
  </tbody>
</table>
</div>
</div>
<br />
<br /><p class="sphx-glr-timing"><strong>Total running time of the script:</strong> (0 minutes 0.462 seconds)</p>
<div class="sphx-glr-footer sphx-glr-footer-example docutils container" id="sphx-glr-download-gallery-output-functionenrichment-plot-specific-annotation-py">
<div class="sphx-glr-download sphx-glr-download-jupyter docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/2bb06d4b584935b7fee09b2dd7f05775/plot_specific_annotation.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_specific_annotation.ipynb</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-python docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/f35bc2d0453fd0e6bd9b64da223cdf82/plot_specific_annotation.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_specific_annotation.py</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-zip docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/729e8a186730a298d9e5fbe7332723cc/plot_specific_annotation.zip"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">zipped:</span> <span class="pre">plot_specific_annotation.zip</span></code></a></p>
</div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.github.io">Gallery generated by Sphinx-Gallery</a></p>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="plot_geneset_enrichment.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Identify enriched gene sets associated with differentially spliced PTMs</p>
      </div>
    </a>
    <a class="right-next"
       href="../Networks/index.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Impact on protein interaction and regulatory networks</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>