
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ptm_pose.analyze &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/ptm_pose/analyze';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.3.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../General_Instructions.html">Running PTM-POSE</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Full Analysis Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">Project PTMs onto ESRP1 knockdown data (Yang et al, 2016)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_in_Prostate.html">Exploring the role of ESRP1 expression in prostate cancer</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../gallery_output/index.html">Types of Analysis Performed with PTM-POSE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/plot_num_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Networks/plot_protein_interactions.html">Identify protein interactions that may be impacted by splicing of PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for ptm_pose.analyze</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="c1">#plotting </span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib.lines</span> <span class="k">as</span> <span class="nn">mlines</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">plots</span> <span class="k">as</span> <span class="n">pose_plots</span>

<span class="c1">#analysis packages</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">PairwiseAligner</span>
<span class="kn">import</span> <span class="nn">gseapy</span> <span class="k">as</span> <span class="nn">gp</span>
<span class="kn">import</span> <span class="nn">networkx</span> <span class="k">as</span> <span class="nn">nx</span>
<span class="kn">import</span> <span class="nn">re</span>


<span class="c1">#custom stat functions</span>
<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">stat_utils</span><span class="p">,</span> <span class="n">pose_config</span><span class="p">,</span> <span class="n">annotate</span><span class="p">,</span> <span class="n">helpers</span>


<span class="c1">#optional analysis packages</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">kinase_library</span> <span class="k">as</span> <span class="nn">kl</span>
    <span class="n">kinase_library</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">kinase_library</span> <span class="o">=</span> <span class="kc">False</span>

<span class="n">package_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>




<div class="viewcode-block" id="get_modification_counts">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.get_modification_counts">[docs]</a>
<span class="k">def</span> <span class="nf">get_modification_counts</span><span class="p">(</span><span class="n">ptms</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given PTM data (either spliced ptms, altered flanks, or combined data), return the counts of each modification class</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events or with altered flanking sequences</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    modification_counts: pd.Series</span>
<span class="sd">        Series with the counts of each modification class</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span>
    <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
    <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">modification_counts</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">modification_counts</span></div>


<div class="viewcode-block" id="get_annotation_col">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.get_annotation_col">[docs]</a>
<span class="k">def</span> <span class="nf">get_annotation_col</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the database of interest and annotation type, return the annotation column that will be found in a annotated spliced_ptm dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTM annotations added from annotate module</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is &#39;Function&#39;.</span>
<span class="sd">    database: str</span>
<span class="sd">        database from which PTMs are pulled. Options include &#39;PhosphoSitePlus&#39;, &#39;ELM&#39;, &#39;PTMInt&#39;, &#39;PTMcode&#39;, &#39;DEPOD&#39;, and &#39;RegPhos&#39;. Default is &#39;PhosphoSitePlus&#39;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    annotation_col: str</span>
<span class="sd">        Column name in spliced_ptms dataframe that contains the requested annotation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="sa">f</span><span class="s1">&#39;Combined:</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Requested annotation data has not yet been added to spliced_ptms dataframe. Please run the annotate.</span><span class="si">{</span><span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_function_dict</span><span class="p">[</span><span class="n">database</span><span class="p">]</span><span class="si">}</span><span class="s1"> function to append this information.&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;Combined:</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">&#39;</span>
    <span class="k">elif</span> <span class="n">annotation_type</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_col_dict</span><span class="p">[</span><span class="n">database</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_col_dict</span><span class="p">[</span><span class="n">database</span><span class="p">][</span><span class="n">annotation_type</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">annotation_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Requested annotation data has not yet been added to spliced_ptms dataframe. Please run the annotate.</span><span class="si">{</span><span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_function_dict</span><span class="p">[</span><span class="n">database</span><span class="p">][</span><span class="n">annotation_type</span><span class="p">]</span><span class="si">}</span><span class="s1"> function to append this information.&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">annotation_col</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid annotation type for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">. Available annotation data for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> includes: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_col_dict</span><span class="p">[</span><span class="n">database</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>



<div class="viewcode-block" id="combine_outputs">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.combine_outputs">[docs]</a>
<span class="k">def</span> <span class="nf">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">include_stop_codon_introduction</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">remove_conflicting</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the spliced_ptms (differentially included) and altered_flanks (altered flanking sequences) dataframes obtained from project and flanking_sequences modules, combine the two into a single dataframe that categorizes each PTM by the impact on the PTM site</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</span>
<span class="sd">    altered_flanks: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases</span>
<span class="sd">    mod_class: str</span>
<span class="sd">        modification class to subset, if any</span>
<span class="sd">    include_stop_codon_introduction: bool</span>
<span class="sd">        Whether to include PTMs that introduce stop codons in the altered flanks. Default is False.</span>
<span class="sd">    remove_conflicting: bool</span>
<span class="sd">        Whether to remove PTMs that are both included and excluded across different splicing events. Default is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#process differentially included PTMs and altered flanking sequences</span>
    <span class="k">if</span> <span class="n">mod_class</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">get_modification_class_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">)</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">get_modification_class_data</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">)</span>

    <span class="c1">#extract specific direction of splicing change and add to dataframe</span>
    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Included&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>

    <span class="c1">#restrict altered flanks to those that are changed and are not disrupted by stop codons</span>
    <span class="k">if</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtypes</span> <span class="o">!=</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">include_stop_codon_introduction</span><span class="p">:</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Stop Codon Introduced&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="k">else</span> <span class="s1">&#39;Altered Flank&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="o">~</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Altered Flank&#39;</span>

    <span class="c1">#identify annotations that are found in both datasets</span>
    <span class="n">annotation_columns_in_spliced_ptms</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
    <span class="n">annotation_columns_in_altered_flanks</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
    <span class="n">annotation_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns</span><span class="p">)</span> <span class="o">!=</span> <span class="n">annotation_columns_in_spliced_ptms</span><span class="p">:</span>
        <span class="n">annotation_columns_only_in_spliced</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">))</span>
        <span class="n">annotation_columns_only_in_altered</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns_only_in_spliced</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Warning: some annotations in spliced ptms dataframe not found in altered flanks dataframe: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_columns_only_in_spliced</span><span class="p">)</span><span class="si">}</span><span class="s1">. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns_only_in_altered</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Warning: some annotations in altered flanks dataframe not found in spliced ptms dataframe: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_columns_only_in_altered</span><span class="p">)</span><span class="si">}</span><span class="s1">. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.&#39;</span><span class="p">)</span>

    <span class="c1">#check if dPSI or sig columns are in both dataframes</span>
    <span class="n">sig_cols</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">sig_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;dPSI&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;Significance&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Significance&#39;</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">sig_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Significance&#39;</span><span class="p">)</span>

    <span class="n">shared_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">sig_cols</span> <span class="o">+</span> <span class="n">annotation_columns</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">spliced_ptms</span><span class="p">[</span><span class="n">shared_columns</span><span class="p">],</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="n">shared_columns</span><span class="p">]])</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s1">&#39;Impact&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>

    <span class="c1">#remove ptms that are both included and excluded across different events</span>
    <span class="k">if</span> <span class="n">remove_conflicting</span><span class="p">:</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="o">~</span><span class="p">((</span><span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Included&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Excluded&#39;</span><span class="p">)))]</span>

    <span class="k">return</span> <span class="n">combined</span></div>


<div class="viewcode-block" id="simplify_annotation">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.simplify_annotation">[docs]</a>
<span class="k">def</span> <span class="nf">simplify_annotation</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given an annotation, remove additional information such as whether or not a function is increasing or decreasing. For example, &#39;cell growth, induced&#39; would be simplified to &#39;cell growth&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    annotation: str</span>
<span class="sd">        Annotation to simplify</span>
<span class="sd">    sep: str</span>
<span class="sd">        Separator that splits the core annotation from additional detail. Default is &#39;,&#39;. Assumes the first element is the core annotation.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    annotation: str</span>
<span class="sd">        Simplified annotation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">annotation</span> <span class="o">==</span> <span class="n">annotation</span> <span class="k">else</span> <span class="n">annotation</span>
    <span class="k">return</span> <span class="n">annotation</span></div>


<span class="k">def</span> <span class="nf">collapse_annotations</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">):</span>
    <span class="n">sep_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:{</span><span class="s1">&#39;Function&#39;</span><span class="p">:</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">:</span><span class="s1">&#39;,&#39;</span><span class="p">,</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;(&#39;</span><span class="p">,</span> <span class="s1">&#39;Disease&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;ELM&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span> <span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;Motif Match&#39;</span><span class="p">:</span> <span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;Intraprotein&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:{</span><span class="s1">&#39;Phosphatase&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;WikiPathway&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;NetPath&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span><span class="s1">&#39;mSigDB&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (DIA2)&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (DIA)&#39;</span><span class="p">:</span> <span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (PRM)&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">}}</span>
    
    <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Kinase&#39;</span> <span class="ow">and</span> <span class="n">database</span> <span class="o">!=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span>
        <span class="n">collapsed</span> <span class="o">=</span> <span class="n">annotations</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sep</span> <span class="o">=</span> <span class="n">sep_dict</span><span class="p">[</span><span class="n">database</span><span class="p">][</span><span class="n">annotation_type</span><span class="p">]</span>
        <span class="n">collapsed</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">annot</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">annot</span> <span class="o">==</span> <span class="n">annot</span><span class="p">:</span>
                <span class="n">collapsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simplify_annotation</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="n">sep</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">collapsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annot</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">collapsed</span>


<span class="k">def</span> <span class="nf">get_modification_class_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">):</span>
    <span class="c1">#check if specific modification class was provided and subset data by modification if so</span>
    <span class="k">if</span> <span class="n">mod_class</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
        <span class="n">ptms_of_interest</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">mod_class</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ptms_of_interest</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms_of_interest</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">ptms_of_interest</span> <span class="o">=</span> <span class="n">ptms_of_interest</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span>
        <span class="n">available_ptms</span> <span class="o">=</span> <span class="n">ptms_of_interest</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Requested modification class not present in the data. The available modifications include </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">available_ptms</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ptms_of_interest</span>

<div class="viewcode-block" id="get_ptm_annotations">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.get_ptm_annotations">[docs]</a>
<span class="k">def</span> <span class="nf">get_ptm_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">dPSI_col</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sig_col</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm information obtained from project and annotate modules, grab PTMs in spliced ptms associated with specific PTM modules</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        PTMs projected onto splicing events and with annotations appended from various databases</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is &#39;Function&#39;.</span>
<span class="sd">    database: str</span>
<span class="sd">        database from which PTMs are pulled. Options include &#39;PhosphoSitePlus&#39;, &#39;ELM&#39;, or &#39;PTMInt&#39;. ELM and PTMInt data will automatically be downloaded, but due to download restrictions, PhosphoSitePlus data must be manually downloaded and annotated in the spliced_ptms data using functions from the annotate module. Default is &#39;PhosphoSitePlus&#39;.</span>
<span class="sd">    mod_class: str</span>
<span class="sd">        modification class to subset </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check to make sure requested annotation is available</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">!=</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:</span>
        <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">get_annotation_col</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">annotation_col</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Combined:</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">&#39;</span>


    <span class="c1">#check if specific modification class was provided and subset data by modification if so</span>
    <span class="k">if</span> <span class="n">mod_class</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ptms_of_interest</span> <span class="o">=</span> <span class="n">get_modification_class_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ptms_of_interest</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="c1">#extract relevant annotation and remove PTMs without an annotation</span>
    <span class="k">if</span> <span class="s1">&#39;Impact&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms_of_interest</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">ptms_of_interest</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">ptms_of_interest</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms_of_interest</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Included&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>

    <span class="n">optional_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ptms_of_interest</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">,</span> <span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="s1">&#39;Significance&#39;</span><span class="p">]</span> <span class="ow">or</span> <span class="n">col</span> <span class="o">==</span> <span class="n">dPSI_col</span> <span class="ow">or</span> <span class="n">col</span> <span class="o">==</span> <span class="n">sig_col</span> <span class="p">]</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">ptms_of_interest</span><span class="p">[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">optional_cols</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">annotation_col</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">annotations</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;No PTMs with associated annotation&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span>
    
    <span class="c1">#combine repeat entries for same PTM (with multiple impacts)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">i</span><span class="p">])))</span>

    <span class="c1">#separate distinct modification annotations in unique rows</span>
    <span class="n">annotations_exploded</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">annotations_exploded</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)</span>
    <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    
    <span class="c1">#if desired collapse similar annotations (for example, same function but increasing or decreasing)</span>
    <span class="k">if</span> <span class="n">collapse_on_similar</span><span class="p">:</span>
        <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">collapse_annotations</span><span class="p">(</span><span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">)</span>
        <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="n">annotation_col</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
    
    <span class="c1">#get the number of annotations found</span>
    <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">annotation_col</span><span class="p">])[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>

    <span class="c1">#additional_counts</span>
    <span class="n">sub_counts</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="s1">&#39;Impact&#39;</span> <span class="ow">in</span> <span class="n">annotations_exploded</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">imp</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Included&#39;</span><span class="p">,</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Flank&#39;</span><span class="p">]:</span>
            <span class="n">tmp_annotations</span> <span class="o">=</span> <span class="n">annotations_exploded</span><span class="p">[</span><span class="n">annotations_exploded</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">imp</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">tmp_annotations</span> <span class="o">=</span> <span class="n">tmp_annotations</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">annotation_col</span><span class="p">])</span>
            <span class="n">sub_counts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tmp_annotations</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">())</span>
    
        <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">annotation_counts</span><span class="p">]</span> <span class="o">+</span> <span class="n">sub_counts</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">annotation_counts</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;All Impacted&#39;</span><span class="p">,</span> <span class="s1">&#39;Included&#39;</span><span class="p">,</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Flank&#39;</span><span class="p">]</span>
        <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">annotation_counts</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    
    <span class="c1">#combine repeat entries for same PTM (with multiple impacts)</span>
    <span class="n">annotations</span> <span class="o">=</span> <span class="n">annotations</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">i</span><span class="p">])))</span>

    <span class="k">return</span> <span class="n">annotations</span><span class="p">,</span> <span class="n">annotation_counts</span></div>


<div class="viewcode-block" id="get_annotation_categories">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.get_annotation_categories">[docs]</a>
<span class="k">def</span> <span class="nf">get_annotation_categories</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm information, return the available annotation categories that have been appended to dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        PTMs projected onto splicing events and with annotations appended from various databases</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    annot_categories: pd.DataFrame</span>
<span class="sd">        Dataframe that indicates the available databases, annotations from each database, and column associated with that annotation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">database_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">type_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">column_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#get available phosphositeplus annotations</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">:</span>
            <span class="n">database</span> <span class="o">=</span> <span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;PSP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span> <span class="k">else</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span>
            <span class="k">if</span> <span class="n">database</span> <span class="o">!=</span> <span class="s1">&#39;Combined&#39;</span> <span class="ow">and</span> <span class="n">database</span> <span class="o">!=</span> <span class="s1">&#39;Unnamed&#39;</span><span class="p">:</span>
                <span class="n">col_dict</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">annotation_col_dict</span><span class="p">[</span><span class="n">database</span><span class="p">]</span>

                <span class="c1">#flip through annotation types in col_dict and add the one that matches the column</span>
                <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">col_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="k">if</span> <span class="n">value</span> <span class="o">==</span> <span class="n">col</span><span class="p">:</span>
                        <span class="n">type_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
                        <span class="n">database_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">database</span><span class="p">)</span>
                        <span class="n">column_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:</span>
                <span class="n">type_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
                <span class="n">database_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Combined&#39;</span><span class="p">)</span>
                <span class="n">column_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">type_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">annot_categories</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;database&#39;</span><span class="p">:</span><span class="n">database_list</span><span class="p">,</span> <span class="s1">&#39;annotation_type&#39;</span><span class="p">:</span><span class="n">type_list</span><span class="p">,</span> <span class="s1">&#39;column&#39;</span><span class="p">:</span> <span class="n">column_list</span><span class="p">})</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;database&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">annot_categories</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;No annotation information found. Please run functions from annotate module to append annotation information&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span></div>

    

<span class="k">def</span> <span class="nf">construct_background</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">modification</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s1">&#39;Gene name&#39;</span><span class="p">:</span><span class="s1">&#39;Gene&#39;</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">modification</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">modification</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No PTMs found with modification class </span><span class="si">{</span><span class="n">modification</span><span class="si">}</span><span class="s1">. Please provide a valid modification class. Examples include Phosphorylation, Glycosylation, Ubiquitination, etc.&#39;</span><span class="p">)</span>
    
        
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Please provide PhosphoSitePlus source file to construct the background dataframe&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">]:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_regulatory_site_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_kinase_substrate_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Disease&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_disease_association</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMsigDB_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;ELM&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_ELM_interactions</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Motif Match&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">:</span>
        <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMInt_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Intraprotein&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMcode_intraprotein</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
            <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
        <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:</span>
        <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_DEPOD_phosphatase_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span>
        <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMsigDB_data</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Combined information is not supported for constructing background data from entire proteome at this time. Please provide a specific database to construct background data.&#39;</span><span class="p">)</span>
    

    <span class="n">_</span><span class="p">,</span> <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="n">collapse_on_similar</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">package_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">collapse_on_similar</span> <span class="ow">and</span> <span class="n">modification</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">annotation_counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">package_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/Resource_Files/background_annotations/</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">modification</span><span class="si">}</span><span class="s1">_collapsed.csv&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">collapse_on_similar</span><span class="p">:</span>
            <span class="n">annotation_counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">package_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/Resource_Files/background_annotations/</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">_collapsed.csv&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">modification</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">annotation_counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">package_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/Resource_Files/background_annotations/</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">modification</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">annotation_counts</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">package_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/Resource_Files/background_annotations/</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">_</span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s1">.csv&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">annotation_counts</span>

    

    
<div class="viewcode-block" id="get_enrichment_inputs">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.get_enrichment_inputs">[docs]</a>
<span class="k">def</span> <span class="nf">get_enrichment_inputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span>  <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">background_type</span> <span class="o">=</span> <span class="s1">&#39;pregenerated&#39;</span><span class="p">,</span> <span class="n">background</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">min_dPSI</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">annotation_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">save_background</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the spliced ptms, altered_flanks, or combined PTMs dataframe, identify the number of PTMs corresponding to specific annotations in the foreground (PTMs impacted by splicing) and the background (all PTMs in the proteome or all PTMs in dataset not impacted by splicing). This information can be used to calculate the enrichment of specific annotations among PTMs impacted by splicing. Several options are provided for constructing the background data: pregenerated (based on entire proteome in the ptm_coordinates dataframe) or significance (foreground PTMs are extracted from provided spliced PTMs based on significance and minimum delta PSI)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">background_type</span> <span class="o">==</span> <span class="s1">&#39;pregenerated&#39;</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Using pregenerated background information on all PTMs in the proteome.&#39;</span><span class="p">)</span>
        <span class="c1">#first look for pregenerated background data</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">background_annotation_count</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">download_background</span><span class="p">(</span><span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="n">mod_class</span><span class="p">,</span> <span class="n">collapsed</span><span class="o">=</span><span class="n">collapse_on_similar</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">annotation_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Note: To avoid having to constructing background each time (which is slower), you can choose to set save_background = True to save the background data to Resource Files in package directory.&#39;</span><span class="p">)</span>
            <span class="n">background_annotation_count</span> <span class="o">=</span> <span class="n">construct_background</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="n">annotation_file</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="n">collapse_on_similar</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="n">save_background</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">mod_class</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">background_size</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">background_size</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">mod_class</span><span class="p">)]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

    <span class="k">elif</span> <span class="n">background_type</span> <span class="o">==</span> <span class="s1">&#39;significance&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;Significance&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Significance and dPSI columns must be present in spliced_ptms dataframe to construct a background based on significance (these columns must be provided during projection).&#39;</span><span class="p">)</span>
        
        <span class="n">background</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1">#restrict sample to significantly spliced ptms</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[(</span><span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Significance&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">alpha</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">min_dPSI</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>


        <span class="c1">#check to make sure there are significant PTMs in the data and that there is a difference in the number of significant and background PTMs</span>
        <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;No significantly spliced PTMs found in the data&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">background</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The foreground and background PTM sets are the same size when considering significance. Please provide a different background set with the background_ptms parameter, or make sure spliced_ptms also includes non-significant PTMs. Instead using pregenerated background sets of the whole proteome.&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mod_class</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">background</span> <span class="o">=</span> <span class="n">get_modification_class_data</span><span class="p">(</span><span class="n">background</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">)</span>
            <span class="c1">#remove impact from background columns if present</span>
            <span class="n">cols_to_remove</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="s1">&#39;Significance&#39;</span><span class="p">,</span> <span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">background</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols_to_remove</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">background</span> <span class="o">=</span> <span class="n">background</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">cols_to_remove</span><span class="p">)</span>
        <span class="c1">#get background counts</span>
            <span class="n">background_size</span> <span class="o">=</span> <span class="n">background</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">background_annotation_count</span> <span class="o">=</span> <span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">background</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="n">collapse_on_similar</span><span class="p">)</span>
    <span class="c1">#elif background is not None: #if custom background is provided</span>
    <span class="c1">#    print(&#39;Using the provided custom background&#39;)</span>
    <span class="c1">#    if isinstance(background, list) or isinstance(background, np.ndarray):</span>
    <span class="c1">#        #from list of PTM strings, separate into uniprot id, residue, and position</span>
    <span class="c1">#        uniprot_id = [ptm.split(&#39;_&#39;)[0] for ptm in background]</span>
    <span class="c1">#        residue = [ptm.split(&#39;_&#39;)[1][0] for ptm in background]</span>
    <span class="c1">#        position = [int(ptm.split(&#39;_&#39;)[1][1:]) for ptm in background]</span>
    <span class="c1">#        background = pd.DataFrame({&#39;UniProtKB Accession&#39;:uniprot_id, &#39;Residue&#39;:residue, &#39;PTM Position in Isoform&#39;:position, &#39;Modification Class&#39;:mod_class})</span>
    <span class="c1">#    if isinstance(background, pd.DataFrame):</span>
    <span class="c1">#        #check to make sure ptm data has key columns to identify ptms</span>
    <span class="c1">#        if &#39;UniProtKB Accession&#39; not in background.columns or &#39;Residue&#39; not in background.columns or &#39;PTM Position in Isoform&#39; not in background.columns or #&#39;Modification Class&#39; not in background.columns:</span>
    <span class="c1">#            raise ValueError(&#39;Background dataframe must have UniProtKB Accession, Residue, PTM Position in Isoform, and Modification Class columns to identify PTMs&#39;)</span>
            
    <span class="c1">#        #restrict to specific modification class</span>
    <span class="c1">#        if mod_class is not None and &#39;Modification Class&#39; in background.columns:</span>
    <span class="c1">#            background = get_modification_class_data(background, mod_class)</span>
    <span class="c1">#        elif mod_class is not None:</span>
    <span class="c1">#            raise ValueError(&#39;Custom background dataframe must have a Modification Class column to subset by modification class.&#39;)</span>
    <span class="c1">#    else:</span>
    <span class="c1">#        raise ValueError(&#39;Custom backgrounds must be provided as a list/array of PTMs in the form of &quot;P00533_Y1068&quot; (Uniprot ID followed by site number) or as a custom background dataframe with UniProtKB Accession, Residue, PTM Position in Isoform, and Modification Class columns.&#39;)</span>
        
    <span class="c1">#    background = annotate.add_annotation(background, annotation_type = annotation_type, database = database, check_existing = True, file = annotation_file)    </span>
    <span class="c1">#    background_size = background.drop_duplicates([&#39;UniProtKB Accession&#39;, &#39;Residue&#39;, &#39;PTM Position in Isoform&#39;]).shape[0]</span>

        <span class="c1">#get background counts</span>
    <span class="c1">#    _, background_annotation_count = get_ptm_annotations(background, annotation_type = annotation_type, database = database, collapse_on_similar = collapse_on_similar)</span>
    <span class="c1">#elif background_type == &#39;custom&#39;:</span>
    <span class="c1">#    raise ValueError(&#39;Please provide a custom background dataframe or list of PTMs to use as the background if wanting to use custom background data.&#39;)    </span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Invalid background type. Must be pregenerated (default) or significance&#39;</span><span class="p">)</span>

    <span class="c1">#get counts</span>
    <span class="n">foreground_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">annotation_details</span><span class="p">,</span> <span class="n">foreground_annotation_count</span> <span class="o">=</span> <span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">collapse_on_similar</span><span class="o">=</span><span class="n">collapse_on_similar</span><span class="p">)</span>

    <span class="c1">#process annotation details into usable format</span>
    <span class="k">if</span> <span class="n">annotation_details</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;No PTMs with requested annotation type, so could not perform enrichment analysis&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">repeat</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">get_annotation_col</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">)</span>
        <span class="n">annotation_details</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotation_details</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">annotation_details</span> <span class="o">=</span> <span class="n">annotation_details</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)</span>
        <span class="n">annotation_details</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotation_details</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">annotation_details</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotation_details</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">annotation_details</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">annotation_details</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">annotation_details</span> <span class="o">=</span> <span class="n">annotation_details</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">foreground_annotation_count</span><span class="p">,</span> <span class="n">foreground_size</span><span class="p">,</span> <span class="n">background_annotation_count</span><span class="p">,</span> <span class="n">background_size</span><span class="p">,</span> <span class="n">annotation_details</span></div>



<div class="viewcode-block" id="annotation_enrichment">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.annotation_enrichment">[docs]</a>
<span class="k">def</span> <span class="nf">annotation_enrichment</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">background_type</span> <span class="o">=</span> <span class="s1">&#39;pregenerated&#39;</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_dPSI</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">annotation_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">save_background</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span><span class="c1">#</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    In progress, needs to be tested</span>

<span class="sd">    Given spliced ptm information (differential inclusion, altered flanking sequences, or both), calculate the enrichment of specific annotations in the dataset using a hypergeometric test. Background data can be provided/constructed in a few ways:</span>

<span class="sd">    1. Use preconstructed background data for the annotation of interest, which considers the entire proteome present in the ptm_coordinates dataframe. While this is the default, it may not be the most accurate representation of your data, so you may alternative decide to use the other options which will be more specific to your context.</span>
<span class="sd">    2. Use the alpha and min_dPSI parameter to construct a foreground that only includes significantly spliced PTMs, and use the entire provided spliced_ptms dataframe as the background. This will allow you to compare the enrichment of specific annotations in the significantly spliced PTMs compared to the entire dataset. Will do this automatically if alpha or min_dPSI is provided.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</span>
<span class="sd">    database: str</span>
<span class="sd">        database from which PTMs are pulled. Options include &#39;PhosphoSitePlus&#39;, &#39;ELM&#39;, &#39;PTMInt&#39;, &#39;PTMcode&#39;, &#39;DEPOD&#39;, &#39;RegPhos&#39;, &#39;PTMsigDB&#39;. Default is &#39;PhosphoSitePlus&#39;.</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation to pull from spliced_ptms dataframe. Available information depends on the selected database. Default is &#39;Function&#39;.</span>
<span class="sd">    background_type: str</span>
<span class="sd">        how to construct the background data. Options include &#39;pregenerated&#39; (default) and &#39;significance&#39;. If &#39;significance&#39; is selected, the alpha and min_dPSI parameters must be provided. Otherwise, will use whole proteome in the ptm_coordinates dataframe as the background.</span>
<span class="sd">    collapse_on_similar: bool</span>
<span class="sd">        Whether to collapse similar annotations (for example, increasing and decreasing functions) into a single category. Default is False.</span>
<span class="sd">    mod_class: str</span>
<span class="sd">        modification class to subset, if any</span>
<span class="sd">    alpha: float</span>
<span class="sd">        significance threshold to use to subset foreground PTMs. Default is None.</span>
<span class="sd">    min_dPSI: float</span>
<span class="sd">        minimum delta PSI value to use to subset foreground PTMs. Default is None.</span>
<span class="sd">    annotation_file: str</span>
<span class="sd">        file to use to annotate custom background data. Default is None.</span>
<span class="sd">    save_background: bool</span>
<span class="sd">        Whether to save the background data constructed from the ptm_coordinates dataframe into Resource_Files within package. Default is False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">foreground_annotation_count</span><span class="p">,</span> <span class="n">foreground_size</span><span class="p">,</span> <span class="n">background_annotations</span><span class="p">,</span> <span class="n">background_size</span><span class="p">,</span> <span class="n">annotation_details</span> <span class="o">=</span> <span class="n">get_enrichment_inputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">background_type</span> <span class="o">=</span> <span class="n">background_type</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">collapse_on_similar</span> <span class="o">=</span> <span class="n">collapse_on_similar</span><span class="p">,</span> <span class="n">mod_class</span> <span class="o">=</span> <span class="n">mod_class</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">min_dPSI</span> <span class="o">=</span> <span class="n">min_dPSI</span><span class="p">,</span> <span class="n">annotation_file</span> <span class="o">=</span> <span class="n">annotation_file</span><span class="p">,</span> <span class="n">save_background</span> <span class="o">=</span> <span class="n">save_background</span><span class="p">)</span>
    

    <span class="k">if</span> <span class="n">foreground_annotation_count</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1">#iterate through all annotations and calculate enrichment with a hypergeometric test</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Fraction Impacted&#39;</span><span class="p">,</span> <span class="s1">&#39;p-value&#39;</span><span class="p">],</span> <span class="n">index</span> <span class="o">=</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">n</span> <span class="ow">in</span> <span class="n">background_annotations</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1">#number of PTMs in the foreground with the annotation</span>
            <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">foreground_annotation_count</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
                    <span class="n">k</span> <span class="o">=</span> <span class="n">foreground_annotation_count</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">k</span> <span class="o">=</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;count&#39;</span><span class="p">]</span>
                <span class="k">elif</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="n">k</span> <span class="o">=</span> <span class="n">foreground_annotation_count</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;All Impacted&#39;</span><span class="p">]</span>

                <span class="n">p</span> <span class="o">=</span> <span class="n">stat_utils</span><span class="o">.</span><span class="n">getEnrichment</span><span class="p">(</span><span class="n">background_size</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="n">foreground_size</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">fishers</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
                <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;Fraction Impacted&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">n</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;p-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">p</span>

        <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;p-value&#39;</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Adjusted p-value&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stat_utils</span><span class="o">.</span><span class="n">adjustP</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;p-value&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">results</span><span class="p">,</span> <span class="n">annotation_details</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">results</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">return</span> <span class="n">results</span></div>



<div class="viewcode-block" id="gene_set_enrichment">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.gene_set_enrichment">[docs]</a>
<span class="k">def</span> <span class="nf">gene_set_enrichment</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">combined</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sig_col</span> <span class="o">=</span> <span class="s1">&#39;Significance&#39;</span><span class="p">,</span> <span class="n">dpsi_col</span> <span class="o">=</span> <span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">min_dPSI</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gene_sets</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;KEGG_2021_Human&#39;</span><span class="p">,</span> <span class="s1">&#39;GO_Biological_Process_2023&#39;</span><span class="p">,</span> <span class="s1">&#39;GO_Cellular_Component_2023&#39;</span><span class="p">,</span> <span class="s1">&#39;GO_Molecular_Function_2023&#39;</span><span class="p">,</span><span class="s1">&#39;Reactome_2022&#39;</span><span class="p">],</span> <span class="n">background</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">return_sig_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">max_retries</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">delay</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced_ptms and/or altered_flanks dataframes (or the dataframes combined from combine_outputs()), perform gene set enrichment analysis using the enrichr API</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with differentially included PTMs projected onto splicing events and with annotations appended from various databases. Default is None (will not be considered in analysis). If combined dataframe is provided, this dataframe will be ignored. </span>
<span class="sd">    altered_flanks: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases. Default is None (will not be considered). If combined dataframe is provided, this dataframe will be ignored.</span>
<span class="sd">    combined: pd.DataFrame</span>
<span class="sd">        Combined dataframe with spliced_ptms and altered_flanks dataframes. Default is None. If provided, spliced_ptms and altered_flanks dataframes will be ignored.</span>
<span class="sd">    gene_sets: list</span>
<span class="sd">        List of gene sets to use in enrichment analysis. Default is [&#39;KEGG_2021_Human&#39;, &#39;GO_Biological_Process_2023&#39;, &#39;GO_Cellular_Component_2023&#39;, &#39;GO_Molecular_Function_2023&#39;,&#39;Reactome_2022&#39;]. Look at gseapy and enrichr documentation for other available gene sets</span>
<span class="sd">    background: list</span>
<span class="sd">        List of genes to use as background in enrichment analysis. Default is None (all genes in the gene set database will be used).</span>
<span class="sd">    return_sig_only: bool</span>
<span class="sd">        Whether to return only significantly enriched gene sets. Default is True.</span>
<span class="sd">    max_retries: int</span>
<span class="sd">        Number of times to retry downloading gene set enrichment data from enrichr API. Default is 5.</span>
<span class="sd">    delay: int</span>
<span class="sd">        Number of seconds to wait between retries. Default is 10.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    results: pd.DataFrame</span>
<span class="sd">        Dataframe with gene set enrichment results from enrichr API</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">background</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Background data not supported at this time, but will be added in future versions. Please set background = None, which will use all genes in the gene set database as the background.&#39;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">combined</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;If combined dataframe is provided, you do not need to include spliced_ptms or altered_flanks dataframes. Ignoring these inputs.&#39;</span><span class="p">)</span>

        <span class="n">foreground</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;Differentially Included + Altered Flanking Sequences&#39;</span>

        <span class="c1">#isolate the type of impact on the gene</span>
        <span class="n">combined_on_gene</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
        <span class="n">included</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Included&#39;</span><span class="p">)</span>
        <span class="n">excluded</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>
        <span class="n">differential</span> <span class="o">=</span> <span class="n">included</span> <span class="o">|</span> <span class="n">excluded</span>
        <span class="n">altered_flank</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Altered Flank&#39;</span><span class="p">)</span>

        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="n">altered_flank</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">differential</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="n">differential</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">altered_flank</span>
        <span class="n">both</span> <span class="o">=</span> <span class="n">differential</span> <span class="o">&amp;</span> <span class="n">altered_flank</span>
        
        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">altered_flank_only</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">differential_only</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">both</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">both</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1">#gene information (total and spliced genes)</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">)</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;Differentially Included + Altered Flanking Sequences&#39;</span>

        <span class="c1">#isolate the type of impact on the gene</span>
        <span class="n">combined_on_gene</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
        <span class="n">included</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Included&#39;</span><span class="p">)</span>
        <span class="n">excluded</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>
        <span class="n">differential</span> <span class="o">=</span> <span class="n">included</span> <span class="o">|</span> <span class="n">excluded</span>
        <span class="n">altered_flank</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Altered Flank&#39;</span><span class="p">)</span>

        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="n">altered_flank</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">differential</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="n">differential</span> <span class="o">&amp;</span> <span class="o">~</span><span class="n">altered_flank</span>
        <span class="n">both</span> <span class="o">=</span> <span class="n">differential</span> <span class="o">&amp;</span> <span class="n">altered_flank</span>

        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">altered_flank_only</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">differential_only</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">both</span> <span class="o">=</span> <span class="n">combined_on_gene</span><span class="p">[</span><span class="n">both</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;Differentially Included&#39;</span>

        <span class="c1">#isolate the type of impact on the gene</span>
        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">both</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">elif</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;Altered Flanking Sequences&#39;</span>

        <span class="c1">#isolate the type of impact on the gene</span>
        <span class="n">altered_flank_only</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="n">differential_only</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">both</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;No dataframes provided. Please provide spliced_ptms, altered_flanks, or the combined dataframe.&#39;</span><span class="p">)</span>
    
    <span class="c1">#restrict to significant ptms, if available</span>
    <span class="k">if</span> <span class="n">sig_col</span> <span class="ow">in</span> <span class="n">foreground</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="p">(</span><span class="n">min_dPSI</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">dpsi_col</span> <span class="ow">in</span> <span class="n">foreground</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">foreground</span><span class="p">[</span><span class="n">foreground</span><span class="p">[</span><span class="n">sig_col</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">alpha</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">foreground</span><span class="p">[</span><span class="n">foreground</span><span class="p">[</span><span class="n">dpsi_col</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">min_dPSI</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">sig_col</span> <span class="ow">in</span> <span class="n">foreground</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">min_dPSI</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;`min_dpsi` provided but `dpsi_col` not found in dataframe. Ignoring `min_dpsi` parameter. Please indicate correct column name for dPSI values and rerun if desired.&#39;</span><span class="p">)</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">foreground</span><span class="p">[</span><span class="n">foreground</span><span class="p">[</span><span class="n">sig_col</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">alpha</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">min_dPSI</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;`sig_col` not found in dataframe. Ignoring `alpha` parameter. Please indicate correct column name for Significance and rerun if desired.&#39;</span><span class="p">)</span>
        <span class="n">foreground</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="n">combined</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span> <span class="o">&gt;=</span> <span class="n">min_dPSI</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Significance column not found and min_dPSI not provided. All PTMs in dataframe will be considered as the foreground&#39;</span><span class="p">)</span>

    <span class="n">foreground</span> <span class="o">=</span> <span class="n">foreground</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>   

    <span class="c1">#construct background</span>
    <span class="c1">#if isinstance(background, list):</span>
    <span class="c1">#    pass</span>
    <span class="c1">#elif isinstance(background, np.ndarray):</span>
    <span class="c1">#    background = list(background)</span>
    <span class="c1">#elif isinstance(background, pd.DataFrame):</span>
    <span class="c1">#    if &#39;Gene&#39; not in background.columns:</span>
    <span class="c1">#        raise ValueError(&#39;Gene column not found in dataframe. Please provide a valid column name for genes.&#39;)</span>
    <span class="c1">#    background = background[&#39;Gene&#39;].unique().tolist()</span>
    <span class="c1">#elif background == &#39;Significance&#39;:</span>
    <span class="c1">#    if sig_col not in foreground.columns:</span>
    <span class="c1">#        raise ValueError(&#39;Significance column not found in dataframe. Please provide a valid column name for significance.&#39;)</span>
    <span class="c1">#    background = combined.copy()</span>
    <span class="c1">#    background = background[&#39;Gene&#39;].unique().tolist()  </span>

    

    
    <span class="c1">#perform gene set enrichment analysis and save data</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">enr</span> <span class="o">=</span> <span class="n">gp</span><span class="o">.</span><span class="n">enrichr</span><span class="p">(</span><span class="n">foreground</span><span class="p">,</span> <span class="n">gene_sets</span> <span class="o">=</span> <span class="n">gene_sets</span><span class="p">)</span>
            <span class="k">break</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span> 
            <span class="n">enrichr_error</span> <span class="o">=</span> <span class="n">e</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">delay</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Failed to run enrichr analysis after &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">max_retries</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39; attempts. Please try again later. Error given by EnrichR: &#39;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">enrichr_error</span><span class="p">))</span>
    
    <span class="n">results</span> <span class="o">=</span> <span class="n">enr</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">type</span>

    <span class="c1">#indicate the genes in each gene set associated with each type of impact</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes with Differentially Included PTMs only&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">differential_only</span><span class="p">))))</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes with PTM with Altered Flanking Sequence only&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">altered_flank_only</span><span class="p">))))</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes with Both&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">both</span><span class="p">))))</span>


    <span class="k">if</span> <span class="n">return_sig_only</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">results</span><span class="p">[</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;Adjusted P-value&#39;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="mf">0.05</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">results</span></div>

    
<span class="k">def</span> <span class="nf">compare_flanking_sequences</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">):</span>
    <span class="n">sequence_identity_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">altered_positions_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">residue_change_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">flank_side_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1">#if there is sequence info for both and does not introduce stop codons, compare sequence identity</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">]:</span>
            <span class="c1">#compare sequence identity</span>
            <span class="n">sequence_identity</span> <span class="o">=</span> <span class="n">getSequenceIdentity</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">])</span>
            <span class="c1">#identify where flanking sequence changes</span>
            <span class="n">altered_positions</span><span class="p">,</span> <span class="n">residue_change</span><span class="p">,</span> <span class="n">flank_side</span> <span class="o">=</span> <span class="n">findAlteredPositions</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="n">flank_size</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sequence_identity</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">altered_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">residue_change</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">flank_side</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>



        <span class="c1">#add to lists</span>
        <span class="n">sequence_identity_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence_identity</span><span class="p">)</span>
        <span class="n">altered_positions_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">altered_positions</span><span class="p">)</span>
        <span class="n">residue_change_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">residue_change</span><span class="p">)</span>
        <span class="n">flank_side_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">flank_side</span><span class="p">)</span>

    <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Sequence Identity&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sequence_identity_list</span>
    <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Altered Positions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">altered_positions_list</span>
    <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Residue Change&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">residue_change_list</span>
    <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Altered Flank Side&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">flank_side_list</span>
    <span class="k">return</span> <span class="n">altered_flanks</span>



<div class="viewcode-block" id="compare_inclusion_motifs">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.compare_inclusion_motifs">[docs]</a>
<span class="k">def</span> <span class="nf">compare_inclusion_motifs</span><span class="p">(</span><span class="n">flanking_sequences</span><span class="p">,</span> <span class="n">elm_classes</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a DataFrame containing flanking sequences with changes and a DataFrame containing ELM class information, identify motifs that are found in the inclusion and exclusion events, identifying motifs unique to each case. This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (http://elm.eu.org/elms/elms_index.tsv).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    flanking_sequences: pandas.DataFrame</span>
<span class="sd">        DataFrame containing flanking sequences with changes, obtained from get_flanking_changes_from_splice_data()</span>
<span class="sd">    elm_classes: pandas.DataFrame</span>
<span class="sd">        DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (http://elm.eu.org/elms/elms_index.tsv). Recommended to download this file and input it manually, but will download from ELM otherwise</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    flanking_sequences: pandas.DataFrame</span>
<span class="sd">        DataFrame containing flanking sequences with changes and motifs found in the inclusion and exclusion events</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">elm_classes</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://elm.eu.org/elms/elms_index.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>

        

    <span class="n">only_in_inclusion</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">only_in_exclusion</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">flanking_sequences</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1">#check if there is a stop codon introduced and both flanking sequences are present</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">]:</span>
            <span class="c1">#get elm motifs that match inclusion or Exclusion Flanking Sequences</span>
            <span class="n">inclusion_matches</span> <span class="o">=</span> <span class="n">find_motifs</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">elm_classes</span><span class="p">)</span>
            <span class="n">exclusion_matches</span> <span class="o">=</span> <span class="n">find_motifs</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">elm_classes</span><span class="p">)</span>

            <span class="c1">#get motifs that are unique to each case</span>
            <span class="n">only_in_inclusion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">inclusion_matches</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">exclusion_matches</span><span class="p">)))</span>
            <span class="n">only_in_exclusion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">exclusion_matches</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">inclusion_matches</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">only_in_inclusion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="n">only_in_exclusion</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="c1">#save data</span>
    <span class="n">flanking_sequences</span><span class="p">[</span><span class="s2">&quot;Motif only in Inclusion&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">only_in_inclusion</span>
    <span class="n">flanking_sequences</span><span class="p">[</span><span class="s2">&quot;Motif only in Exclusion&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">only_in_exclusion</span>
    <span class="k">return</span> <span class="n">flanking_sequences</span></div>


<span class="k">def</span> <span class="nf">identify_change_to_specific_motif</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">elm_motif_name</span><span class="p">,</span> <span class="n">elm_classes</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">modification_class</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">residues</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">dPSI_col</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="k">if</span> <span class="s1">&#39;Altered Positions&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">compare_flanking_sequences</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">)</span>
    
    <span class="c1">#grab elm motifs that match inclusion or Exclusion Flanking Sequences</span>
    <span class="k">if</span> <span class="s1">&#39;Motif only in Inclusion&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">compare_inclusion_motifs</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">elm_classes</span><span class="p">)</span>

    <span class="c1">#grab only needed info</span>
    <span class="n">motif_data</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">cols_to_keep</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Positions&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue Change&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">dPSI_col</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cols_to_keep</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">dPSI_col</span><span class="p">)</span>

    <span class="c1">#go through motif data and identify motifs matching elm motif of interest</span>
    <span class="n">motif_data</span> <span class="o">=</span> <span class="n">motif_data</span><span class="p">[</span><span class="n">cols_to_keep</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">motif_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">elm_motif_name</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]:</span>
                <span class="n">motif_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">motif</span> <span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">elm_motif_name</span> <span class="ow">in</span> <span class="n">motif</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">motif_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">elm_motif_name</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]:</span>
                <span class="n">motif_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">motif</span> <span class="k">for</span> <span class="n">motif</span> <span class="ow">in</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">elm_motif_name</span> <span class="ow">in</span> <span class="n">motif</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">motif_data</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

    <span class="c1">#restrict to events that are specific modification types or residues (for example, SH2 domain motifs should be phosphotyrosine)</span>
    <span class="n">motif_data</span> <span class="o">=</span> <span class="n">motif_data</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">],</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">modification_class</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">motif_data</span> <span class="o">=</span> <span class="n">motif_data</span><span class="p">[</span><span class="n">motif_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">modification_class</span><span class="p">)]</span>

    <span class="k">if</span> <span class="n">residues</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">residues</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">motif_data</span> <span class="o">=</span> <span class="n">motif_data</span><span class="p">[</span><span class="n">motif_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">residues</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">residues</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">residues</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="n">motif_data</span> <span class="o">=</span> <span class="n">motif_data</span><span class="p">[</span><span class="n">motif_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">residues</span><span class="p">)]</span>
    <span class="k">elif</span> <span class="n">residues</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;residues parameter must be a string or list of strings&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">motif_data</span>

    



<div class="viewcode-block" id="findAlteredPositions">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.findAlteredPositions">[docs]</a>
<span class="k">def</span> <span class="nf">findAlteredPositions</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">,</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given two sequences, identify the location of positions that have changed</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq1, seq2: str</span>
<span class="sd">        sequences to compare (order does not matter)</span>
<span class="sd">    flank_size: int</span>
<span class="sd">        size of the flanking sequences (default is 5). This is used to make sure the provided sequences are the correct length</span>
<span class="sd">    </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    altered_positions: list</span>
<span class="sd">        list of positions that have changed</span>
<span class="sd">    residue_change: list</span>
<span class="sd">        list of residues that have changed associated with that position</span>
<span class="sd">    flank_side: str</span>
<span class="sd">        indicates which side of the flanking sequence the change has occurred (N-term, C-term, or Both)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">desired_seq_size</span> <span class="o">=</span> <span class="n">flank_size</span><span class="o">*</span><span class="mi">2</span><span class="o">+</span><span class="mi">1</span>
    <span class="n">altered_positions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">residue_change</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">flank_side</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">seq_size</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
    <span class="n">flank_size</span> <span class="o">=</span> <span class="p">(</span><span class="n">seq_size</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
    <span class="k">if</span> <span class="n">seq_size</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span> <span class="ow">and</span> <span class="n">seq_size</span> <span class="o">==</span> <span class="n">desired_seq_size</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">seq_size</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">seq1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">!=</span> <span class="n">seq2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                <span class="n">altered_positions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="o">-</span><span class="p">(</span><span class="n">flank_size</span><span class="p">))</span>
                <span class="n">residue_change</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">seq1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">-&gt;</span><span class="si">{</span><span class="n">seq2</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1">#check to see which side flanking sequence</span>
        <span class="n">altered_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">altered_positions</span><span class="p">)</span>
        <span class="n">n_term</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="n">altered_positions</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">c_term</span> <span class="o">=</span> <span class="nb">any</span><span class="p">(</span><span class="n">altered_positions</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">n_term</span> <span class="ow">and</span> <span class="n">c_term</span><span class="p">:</span>
            <span class="n">flank_side</span> <span class="o">=</span> <span class="s1">&#39;Both&#39;</span>
        <span class="k">elif</span> <span class="n">n_term</span><span class="p">:</span>
            <span class="n">flank_side</span> <span class="o">=</span> <span class="s1">&#39;N-term only&#39;</span>
        <span class="k">elif</span> <span class="n">c_term</span><span class="p">:</span>
            <span class="n">flank_side</span> <span class="o">=</span> <span class="s1">&#39;C-term only&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">flank_side</span> <span class="o">=</span> <span class="s1">&#39;Unclear&#39;</span>
        <span class="k">return</span> <span class="n">altered_positions</span><span class="p">,</span> <span class="n">residue_change</span><span class="p">,</span> <span class="n">flank_side</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span></div>

        
<div class="viewcode-block" id="getSequenceIdentity">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.getSequenceIdentity">[docs]</a>
<span class="k">def</span> <span class="nf">getSequenceIdentity</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given two flanking sequences, calculate the sequence identity between them using Biopython and parameters definded by Pillman et al. BMC Bioinformatics 2011</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq1, seq2: str</span>
<span class="sd">        flanking sequence </span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    normalized_score: float</span>
<span class="sd">        normalized score of sequence similarity between flanking sequences (calculated similarity/max possible similarity)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#make pairwise aligner object</span>
    <span class="n">aligner</span> <span class="o">=</span> <span class="n">PairwiseAligner</span><span class="p">()</span>
    <span class="c1">#set parameters, with match score of 10 and mismatch score of -2</span>
    <span class="n">aligner</span><span class="o">.</span><span class="n">mode</span> <span class="o">=</span> <span class="s1">&#39;global&#39;</span>
    <span class="n">aligner</span><span class="o">.</span><span class="n">match_score</span> <span class="o">=</span> <span class="mi">10</span>
    <span class="n">aligner</span><span class="o">.</span><span class="n">mismatch_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">2</span>
    <span class="c1">#calculate sequence alignment score between two sequences</span>
    <span class="n">actual_similarity</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>
    <span class="c1">#calculate sequence alignment score between the same sequence</span>
    <span class="n">control_similarity</span> <span class="o">=</span> <span class="n">aligner</span><span class="o">.</span><span class="n">align</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span> <span class="n">seq1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">score</span>
    <span class="c1">#normalize score</span>
    <span class="n">normalized_score</span> <span class="o">=</span> <span class="n">actual_similarity</span><span class="o">/</span><span class="n">control_similarity</span>
    <span class="k">return</span> <span class="n">normalized_score</span></div>


<div class="viewcode-block" id="find_motifs">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.find_motifs">[docs]</a>
<span class="k">def</span> <span class="nf">find_motifs</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">elm_classes</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a sequence and a dataframe containinn ELM class information, identify motifs that can be found in the provided sequence using the RegEx expression provided by ELM (PTMs not considered). This does not take into account the position of the motif in the sequence or additional information that might validate any potential interaction (i.e. structural information that would indicate whether the motif is accessible or not). ELM class information can be downloaded from the download page of elm (http://elm.eu.org/elms/elms_index.tsv).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq: str</span>
<span class="sd">        sequence to search for motifs</span>
<span class="sd">    elm_classes: pandas.DataFrame</span>
<span class="sd">        DataFrame containing ELM class information (ELMIdentifier, Regex, etc.), downloaded directly from ELM (http://elm.eu.org/elms/elms_index.tsv)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">matches</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">elm_row</span> <span class="ow">in</span> <span class="n">elm_classes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">reg_ex</span> <span class="o">=</span> <span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;Regex&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">reg_ex</span><span class="p">,</span> <span class="n">seq</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;ELMIdentifier&#39;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">matches</span></div>

    

<div class="viewcode-block" id="protein_interactions">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions">[docs]</a>
<span class="k">class</span> <span class="nc">protein_interactions</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class to assess interactions facilitated by PTMs in splicing network</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</span>
<span class="sd">    include_enzyme_interactions: bool</span>
<span class="sd">        Whether to include interactions with enzymes in the network, such as kinase-substrate interactions. Default is True</span>
<span class="sd">    interaction_databases: list</span>
<span class="sd">        List of databases whose information to include in the network. Default is [&#39;PhosphoSitePlus&#39;, &#39;PTMcode&#39;, &#39;PTMInt&#39;, &#39;RegPhos&#39;, &#39;DEPOD&#39;, &#39;ELM&#39;]</span>

<span class="sd">    Attributes</span>
<span class="sd">    ----------</span>
<span class="sd">    interaction_graph: nx.Graph</span>
<span class="sd">        NetworkX graph object representing the interaction network, created from analyze.get_interaction_network</span>
<span class="sd">    network_data: pd.DataFrame</span>
<span class="sd">        Dataframe containing details about specifici protein interactions (including which protein contains the spliced PTMs)</span>
<span class="sd">    network_stats: pd.DataFrame</span>
<span class="sd">        Dataframe containing network statistics for each protein in the interaction network, obtained from analyze.get_interaction_stats(). Default is None, which will not label any proteins in the network.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">include_enzyme_interactions</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">interaction_databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM&#39;</span><span class="p">]):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">include_enzyme_interactions</span> <span class="o">=</span> <span class="n">include_enzyme_interactions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_databases</span> <span class="o">=</span> <span class="n">interaction_databases</span>


<div class="viewcode-block" id="protein_interactions.get_interaction_network">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions.get_interaction_network">[docs]</a>
    <span class="k">def</span> <span class="nf">get_interaction_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">node_type</span> <span class="o">=</span> <span class="s1">&#39;Gene&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given the spliced PTM data, extract interaction information and construct a dataframe containing all possible interactions driven by PTMs, either centered around specific PTMs or specific genes.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        node_type: str</span>
<span class="sd">            What to define interactions by. Can either be by &#39;PTM&#39;, which will consider each PTM as a separate node, or by &#39;Gene&#39;, which will aggregate information across all PTMs of a single gene into a single node. Default is &#39;Gene&#39;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">node_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;node_type parameter (which dictates whether to consider interactions at PTM or gene level) can be either Gene or PTM&quot;</span><span class="p">)</span>
        
        <span class="c1">#extract interaction information in provided data</span>
        <span class="n">interactions</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">combine_interaction_data</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">include_enzyme_interactions</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">include_enzyme_interactions</span><span class="p">,</span> <span class="n">interaction_databases</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_databases</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">interactions</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;No interaction data found in spliced PTM data. Please provide spliced PTM data with interaction information to generate network. This can be done by using the annotate module&#39;</span><span class="p">)</span>
        <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">interactions</span> <span class="o">=</span> <span class="n">interactions</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span>

        <span class="c1">#add regulation change information</span>
        <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Regulation Change&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interactions</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;+&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;DISRUPTS&#39;</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;+&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;DISRUPTS&#39;</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">grouping_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Source&#39;</span><span class="p">,</span> <span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="s1">&#39;Regulation Change&#39;</span><span class="p">]</span>
            <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">grouping_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Source&#39;</span><span class="p">]</span>

        <span class="c1">#extract gene_specific network information</span>
        <span class="k">if</span> <span class="n">node_type</span> <span class="o">==</span> <span class="s1">&#39;Gene&#39;</span><span class="p">:</span>
            <span class="n">network_data</span> <span class="o">=</span> <span class="n">interactions</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="n">grouping_cols</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">helpers</span><span class="o">.</span><span class="n">join_unique_entries</span><span class="p">)</span>
            <span class="c1">#generate network with all possible PTM-associated interactions</span>
            <span class="n">interaction_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_pandas_edgelist</span><span class="p">(</span><span class="n">network_data</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="s1">&#39;Modified Gene&#39;</span><span class="p">,</span> <span class="n">target</span> <span class="o">=</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Spliced PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">interactions</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span>
            <span class="n">network_data</span> <span class="o">=</span> <span class="n">interactions</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;Spliced PTM&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="n">grouping_cols</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">helpers</span><span class="o">.</span><span class="n">join_unique_entries</span><span class="p">)</span>
            <span class="n">network_data</span> <span class="o">=</span> <span class="n">network_data</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">])</span>
            
            <span class="c1">#generate network with all possible PTM-associated interactions</span>
            <span class="n">interaction_graph</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_pandas_edgelist</span><span class="p">(</span><span class="n">network_data</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="s1">&#39;Spliced PTM&#39;</span><span class="p">,</span> <span class="n">target</span> <span class="o">=</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span> <span class="o">=</span> <span class="n">network_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span> <span class="o">=</span> <span class="n">interaction_graph</span></div>



<div class="viewcode-block" id="protein_interactions.get_interaction_stats">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions.get_interaction_stats">[docs]</a>
    <span class="k">def</span> <span class="nf">get_interaction_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given the networkx interaction graph, calculate various network centrality measures to identify the most relevant PTMs or genes in the network</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#calculate network centrality measures</span>
        <span class="n">degree_centrality</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">degree_centrality</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span><span class="p">)</span>
        <span class="n">closeness_centrality</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">closeness_centrality</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span><span class="p">)</span>
        <span class="n">betweenness_centrality</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">betweenness_centrality</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span><span class="p">)</span>
        <span class="n">network_stats</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Degree&#39;</span><span class="p">:</span> <span class="nb">dict</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span><span class="o">.</span><span class="n">degree</span><span class="p">()),</span> <span class="s1">&#39;Degree Centrality&#39;</span><span class="p">:</span><span class="n">degree_centrality</span><span class="p">,</span> <span class="s1">&#39;Closeness&#39;</span><span class="p">:</span><span class="n">closeness_centrality</span><span class="p">,</span><span class="s1">&#39;Betweenness&#39;</span><span class="p">:</span><span class="n">betweenness_centrality</span><span class="p">})</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span> <span class="o">=</span> <span class="n">network_stats</span></div>


<div class="viewcode-block" id="protein_interactions.get_protein_interaction_network">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions.get_protein_interaction_network">[docs]</a>
    <span class="k">def</span> <span class="nf">get_protein_interaction_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a specific protein, return the network data for that protein</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        protein: str</span>
<span class="sd">            Gene name of the protein of interest</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        protein_network: pd.DataFrame</span>
<span class="sd">            Dataframe containing network data for the protein of interest</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;network_data&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_network</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">protein</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">protein</span><span class="si">}</span><span class="s1"> is not found in the network data. Please provide a valid gene name.&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
        
        <span class="n">protein_network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">protein</span><span class="p">]</span>
        <span class="n">protein_network</span> <span class="o">=</span> <span class="n">protein_network</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">])</span>
        <span class="n">protein_network</span> <span class="o">=</span> <span class="n">protein_network</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Residue&#39;</span><span class="p">:</span> <span class="s1">&#39;Spliced PTMs facilitating Interacting&#39;</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">protein_network</span></div>


<div class="viewcode-block" id="protein_interactions.summarize_protein_network">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions.summarize_protein_network">[docs]</a>
    <span class="k">def</span> <span class="nf">summarize_protein_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a protein of interest, summarize the network data for that protein</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        protein: str</span>
<span class="sd">            Gene name of the protein of interest</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;network_data&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_network</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;network_stats&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_stats</span><span class="p">()</span>

        <span class="n">protein_network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">protein</span><span class="p">]</span>
        <span class="n">increased_interactions</span> <span class="o">=</span> <span class="n">protein_network</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_network</span><span class="p">[</span><span class="s1">&#39;Regulation Change&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;+&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">decreased_interactions</span> <span class="o">=</span> <span class="n">protein_network</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_network</span><span class="p">[</span><span class="s1">&#39;Regulation Change&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
        <span class="n">ambiguous_interactions</span> <span class="o">=</span> <span class="n">protein_network</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_network</span><span class="p">[</span><span class="s1">&#39;Regulation Change&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">),</span> <span class="s1">&#39;Interacting Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

        <span class="c1">#print interactions</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">increased_interactions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Increased interaction likelihoods: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">increased_interactions</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">decreased_interactions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Decreased interaction likelihoods: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">decreased_interactions</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ambiguous_interactions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Ambiguous interaction impact: </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ambiguous_interactions</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">network_ranks</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="o">.</span><span class="n">rank</span><span class="p">(</span><span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Number of interactions: </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Degree&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> (Rank: </span><span class="si">{</span><span class="n">network_ranks</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Degree&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Centrality measures - </span><span class="se">\t</span><span class="s1"> Degree = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Degree Centrality&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> (Rank: </span><span class="si">{</span><span class="n">network_ranks</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Degree Centrality&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;                      </span><span class="se">\t</span><span class="s1"> Betweenness = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Betweenness&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> (Rank: </span><span class="si">{</span><span class="n">network_ranks</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Betweenness&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;                      </span><span class="se">\t</span><span class="s1"> Closeness = </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Closeness&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1"> (Rank: </span><span class="si">{</span><span class="n">network_ranks</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein</span><span class="p">,</span><span class="w"> </span><span class="s2">&quot;Closeness&quot;</span><span class="p">]</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="protein_interactions.plot_interaction_network">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.protein_interactions.plot_interaction_network">[docs]</a>
    <span class="k">def</span> <span class="nf">plot_interaction_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">modified_color</span> <span class="o">=</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">modified_node_size</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">interacting_color</span> <span class="o">=</span> <span class="s1">&#39;lightblue&#39;</span><span class="p">,</span> <span class="n">interacting_node_size</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">defaultedgecolor</span> <span class="o">=</span> <span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">color_edges_by</span> <span class="o">=</span> <span class="s1">&#39;Same&#39;</span><span class="p">,</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">proteins_to_label</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">labelcolor</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given the interactiong graph and network data outputted from analyze.get_interaction_network, plot the interaction network, signifying which proteins or ptms are altered by splicing and the specific regulation change that occurs. by default, will only label proteins </span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        modified_color: str</span>
<span class="sd">            Color to use for nodes that are modified by splicing. Default is &#39;red&#39;</span>
<span class="sd">        modified_node_size: int</span>
<span class="sd">            Size of nodes that are modified by splicing. Default is 10</span>
<span class="sd">        interacting_color: str</span>
<span class="sd">            Color to use for nodes that interact with modified nodes. Default is &#39;lightblue&#39;</span>
<span class="sd">        interacting_node_size: int</span>
<span class="sd">            Size of nodes that interact with modified nodes. Default is 5</span>
<span class="sd">        edgecolor: str</span>
<span class="sd">            Color to use for edges in the network. Default is &#39;gray&#39;. Can choose to color by database by providing a dictionary with database names as keys and colors as values or by specifying &#39;database&#39; as color</span>
<span class="sd">        seed: int</span>
<span class="sd">            Seed to use for random number generator. Default is 200</span>
<span class="sd">        ax: matplotlib.pyplot.Axes</span>
<span class="sd">            Axes object to plot the network. Default is None</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;interaction_graph&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_network</span><span class="p">()</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;network_stats&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_stats</span><span class="p">()</span>

        <span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_interaction_network</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">interaction_graph</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="p">,</span> <span class="n">modified_color</span> <span class="o">=</span> <span class="n">modified_color</span><span class="p">,</span> <span class="n">modified_node_size</span> <span class="o">=</span> <span class="n">modified_node_size</span><span class="p">,</span> <span class="n">interacting_color</span> <span class="o">=</span> <span class="n">interacting_color</span><span class="p">,</span> <span class="n">interacting_node_size</span> <span class="o">=</span> <span class="n">interacting_node_size</span><span class="p">,</span> <span class="n">defaultedgecolor</span> <span class="o">=</span> <span class="n">defaultedgecolor</span><span class="p">,</span> <span class="n">color_edges_by</span><span class="o">=</span><span class="n">color_edges_by</span><span class="p">,</span> <span class="n">seed</span> <span class="o">=</span> <span class="n">seed</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">,</span> <span class="n">proteins_to_label</span> <span class="o">=</span> <span class="n">proteins_to_label</span><span class="p">,</span> <span class="n">labelcolor</span> <span class="o">=</span> <span class="n">labelcolor</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">plot_network_centrality</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>  <span class="n">centrality_measure</span> <span class="o">=</span> <span class="s1">&#39;Degree&#39;</span><span class="p">,</span> <span class="n">top_N</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">modified_color</span> <span class="o">=</span> <span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">interacting_color</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;interaction_graph&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_network</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;network_stats&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">get_interaction_stats</span><span class="p">()</span>

        <span class="n">pose_plots</span><span class="o">.</span><span class="n">plot_network_centrality</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">network_stats</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">network_data</span><span class="p">,</span> <span class="n">centrality_measure</span><span class="o">=</span><span class="n">centrality_measure</span><span class="p">,</span><span class="n">top_N</span> <span class="o">=</span> <span class="n">top_N</span><span class="p">,</span> <span class="n">modified_color</span> <span class="o">=</span> <span class="n">modified_color</span><span class="p">,</span> <span class="n">interacting_color</span> <span class="o">=</span> <span class="n">interacting_color</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">)</span></div>


<div class="viewcode-block" id="edit_sequence_for_kinase_library">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.analyze.edit_sequence_for_kinase_library">[docs]</a>
<span class="k">def</span> <span class="nf">edit_sequence_for_kinase_library</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert flanking sequence to version accepted by kinase library (modified residue denoted by asterick)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">seq</span> <span class="o">==</span> <span class="n">seq</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;t&#39;</span><span class="p">,</span><span class="s1">&#39;t*&#39;</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;s&#39;</span><span class="p">,</span><span class="s1">&#39;s*&#39;</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;y&#39;</span><span class="p">,</span><span class="s1">&#39;y*&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
    <span class="k">return</span> <span class="n">seq</span></div>



<span class="k">def</span> <span class="nf">compare_KL_for_sequence</span><span class="p">(</span><span class="n">inclusion_seq</span><span class="p">,</span> <span class="n">exclusion_seq</span><span class="p">,</span> <span class="n">dpsi</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">comparison_type</span> <span class="o">=</span> <span class="s1">&#39;percentile&#39;</span><span class="p">):</span>

    <span class="c1">#get percentiles for inclusion sequence</span>
    <span class="n">s_inc</span> <span class="o">=</span> <span class="n">kl</span><span class="o">.</span><span class="n">Substrate</span><span class="p">(</span><span class="n">inclusion_seq</span><span class="p">)</span>
    <span class="n">inn_perc</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>

    <span class="c1">#get percentiles for exclusion sequence</span>
    <span class="n">s_exc</span> <span class="o">=</span> <span class="n">kl</span><span class="o">.</span><span class="n">Substrate</span><span class="p">(</span><span class="n">exclusion_seq</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;percentile&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;score&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">score</span><span class="p">()</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">score</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;rank&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">rank</span><span class="p">()</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">rank</span><span class="p">()</span>

    <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">inc_results</span><span class="p">,</span> <span class="n">ex_results</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">results</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Exclusion  </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">results</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Absolute Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">dpsi</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dpsi</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Relative Change in Preference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">dpsi</span>

    <span class="k">return</span> <span class="n">results</span>

<span class="k">class</span> <span class="nc">KL_flank_analysis</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">):</span>
        <span class="c1">#check to make sure kinase library is installed</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">kinase_library</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Kinase library package not installed. To use this functionality, please install the kinase library package by running `pip install kinase-library`&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">analyze_single_ptm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">loc</span><span class="p">):</span>
        <span class="n">ptm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">loc</span><span class="p">]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>

        <span class="n">dpsi</span> <span class="o">=</span> <span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">ptm_data</span> <span class="k">else</span> <span class="kc">None</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">compare_KL_for_sequence</span><span class="p">(</span><span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">dpsi</span> <span class="o">=</span> <span class="n">dpsi</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">analyze_all_ptms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">min_MS</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">min_compendia</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
        <span class="k">pass</span>

<span class="c1">#class KL_flank_analysis:</span>
<span class="c1">#    def __init__(self, altered_flanks, odir):</span>
<span class="c1">#        self.altered_flanks = altered_flanks</span>
<span class="c1">#        self.odir = odir</span>

<span class="c1">#    def identify_sequences_of_interest(self):</span>
<span class="c1">#        self.sequences_of_interest = self.altered_flanks[(~self.altered_flanks[&#39;Matched&#39;]) &amp; (~self.altered_flanks[&#39;Stop Codon Introduced&#39;]) &amp; (self.altered_flanks[&#39;Modification Class&#39;].str.contains(&#39;Phosphorylation&#39;))].copy() </span>

    
<span class="c1">#    def process_data_for_kinase_library(self):</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        Extract flanking sequence information for </span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        #restrict to events with changed flanking sequences, no introduced stop codons, and phosphorylation modifications</span>
<span class="c1">#        if not hasattr(self, &#39;sequences_of_interest&#39;):</span>
<span class="c1">#            self.identify_sequences_of_interest()</span>

        <span class="c1">#generate files to input into Kinase Library (inclusion first then exclusion)</span>
<span class="c1">#        inclusion_sequences = self.sequences_of_interest[[&#39;PTM&#39;, &#39;Inclusion Flanking Sequence&#39;]].drop_duplicates()</span>
<span class="c1">#        inclusion_sequences[&#39;Inclusion Flanking Sequence&#39;] = inclusion_sequences[&#39;Inclusion Flanking Sequence&#39;].apply(edit_sequence_for_kinase_library)</span>
<span class="c1">#        inclusion_sequences = inclusion_sequences.dropna(subset = &#39;Inclusion Flanking Sequence&#39;)</span>
        <span class="c1">#write sequences to text file</span>
<span class="c1">#        with open(self.odir + &#39;inclusion_sequences_input.txt&#39;, &#39;w&#39;) as f:</span>
<span class="c1">#            for _, row in inclusion_sequences.iterrows():</span>
<span class="c1">#                f.write(row[&#39;Inclusion Flanking Sequence&#39;]+&#39;\n&#39;)</span>

<span class="c1">#        exclusion_sequences = self.sequences_of_interest[[&#39;PTM&#39;, &#39;Exclusion Flanking Sequence&#39;]].drop_duplicates()</span>
<span class="c1">#        exclusion_sequences[&#39;Exclusion Flanking Sequence&#39;] = exclusion_sequences[&#39;Exclusion Flanking Sequence&#39;].apply(edit_sequence_for_kinase_library)</span>
<span class="c1">#        exclusion_sequences = exclusion_sequences.dropna(subset = &#39;Exclusion Flanking Sequence&#39;)</span>
<span class="c1">#        #write sequences to text file</span>
<span class="c1">#        with open(self.odir + &#39;exclusion_sequences_input.txt&#39;, &#39;w&#39;) as f:</span>
<span class="c1">#            for _, row in exclusion_sequences.iterrows():</span>
<span class="c1">#                f.write(row[&#39;Exclusion Flanking Sequence&#39;]+&#39;\n&#39;)</span>

<span class="c1">#        print(&#39;Input files for Kinase Library generated. Please run upload the file to the &quot;score sites&quot; tab of Kinase Library (https://kinase-library.mit.edu/sites) and download the full results.&#39;)</span>

<span class="c1">#    def format_sequences_to_match_output(self, sequence_type = &#39;Inclusion&#39;):</span>
<span class="c1">#        if not hasattr(self, &#39;sequences_of_interest&#39;):</span>
<span class="c1">#            self.identify_sequences_of_interest()</span>

<span class="c1">#        sequences = self.sequences_of_interest[[&#39;Region ID&#39;,&#39;PTM&#39;, f&#39;{sequence_type} Flanking Sequence&#39;]].drop_duplicates().copy()</span>
<span class="c1">#        sequences = sequences.dropna(subset = &#39;Inclusion Flanking Sequence&#39;)</span>
<span class="c1">#        sequences[&#39;Label&#39;] = sequences[&#39;Region ID&#39;] + &#39;;&#39; + sequences[&#39;PTM&#39;]</span>
<span class="c1">#        sequences[f&#39;{sequence_type} Flanking Sequence&#39;] = sequences[f&#39;{sequence_type} Flanking Sequence&#39;].apply(lambda x: x.upper().replace(&#39; &#39;, &#39;_&#39;)+&#39;_&#39;)</span>
<span class="c1">#        return sequences</span>

<span class="c1">#    def process_kinase_library_output(self, scores, sequence_type = &#39;Inclusion&#39;):</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        Process output from Kinase Library to connect kinase library scores back to the PTMs in the altered flanks dataframe</span>

<span class="c1">#        Parameters</span>
<span class="c1">#        ----------</span>
<span class="c1">#        altered_flanks: pd.DataFrame</span>
<span class="c1">#            Dataframe with PTMs associated with altered flanking sequences</span>
<span class="c1">#        scores: pd.DataFrame</span>
<span class="c1">#            Dataframe with kinase library scores for flanking sequences (loaded #from downloaded .tsv outputs from kinase library)</span>
<span class="c1">#        flanking_sequence_col: str</span>
<span class="c1">#            Column in altered_flanks dataframe that contains the flanking sequence to match with the kinase library scores. Default is &#39;Inclusion Flanking Sequence&#39;. Can also be &#39;Exclusion Flanking Sequence&#39;</span>

<span class="c1">#        Returns</span>
<span class="c1">#        -------</span>
<span class="c1">#        percentiles_y: pd.DataFrame</span>
<span class="c1">#            Dataframe with kinase library scores for tyrosine sites</span>
<span class="c1">#        percentiles_st: pd.DataFrame</span>
<span class="c1">#            Dataframe with kinase library scores for serine/threonine sites</span>

<span class="c1">#        &quot;&quot;&quot;</span>
        <span class="c1">#restrict to events with changed flanking sequences, no introduced stop codons, and phosphorylation modifications</span>
<span class="c1">#        if not hasattr(self, &#39;sequences_of_interest&#39;):</span>
<span class="c1">#            self.identify_sequences_of_interest()</span>

<span class="c1">#        sequences = self.format_sequences_to_match_output(sequence_type = sequence_type)</span>


<span class="c1">#        sequences = sequences.merge(scores, left_on = f&#39;{sequence_type} Flanking Sequence&#39;, right_on = &#39;sequence&#39;, how = &#39;left&#39;)</span>
        <span class="c1">#split info into tyrosine vs. serine/threonine</span>
<span class="c1">#        sequences_y = sequences[sequences[&#39;Label&#39;].str.contains(&#39;_Y&#39;)]</span>
<span class="c1">#        sequences_st = sequences[(sequences[&#39;Label&#39;].str.contains(&#39;_S&#39;)) | (sequences[&#39;Label&#39;].str.contains(&#39;_T&#39;))]</span>

        <span class="c1">#pivot table to get scores for each kinase</span>
<span class="c1">#        percentiles_y = sequences_y.pivot_table(index = &#39;Label&#39;, columns = &#39;kinase&#39;, values = &#39;site_percentile&#39;)</span>
<span class="c1">#        percentiles_st = sequences_st.pivot_table(index = &#39;Label&#39;, columns = &#39;kinase&#39;, values = &#39;site_percentile&#39;)</span>

<span class="c1">#        return percentiles_y, percentiles_st</span>

<span class="c1">#    def get_kinase_library_differences(self, inclusion_scores_file, exclusion_scores_file):</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        Given altered flanking sequences and kinase library scores for inclusion and Exclusion Flanking Sequences, calculate the difference in kinase library site percentiles between the two</span>

<span class="c1">#        Parameters</span>
<span class="c1">#        ----------</span>
<span class="c1">#        altered_flanks: pd.DataFrame</span>
<span class="c1">#            Dataframe with PTMs associated with altered flanking sequences</span>
<span class="c1">#        inclusion_scores: pd.DataFrame</span>
<span class="c1">#            Dataframe with kinase library scores for Inclusion Flanking Sequences (loaded from downloaded .tsv outputs from kinase library)</span>
<span class="c1">#        exclusion_scores: pd.DataFrame</span>
<span class="c1">#            Dataframe with kinase library scores for Exclusion Flanking Sequences (loaded from downloaded .tsv outputs from kinase library)</span>
        
<span class="c1">#        Returns</span>
<span class="c1">#        -------</span>
<span class="c1">#        percentiles_diff_y: pd.DataFrame</span>
<span class="c1">#            Dataframe with the difference in kinase library scores for tyrosine sites</span>
<span class="c1">#        percentiles_diff_st: pd.DataFrame</span>
<span class="c1">#            Dataframe with the difference in kinase library scores for serine/threonine sites</span>
<span class="c1">#        &quot;&quot;&quot;</span>
<span class="c1">#        inclusion_scores = pd.read_csv(inclusion_scores_file, sep = &#39;\t&#39;)</span>
<span class="c1">#        inclusion_percentiles_y, inclusion_percentiles_st = self.process_kinase_library_output(inclusion_scores, sequence_type = &#39;Inclusion&#39;)</span>
<span class="c1">#        exclusion_scores = pd.read_csv(exclusion_scores_file, sep = &#39;\t&#39;)</span>
<span class="c1">#        exclusion_percentiles_y, exclusion_percentiles_st = self.process_kinase_library_output(exclusion_scores, sequence_type = &#39;Exclusion&#39;)</span>

        <span class="c1">#calculate the difference in percentiles</span>
<span class="c1">#        labels= list(set(inclusion_percentiles_y.index).intersection(exclusion_percentiles_y.index))</span>
<span class="c1">#        percentiles_diff_y = inclusion_percentiles_y.loc[labels].copy()</span>
<span class="c1">#        percentiles_diff_y = percentiles_diff_y[exclusion_percentiles_y.columns]</span>
<span class="c1">#        for i, row in percentiles_diff_y.iterrows():</span>
<span class="c1">#            percentiles_diff_y.loc[i] = row - exclusion_percentiles_y.loc[i]</span>

<span class="c1">#        labels= list(set(inclusion_percentiles_st.index).intersection(exclusion_percentiles_st.index))</span>
<span class="c1">#        percentiles_diff_st = inclusion_percentiles_st.loc[labels].copy()</span>
<span class="c1">#        percentiles_diff_st = percentiles_diff_st[exclusion_percentiles_st.#columns]</span>
<span class="c1">#        for i, row in percentiles_diff_st.iterrows():</span>
<span class="c1">#            percentiles_diff_st.loc[i] = row - exclusion_percentiles_st.loc[i]</span>

<span class="c1">#        #save all data</span>
<span class="c1">#        self.inclusion_percentiles = {}</span>
<span class="c1">#        self.inclusion_percentiles[&#39;Y&#39;] = inclusion_percentiles_y</span>
<span class="c1">#        self.inclusion_percentiles[&#39;ST&#39;] = inclusion_percentiles_st</span>

<span class="c1">#        self.exclusion_percentiles = {}</span>
<span class="c1">#        self.exclusion_percentiles[&#39;Y&#39;] = exclusion_percentiles_y</span>
<span class="c1">#        self.exclusion_percentiles[&#39;ST&#39;] = exclusion_percentiles_st</span>

<span class="c1">#        self.percentile_difference = {}</span>
<span class="c1">#        self.percentile_difference[&#39;Y&#39;] = percentiles_diff_y</span>
<span class="c1">#        self.percentile_difference[&#39;ST&#39;] = percentiles_diff_st</span>

    
<span class="c1">#def process_data_for_exon_ontology(odir, spliced_ptms = None, altered_flanks = None):</span>
<span class="c1">#    pass</span>



    

<span class="k">class</span> <span class="nc">kstar_enrichment</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">significant_ptms</span><span class="p">,</span> <span class="n">network_dir</span><span class="p">,</span> <span class="n">background_ptms</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given spliced ptm or PTMs with altered flanks and a single kstar network, get enrichment for each kinase in the network using a hypergeometric. Assumes the  data has already been reduced to the modification of interest (phosphotyrosine or phoshoserine/threonine)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        network_dir : dict</span>
<span class="sd">            dictionary of networks with kinase-substrate information</span>
<span class="sd">        spliced_ptms : pandas dataframe</span>
<span class="sd">            all PTMs of interest</span>
<span class="sd">        background_ptms: pd.DataFrame</span>
<span class="sd">            PTMs to consider as the background for enrichment purposes, which should overlap with the spliced ptms information provided (an example might be all identified events, whether or not they are significant). If not provided, will use all ptms in the phosphoproteome.</span>
<span class="sd">        phospho_type : str </span>
<span class="sd">            type of phosphorylation event to extract. Can either by phosphotyrosine (&#39;Y&#39;) or phosphoserine/threonine (&#39;ST&#39;). Default is &#39;Y&#39;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#change phospho_type to list if not already</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">phospho_type</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;ST&#39;</span><span class="p">]:</span>
            <span class="n">phospho_type</span> <span class="o">=</span> <span class="p">[</span><span class="n">phospho_type</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
            <span class="n">phospho_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;ST&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;phospho_type must be either Y, ST, or All&#39;</span><span class="p">)</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">networks</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="n">phospho_type</span><span class="p">:</span>
            <span class="c1">#process ptms to only include specific phosphorylation data needed</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">significant_ptms</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">)</span>
            <span class="c1">#load background, either from provided data or from all phosphoproteome data</span>
            <span class="k">if</span> <span class="n">background_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">background_ptms</span><span class="p">,</span> <span class="n">phospho_type</span><span class="o">=</span><span class="n">ptype</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">)</span>

            <span class="c1">#load kstar networks</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_networks</span><span class="p">(</span><span class="n">network_dir</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">)</span>


        <span class="c1">#save info</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span> <span class="o">=</span> <span class="n">phospho_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span> <span class="nf">load_networks</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">network_dir</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given network directory, load all kstar networks for specific phosphorylation type</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#check if file exists and whether a pickle has been generated: if not, load each network file individually</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">network_dir</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Network directory not found&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_dir</span><span class="si">}</span><span class="s2">/*.p&quot;</span><span class="p">):</span>
            <span class="n">networks</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_dir</span><span class="si">}</span><span class="s2">/network_</span><span class="si">{</span><span class="n">phospho_type</span><span class="si">}</span><span class="s2">.p&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span> <span class="p">)</span> <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">network_directory</span> <span class="o">=</span> <span class="n">network_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/</span><span class="si">{</span><span class="n">phospho_type</span><span class="si">}</span><span class="s1">/INDIVIDUAL_NETWORKS/&#39;</span>
            <span class="n">networks</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">network_directory</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.tsv&#39;</span><span class="p">):</span>
                    <span class="c1">#get the value of the network number</span>
                    <span class="n">file_noext</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
                    <span class="n">key_name</span> <span class="o">=</span> <span class="s1">&#39;nkin&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">file_noext</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                    <span class="c1">#print(&quot;Debug: key name is %s&quot;%(key_name))</span>
                    <span class="n">networks</span><span class="p">[</span><span class="n">key_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_directory</span><span class="si">}{</span><span class="n">file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">networks</span>


    <span class="k">def</span> <span class="nf">process_ptms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptms</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given ptm information, restrict data to include only the phosphorylation type of interest and add a PTM column for matching information from KSTAR</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptms: pd.DataFrame</span>
<span class="sd">            ptm information containing modification type and ptm locatin information, such as the output from projection or altered flanking sequence analysis</span>
<span class="sd">        phospho_type: str</span>
<span class="sd">            type of phosphorylation event to extract. Can either by phosphotyrosine (&#39;Y&#39;) or phosphoserine/threonine (&#39;ST&#39;)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        ptms: pd.DataFrame</span>
<span class="sd">            trimmed dataframe containing only modifications of interest and new &#39;PTM&#39; column</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1">#restrict to ptms to phosphorylation type of interest</span>
        <span class="k">if</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Phosphotyrosine&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[(</span><span class="n">ptms</span><span class="p">[</span><span class="s2">&quot;Modification&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Phosphoserine&#39;</span><span class="p">))</span> <span class="o">|</span> <span class="p">(</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Phosphothreonine&#39;</span><span class="p">))]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1">#construct PTM column that matches KSTAR information</span>
        <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>

        <span class="c1">#filter out any PTMs that come from alternative isoforms</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="o">~</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">ptms</span>

    
    <span class="k">def</span> <span class="nf">get_enrichment_single_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">network_key</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        in progress</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="n">network_key</span><span class="p">]</span>
        <span class="n">network</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">network</span><span class="p">[</span><span class="s1">&#39;KSTAR_ACCESSION&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">network</span><span class="p">[</span><span class="s1">&#39;KSTAR_SITE&#39;</span><span class="p">]</span>

        <span class="c1">#add network information to all significant data</span>
        <span class="n">sig_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="n">sig_ptms_kstar</span> <span class="o">=</span> <span class="n">sig_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">network</span><span class="p">[[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">,</span><span class="s1">&#39;PTM&#39;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;PTM&#39;</span><span class="p">)</span>

        <span class="c1">#repeat for background data</span>
        <span class="n">bg_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="n">bg_ptms_kstar</span> <span class="o">=</span> <span class="n">bg_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">network</span><span class="p">[[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">,</span><span class="s1">&#39;PTM&#39;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;PTM&#39;</span><span class="p">)</span>

        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">,</span><span class="s1">&#39;n&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">,</span><span class="s1">&#39;N&#39;</span><span class="p">,</span><span class="s1">&#39;p&#39;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="c1">#get numbers for a hypergeometric test to look for enrichment of kinase substrates</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">n</span> <span class="o">=</span> <span class="n">bg_ptms_kstar</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">bg_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">M</span> <span class="o">=</span> <span class="n">bg_ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">N</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>

            <span class="c1">#run hypergeometric test</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span><span class="s1">&#39;p&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stat_utils</span><span class="o">.</span><span class="n">hypergeom</span><span class="p">(</span><span class="n">M</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="n">k</span><span class="p">)</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">M</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;k&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;n&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n</span>

        <span class="k">return</span> <span class="n">results</span>
    
    <span class="k">def</span> <span class="nf">get_enrichment_all_networks</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given prostate data and a dictionary of kstar networks, get enrichment for each kinase in each network in the prostate data. Assumes the prostate data has already been reduced to the modification of interest (phosphotyrosine or phoshoserine/threonine)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        networks : dict</span>
<span class="sd">            dictionary of kstar networks</span>
<span class="sd">        prostate : pandas dataframe</span>
<span class="sd">            all PTMs identified in tCGA prostate data, regardless of significance (reduced to only include mods of interest)</span>
<span class="sd">        sig_prostate : pandas dataframe</span>
<span class="sd">            significant PTMs identified in tCGA prostate data, p &lt; 0.05 and effect size &gt; 0.25 (reduced to only include mods of interest)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">network</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">]:</span>
            <span class="n">results</span><span class="p">[</span><span class="n">network</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_enrichment_single_network</span><span class="p">(</span><span class="n">network_key</span><span class="o">=</span><span class="n">network</span><span class="p">,</span> <span class="n">phospho_type</span><span class="o">=</span><span class="n">phospho_type</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>

    <span class="k">def</span> <span class="nf">extract_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a dictionary of results from get_enrichment_all_networks, extract the p-values for each network and kinase, and then calculate the median p-value across all networks for each kinase</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        results : dict</span>
<span class="sd">            dictionary of results from get_enrichment_all_networks</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;nkin0&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">network</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">enrichment</span><span class="p">[</span><span class="n">network</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="n">network</span><span class="p">][</span><span class="s1">&#39;p&#39;</span><span class="p">]</span>
        <span class="n">enrichment</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">enrichment</span>
    
    <span class="k">def</span> <span class="nf">run_kstar_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Run full kstar analysis to generate substrate enrichment across each of the 50 KSTAR networks and calculate the median p-value for each kinase across all networks</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">median</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span><span class="p">:</span>
            <span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_enrichment_all_networks</span><span class="p">(</span><span class="n">phospho_type</span><span class="o">=</span><span class="n">ptype</span><span class="p">)</span>
            <span class="n">enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extract_enrichment</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
            <span class="n">median</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="n">enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="s1">&#39;median&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">enrichment_all</span> <span class="o">=</span> <span class="n">enrichment</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="o">=</span> <span class="n">median</span>

    <span class="k">def</span> <span class="nf">return_enriched_kinases</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return kinases with a median p-value less than the provided alpha value (substrates are enriched among the significant PTMs)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        alpha : float</span>
<span class="sd">            significance threshold to use to subset kinases. Default is 0.05.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">run_kstar_enrichment</span><span class="p">()</span>

        <span class="n">sig_enrichment</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span><span class="p">:</span>
            <span class="n">sig_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
        <span class="k">return</span> <span class="n">sig_enrichment</span>
    
<span class="k">class</span> <span class="nc">comprehensive_analysis</span><span class="p">:</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">min_MS_studies</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">min_compendia</span> <span class="o">=</span> <span class="mi">0</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">filter_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">min_MS_studies</span> <span class="o">=</span> <span class="n">min_MS_studies</span><span class="p">,</span> <span class="n">min_compendia</span> <span class="o">=</span> <span class="n">min_compendia</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">filter_ptms</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">min_MS_studies</span> <span class="o">=</span> <span class="n">min_MS_studies</span><span class="p">,</span> <span class="n">min_compendia</span> <span class="o">=</span> <span class="n">min_compendia</span><span class="p">)</span>
    
    


</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>