
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ptm_pose.annotate &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- So that users can add custom icons -->
  <script src="../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/ptm_pose/annotate';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.3.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../gallery_output/index.html">Gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/Networks/plot_protein_interactions.html">Protein interactions network</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../gallery_output/flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../gallery_output/flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for ptm_pose.annotate</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">pose_config</span><span class="p">,</span> <span class="n">helpers</span>

<span class="c1">#try importing omnipath, if not print error message prompting user to install omnipath</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">omnipath</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">omnipath</span> <span class="o">=</span> <span class="kc">None</span>


<span class="c1">#dictionaries for converting modification codes to modification names in PhosphoSitePlus data</span>
<span class="n">mod_shorthand_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;p&#39;</span><span class="p">:</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ca&#39;</span><span class="p">:</span><span class="s1">&#39;Caspase Cleavage&#39;</span><span class="p">,</span> <span class="s1">&#39;hy&#39;</span><span class="p">:</span><span class="s1">&#39;Hydroxylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sn&#39;</span><span class="p">:</span><span class="s1">&#39;S-Nitrosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">:</span><span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ub&#39;</span><span class="p">:</span> <span class="s1">&#39;Ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;pa&#39;</span><span class="p">:</span> <span class="s2">&quot;Palmitoylation&quot;</span><span class="p">,</span><span class="s1">&#39;ne&#39;</span><span class="p">:</span><span class="s1">&#39;Neddylation&#39;</span><span class="p">,</span><span class="s1">&#39;sc&#39;</span><span class="p">:</span><span class="s1">&#39;Succinylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sm&#39;</span><span class="p">:</span> <span class="s1">&#39;Sumoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ga&#39;</span><span class="p">:</span> <span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;gl&#39;</span><span class="p">:</span> <span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ac&#39;</span><span class="p">:</span> <span class="s1">&#39;Acetylation&#39;</span><span class="p">,</span> <span class="s1">&#39;me&#39;</span><span class="p">:</span><span class="s1">&#39;Methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m1&#39;</span><span class="p">:</span><span class="s1">&#39;Methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m2&#39;</span><span class="p">:</span> <span class="s1">&#39;Dimethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m3&#39;</span><span class="p">:</span><span class="s1">&#39;Trimethylation&#39;</span><span class="p">}</span>
<span class="n">residue_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;P&#39;</span><span class="p">:</span> <span class="s1">&#39;proline&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;tyrosine&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">:</span><span class="s1">&#39;serine&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;threonine&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">:</span><span class="s1">&#39;histidine&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">:</span><span class="s1">&#39;aspartic acid&#39;</span><span class="p">,</span> <span class="s1">&#39;I&#39;</span><span class="p">:</span><span class="s1">&#39;isoleucine&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">:</span><span class="s1">&#39;lysine&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">:</span><span class="s1">&#39;arginine&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;glycine&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span><span class="s1">&#39;asparagine&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;methionine&#39;</span><span class="p">}</span>



<div class="viewcode-block" id="get_available_gmt_annotations">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.get_available_gmt_annotations">[docs]</a>
<span class="k">def</span> <span class="nf">get_available_gmt_annotations</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;dict&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Get the annotations available in resource files in GMT format. Can be outputted as either a dictionary or pandas DataFrame</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    format: str</span>
<span class="sd">        Format to output the available annotations. Options are &#39;dict&#39; or &#39;dataframe&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#get available annotations</span>
    <span class="n">annotation_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">)</span>
    <span class="c1">#grab available directories in the annotation directory</span>
    <span class="n">available_databases</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">annotation_dir</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;dict&#39;</span><span class="p">:</span>
        <span class="n">available_annots</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">database</span> <span class="ow">in</span> <span class="n">available_databases</span><span class="p">:</span>
            <span class="c1">#grab available annotation types for each database</span>
            <span class="n">annot_types</span> <span class="o">=</span> <span class="p">[</span><span class="n">annot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">annot</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_dir</span><span class="p">,</span> <span class="n">database</span><span class="p">))]</span>
            <span class="n">available_annots</span><span class="p">[</span><span class="n">database</span><span class="p">]</span> <span class="o">=</span> <span class="n">annot_types</span>
    <span class="k">elif</span> <span class="nb">format</span> <span class="o">==</span> <span class="s1">&#39;dataframe&#39;</span><span class="p">:</span>
        <span class="n">database_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">annot_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">database</span> <span class="ow">in</span> <span class="n">available_databases</span><span class="p">:</span>
            <span class="c1">#grab available annotation types for each database</span>
            <span class="n">annot_types</span> <span class="o">=</span> <span class="p">[</span><span class="n">annot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">annot</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_dir</span><span class="p">,</span> <span class="n">database</span><span class="p">))]</span>
            <span class="k">for</span> <span class="n">annot</span> <span class="ow">in</span> <span class="n">annot_types</span><span class="p">:</span>
                <span class="n">database_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">database</span><span class="p">)</span>
                <span class="n">annot_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annot</span><span class="p">)</span>
        <span class="n">available_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Database&#39;</span><span class="p">:</span> <span class="n">database_list</span><span class="p">,</span> <span class="s1">&#39;Annotation Type&#39;</span><span class="p">:</span> <span class="n">annot_list</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;format must be either &#39;dict&#39; or &#39;dataframe&#39;&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">available_annots</span></div>



<span class="k">def</span> <span class="nf">get_available_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">):</span>
    <span class="n">available_gmt</span> <span class="o">=</span> <span class="n">get_available_gmt_annotations</span><span class="p">(</span><span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;dataframe&#39;</span><span class="p">)</span>
    <span class="n">available_gmt</span><span class="p">[</span><span class="s1">&#39;Appended to PTM data?&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;No&#39;</span>


    <span class="c1">#check to see if any annotations have been added to the spliced_ptms dataframe</span>
    <span class="n">annot_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
    <span class="n">database_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">annot_type_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">annot_cols</span><span class="p">:</span>
        <span class="n">database_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">annot_type_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">available_in_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;Database&#39;</span><span class="p">:</span><span class="n">database_list</span><span class="p">,</span> <span class="s1">&#39;Annotation Type&#39;</span><span class="p">:</span><span class="n">annot_type_list</span><span class="p">,</span> <span class="s1">&#39;Appended to PTM data?&#39;</span><span class="p">:</span> <span class="s1">&#39;Yes&#39;</span><span class="p">})</span>

    <span class="c1">#combine the two dataframes and remove duplicates</span>
    <span class="n">available_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">available_in_df</span><span class="p">,</span> <span class="n">available_gmt</span><span class="p">])</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;first&#39;</span><span class="p">,</span> <span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotation Type&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">available_annots</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Annotation Type&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">available_annots</span>



<div class="viewcode-block" id="add_custom_annotation">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_custom_annotation">[docs]</a>
<span class="k">def</span> <span class="nf">add_custom_annotation</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">annotation_data</span><span class="p">,</span> <span class="n">source_name</span><span class="p">,</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">annotation_col</span><span class="p">,</span> <span class="n">accession_col</span> <span class="o">=</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="n">residue_col</span> <span class="o">=</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="n">position_col</span> <span class="o">=</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add custom annotation data to ptms or altered flanking sequence dataframes</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    annotation_data: pandas.DataFrame</span>
<span class="sd">        Dataframe containing the annotation data to be added to the ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added</span>
<span class="sd">    source_name: str</span>
<span class="sd">        Name of the source of the annotation data, will be used to label the columns in the ptms dataframe</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation data being added, will be used to label the columns in the ptms dataframe</span>
<span class="sd">    annotation_col: str</span>
<span class="sd">        Column name in the annotation data that contains the annotation data to be added to the ptms dataframe</span>
<span class="sd">    accession_col: str</span>
<span class="sd">        Column name in the annotation data that contains the UniProtKB Accession information. Default is &#39;UniProtKB Accession&#39;</span>
<span class="sd">    residue_col: str</span>
<span class="sd">        Column name in the annotation data that contains the residue information</span>
<span class="sd">    position_col: str</span>
<span class="sd">        Column name in the annotation data that contains the PTM position information</span>
<span class="sd">    </span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with an additional column for the custom annotation data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if annotation data contains the annotation col</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">annotation_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find column indicated to contain </span><span class="si">{</span><span class="n">annotation_col</span><span class="si">}</span><span class="s1"> in annotation data. Please either change the name of your annotation data column with this information or indicate the correct column name with the annotation_col parameter&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#make annotation col name based on source and annotation type</span>
            <span class="n">annotation_col_name</span> <span class="o">=</span> <span class="n">source_name</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">annotation_type</span>
            <span class="n">annotation_data</span> <span class="o">=</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">annotation_col</span><span class="p">:</span> <span class="n">annotation_col_name</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;annotation_col must be a string indicating column with annotation data to be added to the ptms dataframe&#39;</span><span class="p">)</span>

    <span class="c1">#check to make sure annotation data has the necessary columns</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">]]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find columns containing ptm information: </span><span class="si">{</span><span class="n">accession_col</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">residue_col</span><span class="si">}</span><span class="s1">, and </span><span class="si">{</span><span class="n">position_col</span><span class="si">}</span><span class="s1">. Please either change the name of your annotation data columns containing this information or indicate the correct column names with the accession_col, residue_col, and position_col parameters&#39;</span><span class="p">)</span>

    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="n">annotation_col_name</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotation_col_name</span><span class="p">])</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annotation_data</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">],</span> <span class="n">right_on</span> <span class="o">=</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms or annotation dataframe&#39;</span><span class="p">)</span>
    
    <span class="c1">#report the number of PTMs identified</span>
    <span class="n">num_ptms_with_custom_data</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">annotation_col</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source_name</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> data added: </span><span class="si">{</span><span class="n">num_ptms_with_custom_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with </span><span class="si">{</span><span class="n">source_name</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ptms</span></div>


<div class="viewcode-block" id="simplify_annotation">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.simplify_annotation">[docs]</a>
<span class="k">def</span> <span class="nf">simplify_annotation</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given an annotation, remove additional information such as whether or not a function is increasing or decreasing. For example, &#39;cell growth, induced&#39; would be simplified to &#39;cell growth&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    annotation: str</span>
<span class="sd">        Annotation to simplify</span>
<span class="sd">    sep: str</span>
<span class="sd">        Separator that splits the core annotation from additional detail. Default is &#39;,&#39;. Assumes the first element is the core annotation.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    annotation: str</span>
<span class="sd">        Simplified annotation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">sep</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">annotation</span> <span class="o">==</span> <span class="n">annotation</span> <span class="k">else</span> <span class="n">annotation</span>
    <span class="k">return</span> <span class="n">annotation</span></div>


<span class="k">def</span> <span class="nf">collapse_annotations</span><span class="p">(</span><span class="n">annotations</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">):</span>
    <span class="n">sep_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:{</span><span class="s1">&#39;Function&#39;</span><span class="p">:</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">:</span><span class="s1">&#39;,&#39;</span><span class="p">,</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;(&#39;</span><span class="p">,</span> <span class="s1">&#39;Disease&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;ELM&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span> <span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;Motif Match&#39;</span><span class="p">:</span> <span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;Intraprotein&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:{</span><span class="s1">&#39;Phosphatase&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">},</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;Pathway-WikiPathway&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;NetPath&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span><span class="s1">&#39;mSigDB&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (DIA2)&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (DIA)&#39;</span><span class="p">:</span> <span class="s1">&#39;-&gt;&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (PRM)&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">},</span> <span class="s2">&quot;iKiP&quot;</span><span class="p">:{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">}}</span>
    
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sep</span> <span class="o">=</span> <span class="n">sep_dict</span><span class="p">[</span><span class="n">database</span><span class="p">][</span><span class="n">annot_type</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">annotations</span>

    <span class="n">collapsed</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">annot</span> <span class="ow">in</span> <span class="n">annotations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annot</span> <span class="o">==</span> <span class="n">annot</span><span class="p">:</span>
            <span class="n">collapsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">simplify_annotation</span><span class="p">(</span><span class="n">annot</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="n">sep</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">collapsed</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">annot</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">collapsed</span>

<div class="viewcode-block" id="load_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.load_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">load_gmt_file</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Load a GMT file into a pandas DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Read the GMT file into a DataFrame</span>
    <span class="n">gmt_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    
    <span class="c1"># Remove any empty columns</span>
    <span class="n">gmt_df</span> <span class="o">=</span> <span class="n">gmt_df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">gmt_df</span></div>


<div class="viewcode-block" id="construct_annotation_dict_from_gmt">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_annotation_dict_from_gmt">[docs]</a>
<span class="k">def</span> <span class="nf">construct_annotation_dict_from_gmt</span><span class="p">(</span><span class="n">gmt_df</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;annotation&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a gmt annotation file format, construct a dictionary mapping each item to its annotations, with either the annotation as key or PTM as the key</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">key_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="s1">&#39;ptm&#39;</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;key_type must be either &#39;annotation&#39; or &#39;ptm&#39;&quot;</span><span class="p">)</span>
    
    <span class="c1"># Create a dictionary mapping each item to its annotations</span>
    <span class="n">annotation_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">gmt_df</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">key_type</span> <span class="o">==</span> <span class="s1">&#39;ptm&#39;</span><span class="p">:</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">ptm</span> <span class="ow">in</span> <span class="n">ptms</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ptm</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation_dict</span><span class="p">:</span>
                    <span class="n">annotation_dict</span><span class="p">[</span><span class="n">ptm</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
                <span class="n">annotation_dict</span><span class="p">[</span><span class="n">ptm</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">annotation</span><span class="p">)</span>
            
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">))</span>
            <span class="n">annotation</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">annotation_dict</span><span class="p">[</span><span class="n">annotation</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span>

    <span class="k">return</span> <span class="n">annotation_dict</span></div>


<div class="viewcode-block" id="check_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.check_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">check_gmt_file</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span><span class="p">,</span> <span class="n">automatic_download</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a gmt file path, check to make sure it exists. If it doesn&#39;t, either raise error or download and save a gmt file in the provided directory.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gmt_file : str</span>
<span class="sd">        file path to gmt file</span>
<span class="sd">    database : str</span>
<span class="sd">        name of database associated with gmt file</span>
<span class="sd">    annot_type : str</span>
<span class="sd">        type of annotation to check for. This is used to provide more specific error messages</span>
<span class="sd">    automatic download: bool</span>
<span class="sd">        whether to automatically download data and process into gmt file if it does not exist and can be done. Default is false</span>
<span class="sd">    odir : str or None</span>
<span class="sd">        location to save annotations, if automatic download is true</span>
<span class="sd">    kwargs : additional keyword arguments</span>
<span class="sd">        Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or use `construct_PhosphoSitePlus_gmt_file()` to create file in resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or use `construct_PTMsigDB_gmt_file()` to create file in resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or use `construct_RegPhos_gmt_file()` to create file in resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">automatic_download</span><span class="p">:</span>
                <span class="n">construct_PTMInt_gmt_file</span><span class="p">(</span><span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or set automatic_download=True to automatically create gmt file in provided resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">automatic_download</span><span class="p">:</span>
                <span class="n">construct_PTMcode_interprotein_gmt_file</span><span class="p">(</span><span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or set automatic_download=True to automatically create gmt file in provided resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">automatic_download</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">omnipath</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s2">&quot;OmniPath is not installed. Please install OmniPath to use this database.&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">construct_omnipath_gmt_file</span><span class="p">(</span><span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or set automatic_download=True to automatically create gmt file in provided resource directory.&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">automatic_download</span><span class="p">:</span>
                <span class="n">construct_DEPOD_gmt_file</span><span class="p">(</span><span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s2"> not found at </span><span class="si">{</span><span class="n">gmt_file</span><span class="si">}</span><span class="s2">. Please check the resource directory or set automatic_download=True to automatically create gmt file in provided resource directory.&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="process_database_annotations">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.process_database_annotations">[docs]</a>
<span class="k">def</span> <span class="nf">process_database_annotations</span><span class="p">(</span><span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;annotation&#39;</span><span class="p">,</span> <span class="n">collapsed</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">resource_dir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">automatic_download</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a database and annotation type, find and process the annotations into a dictionary mapping each PTM to its annotations, or vice versa</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    database: str</span>
<span class="sd">        source of annotation</span>
<span class="sd">    annot_type : str</span>
<span class="sd">        type of annotation to retrieve</span>
<span class="sd">    key_type : str</span>
<span class="sd">        whether the annotation or ptm should be the key of the output dictionary. Default is annotation</span>
<span class="sd">    collapsed : bool</span>
<span class="sd">        whether to combine annotations for similar types into a single annotation. For example, &#39;cell growth, induced&#39; and &#39;cell growth, inhibited&#39; would be simplified to &#39;cell growth&#39;. Default is False.</span>
<span class="sd">    resource_dir : str or None</span>
<span class="sd">        location of annotations. By default, this will look for annotations in PTM-POSE resource directory</span>
<span class="sd">    automatic_download: bool</span>
<span class="sd">        Whether to automatically download annotations that are not yet present in resource files directory</span>
<span class="sd">    kwargs : additional keyword arguments</span>
<span class="sd">        Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">resource_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">resource_dir</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span>
    
    <span class="c1">#check for valid database and annotation type, provide clarification where needed</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMsigDB&#39;</span> <span class="ow">and</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;PTMsigDB has multiple perturbation annotations. Please specify which perturbation you would like to use with the annot_type parameter. Options are: &#39;Perturbation (DIA2)&#39;, &#39;Perturbation (DIA)&#39;, or &#39;Perturbation (PRM)&#39;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMsigDB&#39;</span> <span class="ow">and</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Pathway&#39;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;PTMsigDB has multiple pathway annotations. Please specify which pathway you would like to use with the annot_type parameter. Options are: &#39;Pathway-WikiPathway&#39;, &#39;Pathway-NetPath&#39;, or &#39;Pathway-BI&#39;&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span> <span class="ow">and</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;OmniPath has two enzyme annotations, one for writer enzymes and one for eraser enzymes. Please specify which enzyme you would like to use with the annot_type parameter. Options are: &#39;Writer_Enzyme&#39; or &#39;Eraser_Enzyme&#39;&quot;</span><span class="p">)</span>
    
    <span class="c1">#check existence of the gmt file, download if automatic_download is True and does not exist</span>
    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Writer Enzyme&#39;</span> <span class="ow">or</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Writer_Enzyme&#39;</span><span class="p">:</span>
            <span class="n">gmt_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">resource_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="s1">&#39;Writer_Enzyme.gmt.gz&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Eraser Enzyme&#39;</span> <span class="ow">or</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;Eraser_Enzyme&#39;</span><span class="p">:</span>
            <span class="n">gmt_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">resource_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="s1">&#39;Eraser_Enzyme.gmt.gz&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">gmt_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">resource_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s1">.gmt.gz&#39;</span><span class="p">)</span>

    <span class="n">check_gmt_file</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">annot_type</span><span class="p">,</span> <span class="n">automatic_download</span> <span class="o">=</span> <span class="n">automatic_download</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">resource_dir</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>        

    
    <span class="c1">#load gmt df</span>
    <span class="n">gmt_df</span> <span class="o">=</span> <span class="n">load_gmt_file</span><span class="p">(</span><span class="n">gmt_file</span><span class="p">)</span>

    <span class="c1">#collapse annotations if specified</span>
    <span class="k">if</span> <span class="n">collapsed</span> <span class="ow">or</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;iKiP&#39;</span><span class="p">:</span>
        <span class="n">gmt_df</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">collapse_annotations</span><span class="p">(</span><span class="n">gmt_df</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">annot_type</span><span class="p">)</span>
        
    <span class="c1"># Process the GMT file to create a dictionary mapping each item to its annotations</span>
    <span class="n">annotation_dict</span> <span class="o">=</span> <span class="n">construct_annotation_dict_from_gmt</span><span class="p">(</span><span class="n">gmt_df</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="n">key_type</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">annotation_dict</span></div>



<div class="viewcode-block" id="append_from_gmt">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.append_from_gmt">[docs]</a>
<span class="k">def</span> <span class="nf">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">gmt_df</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">column_name</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a gmt annotation file format, add the annotations to the ptms dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms : pd.DataFrame</span>
<span class="sd">        dataframe containing ptm information, which can be the spliced_ptms or altered_flanks dataframe generated during projection</span>
<span class="sd">    database : str</span>
<span class="sd">        Name of the database for the annotation. Used to identify proper annotation if gmt_df not provided</span>
<span class="sd">    annot_type : str</span>
<span class="sd">        Type of annotation to append to the ptms dataframe</span>
<span class="sd">    gmt_df : pd.DataFrame </span>
<span class="sd">        If using custom gmt file, provide the dataframe loaded from the GMT file. This will override the database and annot_type parameters if provided.</span>
<span class="sd">    column_name : str or None</span>
<span class="sd">        </span>
<span class="sd">        Name of the column to use for the annotations in the ptms dataframe. If None, will use a default name based on the database and annot_type. Default is None.</span>
<span class="sd">    **kwargs : additional keyword arguments</span>
<span class="sd">        Passes additional keyword arguments to annotation specific functions. For example, you could pass min_sources for the construct_omnipath_gmt() function </span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#create dictionary mapping each ptm to its annotations</span>
    <span class="k">if</span> <span class="n">gmt_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_dict</span> <span class="o">=</span> <span class="n">construct_annotation_dict_from_gmt</span><span class="p">(</span><span class="n">gmt_df</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;ptm&#39;</span><span class="p">)</span> <span class="c1">#from provided gmt_df</span>
        <span class="k">if</span> <span class="n">column_name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">column_name</span> <span class="o">=</span> <span class="s1">&#39;Annotation&#39;</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">annot_type</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_dict</span> <span class="o">=</span> <span class="n">process_database_annotations</span><span class="p">(</span><span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">annot_type</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;ptm&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">column_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">annot_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span><span class="w"> </span><span class="s1">&#39;_&#39;</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">if</span> <span class="n">column_name</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">column_name</span> <span class="c1">#from database and annot_type</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;You must either provide the gmt-formatted dataframe or both the database and annot_type you would like to use.&quot;</span><span class="p">)</span>

    <span class="c1">#construct label to use for mapping to ptm data</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">add_ptm_column</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>
    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>
    <span class="c1"># Create a new column in the ptms DataFrame for the annotations</span>
    <span class="n">ptms</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">annotation_dict</span><span class="p">)</span>
    <span class="c1"># Convert the list of annotations to a semicolon-separated string</span>
    <span class="n">ptms</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="c1">#drop the label and PTM columns</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">])</span>

    <span class="c1">#report the number of ptms added</span>
    <span class="n">total</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">column_name</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">total</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;No PTMs found for </span><span class="si">{</span><span class="n">column_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">total</span><span class="si">}</span><span class="s1"> PTMs found with annotations from </span><span class="si">{</span><span class="n">column_name</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ptms</span></div>


<div class="viewcode-block" id="construct_gmt_df">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_gmt_df">[docs]</a>
<span class="k">def</span> <span class="nf">construct_gmt_df</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">annotation_col</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">annotation_separator</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given annotation data, construct a dataframe in the gmt file format. Save if odir and fname are provided</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    df : pd.DataFrame</span>
<span class="sd">        </span>
<span class="sd">        Dataframe containing the annotation data to be converted to GMT format. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</span>
<span class="sd">    annotation_col : str</span>
<span class="sd">        </span>
<span class="sd">        Column name in the dataframe that contains the annotation data to be added to the GMT file. This will be used as the annotation column in the output GMT file.</span>
<span class="sd">    description : str or np.nan</span>
<span class="sd">        description to add to description column</span>
<span class="sd">    annotation_separator : str or None</span>
<span class="sd">        what separator to use for splitting annotations in the annotation_col. If None, will not split annotations. Default is None.</span>
<span class="sd">    odir : str or None</span>
<span class="sd">        file path to output directory where the GMT file will be saved. If None, will not save. Default is None.</span>
<span class="sd">    fname : str or None:</span>
<span class="sd">        name of output file. If None, will use the annotation_col as the file name. Default is None.</span>
<span class="sd">    compressed : bool</span>
<span class="sd">        whether to save gmt file in gzip format. Default is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># add label columns to give each row a unique PTM + Modification identifier</span>
    <span class="k">if</span> <span class="s1">&#39;Label&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>

    <span class="c1"># if annotations are grouped by PTM in original df, separate them by the specified separator</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[[</span><span class="s1">&#39;Label&#39;</span><span class="p">,</span> <span class="n">annotation_col</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="c1">#remove entries that don&#39;t match the ptm_coordinates dataframe</span>
    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">add_ptm_column</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">,</span> <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="n">consider_isoforms</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> 
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">])]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1">#check if annotation_col is a list, if so, separate and split into unique rows</span>
    <span class="k">if</span> <span class="n">annotation_separator</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1">#separate the processes in each dataframe</span>
        <span class="n">df</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">annotation_separator</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">df</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">annotation_col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>  <span class="c1"># remove leading/trailing whitespace</span>

    

    <span class="c1">#convert dataframe to gmt annotation file and save</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">)[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="p">))</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="c1">#add description column to second column</span>
    <span class="n">df</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;Description&#39;</span><span class="p">,</span> <span class="n">description</span><span class="p">)</span>

    <span class="c1">#save to file</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">fname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fname</span> <span class="o">=</span> <span class="n">annotation_col</span>
        
        <span class="k">if</span> <span class="n">compressed</span><span class="p">:</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1">.gmt.gz&#39;</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1">.gmt&#39;</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">annotation_col</span><span class="p">:</span><span class="s1">&#39;Annotation&#39;</span><span class="p">})</span>
    <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="construct_custom_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_custom_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_custom_gmt_file</span><span class="p">(</span><span class="n">annotation_df</span><span class="p">,</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span><span class="p">,</span> <span class="n">annot_col</span><span class="p">,</span> <span class="n">accession_col</span> <span class="o">=</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="n">residue_col</span> <span class="o">=</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="n">position_col</span> <span class="o">=</span> <span class="s1">&#39;PTM Position in Isofrom&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function for constructing a gmt file for annotations not currently provided by PTM-POSE. Ideally, these annotations should be partially processed to have the same format as PTM-POSE annotations. For example, they should have columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    annotation_df: pandas.DataFrame</span>
<span class="sd">        Dataframe containing the annotation data to be added to the ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Isoform, and the annotation data to be added.</span>
<span class="sd">    database : str</span>
<span class="sd">        Name of the database for the annotation. This will be used to create the output directory and file name.</span>
<span class="sd">    annot_type : str</span>
<span class="sd">        </span>
<span class="sd">        Type of annotation data being added. This will be used to create the output file name and description.</span>
<span class="sd">    annot_col : str  </span>
<span class="sd">        Column name in the annotation data that contains the annotation data to be added to the ptms dataframe. This will be used as the annotation column in the output GMT file.</span>
<span class="sd">    accession_col : str</span>
<span class="sd">        </span>
<span class="sd">        Column name in the annotation data that contains the UniProtKB Accession information. Default is &#39;UniProtKB Accession&#39;.</span>
<span class="sd">    residue_col : str</span>
<span class="sd">        </span>
<span class="sd">        Column name in the annotation data that contains the residue information. Default is &#39;Residue&#39;.</span>
<span class="sd">    position_col : str</span>
<span class="sd">        Column name in the annotation data that contains the PTM position information. Default is &#39;PTM Position in Isoform&#39;.</span>
<span class="sd">    odir : str or None</span>
<span class="sd">        </span>
<span class="sd">        Path to the output directory where the GMT file will be saved. If None, will save to the default resource directory for annotations. Default is None.</span>
<span class="sd">    kwargs : additional keyword arguments</span>
<span class="sd">        additional keywords to pass to construct_gmt_df function. This can include parameters such as annotation_separator, description, and compressed.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check to make sure annotation data has the necessary columns</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">]]):</span>
        <span class="n">missing_cols</span> <span class="o">=</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find the following columns with PTM information: </span><span class="si">{</span><span class="n">missing_cols</span><span class="si">}</span><span class="s1">. Please either change the name of your annotation data columns containing this information or indicate the correct column names with the accession_col, residue_col, and position_col parameters&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">annot_col</span> <span class="ow">in</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find column indicated to contain </span><span class="si">{</span><span class="n">annot_col</span><span class="si">}</span><span class="s1"> in annotation data. Please either change the name of your annotation data column with this information or indicate the correct column name with the annot_col parameter&#39;</span><span class="p">)</span>
    
    <span class="c1">#rename PTM columns to match PTM-POSE format</span>
    <span class="k">if</span> <span class="n">accession_col</span> <span class="o">!=</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">:</span>
        <span class="n">annotation_df</span> <span class="o">=</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">accession_col</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">})</span>
    
    <span class="k">if</span> <span class="n">residue_col</span> <span class="o">!=</span> <span class="s1">&#39;Residue&#39;</span><span class="p">:</span>
        <span class="n">annotation_df</span> <span class="o">=</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">residue_col</span><span class="p">:</span><span class="s1">&#39;Residue&#39;</span><span class="p">})</span>
    
    <span class="k">if</span> <span class="n">position_col</span> <span class="o">!=</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">:</span>
        <span class="n">annotation_df</span> <span class="o">=</span> <span class="n">annotation_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">position_col</span><span class="p">:</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">})</span>
        <span class="n">annotation_df</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">annotation_df</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

    <span class="c1">#set odir and file name</span>
    <span class="n">fname</span> <span class="o">=</span> <span class="n">annot_type</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="n">database</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="n">database</span><span class="p">)</span>


    <span class="c1">#construct gmt df</span>
    <span class="n">gmt_df</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">annotation_df</span><span class="p">,</span> <span class="n">annotation_col</span> <span class="o">=</span> <span class="n">annot_col</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">annot_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="n">fname</span><span class="p">,</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>



<div class="viewcode-block" id="construct_PhosphoSitePlus_gmt_files">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_PhosphoSitePlus_gmt_files">[docs]</a>
<span class="k">def</span> <span class="nf">construct_PhosphoSitePlus_gmt_files</span><span class="p">(</span><span class="n">regulatory_site_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">kinase_substrate_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">disease_association_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given three PhosphoSitePlus annotation files, convert to readily usable format with PTM-POSE in gmt file format</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    regulatory_site_file: str or None</span>
<span class="sd">        </span>
<span class="sd">        Path to the PhosphoSitePlus regulatory site file (gzipped). If None, will skip creating function annotations.</span>
<span class="sd">    kinase_substrate_file: str or None</span>
<span class="sd">        Path to the PhosphoSitePlus kinase-substrate file (gzipped). If None, will skip creating kinase-substrate annotations.</span>
<span class="sd">    disease_association_file: str or None</span>
<span class="sd">        Path to the PhosphoSitePlus disease association file (gzipped). If None, will skip creating disease association annotations.</span>
<span class="sd">    odir : str or None</span>
<span class="sd">        Path to the output directory where the GMT files will be saved. If None, will save to the default resource directory for PhosphoSitePlus annotations.</span>
<span class="sd">    overwrite : bool</span>
<span class="sd">        If True, will overwrite existing GMT files if they already exist. If False, will skip creating the GMT files if they already exist. Default is False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">)</span>
        

    <span class="c1">#check if output directory exists, if not create it</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>


    <span class="c1">#check if regulatory site file is provided</span>
    <span class="k">if</span> <span class="n">regulatory_site_file</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Function.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT files from regulatory site data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">regulatory_site_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

            <span class="c1">#read in the kinase substrate data and add to spliced ptm info</span>
            <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">regulatory_site_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
            <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ACC_ID&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">})</span>
            <span class="c1">#drop extra modification information that is not needed</span>
            <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>
            <span class="c1">#add modification type</span>
            <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mod_shorthand_dict</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]])</span>

            <span class="c1">#restrict to human data</span>
            <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
            <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_FUNCTION&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

            <span class="c1">#add labels that match what PTM-POSE will expect (&lt;accession&gt;_&lt;residue&gt;&lt;position&gt;-{&lt;modification class&gt;})</span>
            <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>

            <span class="c1">#for each available type of annotation, create a gmt file</span>
            <span class="n">function</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">regulatory_site_data</span><span class="p">,</span> <span class="s1">&#39;ON_FUNCTION&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PSP:ON_FUNCTION&#39;</span><span class="p">,</span> <span class="n">annotation_separator</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus Function gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;/Function.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1">#biological processes</span>
            <span class="n">process</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">regulatory_site_data</span><span class="p">,</span> <span class="s1">&#39;ON_PROCESS&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PSP:ON_PROCESS&#39;</span><span class="p">,</span> <span class="n">annotation_separator</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus Biological Process gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;/Process.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="c1">#protein interactions</span>
            <span class="n">interactions</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">regulatory_site_data</span><span class="p">,</span> <span class="s1">&#39;ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s2">&quot;PSP:ON_PROT_INTERACT&quot;</span><span class="p">,</span> <span class="n">annotation_separator</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus Protein Interactions gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;/Interactions.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1">#check if kinase substrate file is provided</span>
    <span class="k">if</span> <span class="n">kinase_substrate_file</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Enzyme.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from kinase substrate data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">kinase_substrate_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

            <span class="c1">#load data</span>
            <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">kinase_substrate_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">encoding</span> <span class="o">=</span> <span class="s2">&quot;cp1252&quot;</span><span class="p">)</span>
            <span class="c1">#restrict to human data</span>
            <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;KIN_ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
            <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;SUB_ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>

            <span class="c1">#add labels that match what PTM-POSE will expect (&lt;accession&gt;_&lt;residue&gt;&lt;position&gt;-{&lt;modification class&gt;})</span>
            
            <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;SUB_ACC_ID&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;SUB_MOD_RSD&#39;</span><span class="p">]</span>
            <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-Phosphorylation&#39;</span>

            <span class="n">kinase_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">ks_dataset</span><span class="p">,</span> <span class="s1">&#39;GENE&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus Kinase-Substrate gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;/Enzyme.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1">#check if disease association file is provided</span>
    <span class="k">if</span> <span class="n">disease_association_file</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Disease.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from disease association data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">disease_association_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

            <span class="c1">#load data</span>
            <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">disease_association_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
            <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>

            <span class="c1">#removes sites without a specific disease annotation</span>
            <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">])</span>


            <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ACC_ID&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">})</span>
            <span class="c1">#drop extra modification information that is not needed</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>
            <span class="c1">#add modification type</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mod_shorthand_dict</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]])</span>
            <span class="c1">#if phosphorylation, add specific residue</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">residue_dict</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phospho&#39;</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="c1">#change O-GalNac occurring on N to N-glycosylation</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;N-Glycosylation&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;O-Glycosylation&#39;</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N&#39;</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>

            <span class="c1">#expand disease column to allow for multiple disease associations</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
            <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">)</span>
            <span class="c1">#remove any extra whitespace</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>


            <span class="c1">#combine disease and alteration</span>
            <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>

            <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">disease_associated_sites</span><span class="p">,</span> <span class="n">annotation_col</span> <span class="o">=</span> <span class="s1">&#39;ALTERATION&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PSP:Disease_Association&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Disease&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus Disease-association gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="s1">&#39;/Disease.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>

    




<div class="viewcode-block" id="add_ELM_interactions">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_ELM_interactions">[docs]</a>
<span class="k">def</span> <span class="nf">add_ELM_interactions</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a spliced ptms or altered flanks dataframe from the project module, add ELM interaction data to the dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events, either differentially included events, or altered flanking sequences</span>
<span class="sd">    file: str</span>
<span class="sd">        Path to the ELM data file. If not provided, the data will be downloaded directly from the ELM website</span>
<span class="sd">    report_success: bool</span>
<span class="sd">        If True, will print out the number of PTMs identified in the dataset that have ELM interaction information</span>
<span class="sd">    </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with additional columns for ELM interaction data</span>
<span class="sd">    </span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#load data</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://elm.eu.org/interactions/as_tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
        <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

    <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[(</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;taxonomyElm&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;9606(Homo sapiens)&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;taxonomyDomain&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;9606(Homo sapiens)&#39;</span><span class="p">)]</span>

    <span class="n">elm_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">elm_type</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">elm_interactor</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1">#grab ptm location from residue column (gives residue and position (S981), so need to remove residue and convert to int)</span>
        <span class="n">ptm_loc</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;none&#39;</span> <span class="k">else</span> <span class="kc">None</span>

        <span class="c1">#if data does not have position information, move to the next</span>
        <span class="k">if</span> <span class="n">ptm_loc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1">#find if any of the linear motifs match ptm loc</span>
        <span class="n">protein_match</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;interactorElm&#39;</span><span class="p">]</span>
        <span class="n">region_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&gt;=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StartElm&#39;</span><span class="p">])</span>  <span class="o">&amp;</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&lt;=</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StopElm&#39;</span><span class="p">])</span>
        <span class="n">elm_subset_motif</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="n">protein_match</span> <span class="o">&amp;</span> <span class="n">region_match</span><span class="p">]</span>
        <span class="c1">#if any interactions were found, record and continue to the next (assumes a single ptm won&#39;t be found as both a SLiM and domain)</span>
        <span class="k">if</span> <span class="n">elm_subset_motif</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_motif</span><span class="p">[</span><span class="s1">&#39;Elm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;SLiM&#39;</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_motif</span><span class="p">[</span><span class="s1">&#39;interactorDomain&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="k">continue</span>


        <span class="c1">#domain</span>
        <span class="n">protein_match</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;interactorDomain&#39;</span><span class="p">]</span>
        <span class="n">region_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&gt;=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StartDomain&#39;</span><span class="p">])</span>  <span class="o">&amp;</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&lt;=</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StopDomain&#39;</span><span class="p">])</span>
        <span class="n">elm_subset_domain</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="n">protein_match</span> <span class="o">&amp;</span> <span class="n">region_match</span><span class="p">]</span>
        <span class="c1">#if any interactions were found, record and continue to the next (assumes a single ptm won&#39;t be found as both a SLiM and domain)</span>
        <span class="k">if</span> <span class="n">elm_subset_domain</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_domain</span><span class="p">[</span><span class="s1">&#39;Elm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Domain&#39;</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_domain</span><span class="p">[</span><span class="s1">&#39;interactorElm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="k">continue</span>

        <span class="c1">#if no interactions wer found, record as np.nan</span>
        <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_interactor</span>
    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Location of PTM for Interaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_type</span>
    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Motifs Associated with Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_list</span>
    
    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_ELM_instance</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ELM interaction instances added: </span><span class="si">{</span><span class="n">num_ptms_with_ELM_instance</span><span class="si">}</span><span class="s2"> PTMs in dataset found associated with at least one known ELM instance&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ptms</span></div>



<div class="viewcode-block" id="add_ELM_matched_motifs">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_ELM_matched_motifs">[docs]</a>
<span class="k">def</span> <span class="nf">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptms or altered flanks dataframes, compare the canonical flanking sequences of each PTM to motifs recorded in the ELM database. If a match is found, the ELM motif will be recorded in the ELM:Motif Matches column</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events, either differentially included events, or altered flanking sequences</span>
<span class="sd">    flank_size: int</span>
<span class="sd">        Number of residues to include on either side of the PTM for the motif search. Default is 7</span>
<span class="sd">    file: str</span>
<span class="sd">        Path to the ELM data file. If not provided, the data will be downloaded directly from the ELM website</span>
<span class="sd">    report_success: bool</span>
<span class="sd">        If True, will print out the number of PTMs identified in the dataset that have ELM motif data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://elm.eu.org/elms/elms_index.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
        <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>

    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1">#create corresponding label for ptm_coordinate data</span>
    <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    
    <span class="n">match_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1">#grab ptm information</span>
        <span class="c1">#grab flanking sequence for the ptm</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;PTM Position in Isoform&quot;</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="n">ptm</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>

        
        <span class="k">if</span> <span class="n">ptm</span> <span class="ow">in</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">ptm_flanking_seq</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ptm</span><span class="p">,</span> <span class="s1">&#39;Canonical Flanking Sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="c1">#make sure flanking sequence is present</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ptm_flanking_seq</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>

                <span class="c1">#default flanking sequence is 10, if requested flanking sequence is different, then adjust</span>
                <span class="k">if</span> <span class="n">flank_size</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Flanking size must be equal to or less than 10&#39;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">flank_size</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>
                    <span class="n">ptm_flanking_seq</span> <span class="o">=</span> <span class="n">ptm_flanking_seq</span><span class="p">[</span><span class="mi">10</span><span class="o">-</span><span class="n">flank_size</span><span class="p">:</span><span class="mi">10</span><span class="o">+</span><span class="n">flank_size</span><span class="p">]</span>

                <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">elm_row</span> <span class="ow">in</span> <span class="n">elm_classes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">reg_ex</span> <span class="o">=</span> <span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;Regex&#39;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">reg_ex</span><span class="p">,</span> <span class="n">ptm_flanking_seq</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;ELMIdentifier&#39;</span><span class="p">])</span>

                <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">matches</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#print(f&#39;PTM {ptm} not found in PTM info file&#39;)</span>
            <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    
    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Motif Matches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">match_list</span>

    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_matched_motif</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;ELM:Motif Matches&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ELM Class motif matches found: </span><span class="si">{</span><span class="n">num_ptms_with_matched_motif</span><span class="si">}</span><span class="s2"> PTMs in dataset found with at least one matched motif&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ptms</span></div>


<div class="viewcode-block" id="construct_PTMInt_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_PTMInt_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_PTMInt_gmt_file</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">max_retries</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">delay</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download and process PTMInt interaction data to create gmt files for PTM-POSE</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file : str, optional</span>
<span class="sd">        Path to the PTMInt data file. If not provided, the data will be downloaded directly from the PTMInt website. Default is None.</span>
<span class="sd">    odir : str, optional</span>
<span class="sd">        Output directory for the gmt file. If not provided, will default to the PTM-POSE resource directory for annotations. Default is None.</span>
<span class="sd">    overwrite : bool, optional</span>
<span class="sd">        If True, will overwrite any existing gmt files in the output directory. If False, will skip the creation of the gmt file if it already exists. Default is False.</span>
<span class="sd">    max_retries : int, optional</span>
<span class="sd">        Number of times to retry downloading the PTMInt data if the download fails. Default is 5.</span>
<span class="sd">    delay : int, optional</span>
<span class="sd">        Amount of time to wait (in seconds) before retrying the download if it fails. Default is 10 seconds.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#find output directory for gmt files</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMint&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMint&#39;</span><span class="p">)</span>

    <span class="c1">#check if gmt files already exist, skip if overwrite is False</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Interactions.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from PTMInt data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    
    <span class="c1">#load file</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;No PTMInt data provided, downloading directly from website&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="c1"># Attempt to download the file</span>
                <span class="n">PTMint</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://ptmint.sjtu.edu.cn/data/PTM</span><span class="si">%20e</span><span class="s1">xperimental</span><span class="si">%20e</span><span class="s1">vidence.csv&#39;</span><span class="p">)</span>
                <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">delay</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">downloaded</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Failed to download PTMint data after </span><span class="si">{</span><span class="n">max_retries</span><span class="si">}</span><span class="s2"> attempts.&quot;</span><span class="p">)</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Successfully downloaded, processing...&#39;</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.csv&#39;</span><span class="p">)</span>
        <span class="n">PTMint</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

    <span class="c1">#restrict to human interactions</span>
    <span class="n">PTMint</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">[</span><span class="n">PTMint</span><span class="p">[</span><span class="s2">&quot;Organism&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Human&#39;</span><span class="p">]</span>
    <span class="c1">#convert modification names to match Modification Class names used in PTM-POSE</span>
    <span class="n">mod_conversion</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Phos&#39;</span><span class="p">:</span><span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Me&#39;</span><span class="p">:</span><span class="s1">&#39;Methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Ac&#39;</span><span class="p">:</span><span class="s1">&#39;Acetylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Sumo&#39;</span><span class="p">:</span><span class="s1">&#39;Sumoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Ub&#39;</span><span class="p">:</span><span class="s1">&#39;Ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;Glyco&#39;</span><span class="p">:</span><span class="s1">&#39;Glycosylation&#39;</span><span class="p">}</span>
    <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">mod_conversion</span><span class="p">)</span>

    <span class="c1">#rename columns to match PTM-POSE expectations</span>
    <span class="n">PTMint</span> <span class="o">=</span> <span class="n">PTMint</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Uniprot&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> 
                                    <span class="s1">&#39;AA&#39;</span><span class="p">:</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Site&#39;</span><span class="p">:</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">:</span><span class="s2">&quot;Modification Class&quot;</span><span class="p">})</span>
    <span class="c1">#combine the interacting protein and effect into a single column for the PTMint data</span>
    <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;Interaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;Int_uniprot&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;Effect&#39;</span><span class="p">]</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">)</span>
    
    <span class="c1">#for each available type of annotation, create a gmt file</span>
    <span class="n">PTMint_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">PTMint</span><span class="p">,</span> <span class="s1">&#39;Interaction&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PTMint:Interactions&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>



<div class="viewcode-block" id="extract_ids_PTMcode">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.extract_ids_PTMcode">[docs]</a>
<span class="k">def</span> <span class="nf">extract_ids_PTMcode</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s1">&#39;## Protein1&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Many different ID forms are used in PTMcode, but we rely on UniProt IDs. This function is used to convert between Ensembl Gene IDs to UniProt IDs</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">#add gene name to data</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>
    <span class="n">name_to_uniprot</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB/Swiss-Prot ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">]</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

    <span class="c1">#protein name is provided as either ensemble gene id or gene name check for both</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">translator</span><span class="p">[[</span><span class="s1">&#39;Gene stable ID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(),</span> <span class="n">left_on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span> <span class="n">right_on</span> <span class="o">=</span> <span class="s1">&#39;Gene stable ID&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="s1">&#39;From_ID&#39;</span><span class="p">})</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">name_to_uniprot</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span> <span class="n">right_on</span> <span class="o">=</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;UniProtKB/Swiss-Prot ID&#39;</span><span class="p">:</span> <span class="s1">&#39;From_Name&#39;</span><span class="p">})</span>

    <span class="c1">#grab unique id from &#39;From_ID&#39; and &#39;From_Name&#39; column, if available</span>
    <span class="n">uniprot_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;From_Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">combine_first</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;From_ID&#39;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">uniprot_ids</span><span class="o">.</span><span class="n">values</span></div>


<div class="viewcode-block" id="construct_PTMcode_interprotein_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_PTMcode_interprotein_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_PTMcode_interprotein_gmt_file</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">max_retries</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">delay</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the PTMcode interprotein interaction data, convert to readily usable format with PTM-POSE in gmt file format</span>

<span class="sd">    file: str</span>
<span class="sd">        Path to the PTMcode interprotein interaction data file. If not provided, the data will be downloaded directly from the PTMcode website</span>
<span class="sd">    odir : str</span>
<span class="sd">        Output directory for the gmt file. If not provided, will default to the PTM-POSE resource directory for annotations</span>
<span class="sd">    overwrite : bool, optional</span>
<span class="sd">        If True, will overwrite any existing gmt files in the output directory. If False, will skip the creation of the gmt file if it already exists. Default is False.</span>
<span class="sd">    max_retries : int, optional</span>
<span class="sd">        Number of times to retry downloading the PTMcode data if the initial attempt fails. Default is 5.</span>
<span class="sd">    delay : int, optional</span>
<span class="sd">        Number of seconds to wait between retries if the download fails. Default is 10 seconds.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">)</span>

    <span class="c1">#check if gmt files already exist, skip if overwrite is False</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Interactions.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from PTMcode interprotein interaction data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">return</span>

    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;PTMcode data not provided, downloading directly from website&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://ptmcode.embl.de/data/PTMcode2_associations_between_proteins.txt.gz&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
                <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">break</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">delay</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">downloaded</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Failed to download PTMcode data after </span><span class="si">{}</span><span class="s1"> attempts&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">max_retries</span><span class="p">))</span>
        
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Successfully downloaded, processing...&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
        <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>

    <span class="c1">#grab human interactions</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Species&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Homo sapiens&#39;</span><span class="p">]</span>
    <span class="c1">#ignore intraprotein interactions</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;## Protein1&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Protein2&#39;</span><span class="p">]]</span>

    <span class="c1">#get uniprot id for primary protein and interacting protein</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">extract_ids_PTMcode</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="s1">&#39;## Protein1&#39;</span><span class="p">)</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">extract_ids_PTMcode</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="s1">&#39;Protein2&#39;</span><span class="p">)</span>

    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Protein&#39;</span><span class="p">])</span>
    <span class="c1">#remove duplicate proteins (some entries have different ids but are actually the same protein)</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]]</span>

    <span class="c1">#convert modification names</span>
    <span class="n">convert_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Adp ribosylation&#39;</span><span class="p">:</span> <span class="s1">&#39;ADP Ribosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Glutamine deamidation&#39;</span><span class="p">:</span><span class="s1">&#39;Deamidation&#39;</span><span class="p">}</span>
    <span class="n">new_mod_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">failed_mod</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mod_list</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">mod_list</span><span class="p">:</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;glycosylation&#39;</span> <span class="ow">in</span> <span class="n">mod</span><span class="p">:</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Glycosylation&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">convert_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">convert_dict</span><span class="p">[</span><span class="n">mod</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">new_mod</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="s1">&#39;Modification&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_mod</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">failed_mod</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
                <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
    <span class="n">conversion_dict</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">mod_list</span><span class="p">,</span> <span class="n">new_mod_names</span><span class="p">))</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">conversion_dict</span><span class="p">)</span>



    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Residue1&#39;</span><span class="p">]</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">)</span>

    <span class="n">ptmcode_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="s1">&#39;Interacting Protein&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PTMcode:Interprotein_Interaction&#39;</span><span class="p">,</span> <span class="n">annotation_separator</span> <span class="o">=</span> <span class="s1">&#39;;&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>



<div class="viewcode-block" id="extract_positions_from_DEPOD">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.extract_positions_from_DEPOD">[docs]</a>
<span class="k">def</span> <span class="nf">extract_positions_from_DEPOD</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given string object consisting of multiple modifications in the form of &#39;Residue-Position&#39; separated by &#39;, &#39;, extract the residue and position. Ignore any excess details in the string.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    x : str</span>
<span class="sd">        dephosphosite entry from DEPOD data</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    new_x :str</span>
<span class="sd">        ptm residue and position in format that PTM-POSE recognizes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)</span>
    <span class="c1">#for each residue in list, find location of &#39;Ser&#39;, &#39;Thr&#39; and &#39;Tyr&#39; in the string (should either have &#39;-&#39; or a number immediately after it)</span>
    <span class="n">new_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="c1">#determine type of modification</span>
        <span class="k">if</span> <span class="s1">&#39;Ser&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Ser&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;S&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;Thr&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Thr&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;T&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;Tyr&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Tyr&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;His&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;His&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;H&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

        <span class="c1">#check if multiple locations were found, if so grab last entry</span>
        <span class="k">if</span> <span class="n">loc</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">item</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">True</span>
        
        <span class="c1">#find integer</span>
        <span class="k">if</span> <span class="n">make_string</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">[</span><span class="n">loc</span><span class="p">:]:</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">item</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">res</span> <span class="o">+</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="n">loc</span><span class="o">+</span><span class="mi">3</span><span class="p">:]</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">res</span> <span class="o">+</span> <span class="n">item</span>

        <span class="n">new_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">new_x</span></div>


<div class="viewcode-block" id="construct_DEPOD_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_DEPOD_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_DEPOD_gmt_file</span><span class="p">(</span><span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">max_retries</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">delay</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download and process DEPOD data to create a GMT file for PTM-POSE. DEPOD contains information on dephosphorylation sites and their corresponding substrates.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    odir : str, optional</span>
<span class="sd">        Output directory for the GMT file. If not provided, it will default to the &#39;Resource_Files/Annotations/DEPOD&#39; directory within the PTM-POSE package directory.</span>
<span class="sd">    overwrite : bool, optional</span>
<span class="sd">        If True, will overwrite any existing GMT file in the output directory. If False and the GMT file already exists, the function will skip processing and print a message.</span>
<span class="sd">    max_retries : int, optional</span>
<span class="sd">        Number of times to try downloading data from DEPOD</span>
<span class="sd">    delay : int, optional</span>
<span class="sd">        Delay in seconds between download attempts. Default is 10 seconds.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if gmt file already exists</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">)</span>
        
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Enzyme.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from DEPOD data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    
    <span class="c1">#download data from website</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Downloading DEPOD data...&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">max_retries</span><span class="p">):</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">depod1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;https://depod.bioss.uni-freiburg.de/download/PPase_protSubtrates_201903.xls&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;PSprots&#39;</span><span class="p">)</span>
            <span class="n">depod2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;https://depod.bioss.uni-freiburg.de/download/PPase_protSubtrates_newPairs_201903.xls&#39;</span><span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;newPSprots&#39;</span><span class="p">)</span>
            <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">break</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">downloaded</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">delay</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">downloaded</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;Failed to download DEPOD data after </span><span class="si">{}</span><span class="s1"> attempts&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">max_retries</span><span class="p">))</span>
    
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Successfully downloaded, processing...&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">depod1</span><span class="p">,</span> <span class="n">depod2</span><span class="p">])</span>

    <span class="c1">#separate multiple substrate accessions into their own rows (many of these link back to the same ID, but will keep just in case)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">])</span>

    <span class="c1">##### match entries by flanking sequence first (accounts for small changes in sequence) ####</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;5 amino acid window around the dephosphosite (small letters)&#39;</span><span class="p">:</span> <span class="s1">&#39;5aa flank&#39;</span><span class="p">})</span>

    <span class="c1">#remove any rows with missing sit information</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;N/A&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;5aa flank&#39;</span><span class="p">)</span>

    <span class="c1">#convert multiple sequences to have consistent delimiters and remove any excess whitespace</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">,</span> <span class="s1">&#39;;&#39;</span><span class="p">))</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">[</span><span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">])</span>

    <span class="c1">#separate into unique rows</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>

    <span class="n">ptm_labels</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1">#grab unique phosphorylatio site entries from ptm_coordinates</span>
    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Isoform ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Canonical Flanking Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Isoform Type&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">row</span> <span class="ow">in</span> <span class="n">depod</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1">#grab ptm information matching the flanking sequnce</span>
        <span class="n">tmp_ptm</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">tmp_ptm</span> <span class="o">=</span> <span class="n">tmp_ptm</span><span class="p">[</span><span class="n">tmp_ptm</span><span class="p">[</span><span class="s1">&#39;Canonical Flanking Sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;5aa flank&#39;</span><span class="p">])]</span>
        <span class="c1">#if found, append PTM label to list, else append np.nan</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">tmp_ptm</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">label</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">get_ptm_label</span><span class="p">(</span><span class="n">tmp_ptm</span><span class="o">.</span><span class="n">squeeze</span><span class="p">())</span>
            <span class="n">ptm_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">label</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ptm_labels</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_labels</span>
    <span class="c1">#### For remaining missed PTMs, try using site information ####</span>
    <span class="n">depod_flank</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">depod_flank</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod_flank</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-Phosphorylation&#39;</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="n">depod</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">,</span> <span class="s1">&#39;Dephosphosites&#39;</span><span class="p">,</span> <span class="s1">&#39;Phosphatase entry names&#39;</span><span class="p">]]</span>

    <span class="c1">#remove any rows with missing sit information</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>

    <span class="c1">#remove excess annotations that make parsing difficult</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;in&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;in ref.&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

    <span class="c1">#separate individual sites</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[(</span><span class="o">~</span><span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Isoform&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="o">~</span><span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;isoform&#39;</span><span class="p">))]</span>

    <span class="c1">#process dephosphosite strings to extract residue and position and explode so that each phosphosite is its own row</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">extract_positions_from_DEPOD</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>


    <span class="c1">#combine info from both approaches into one df</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">depod</span><span class="p">,</span> <span class="n">depod_flank</span><span class="p">[[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">,</span> <span class="s1">&#39;Dephosphosites&#39;</span><span class="p">,</span> <span class="s1">&#39;Phosphatase entry names&#39;</span><span class="p">,</span> <span class="s1">&#39;Label&#39;</span><span class="p">]]],</span> <span class="n">ignore_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#construct gmt file </span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">)</span>
    <span class="n">depod_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">depod</span><span class="p">,</span> <span class="s1">&#39;Phosphatase entry names&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;DEPOD:Enzyme&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>





<div class="viewcode-block" id="construct_RegPhos_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_RegPhos_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_RegPhos_gmt_file</span><span class="p">(</span><span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    file : str</span>
<span class="sd">        RegPhos text file path. This file can be downloaded from the RegPhos website. If None, the function will raise an error.</span>
<span class="sd">    odir : str</span>
<span class="sd">        Output directory for the gmt files. If None, will default to the PTM-POSE resource directory.</span>
<span class="sd">    overwrite : bool, optional</span>
<span class="sd">        If True, will overwrite any existing gmt files in the output directory. Default is False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if gmt file already exists</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">)</span>
        
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Enzyme.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from RegPhos data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    <span class="c1">#check file and then load</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.txt&#39;</span><span class="p">)</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>


    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;catalytic kinase&#39;</span><span class="p">)</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;code&#39;</span><span class="p">:</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;position&#39;</span><span class="p">:</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;catalytic kinase&#39;</span><span class="p">:</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">})</span>
    <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Phosphorylation&#39;</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
    <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>



    <span class="c1">#construct gmt file</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">)</span>

    <span class="n">regphos_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">regphos</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;RegPhos:Enzyme&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>



<div class="viewcode-block" id="construct_PTMsigDB_gmt_files">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_PTMsigDB_gmt_files">[docs]</a>
<span class="k">def</span> <span class="nf">construct_PTMsigDB_gmt_files</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">process_PSP_data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the PTMsigDB xlsx file, convert to readily usable format with PTM-POSE in gmt file format. This will also process the PhosphoSitePlus data in PTMsigDB if requested.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file : str</span>
<span class="sd">        PTMsigDB excel file path. This file can be downloaded from the PTMsigDB website.</span>
<span class="sd">    odir : str</span>
<span class="sd">        Output directory for the gmt files. If None, will default to the PTM-POSE resource directory.</span>
<span class="sd">    overwrite : bool, optional</span>
<span class="sd">        If True, will overwrite any existing gmt files in the output directory. Default is False.</span>
<span class="sd">    process_PSP_data : bool, optional</span>
<span class="sd">        If True, will process the PhosphoSitePlus data included in the PTMsigDB file, but only if not already found in odir. Default is True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if gmt file already exists</span>
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ptmsigdb_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">)</span>
        <span class="n">ikip_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;iKiP&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">process_PSP_data</span><span class="p">:</span>
            <span class="n">psp_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ptmsigdb_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">)</span>
        <span class="n">ikip_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;iKiP&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">process_PSP_data</span><span class="p">:</span>
            <span class="n">psp_odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">)</span>
        
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ptmsigdb_odir</span><span class="p">,</span> <span class="s1">&#39;Perturbation-DIA.gmt.gz&#39;</span><span class="p">))</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">overwrite</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;GMT file from PTMsigDB data already exists, skipping. Set overwrite = True to overwrite&quot;</span><span class="p">)</span>
        <span class="k">return</span>
    
    
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.xlsx&#39;</span><span class="p">)</span>
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">)</span>


    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>

    <span class="c1">#filter out excess information in some of the site.ptm column, then convert to modification class details</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;;&#39;</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">mod_shorthand_dict</span><span class="p">)</span>

    <span class="c1">#combine signature and direction for annotation column</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;signature&#39;</span><span class="p">]</span> <span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.direction&#39;</span><span class="p">]</span>

    <span class="c1">#drop unneeded columns</span>
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="s1">&#39;category&#39;</span><span class="p">]]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:&#39;</span> <span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> 
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="c1">#add label for matching to ptm coordinates dataframe</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>

    <span class="c1">#convert column names to desired descriptor</span>
    <span class="n">col_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PTMsigDB:KINASE-iKiP&#39;</span><span class="p">:</span><span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:PERT-P100-PRM&#39;</span><span class="p">:</span><span class="s1">&#39;Perturbation-PRM&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:PERT-P100-DIA&#39;</span><span class="p">:</span><span class="s1">&#39;Perturbation-DIA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;PTMsigDB:PERT-P100-DIA2&#39;</span><span class="p">:</span><span class="s1">&#39;Perturbation-DIA2&#39;</span><span class="p">,</span>   
    <span class="s1">&#39;PTMsigDB:PATH-BI&#39;</span><span class="p">:</span><span class="s1">&#39;Pathway-BI&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:PATH-WP&#39;</span><span class="p">:</span><span class="s1">&#39;Pathway-WikiPathways&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:PATH-NP&#39;</span><span class="p">:</span><span class="s1">&#39;Pathway-NetPath&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:PERT-PSP&#39;</span><span class="p">:</span><span class="s1">&#39;Perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:DISEASE-PSP&#39;</span><span class="p">:</span><span class="s1">&#39;Disease&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB:KINASE-PSP&#39;</span><span class="p">:</span><span class="s1">&#39;Enzyme&#39;</span><span class="p">}</span>
    <span class="c1">#create directories if do not exist</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ikip_odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">ikip_odir</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">ptmsigdb_odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">ptmsigdb_odir</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">psp_odir</span><span class="p">)</span> <span class="ow">and</span> <span class="n">process_PSP_data</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">psp_odir</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">category</span> <span class="ow">in</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">tmp</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">category</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">category</span> <span class="o">!=</span> <span class="s1">&#39;PTMsigDB:KINASE-iKiP&#39;</span> <span class="ow">and</span> <span class="s1">&#39;PSP&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">category</span><span class="p">:</span>
            <span class="c1">#all other ptmsigdb data</span>
            <span class="n">gmt_File</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="n">category</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">ptmsigdb_odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">],</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PTMsigDB </span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s2"> gmt file created at </span><span class="si">{</span><span class="n">ptmsigdb_odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;/</span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s1">.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">category</span> <span class="o">==</span> <span class="s1">&#39;PTMsigDB:KINASE-iKiP&#39;</span><span class="p">:</span>
            <span class="c1">#ikip data</span>
            <span class="n">gmt_File</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:Kinase-iKiP&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">ikip_odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">],</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PTMsigDB Kinase-iKiP gmt file created at </span><span class="si">{</span><span class="n">ikip_odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;/</span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s1">.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="s1">&#39;PSP&#39;</span> <span class="ow">in</span> <span class="n">category</span> <span class="ow">and</span> <span class="n">process_PSP_data</span><span class="p">:</span>
            <span class="c1">#check if already exists, if so skip (even if overwrite is listed)</span>
            <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">psp_odir</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s1">.gmt.gz&#39;</span><span class="p">)):</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;PhosphoSitePlus </span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s1"> gmt file already exists, skipping&#39;</span><span class="p">)</span>
                <span class="k">continue</span>

            <span class="c1">#psp data</span>
            <span class="n">gmt_File</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">tmp</span><span class="p">,</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;PSP:&#39;</span><span class="o">+</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">],</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">psp_odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">col_dict</span><span class="p">[</span><span class="n">category</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">compressed</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>





<div class="viewcode-block" id="construct_omnipath_gmt_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.construct_omnipath_gmt_file">[docs]</a>
<span class="k">def</span> <span class="nf">construct_omnipath_gmt_file</span><span class="p">(</span><span class="n">min_sources</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">min_references</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">convert_to_gene_name</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Download enzyme-substrate interactions from the OmniPath database. The data will be filtered based on the number of sources and references specified. The resulting data will be split into two categories: &#39;Writer&#39; enzymes, which add the modification, and &#39;Eraser&#39; enzymes, which remove the modification. The output will be saved as GMT files in resource files directory.</span>

<span class="sd">    Parameter</span>
<span class="sd">    ---------</span>
<span class="sd">    min_sources : int</span>
<span class="sd">        Minimum number of sources (i.e. database) for an enzyme-substrate interaction to be included. Default is 1.</span>
<span class="sd">    min_references : int</span>
<span class="sd">        Minimum number of literature references for an enzyme-substrate interactino to be included. Default is 1</span>
<span class="sd">    convert_to_gene_name : bool</span>
<span class="sd">        Whether to convert identifier to gene name</span>
<span class="sd">    odir : str or None</span>
<span class="sd">        Output directory for the GMT files. If None, the package resource file will be used. Default is None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#try importing omnipath, if not print error message prompting user to install omnipath</span>
    <span class="k">if</span> <span class="s1">&#39;omnipath&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">globals</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s1">&#39;Optional dependency `omnipath` required to run this package, but is not installed. Please run `pip install omnipath` then reimport the annotate module.&#39;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">odir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">package_dir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span><span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">odir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="s1">&#39;Resource_Files&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotations&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">)</span>
        

    <span class="c1">#check if output directory exists, if not create it</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">odir</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">odir</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    
    <span class="c1">#download enzyme-substrate data from omnipath</span>
    <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">omnipath</span><span class="o">.</span><span class="n">requests</span><span class="o">.</span><span class="n">Enzsub</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">min_references</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">min_sources</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[(</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;n_references&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_references</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;n_sources&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_sources</span><span class="p">)]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtered enzyme-substrate interactions to include only those with at least </span><span class="si">{</span><span class="n">min_references</span><span class="si">}</span><span class="s2"> references and </span><span class="si">{</span><span class="n">min_sources</span><span class="si">}</span><span class="s2"> sources. This resulted in </span><span class="si">{</span><span class="n">enzyme_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> interactions.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">convert_to_gene_name</span><span class="p">:</span>
        <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;enzyme&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
    
    <span class="c1">#format for use with ptm coordinates file</span>
    <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;substrate&#39;</span><span class="p">]</span><span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;residue_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;residue_offset&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;modification&quot;</span><span class="p">:</span> <span class="s2">&quot;Modification Class&quot;</span><span class="p">})</span>

    <span class="c1">#define addition and removal modifications</span>
    <span class="n">addition</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;acetylation&#39;</span><span class="p">,</span><span class="s1">&#39;methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;palmitoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;polyubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;carboxylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sulfation&#39;</span><span class="p">,</span> <span class="s1">&#39;sumoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;trimethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;neddylation&#39;</span><span class="p">,</span> <span class="s1">&#39;monoubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;myristoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;amidation&#39;</span><span class="p">,</span> <span class="s1">&#39;alkylation&#39;</span><span class="p">,</span> <span class="s1">&#39;prenylation&#39;</span><span class="p">,</span> <span class="s1">&#39;glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;hydroxylation&#39;</span><span class="p">]</span>
    <span class="n">removal</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dephosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;demethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deacetylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;deglycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deacylation&#39;</span><span class="p">,</span> <span class="s1">&#39;desumoylation&#39;</span><span class="p">]</span>

    <span class="c1"># grab writer enzymes and reformat for use with ptm coordinates file, then merge</span>
    <span class="n">addition_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">addition</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1">#convert modification class to string and rename</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s1">&#39;polyubiquitination&#39;</span><span class="p">:</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;monoubiquitination&#39;</span><span class="p">:</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;trimethylation&#39;</span><span class="p">:</span> <span class="s1">&#39;methylation&#39;</span><span class="p">})</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>
    <span class="n">addition_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">addition_data</span><span class="p">,</span> <span class="s1">&#39;enzyme&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Writer_Enzyme&#39;</span><span class="p">,</span> <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#grab eraser enzymes and reformat for use with ptm coordinates file</span>
    <span class="n">removal_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">removal</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># remove &#39;de&#39; from descriptor and reformat to match ptm coordinates</span>
    <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;de&#39;</span><span class="p">)</span>
    <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>
    <span class="n">removal_gmt</span> <span class="o">=</span> <span class="n">construct_gmt_df</span><span class="p">(</span><span class="n">removal_data</span><span class="p">,</span> <span class="s1">&#39;enzyme&#39;</span><span class="p">,</span> <span class="n">description</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="n">odir</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;Eraser_Enzyme&#39;</span><span class="p">,</span> <span class="n">compressed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;OmniPath Writer Enzyme gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;/Writer_Enzyme.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;OmniPath Eraser Enzyme gmt file created at </span><span class="si">{</span><span class="n">odir</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="sa">f</span><span class="s1">&#39;/Eraser_Enzyme.gmt.gz&#39;</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>






<div class="viewcode-block" id="add_omnipath_data">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_omnipath_data">[docs]</a>
<span class="k">def</span> <span class="nf">add_omnipath_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">min_sources</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">min_references</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">convert_to_gene_name</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">replace_old_annotations</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptms or altered flanks dataframe, append enzyme-substrate interactions recorded in OmniPath database. These will be split between &#39;Writer&#39; enzymes, or enzymes that add the modification (OmniPath:Writer Enzyme), and &#39;Eraser&#39; enzymes, or enzymes that remove the modification (OmniPath:Eraser Enzyme). Note, we do not consider the &#39;post translational modification&#39; or &#39;cleavage&#39; entries for this purpose.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms : pandas.DataFrame</span>
<span class="sd">        Spliced PTMs or altered flanks dataframe.</span>
<span class="sd">    min_sources : int</span>
<span class="sd">        Minimum number of sources (i.e. database) for enzyme-substrate interaction. Default is 1, or all entries.</span>
<span class="sd">    min_references : int</span>
<span class="sd">        Minimum number of references (i.e. publications) for enzyme-substrate interaction. Default is 1, or all entries.</span>
<span class="sd">    convert_to_gene_name : bool</span>
<span class="sd">        Whether to convert enzyme names from UniProt IDs to gene names using pose_config.uniprot_to_genename. Default is True.</span>
<span class="sd">    report_success : bool</span>
<span class="sd">        Whether to report success message. Default is True.</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#try importing omnipath, if not print error message prompting user to install omnipath</span>
    <span class="k">if</span> <span class="s1">&#39;omnipath&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">globals</span><span class="p">():</span>
        <span class="k">raise</span> <span class="ne">ImportError</span><span class="p">(</span><span class="s1">&#39;Optional dependency `omnipath` required to run this package, but is not installed. Please run `pip install omnipath` then reimport the annotate module.&#39;</span><span class="p">)</span>
    
    <span class="c1">#check if annotations are already present, if so remove them or keep them based on user input (replace_old_annotations)</span>
    <span class="k">if</span> <span class="p">(</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="n">replace_old_annotations</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Found old omnipath annotations, removing them to replace with new data&#39;</span><span class="p">)</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">])</span>
    <span class="k">elif</span> <span class="p">(</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">replace_old_annotations</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Found old omnipath annotations, skipping. To replace them, set replace_old_annotations = True&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ptms</span>
    




    <span class="c1">#add ptm column to add modification information to PTM data for merge</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">add_ptm_column</span><span class="p">(</span><span class="n">ptms</span><span class="p">)</span>
    
    <span class="c1">#download enzyme-substrate data from omnipath</span>
    <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">omnipath</span><span class="o">.</span><span class="n">requests</span><span class="o">.</span><span class="n">Enzsub</span><span class="p">()</span><span class="o">.</span><span class="n">get</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">min_references</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">or</span> <span class="n">min_sources</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[(</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;n_references&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_references</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;n_sources&#39;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">min_sources</span><span class="p">)]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtered enzyme-substrate interactions to include only those with at least </span><span class="si">{</span><span class="n">min_references</span><span class="si">}</span><span class="s2"> references and </span><span class="si">{</span><span class="n">min_sources</span><span class="si">}</span><span class="s2"> sources. This resulted in </span><span class="si">{</span><span class="n">enzyme_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s2"> interactions.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">convert_to_gene_name</span><span class="p">:</span>
        <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;enzyme&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
    
    <span class="c1">#format for use with ptm coordinates file</span>
    <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;substrate&#39;</span><span class="p">]</span><span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;residue_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;residue_offset&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">enzyme_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;modification&quot;</span><span class="p">:</span> <span class="s2">&quot;Modification Class&quot;</span><span class="p">})</span>

    <span class="c1">#define addition and removal modifications</span>
    <span class="n">addition</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;acetylation&#39;</span><span class="p">,</span><span class="s1">&#39;methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;palmitoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;polyubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;carboxylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sulfation&#39;</span><span class="p">,</span> <span class="s1">&#39;sumoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;trimethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;neddylation&#39;</span><span class="p">,</span> <span class="s1">&#39;monoubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;myristoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;amidation&#39;</span><span class="p">,</span> <span class="s1">&#39;alkylation&#39;</span><span class="p">,</span> <span class="s1">&#39;prenylation&#39;</span><span class="p">,</span> <span class="s1">&#39;glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;hydroxylation&#39;</span><span class="p">]</span>
    <span class="n">removal</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dephosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;demethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deacetylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;deglycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;deacylation&#39;</span><span class="p">,</span> <span class="s1">&#39;desumoylation&#39;</span><span class="p">]</span>
    
    <span class="c1">#grab original size of dataset</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># grab writer enzymes and reformat for use with ptm coordinates file, then merge</span>
    <span class="n">addition_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">addition</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1">#convert modification class to string and rename</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">({</span><span class="s1">&#39;polyubiquitination&#39;</span><span class="p">:</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;monoubiquitination&#39;</span><span class="p">:</span> <span class="s1">&#39;ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;trimethylation&#39;</span><span class="p">:</span> <span class="s1">&#39;methylation&#39;</span><span class="p">})</span>
    <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">addition_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="n">addition_data</span> <span class="o">=</span> <span class="n">addition_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">})</span>
    <span class="n">addition_data</span> <span class="o">=</span> <span class="n">addition_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;PTM&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])[</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">addition_data</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">],</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>

    <span class="c1">#grab eraser enzymes and reformat for use with ptm coordinates file, then merge</span>
    <span class="n">removal_data</span> <span class="o">=</span> <span class="n">enzyme_data</span><span class="p">[</span><span class="n">enzyme_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">removal</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># remove &#39;de&#39; from descriptor and reformat to match ptm coordinates</span>
    <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;de&#39;</span><span class="p">)</span>
    <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">removal_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
    <span class="n">removal_data</span> <span class="o">=</span> <span class="n">removal_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">})</span>
    <span class="n">removal_data</span> <span class="o">=</span> <span class="n">removal_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;PTM&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])[</span><span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">removal_data</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">],</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>

    
    <span class="c1">#check size</span>
    <span class="k">if</span> <span class="n">ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataset size changed upon merge, please make sure there are no duplicates in spliced ptms data&#39;</span><span class="p">)</span>

    <span class="c1">#report success message</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="c1">#get number of ptms associated with writer enzymes (specify how many phosphorylation modifications are present)</span>
        <span class="n">num_ptms_with_writer</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;PTM&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;Phosphorylation&#39;</span> <span class="ow">in</span> <span class="n">num_ptms_with_writer</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">):</span>
            <span class="n">num_ptms_with_writer_phospho</span> <span class="o">=</span> <span class="n">num_ptms_with_writer</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s1">&#39;Phosphorylation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">num_ptms_with_writer_phospho</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">num_ptms_with_writer</span> <span class="o">=</span> <span class="n">num_ptms_with_writer</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1">#get number of ptms associated with eraser enzymes (specify how many phosphorylation modifications are present)</span>
        <span class="n">num_ptms_with_eraser</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;PTM&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;Phosphorylation&#39;</span> <span class="ow">in</span> <span class="n">num_ptms_with_eraser</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">get_level_values</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">):</span>
            <span class="n">num_ptms_with_eraser_phospho</span> <span class="o">=</span> <span class="n">num_ptms_with_eraser</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span><span class="s1">&#39;Phosphorylation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">num_ptms_with_eraser_phospho</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">num_ptms_with_eraser</span> <span class="o">=</span> <span class="n">num_ptms_with_eraser</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="c1">#report the number of PTMs associated with writer and eraser enzymes</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;OmniPath Enzyme-Substrate interactions added:</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_writer</span><span class="si">}</span><span class="s2"> PTMs associated with writer enzyme (</span><span class="si">{</span><span class="n">num_ptms_with_writer_phospho</span><span class="si">}</span><span class="s2"> phosphorylation)</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_eraser</span><span class="si">}</span><span class="s2"> PTMs associated with eraser enzyme (</span><span class="si">{</span><span class="n">num_ptms_with_eraser_phospho</span><span class="si">}</span><span class="s2"> phosphorylation)&quot;</span><span class="p">)</span>

    <span class="c1">#remove PTM column</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">ptms</span></div>




<span class="c1">######### Functions for combining annotations from multiple sources ########</span>

<div class="viewcode-block" id="convert_PSP_label_to_UniProt">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.convert_PSP_label_to_UniProt">[docs]</a>
<span class="k">def</span> <span class="nf">convert_PSP_label_to_UniProt</span><span class="p">(</span><span class="n">label</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a label for an interacting protein from PhosphoSitePlus, convert to UniProtKB accession. Required as PhosphoSitePlus interactions are recorded in various ways that aren&#39;t necessarily consistent with other databases (i.e. not always gene name)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    label: str</span>
<span class="sd">        Label for interacting protein from PhosphoSitePlus</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="c1">#using uniprot to gene name dict, construct dict to go the other direction (gene name to uniprot id)</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>


    <span class="c1">#remove isoform label if present</span>
    <span class="k">if</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span> <span class="c1">#if PSP name is gene name found in uniprot</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">:</span> <span class="c1"># if PSP name is not gene name, but is in conversion dictionary</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">[</span><span class="n">label</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span> <span class="c1">#otherwise note that gene was missed</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span></div>

        <span class="c1">#missed_genes.append(gene)</span>

<div class="viewcode-block" id="extract_interaction_details">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.extract_interaction_details">[docs]</a>
<span class="k">def</span> <span class="nf">extract_interaction_details</span><span class="p">(</span><span class="n">interaction</span><span class="p">,</span> <span class="n">column</span> <span class="o">=</span> <span class="s2">&quot;PhosphoSitePlus:Interactions&quot;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given an interaction string from a specific database, extract the type of interaction and the interacting protein. This is required as different databases format their interaction strings differently.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">interaction_types</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PTMcode:Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;INDUCES&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus:Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD:Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;Combined:Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;UNCLEAR&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus:Interactions&#39;</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PTMInt:Interactions&#39;</span><span class="p">:</span>
        <span class="n">ptmint_type_conversion</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Inhibit&#39;</span><span class="p">:</span><span class="s1">&#39;DISRUPTS&#39;</span><span class="p">,</span> <span class="s1">&#39;Enhance&#39;</span><span class="p">:</span><span class="s2">&quot;INDUCES&quot;</span><span class="p">}</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">ptmint_type_conversion</span><span class="p">[</span><span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PTMcode:Interactions&#39;</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;INDUCES&#39;</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">interaction_types</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span>

    <span class="k">return</span> <span class="nb">type</span><span class="p">,</span> <span class="n">protein</span></div>


<div class="viewcode-block" id="unify_interaction_data">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.unify_interaction_data">[docs]</a>
<span class="k">def</span> <span class="nf">unify_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">interaction_col</span><span class="p">,</span> <span class="n">name_dict</span> <span class="o">=</span> <span class="p">{}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm data and a column containing interaction data, extract the interacting protein, type of interaction, and convert to UniProtKB accession. This will be added as a new column labeled &#39;Interacting ID&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Dataframe containing PTM data</span>
<span class="sd">    interaction_col: str</span>
<span class="sd">        column containing interaction information from a specific database</span>
<span class="sd">    name_dict: dict</span>
<span class="sd">        dictionary to convert names within given database to UniProt IDs. For cases when name is not necessarily one of the gene names listed in UniProt</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    interact: pd.DataFrame</span>
<span class="sd">        Contains PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="c1">#using uniprot to gene name dict, construct dict to go the other direction (gene name to uniprot id)</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>

    <span class="c1">#extract PSP data from annotated PTMs, separate cases in which single PTM has multipe interactions</span>
    <span class="n">data_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Significance&#39;</span><span class="p">,</span> <span class="s1">&#39;dPSI&#39;</span><span class="p">]]</span>


    <span class="n">interact</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">interaction_col</span><span class="p">)[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span><span class="n">interaction_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">data_cols</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">interact</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs associated with </span><span class="si">{</span><span class="n">interaction_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">interact</span>
    
    <span class="n">interact</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">interaction_col</span><span class="p">)</span>

    <span class="c1">#extract protein and type of interaction (currently for phosphosite plus)</span>
    <span class="nb">type</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">protein</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">interact</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">processed</span> <span class="o">=</span> <span class="n">extract_interaction_details</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">],</span> <span class="n">interaction_col</span><span class="p">)</span>
        <span class="nb">type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">processed</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">protein</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">processed</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span>  <span class="o">=</span> <span class="nb">type</span>
    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">protein</span>
        

    <span class="c1">#convert interacting protein to uniprot id for databases that are not reported in uniprot ids</span>
    <span class="k">if</span> <span class="n">interaction_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;PTMcode:Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt:Interactions&#39;</span><span class="p">]:</span>
        <span class="n">interacting_id</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">missed_genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]:</span>
            <span class="c1">#remove isoform label if present</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span> <span class="c1">#if PSP name is gene name found in uniprot</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">name_dict</span><span class="p">:</span> <span class="c1"># if PSP name is not gene name, but is in conversion dictionary</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name_dict</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span> <span class="c1">#otherwise note that gene was missed</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
                <span class="n">missed_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1">#save information</span>
        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interacting_id</span>
        <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
        <span class="c1">#make sure there are interactions</span>
        <span class="k">if</span> <span class="n">interact</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">interact</span>

        <span class="c1">#check if there multiple in one row</span>
        <span class="k">if</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
            <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span>
    

    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">interaction_col</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">interact</span></div>


<div class="viewcode-block" id="combine_interaction_data">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.combine_interaction_data">[docs]</a>
<span class="k">def</span> <span class="nf">combine_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">interaction_databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">],</span> <span class="n">include_enzyme_interactions</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given annotated spliced ptm data, extract interaction data from various databases and combine into a single dataframe. This will include the interacting protein, the type of interaction, and the source of the interaction data</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Dataframe containing PTM data and associated interaction annotations from various databases</span>
<span class="sd">    interaction_databases: list</span>
<span class="sd">        List of databases to extract interaction data from. Options include &#39;PhosphoSitePlus&#39;, &#39;PTMcode&#39;, &#39;PTMInt&#39;, &#39;RegPhos&#39;, &#39;DEPOD&#39;. These should already have annotation columns in the ptms dataframe, otherwise they will be ignored. For kinase-substrate interactions, if combined column is present, will use that instead of individual databases</span>
<span class="sd">    include_enzyme_interactions: bool</span>
<span class="sd">        If True, will include kinase-substrate and phosphatase interactions in the output dataframe</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    interact_data: list</span>
<span class="sd">        List of dataframes containing PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES), and the source of the interaction data</span>

<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="c1">#go through and add interaction information for any database that are not appended to the spliced_ptms dataframe</span>
    <span class="n">available_annotations</span> <span class="o">=</span> <span class="n">get_available_annotations</span><span class="p">(</span><span class="n">ptms</span><span class="p">)</span>
    <span class="n">available_interactions</span> <span class="o">=</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">available_annotations</span><span class="p">[</span><span class="s1">&#39;Annotation Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">include_enzyme_interactions</span><span class="p">:</span>
        <span class="n">enzyme_annotations</span> <span class="o">=</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">available_annotations</span><span class="p">[</span><span class="s1">&#39;Annotation Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">]</span>

    <span class="c1">#go through and add interaction information that is not in the ptms dataframe</span>
    <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">interaction_databases</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s2">&quot;OmniPath&quot;</span> <span class="ow">and</span> <span class="n">include_enzyme_interactions</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">db</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Writer Enzyme&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Eraser Enzyme&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">:Interactions&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">available_interactions</span><span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">db</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">)</span>

            <span class="k">if</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">:Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">include_enzyme_interactions</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">enzyme_annotations</span><span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
                    <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">db</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Combining interaction information from </span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">interaction_databases</span><span class="p">)</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">interact_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">database</span> <span class="ow">in</span> <span class="n">interaction_databases</span><span class="p">:</span>
        <span class="k">if</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Interactions&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Writer_Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Eraser_Enzyme&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No interaction data found for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">, skipping&quot;</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">interaction_col</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Interactions&#39;</span>
            <span class="k">if</span> <span class="n">interaction_col</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">ptms</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                    <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">interaction_col</span><span class="p">,</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">)</span>
                    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                    <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">database</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM&#39;</span><span class="p">]:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs with interaction information from </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">, skipping&quot;</span><span class="p">)</span>
        
            <span class="k">if</span> <span class="n">include_enzyme_interactions</span><span class="p">:</span>
                <span class="c1">#dictionary to convert kinase names to gene names</span>
                <span class="n">ks_genes_to_uniprot</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABL1(ABL)&#39;</span><span class="p">:</span><span class="s1">&#39;P00519&#39;</span><span class="p">,</span> <span class="s1">&#39;ACK&#39;</span><span class="p">:</span><span class="s1">&#39;Q07912&#39;</span><span class="p">,</span> <span class="s1">&#39;AURC&#39;</span><span class="p">:</span><span class="s1">&#39;Q9UQB9&#39;</span><span class="p">,</span> <span class="s1">&#39;ERK1(MAPK3)&#39;</span><span class="p">:</span><span class="s1">&#39;P27361&#39;</span><span class="p">,</span><span class="s1">&#39;ERK2(MAPK1)&#39;</span><span class="p">:</span><span class="s1">&#39;P28482&#39;</span><span class="p">,</span>  <span class="s1">&#39;ERK5(MAPK7)&#39;</span><span class="p">:</span><span class="s1">&#39;Q13164&#39;</span><span class="p">,</span><span class="s1">&#39;JNK1(MAPK8)&#39;</span><span class="p">:</span><span class="s1">&#39;P45983&#39;</span><span class="p">,</span> <span class="s1">&#39;CK1A&#39;</span><span class="p">:</span><span class="s1">&#39;P48729&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK2(MAPK9)&#39;</span><span class="p">:</span><span class="s1">&#39;P45984&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK3(MAPK10)&#39;</span><span class="p">:</span><span class="s1">&#39;P53779&#39;</span><span class="p">,</span> <span class="s1">&#39;P38A(MAPK14)&#39;</span><span class="p">:</span><span class="s1">&#39;Q16539&#39;</span><span class="p">,</span><span class="s1">&#39;P38B(MAPK11)&#39;</span><span class="p">:</span><span class="s1">&#39;Q15759&#39;</span><span class="p">,</span> <span class="s1">&#39;P38G(MAPK12)&#39;</span><span class="p">:</span><span class="s1">&#39;P53778&#39;</span><span class="p">,</span><span class="s1">&#39;P70S6K&#39;</span> <span class="p">:</span><span class="s1">&#39;Q9UBS0&#39;</span><span class="p">,</span> <span class="s1">&#39;PAK&#39;</span><span class="p">:</span><span class="s1">&#39;Q13153&#39;</span><span class="p">,</span> <span class="s1">&#39;PKCZ&#39;</span><span class="p">:</span><span class="s1">&#39;Q05513&#39;</span><span class="p">,</span> <span class="s1">&#39;CK2A&#39;</span><span class="p">:</span><span class="s1">&#39;P19784&#39;</span><span class="p">,</span> <span class="s1">&#39;ABL2&#39;</span><span class="p">:</span><span class="s1">&#39;P42684&#39;</span><span class="p">,</span> <span class="s1">&#39;AMPKA1&#39;</span><span class="p">:</span><span class="s1">&#39;Q13131&#39;</span><span class="p">,</span> <span class="s1">&#39;AMPKA2&#39;</span><span class="p">:</span><span class="s1">&#39;Q13131&#39;</span><span class="p">,</span> <span class="s1">&#39;AURB&#39;</span><span class="p">:</span><span class="s1">&#39;Q96GD4&#39;</span><span class="p">,</span> <span class="s1">&#39;CAMK1A&#39;</span><span class="p">:</span><span class="s1">&#39;Q14012&#39;</span><span class="p">,</span> <span class="s1">&#39;CDC42BP&#39;</span><span class="p">:</span><span class="s1">&#39;Q9Y5S2&#39;</span><span class="p">,</span><span class="s1">&#39;CK1D&#39;</span><span class="p">:</span><span class="s1">&#39;P48730&#39;</span><span class="p">,</span><span class="s1">&#39;CK1E&#39;</span><span class="p">:</span><span class="s1">&#39;P49674&#39;</span><span class="p">,</span><span class="s1">&#39;CK2B&#39;</span><span class="p">:</span><span class="s1">&#39;P67870&#39;</span><span class="p">,</span><span class="s1">&#39;DMPK1&#39;</span><span class="p">:</span><span class="s1">&#39;Q09013&#39;</span><span class="p">,</span> <span class="s1">&#39;DNAPK&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span><span class="s1">&#39;DSDNA KINASE&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span> <span class="s1">&#39;EG3 KINASE&#39;</span><span class="p">:</span><span class="s1">&#39;P49840&#39;</span><span class="p">,</span><span class="s1">&#39;ERK3(MAPK6)&#39;</span><span class="p">:</span><span class="s1">&#39;Q16659&#39;</span><span class="p">,</span><span class="s1">&#39;GSK3&#39;</span><span class="p">:</span><span class="s1">&#39;P49840&#39;</span><span class="p">,</span> <span class="s1">&#39;MRCKA&#39;</span><span class="p">:</span><span class="s1">&#39;Q5VT25&#39;</span><span class="p">,</span> <span class="s1">&#39;P38D(MAPK13)&#39;</span><span class="p">:</span><span class="s1">&#39;O15264&#39;</span><span class="p">,</span><span class="s1">&#39;P70S6KB&#39;</span><span class="p">:</span><span class="s1">&#39;Q9UBS0&#39;</span><span class="p">,</span><span class="s1">&#39;PDKC&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span><span class="s1">&#39;PKCH&#39;</span><span class="p">:</span><span class="s1">&#39;P24723&#39;</span><span class="p">,</span><span class="s1">&#39;PKCI&#39;</span><span class="p">:</span><span class="s1">&#39;P41743&#39;</span><span class="p">,</span><span class="s1">&#39;PKCT&#39;</span><span class="p">:</span><span class="s1">&#39;Q04759&#39;</span><span class="p">,</span><span class="s1">&#39;PKD3&#39;</span><span class="p">:</span><span class="s1">&#39;O94806&#39;</span><span class="p">,</span><span class="s1">&#39;PKG1&#39;</span><span class="p">:</span><span class="s1">&#39;Q13976&#39;</span><span class="p">,</span><span class="s1">&#39;PKG2&#39;</span><span class="p">:</span><span class="s1">&#39;Q13237&#39;</span><span class="p">,</span><span class="s1">&#39;SMMLCK&#39;</span><span class="p">:</span><span class="s1">&#39;Q15746&#39;</span><span class="p">}</span>

                <span class="n">enzyme_col</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Enzyme&#39;</span>
                <span class="k">if</span> <span class="n">enzyme_col</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">ptms</span><span class="p">[</span><span class="n">enzyme_col</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">enzyme_col</span><span class="p">,</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">)</span>
                        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                        <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                    <span class="k">elif</span> <span class="n">database</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">]:</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs with enzyme information from </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2">, skipping&quot;</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span> <span class="ow">and</span> <span class="p">(</span><span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">):</span>
                    <span class="k">if</span> <span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="s1">&#39;OmniPath:Writer_Enzyme&#39;</span><span class="p">,</span> <span class="n">ks_genes_to_uniprot</span><span class="p">)</span>
                        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                        <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                    <span class="k">if</span> <span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="s1">&#39;OmniPath:Eraser_Enzyme&#39;</span><span class="p">,</span> <span class="n">ks_genes_to_uniprot</span><span class="p">)</span>
                        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                        <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs with enzyme information from OmniPath found, skipping&quot;</span><span class="p">)</span>



    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">interact_data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">interact_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">interact_data</span><span class="p">)</span>
        <span class="n">extra_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="s1">&#39;Significance&#39;</span><span class="p">]]</span>
        <span class="n">interact_data</span> <span class="o">=</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">extra_cols</span><span class="p">,</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">helpers</span><span class="o">.</span><span class="n">join_unique_entries</span><span class="p">)</span>
    
        <span class="c1">#convert uniprot ids back to gene names for interpretability</span>
        <span class="n">ptm_gene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interacting_gene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">ptm_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">ptm_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">])</span>
            <span class="n">interacting_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">interacting_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">])</span>
        <span class="n">interact_data</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_gene</span>
        <span class="n">interact_data</span><span class="p">[</span><span class="s2">&quot;Interacting Gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interacting_gene</span>
  
  
        <span class="k">return</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span></div>




<div class="viewcode-block" id="combine_enzyme_data">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.combine_enzyme_data">[docs]</a>
<span class="k">def</span> <span class="nf">combine_enzyme_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">enzyme_databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">],</span> <span class="n">regphos_conversion</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ERK1(MAPK3)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK3&#39;</span><span class="p">,</span> <span class="s1">&#39;ERK2(MAPK1)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK2(MAPK9)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK9&#39;</span><span class="p">,</span><span class="s1">&#39;CDC2&#39;</span><span class="p">:</span><span class="s1">&#39;CDK1&#39;</span><span class="p">,</span> <span class="s1">&#39;CK2A1&#39;</span><span class="p">:</span><span class="s1">&#39;CSNK2A1&#39;</span><span class="p">,</span> <span class="s1">&#39;PKACA&#39;</span><span class="p">:</span><span class="s1">&#39;PRKACA&#39;</span><span class="p">,</span> <span class="s1">&#39;ABL1(ABL)&#39;</span><span class="p">:</span><span class="s1">&#39;ABL1&#39;</span><span class="p">}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm information, combine enzyme-substrate data from multiple databases (currently support PhosphoSitePlus, RegPhos, OmniPath, DEPOD, and iKiP downloaded from PTMsigDB), assuming that the enzyme data from these resources has already been added to the spliced ptm data. The combined kinase data will be added as a new column labeled &#39;Combined:Writer Enzyme&#39; and &#39;Combined:Eraser Enzyme&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Spliced PTM data from project module</span>
<span class="sd">    enzyme_databases: list</span>
<span class="sd">        List of databases to combine enzyme data from. Currently support PhosphoSitePlus, RegPhos, OmniPath, and DEPOD</span>
<span class="sd">    regphos_conversion: dict</span>
<span class="sd">        Allows conversion of RegPhos names to matching names in PhosphoSitePlus.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        PTM data with combined kinase data added</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>

    <span class="c1">#remove databases without kinase data in spliced ptms</span>
    <span class="n">writer_databases</span> <span class="o">=</span> <span class="p">[</span><span class="n">db</span> <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">enzyme_databases</span> <span class="k">if</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39;:Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39;:Writer Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">writer_databases</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1 or fewer writer enzyme data columns found in spliced PTMs, skipping&#39;</span><span class="p">)</span>
        <span class="n">skip_writer</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">skip_writer</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">eraser_databases</span> <span class="o">=</span> <span class="p">[</span><span class="n">db</span> <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">enzyme_databases</span> <span class="k">if</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39;:Eraser Enzyme&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">or</span> <span class="p">(</span><span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span> <span class="ow">and</span> <span class="n">db</span> <span class="o">+</span> <span class="s1">&#39;:Enzyme&#39;</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">eraser_databases</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;1 or fewer eraser enzyme data columns found in spliced PTMs, skipping&#39;</span><span class="p">)</span>
        <span class="n">skip_eraser</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">skip_eraser</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="n">writer_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">eraser_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_writer</span><span class="p">:</span>
            <span class="n">combined_writer</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">writer_databases</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:</span>
                    <span class="n">psp</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PhosphoSitePlus:Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PhosphoSitePlus:Enzyme&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PhosphoSitePlus:Enzyme&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="c1">#convert PSP names to a common name (first gene name provided by uniprot)</span>
                    <span class="n">psp</span> <span class="o">=</span> <span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">kin</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>  <span class="k">if</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="k">else</span> <span class="n">kin</span> <span class="k">for</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">psp</span><span class="p">]</span>
                    <span class="n">combined_writer</span> <span class="o">+=</span> <span class="n">psp</span>
                <span class="k">elif</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
                    <span class="n">regphos</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Enzyme&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Enzyme&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rp</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regphos</span><span class="p">):</span>
                        <span class="k">if</span> <span class="n">rp</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                            <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">rp</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="k">elif</span> <span class="n">rp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                            <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">rp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="k">elif</span> <span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">regphos_conversion</span><span class="p">:</span>
                            <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">regphos_conversion</span><span class="p">[</span><span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                    <span class="n">combined_writer</span> <span class="o">+=</span> <span class="n">regphos</span>
                <span class="k">elif</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">:</span>
                    <span class="n">omni</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="n">combined_writer</span> <span class="o">+=</span> <span class="n">omni</span>
                <span class="k">elif</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;iKiP&#39;</span><span class="p">:</span>
                    <span class="n">ikip</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;iKiP:Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;iKiP:Enzyme&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;iKiP:Enzyme&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="c1">#remove directionality from iKiP data (i.e. &#39;kinase-&gt;u&#39; to &#39;kinase&#39; and grab first provided gene name</span>
                    <span class="n">ikip</span> <span class="o">=</span> <span class="p">[</span><span class="n">kin</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">ikip</span><span class="p">]</span>
                    <span class="c1">#convert iKiP names to a common name (first gene name provided by uniprot)</span>
                    <span class="n">ikip</span> <span class="o">=</span> <span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">kin</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>  <span class="k">if</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="k">else</span> <span class="n">kin</span> <span class="k">for</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">ikip</span><span class="p">]</span>
                    <span class="n">combined_writer</span> <span class="o">+=</span> <span class="n">ikip</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">combined_writer</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">writer_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">combined_writer</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">writer_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        

        <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_eraser</span><span class="p">:</span>
            <span class="n">combined_eraser</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">eraser_databases</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:</span>
                    <span class="n">depod</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="n">combined_eraser</span> <span class="o">+=</span> <span class="n">depod</span>
                <span class="k">elif</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">:</span>
                    <span class="n">omni</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                    <span class="n">combined_eraser</span> <span class="o">+=</span> <span class="n">omni</span>

            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">combined_eraser</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">eraser_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">combined_eraser</span><span class="p">)))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">eraser_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_writer</span><span class="p">:</span>
        <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Writer_Enzyme&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">writer_data</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_eraser</span><span class="p">:</span>
        <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Eraser_Enzyme&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">eraser_data</span>

    <span class="k">return</span> <span class="n">ptms</span></div>



<div class="viewcode-block" id="check_file">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.check_file">[docs]</a>
<span class="k">def</span> <span class="nf">check_file</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.tsv&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a file name, check if the file exists and has the expected extension. If the file does not exist or has the wrong extension, raise an error.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fname: str</span>
<span class="sd">        File name to check</span>
<span class="sd">    expected_extension: str</span>
<span class="sd">        Expected file extension. Default is &#39;.tsv&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">fname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Annotation file path must be provided&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;File </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1"> not found&#39;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fname</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">expected_extension</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;File </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1"> does not have the expected extension (</span><span class="si">{</span><span class="n">expected_extension</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span></div>

    

<span class="k">def</span> <span class="nf">annotate_ptms_with_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMint&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="s1">&#39;iKiP&#39;</span><span class="p">],</span> <span class="n">annot_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;Pathway&#39;</span><span class="p">],</span> <span class="n">gmt_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">missing_annotation</span> <span class="o">=</span> <span class="s1">&#39;notify&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="n">available_annotations</span> <span class="o">=</span> <span class="n">get_available_gmt_annotations</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">databases</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">db</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">available_annotations</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">missing_annotation</span> <span class="o">==</span> <span class="s1">&#39;raise&#39;</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Database </span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2"> not found in available annotations&quot;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">missing_annotation</span> <span class="o">==</span> <span class="s1">&#39;notify&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s2"> not found in available annotations, skipping</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">continue</span>
            <span class="k">elif</span> <span class="n">missing_annotation</span> <span class="o">==</span> <span class="s1">&#39;ignore&#39;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid value for missing_annotation: </span><span class="si">{</span><span class="n">missing_annotation</span><span class="si">}</span><span class="s2">. This indicates how to handle cases where annotation information could not be found. Options are &#39;raise&#39;, &#39;notify&#39;, or &#39;ignore&#39;&quot;</span><span class="p">)</span>
        
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">:&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">annot_types</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Perturbation&#39;</span> <span class="ow">or</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Pathway&#39;</span><span class="p">:</span>
                    <span class="n">atype_annots</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">db</span><span class="p">]</span> <span class="k">if</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">a</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">atype_annots</span><span class="p">:</span>
                        <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">db</span><span class="p">,</span> <span class="n">annot_type</span><span class="o">=</span><span class="n">a</span><span class="p">)</span>
                        <span class="c1">#report number of annotations added</span>
                        <span class="n">num_added</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">-&gt;</span><span class="si">{</span><span class="n">num_added</span><span class="si">}</span><span class="s2"> PTMs with </span><span class="si">{</span><span class="n">a</span><span class="si">}</span><span class="s2"> annotations&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">db</span><span class="p">]:</span>
                        <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="n">db</span><span class="p">,</span> <span class="n">annot_type</span><span class="o">=</span><span class="n">atype</span><span class="p">)</span>
                        <span class="c1">#report number of annotations added</span>
                        <span class="n">num_added</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">db</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">-&gt;</span><span class="si">{</span><span class="n">num_added</span><span class="si">}</span><span class="s2"> PTMs with </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s2"> annotations&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">ptms</span>
            
            



<div class="viewcode-block" id="annotate_ptms">
<a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.annotate_ptms">[docs]</a>
<span class="k">def</span> <span class="nf">annotate_ptms</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="n">phosphositeplus</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ptmcode</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ptmint</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">omnipath</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">regphos</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">depod</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">elm</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">interactions_to_combine</span> <span class="o">=</span> <span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="n">enzymes_to_combine</span> <span class="o">=</span> <span class="s2">&quot;All&quot;</span><span class="p">,</span> <span class="n">combine_similar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm data, add annotations from various databases. The annotations that can be added are the following:</span>
<span class="sd">    </span>
<span class="sd">    PhosphoSitePlus: regulatory site data (file must be provided), kinase-substrate data (file must be provided), and disease association data (file must be provided)</span>
<span class="sd">    ELM: interaction data (can be downloaded automatically or provided as a file), motif matches (elm class data can be downloaded automatically or provided as a file)</span>
<span class="sd">    PTMInt: interaction data (will be downloaded automatically)</span>
<span class="sd">    PTMcode: intraprotein interactions (can be downloaded automatically or provided as a file), interprotein interactions (can be downloaded automatically or provided as a file)</span>
<span class="sd">    DEPOD: phosphatase-substrate data (will be downloaded automatically)</span>
<span class="sd">    RegPhos: kinase-substrate data (will be downloaded automatically)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Spliced PTM data from project module</span>
<span class="sd">    psp_regulatory : bool</span>

<span class="sd">    interactions_to_combine: list</span>
<span class="sd">        List of databases to combine interaction data from. Default is [&#39;PTMcode&#39;, &#39;PhosphoSitePlus&#39;, &#39;RegPhos&#39;, &#39;PTMInt&#39;]</span>
<span class="sd">    kinases_to_combine: list</span>
<span class="sd">        List of databases to combine kinase-substrate data from. Default is [&#39;PhosphoSitePlus&#39;, &#39;RegPhos&#39;]</span>
<span class="sd">    combine_similar: bool</span>
<span class="sd">        Whether to combine annotations of similar information (kinase, interactions, etc) from multiple databases into another column labeled as &#39;Combined&#39;. Default is True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#go through and add interaction information for any database that are not appended to the spliced_ptms dataframe</span>
    <span class="n">available_annotations</span> <span class="o">=</span> <span class="n">get_available_annotations</span><span class="p">(</span><span class="n">ptms</span><span class="p">)</span>
    <span class="n">available_annotation_dict</span> <span class="o">=</span> <span class="n">available_annotations</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Database&#39;</span><span class="p">)[</span><span class="s1">&#39;Annotation Type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">annot_type</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
        <span class="n">annot_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;Pathway&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif&#39;</span><span class="p">]</span>

    <span class="c1">#go through and add interaction information that is not in the ptms dataframe</span>
    <span class="k">for</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">annot_type</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">omnipath</span> <span class="ow">and</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Writer Enzyme&#39;</span><span class="p">)</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Eraser Enzyme&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">ptmsigdb</span> <span class="ow">and</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Perturbation-DIA&#39;</span><span class="p">)</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Perturbation-DIA2&#39;</span><span class="p">)</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Perturbation-PRM&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">ptmsigdb</span> <span class="ow">and</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Pathway&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Pathway-WikiPathways&#39;</span><span class="p">)</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Pathway-NetPath&#39;</span><span class="p">)</span>
        
        <span class="c1">#if phosphositeplus data available, add</span>
        <span class="k">if</span> <span class="n">phosphositeplus</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PhosphoSitePlus </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">ptmsigdb</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;PTMsigDB&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMsigDB </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ptmint</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;PTMInt&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMint&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMint </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ptmcode</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;PTMcode&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span> 
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMcode </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">regphos</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;RegPhos&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding RegPhos </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">depod</span> <span class="ow">and</span> <span class="n">atype</span> <span class="ow">in</span> <span class="n">available_annotation_dict</span><span class="p">[</span><span class="s1">&#39;DEPOD&#39;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">append_from_gmt</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="n">atype</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding DEPOD </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">elm</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">ptms</span> <span class="o">=</span> <span class="n">add_ELM_interactions</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>  <span class="c1"># download ELM interaction data automatically</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding ELM </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">atype</span> <span class="o">==</span> <span class="s1">&#39;Motif&#39;</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">ptms</span> <span class="o">=</span> <span class="n">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">report_success</span><span class="o">=</span><span class="n">report_success</span><span class="p">)</span>  <span class="c1"># download ELM motif matches automatically</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding ELM </span><span class="si">{</span><span class="n">atype</span><span class="si">}</span><span class="s1"> data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>


    <span class="k">if</span> <span class="n">combine_similar</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">interactions_to_combine</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
            <span class="n">interaction_annotations</span> <span class="o">=</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">available_annotations</span><span class="p">[</span><span class="s1">&#39;Annotation Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">]</span>
            <span class="n">interactions_to_combine</span> <span class="o">=</span> <span class="n">interaction_annotations</span><span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">interactions_to_combine</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">&#39;`interactions_to_combine` must either be &quot;All&quot; or a list of database names&#39;</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Combining interaction data from multiple databases&#39;</span><span class="p">)</span>
        <span class="n">interact</span> <span class="o">=</span> <span class="n">combine_interaction_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">interaction_databases</span> <span class="o">=</span> <span class="n">interactions_to_combine</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">interact</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Gene&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span>
            <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">],</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
            <span class="k">if</span> <span class="s1">&#39;Combined:Interactions&#39;</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">])</span>

            <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interact</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        <span class="k">if</span> <span class="n">enzymes_to_combine</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
            <span class="n">enzyme_annotations</span> <span class="o">=</span> <span class="n">available_annotations</span><span class="p">[</span><span class="n">available_annotations</span><span class="p">[</span><span class="s1">&#39;Annotation Type&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">]</span>
            <span class="n">enzymes_to_combine</span> <span class="o">=</span> <span class="n">enzyme_annotations</span><span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">elif</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">enzymes_to_combine</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">&#39;`enzymes_to_combine` must either be &quot;All&quot; or a list of database names&#39;</span><span class="p">)</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">combine_enzyme_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">enzyme_databases</span> <span class="o">=</span> <span class="n">enzymes_to_combine</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ptms</span></div>



</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>