

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ptm_pose.annotate &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/ptm_pose/annotate';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../General_Instructions.html">Running PTM-POSE</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../gallery_output/index.html">Types of Analysis Performed with PTM-POSE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../gallery_output/plot_protein_interactions.html">Identify protein interactions that may be impacted by splicing of PTMs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gallery_output/plot_num_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gallery_output/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../gallery_output/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Full Analysis Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for ptm_pose.annotate</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>

<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">pose_config</span><span class="p">,</span> <span class="n">helpers</span>


<span class="c1">#dictionaries for converting modification codes to modification names in PhosphoSitePlus data</span>
<span class="n">mod_shorthand_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;p&#39;</span><span class="p">:</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ca&#39;</span><span class="p">:</span><span class="s1">&#39;Caspase Cleavage&#39;</span><span class="p">,</span> <span class="s1">&#39;hy&#39;</span><span class="p">:</span><span class="s1">&#39;Hydroxylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sn&#39;</span><span class="p">:</span><span class="s1">&#39;S-Nitrosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ng&#39;</span><span class="p">:</span><span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ub&#39;</span><span class="p">:</span> <span class="s1">&#39;Ubiquitination&#39;</span><span class="p">,</span> <span class="s1">&#39;pa&#39;</span><span class="p">:</span> <span class="s2">&quot;Palmitoylation&quot;</span><span class="p">,</span><span class="s1">&#39;ne&#39;</span><span class="p">:</span><span class="s1">&#39;Neddylation&#39;</span><span class="p">,</span><span class="s1">&#39;sc&#39;</span><span class="p">:</span><span class="s1">&#39;Succinylation&#39;</span><span class="p">,</span> <span class="s1">&#39;sm&#39;</span><span class="p">:</span> <span class="s1">&#39;Sumoylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ga&#39;</span><span class="p">:</span> <span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;gl&#39;</span><span class="p">:</span> <span class="s1">&#39;Glycosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;ac&#39;</span><span class="p">:</span> <span class="s1">&#39;Acetylation&#39;</span><span class="p">,</span> <span class="s1">&#39;me&#39;</span><span class="p">:</span><span class="s1">&#39;Methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m1&#39;</span><span class="p">:</span><span class="s1">&#39;Methylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m2&#39;</span><span class="p">:</span> <span class="s1">&#39;Dimethylation&#39;</span><span class="p">,</span> <span class="s1">&#39;m3&#39;</span><span class="p">:</span><span class="s1">&#39;Trimethylation&#39;</span><span class="p">}</span>
<span class="n">residue_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;P&#39;</span><span class="p">:</span> <span class="s1">&#39;proline&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span><span class="s1">&#39;tyrosine&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">:</span><span class="s1">&#39;serine&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span><span class="s1">&#39;threonine&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">:</span><span class="s1">&#39;histidine&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">:</span><span class="s1">&#39;aspartic acid&#39;</span><span class="p">,</span> <span class="s1">&#39;I&#39;</span><span class="p">:</span><span class="s1">&#39;isoleucine&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">:</span><span class="s1">&#39;lysine&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">:</span><span class="s1">&#39;arginine&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span><span class="s1">&#39;glycine&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span><span class="s1">&#39;asparagine&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;methionine&#39;</span><span class="p">}</span>
<span class="n">annotation_col_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:{</span><span class="s1">&#39;Function&#39;</span><span class="p">:</span><span class="s1">&#39;PSP:ON_FUNCTION&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">:</span><span class="s1">&#39;PSP:ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;Disease&#39;</span><span class="p">:</span><span class="s1">&#39;PSP:Disease_Association&#39;</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">,</span><span class="s1">&#39;Perturbation&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PSP-PERT&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;ELM&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif Match&#39;</span><span class="p">:</span><span class="s1">&#39;ELM:Motif Matches&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;PTMcode&#39;</span><span class="p">:{</span><span class="s1">&#39;Intraprotein&#39;</span><span class="p">:</span><span class="s1">&#39;PTMcode:Intraprotein_Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;PTMInt&#39;</span><span class="p">:{</span><span class="s1">&#39;Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;PTMInt:Interactions&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;RegPhos&#39;</span><span class="p">:{</span><span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;DEPOD&#39;</span><span class="p">:{</span><span class="s1">&#39;Phosphatase&#39;</span><span class="p">:</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">},</span>
                        <span class="s1">&#39;PTMsigDB&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;WikiPathway&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PATH-WP&#39;</span><span class="p">,</span> <span class="s1">&#39;NetPath&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PATH-NP&#39;</span><span class="p">,</span><span class="s1">&#39;mSigDB&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PATH-BI&#39;</span><span class="p">,</span> <span class="s1">&#39;Pertubation (DIA2)&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PERT-P100-DIA2&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (DIA)&#39;</span><span class="p">:</span> <span class="s1">&#39;PTMsigDB:PERT-P100-DIA&#39;</span><span class="p">,</span> <span class="s1">&#39;Perturbation (PRM)&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:PERT-P100-PRM&#39;</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;PTMsigDB:Kinase-iKiP&#39;</span><span class="p">}}</span>



<div class="viewcode-block" id="add_custom_annotation"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_custom_annotation">[docs]</a><span class="k">def</span> <span class="nf">add_custom_annotation</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">annotation_data</span><span class="p">,</span> <span class="n">source_name</span><span class="p">,</span> <span class="n">annotation_type</span><span class="p">,</span> <span class="n">annotation_col</span><span class="p">,</span> <span class="n">accession_col</span> <span class="o">=</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="n">residue_col</span> <span class="o">=</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="n">position_col</span> <span class="o">=</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add custom annotation data to spliced_ptms or altered flanking sequence dataframes</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    annotation_data: pandas.DataFrame</span>
<span class="sd">        Dataframe containing the annotation data to be added to the spliced_ptms dataframe. Must contain columns for UniProtKB Accession, Residue, PTM Position in Canonical Isoform, and the annotation data to be added</span>
<span class="sd">    source_name: str</span>
<span class="sd">        Name of the source of the annotation data, will be used to label the columns in the spliced_ptms dataframe</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation data being added, will be used to label the columns in the spliced_ptms dataframe</span>
<span class="sd">    annotation_col: str</span>
<span class="sd">        Column name in the annotation data that contains the annotation data to be added to the spliced_ptms dataframe</span>
<span class="sd">    </span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    spliced_ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with an additional column for the custom annotation data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if annotation data contains the annotation col</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">annotation_col</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">annotation_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find column indicated to contain </span><span class="si">{</span><span class="n">annotation_col</span><span class="si">}</span><span class="s1"> in annotation data. Please either change the name of your annotation data column with this information or indicate the correct column name with the annotation_col parameter&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#make annotation col name based on source and annotation type</span>
            <span class="n">annotation_col_name</span> <span class="o">=</span> <span class="n">source_name</span> <span class="o">+</span> <span class="s1">&#39;:&#39;</span> <span class="o">+</span> <span class="n">annotation_type</span>
            <span class="n">annotation_data</span> <span class="o">=</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="n">annotation_col</span><span class="p">:</span> <span class="n">annotation_col_name</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;annotation_col must be a string indicating column with annotation data to be added to the spliced_ptms dataframe&#39;</span><span class="p">)</span>

    <span class="c1">#check to make sure annotation data has the necessary columns</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="n">annotation_data</span><span class="o">.</span><span class="n">columns</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">]]):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Could not find columns containing ptm information: </span><span class="si">{</span><span class="n">accession_col</span><span class="si">}</span><span class="s1">, </span><span class="si">{</span><span class="n">residue_col</span><span class="si">}</span><span class="s1">, and </span><span class="si">{</span><span class="n">position_col</span><span class="si">}</span><span class="s1">. Please either change the name of your annotation data columns containing this information or indicate the correct column names with the accession_col, residue_col, and position_col parameters&#39;</span><span class="p">)</span>

    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="n">annotation_col_name</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">annotation_col_name</span><span class="p">])</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">annotation_data</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">],</span> <span class="n">right_on</span> <span class="o">=</span> <span class="p">[</span><span class="n">accession_col</span><span class="p">,</span> <span class="n">residue_col</span><span class="p">,</span> <span class="n">position_col</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms or annotation dataframe&#39;</span><span class="p">)</span>
    
    <span class="c1">#report the number of PTMs identified</span>
    <span class="n">num_ptms_with_custom_data</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">annotation_col</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">source_name</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> data added: </span><span class="si">{</span><span class="n">num_ptms_with_custom_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with </span><span class="si">{</span><span class="n">source_name</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">spliced_ptms</span></div>

<div class="viewcode-block" id="add_PSP_regulatory_site_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_PSP_regulatory_site_data">[docs]</a><span class="k">def</span> <span class="nf">add_PSP_regulatory_site_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;Regulatory_sites.gz&#39;</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add functional information from PhosphoSitePlus (Regulatory_sites.gz) to spliced_ptms dataframe from project_ptms_onto_splice_events() function</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file: str</span>
<span class="sd">        Path to the PhosphoSitePlus Regulatory_sites.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    spliced_ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with additional columns for regulatory site information, including domains, biological process, functions, and protein interactions associated with the PTMs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check to make sure file exists</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

    <span class="c1">#read in the kinase substrate data and add to spliced ptm info</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ACC_ID&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">})</span>
    <span class="c1">#drop extra modification information that is not needed</span>
    <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="c1">#add modification type</span>
    <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mod_shorthand_dict</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]])</span>

    <span class="c1">#restrict to human data</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[</span><span class="n">regulatory_site_data</span><span class="p">[</span><span class="s1">&#39;ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_OTHER_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;ON_FUNCTION&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    
    <span class="c1">#group like modifications into a single column</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">y</span> <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">if</span> <span class="n">y</span> <span class="o">==</span> <span class="n">y</span><span class="p">]))</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">regulatory_site_data</span> <span class="o">=</span> <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    
    <span class="c1">#add &#39;PSP:&#39; in front of each column</span>
    <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PSP:&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">regulatory_site_data</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    
    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PSP:ON_FUNCTION&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PSP:ON_FUNCTION&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:ON_PROCESS&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:ON_OTHER_INTERACT&#39;</span><span class="p">])</span>

    <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#merge with spliced_ptm info</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">regulatory_site_data</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataset size changed upon merge, please make sure there are no duplicates in spliced ptms data&#39;</span><span class="p">)</span>

    
    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_known_function</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PSP:ON_FUNCTION&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_known_process</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PSP:ON_PROCESS&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_known_interaction</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus regulatory_site information added:</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_known_function</span><span class="si">}</span><span class="s2"> PTMs in dataset found associated with a molecular function </span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_known_process</span><span class="si">}</span><span class="s2"> PTMs in dataset found associated with a biological process</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_known_interaction</span><span class="si">}</span><span class="s2"> PTMs in dataset found associated with a protein interaction&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>

<div class="viewcode-block" id="add_PSP_kinase_substrate_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_PSP_kinase_substrate_data">[docs]</a><span class="k">def</span> <span class="nf">add_PSP_kinase_substrate_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;Kinase_Substrate_Dataset.gz&#39;</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add kinase substrate data from PhosphoSitePlus (Kinase_Substrate_Dataset.gz) to spliced_ptms dataframe from project_ptms_onto_splice_events() function</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file: str</span>
<span class="sd">        Path to the PhosphoSitePlus Kinase_Substrate_Dataset.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    spliced_ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with an additional column indicating the kinases known to phosphorylate that site (not relevant to non-phosphorylation PTMs)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check to make sure provided file exists</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

    <span class="c1">#load data</span>
    <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">,</span> <span class="n">encoding</span> <span class="o">=</span> <span class="s2">&quot;cp1252&quot;</span><span class="p">)</span>
    <span class="c1">#restrict to human data</span>
    <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;KIN_ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>
    <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;SUB_ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>

    <span class="n">ks_dataset</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[[</span><span class="s1">&#39;GENE&#39;</span><span class="p">,</span> <span class="s1">&#39;SUB_ACC_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;SUB_MOD_RSD&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;SUB_ACC_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;SUB_MOD_RSD&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">ks_dataset</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">]</span>

    <span class="c1">#separate residue and position</span>
    <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ks_dataset</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

    
    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PSP:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">])</span>

    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ks_dataset</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataset size changed upon merge, please make sure there are no duplicates in spliced ptms data&#39;</span><span class="p">)</span>
    
    
        <span class="c1">#report the number of ptms with kinase substrate information</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_KS</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus kinase-substrate interactions added: </span><span class="si">{</span><span class="n">num_ptms_with_KS</span><span class="si">}</span><span class="s2"> phosphorylation sites in dataset found associated with a kinase in PhosphoSitePlus&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>

<div class="viewcode-block" id="add_PSP_disease_association"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_PSP_disease_association">[docs]</a><span class="k">def</span> <span class="nf">add_PSP_disease_association</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="s1">&#39;Disease-associated_sites.gz&#39;</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Process disease asociation data from PhosphoSitePlus (Disease-associated_sites.gz), and add to spliced_ptms dataframe from project_ptms_onto_splice_events() function</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    file: str</span>
<span class="sd">        Path to the PhosphoSitePlus Kinase_Substrate_Dataset.gz file. Should be downloaded from PhosphoSitePlus in the zipped format</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    spliced_ptms: pandas.DataFrame</span>
<span class="sd">        Contains the PTMs identified across the different splice events with an additional column indicating the kinases known to phosphorylate that site (not relevant to non-phosphorylation PTMs)</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check to make sure provided file exists</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>

    <span class="c1">#load data</span>
    <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">on_bad_lines</span><span class="o">=</span><span class="s1">&#39;skip&#39;</span><span class="p">,</span><span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;ORGANISM&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;human&#39;</span><span class="p">]</span>

    <span class="c1">#removes sites without a specific disease annotation</span>
    <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">])</span>

    <span class="c1">#drop extra modification information that is not needed</span>
    <span class="c1">#drop extra modification information that is not needed</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="c1">#add modification type</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;MOD_RSD&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">mod_shorthand_dict</span><span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]])</span>
    <span class="c1">#if phosphorylation, add specific residue</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">residue_dict</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phospho&#39;</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="c1">#change O-GalNac occurring on N to N-glycosylation</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;N-Glycosylation&#39;</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;O-Glycosylation&#39;</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;N&#39;</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>


    <span class="c1">#combine disease and alteration</span>
    <span class="n">disease_associated_sites</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;ALTERATION&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;DISEASE&#39;</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="c1">#grab only necessary columns and rename</span>
    <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="p">[[</span><span class="s1">&#39;ACC_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;ALTERATION&#39;</span><span class="p">]]</span>
    <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:Disease_Association&#39;</span><span class="p">]</span>

    <span class="c1">#aggregate multiple disease associations</span>
    <span class="n">disease_associated_sites</span> <span class="o">=</span> <span class="n">disease_associated_sites</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PSP:Disease_Association&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PSP:Disease_Association&#39;</span><span class="p">])</span>

    <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>


    <span class="c1">#merge with spliced_ptm info</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">disease_associated_sites</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataset size changed upon merge, please make sure there are no duplicates in spliced ptms data&#39;</span><span class="p">)</span>
    
    <span class="c1">#</span>
    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_disease</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PSP:Disease_Association&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PhosphoSitePlus disease associations added: </span><span class="si">{</span><span class="n">num_ptms_with_disease</span><span class="si">}</span><span class="s2"> PTM sites in dataset found associated with a disease in PhosphoSitePlus&quot;</span><span class="p">)</span>
    
    
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>


<div class="viewcode-block" id="add_ELM_interactions"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_ELM_interactions">[docs]</a><span class="k">def</span> <span class="nf">add_ELM_interactions</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a spliced ptms dataframe from the project module, add ELM interaction data to the dataframe</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#load data</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://elm.eu.org/interactions/as_tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
        <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>

    <span class="n">elm_interactions</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[(</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;taxonomyElm&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;9606(Homo sapiens)&#39;</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;taxonomyDomain&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;9606(Homo sapiens)&#39;</span><span class="p">)]</span>

    <span class="n">elm_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">elm_type</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">elm_interactor</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="c1">#grab ptm location from residue column (gives residue and position (S981), so need to remove residue and convert to int)</span>
        <span class="n">ptm_loc</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;none&#39;</span> <span class="k">else</span> <span class="kc">None</span>

        <span class="c1">#if data does not have position information, move to the next</span>
        <span class="k">if</span> <span class="n">ptm_loc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
            <span class="k">continue</span>

        <span class="c1">#find if any of the linear motifs match ptm loc</span>
        <span class="n">protein_match</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;interactorElm&#39;</span><span class="p">]</span>
        <span class="n">region_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&gt;=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StartElm&#39;</span><span class="p">])</span>  <span class="o">&amp;</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&lt;=</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StopElm&#39;</span><span class="p">])</span>
        <span class="n">elm_subset_motif</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="n">protein_match</span> <span class="o">&amp;</span> <span class="n">region_match</span><span class="p">]</span>
        <span class="c1">#if any interactions were found, record and continue to the next (assumes a single ptm won&#39;t be found as both a SLiM and domain)</span>
        <span class="k">if</span> <span class="n">elm_subset_motif</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_motif</span><span class="p">[</span><span class="s1">&#39;Elm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;SLiM&#39;</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_motif</span><span class="p">[</span><span class="s1">&#39;interactorDomain&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="k">continue</span>


        <span class="c1">#domain</span>
        <span class="n">protein_match</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;interactorDomain&#39;</span><span class="p">]</span>
        <span class="n">region_match</span> <span class="o">=</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&gt;=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StartDomain&#39;</span><span class="p">])</span>  <span class="o">&amp;</span> <span class="p">(</span><span class="n">ptm_loc</span> <span class="o">&lt;=</span><span class="n">elm_interactions</span><span class="p">[</span><span class="s1">&#39;StopDomain&#39;</span><span class="p">])</span>
        <span class="n">elm_subset_domain</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">[</span><span class="n">protein_match</span> <span class="o">&amp;</span> <span class="n">region_match</span><span class="p">]</span>
        <span class="c1">#if any interactions were found, record and continue to the next (assumes a single ptm won&#39;t be found as both a SLiM and domain)</span>
        <span class="k">if</span> <span class="n">elm_subset_domain</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_domain</span><span class="p">[</span><span class="s1">&#39;Elm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Domain&#39;</span><span class="p">)</span>
            <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">elm_subset_domain</span><span class="p">[</span><span class="s1">&#39;interactorElm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="k">continue</span>

        <span class="c1">#if no interactions wer found, record as np.nan</span>
        <span class="n">elm_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">elm_type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="n">elm_interactor</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_interactor</span>
    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Location of PTM for Interaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_type</span>
    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Motifs Associated with Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">elm_list</span>
    
    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_ELM_instance</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ELM interaction instances added: </span><span class="si">{</span><span class="n">num_ptms_with_ELM_instance</span><span class="si">}</span><span class="s2"> PTMs in dataset found associated with at least one known ELM instance&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>


<span class="k">def</span> <span class="nf">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">flank_size</span> <span class="o">=</span> <span class="mi">7</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://elm.eu.org/elms/elms_index.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
        <span class="n">elm_classes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>

    <span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1">#create corresponding label for ptm_coordinate data</span>
    <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    
    <span class="n">match_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">matches</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1">#grab ptm information</span>
        <span class="c1">#grab flanking sequence for the ptm</span>
        <span class="n">loc</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;PTM Position in Canonical Isoform&quot;</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        <span class="n">ptm</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>

        
        <span class="k">if</span> <span class="n">ptm</span> <span class="ow">in</span> <span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">ptm_flanking_seq</span> <span class="o">=</span> <span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">ptm_coordinates</span><span class="p">[</span><span class="s1">&#39;PTM Label&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ptm</span><span class="p">,</span> <span class="s1">&#39;Expected Flanking Sequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="c1">#make sure flanking sequence is present</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ptm_flanking_seq</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>

                <span class="c1">#default flanking sequence is 10, if requested flanking sequence is different, then adjust</span>
                <span class="k">if</span> <span class="n">flank_size</span> <span class="o">&gt;</span> <span class="mi">10</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Flanking size must be equal to or less than 10&#39;</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">flank_size</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>
                    <span class="n">ptm_flanking_seq</span> <span class="o">=</span> <span class="n">ptm_flanking_seq</span><span class="p">[</span><span class="mi">10</span><span class="o">-</span><span class="n">flank_size</span><span class="p">:</span><span class="mi">10</span><span class="o">+</span><span class="n">flank_size</span><span class="p">]</span>

                <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">elm_row</span> <span class="ow">in</span> <span class="n">elm_classes</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                    <span class="n">reg_ex</span> <span class="o">=</span> <span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;Regex&#39;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">reg_ex</span><span class="p">,</span> <span class="n">ptm_flanking_seq</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">matches</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">elm_row</span><span class="p">[</span><span class="s1">&#39;ELMIdentifier&#39;</span><span class="p">])</span>

                <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">matches</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1">#print(f&#39;PTM {ptm} not found in PTM info file&#39;)</span>
            <span class="n">match_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    
    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Motif Matches&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">match_list</span>

    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_matched_motif</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;ELM:Motif Matches&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ELM Class motif matches found: </span><span class="si">{</span><span class="n">num_ptms_with_matched_motif</span><span class="si">}</span><span class="s2"> PTMs in dataset found with at least one matched motif&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span>

<div class="viewcode-block" id="add_PTMInt_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_PTMInt_data">[docs]</a><span class="k">def</span> <span class="nf">add_PTMInt_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced_ptms data from project module, add PTMInt interaction data, which will include the protein that is being interacted with, whether it enchances or inhibits binding, and the localization of the interaction. This will be added as a new column labeled PTMInt:Interactions and each entry will be formatted like &#39;Protein-&gt;Effect|Localization&#39;. If multiple interactions, they will be separated by a semicolon</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#load file</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">PTMint</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://ptmint.sjtu.edu.cn/data/PTM</span><span class="si">%20e</span><span class="s1">xperimental</span><span class="si">%20e</span><span class="s1">vidence.csv&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.csv&#39;</span><span class="p">)</span>
        <span class="n">PTMint</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>

    <span class="n">PTMint</span> <span class="o">=</span> <span class="n">PTMint</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Uniprot&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;AA&#39;</span><span class="p">:</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Site&#39;</span><span class="p">:</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">})</span>
    <span class="c1">#PTMint[&#39;Site&#39;] = PTMint[&#39;AA&#39;] + PTMint[&#39;Site&#39;].astype(str)</span>
    <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;Int_gene&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">PTMint</span><span class="p">[</span><span class="s1">&#39;Effect&#39;</span><span class="p">]</span>
    <span class="n">PTMint</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">]]</span>
    <span class="c1">#PTMint[&#39;PTM Position in Canonical Isoform&#39;] = PTMint[&#39;PTM Position in Canonical Isoform&#39;].astype(str)</span>

    <span class="c1">#aggregate PTMint data on the same PTMs</span>
    <span class="n">PTMint</span> <span class="o">=</span> <span class="n">PTMint</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span>

    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">])</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">PTMint</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">],</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms dataframe&#39;</span><span class="p">)</span>

    <span class="c1">#report the number of PTMs identified</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_PTMInt_data</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PTMInt data added: </span><span class="si">{</span><span class="n">num_ptms_with_PTMInt_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with PTMInt interaction information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">spliced_ptms</span></div>
    <span class="c1">#delete source PTMint data</span>
    <span class="c1">#os.remove(pdir + &#39;./Data/PTM_experimental_evidence.csv&#39;)</span>

<span class="c1">#def add_PTMcode_intraprotein(spliced_ptms, fname = None, report_success = True):</span>
<span class="c1">#    #load ptmcode info</span>
<span class="c1">#    if fname is None:</span>
<span class="c1">#        ptmcode = pd.read_csv(&#39;https://ptmcode.embl.de/data/PTMcode2_associations_within_proteins.txt.gz&#39;, sep = &#39;\t&#39;, header = 2, compression=&#39;gzip&#39;)</span>
<span class="c1">#    else:</span>
<span class="c1">#        check_file(fname, expected_extension = &#39;.gz&#39;)</span>
<span class="c1">#        ptmcode = pd.read_csv(fname, sep = &#39;\t&#39;, header = 2, compression = &#39;gzip&#39;)</span>
<span class="c1">#    </span>
<span class="c1">#    #grab humn data</span>
<span class="c1">#    ptmcode = ptmcode[ptmcode[&#39;Species&#39;] == &#39;Homo sapiens&#39;]</span>
<span class="c1">#</span>
<span class="c1">#    #add gene name to data</span>
<span class="c1">#    translator = pd.DataFrame(pose_config.uniprot_to_genename, index = [&#39;Gene&#39;]).T</span>
<span class="c1">#    translator[&#39;Gene&#39;] = translator[&#39;Gene&#39;].apply(lambda x: x.split(&#39; &#39;))</span>
<span class="c1">#    translator = translator.explode(&#39;Gene&#39;)</span>
<span class="c1">#    translator = translator.reset_index()</span>
<span class="c1">#    translator.columns = [&#39;UniProtKB/Swiss-Prot ID&#39;, &#39;Gene name&#39;]</span>
<span class="c1">#</span>
<span class="c1">#    #add uniprot ID information</span>
<span class="c1">#    ptmcode = ptmcode.merge(translator.dropna().drop_duplicates(), left_on = &#39;## Protein&#39;, right_on = &#39;Gene name&#39;, how = &#39;left&#39;)</span>
<span class="c1">#</span>
<span class="c1">#    #convert modification names to match annotation data</span>
<span class="c1">#    convert_dict = {&#39;Adp ribosylation&#39;: &#39;ADP Ribosylation&#39;, &#39;Glutamine deamidation&#39;:&#39;Deamidation&#39;}</span>
<span class="c1">#    new_mod_names = []</span>
<span class="c1">#    failed_mod = []</span>
<span class="c1">#    mod_list = ptmcode[&#39;PTM1&#39;].unique()</span>
<span class="c1">#    for mod in mod_list:</span>
<span class="c1">#        mod = mod.capitalize()</span>
<span class="c1">#        if &#39;glycosylation&#39; in mod: #if glycosylation, group into one gorup</span>
<span class="c1">#            new_mod_names.append(&#39;Glycosylation&#39;)</span>
<span class="c1">#        elif mod in pose_config.modification_conversion[&#39;Modification Class&#39;].values: #if already in modification class data, keep</span>
<span class="c1">#            new_mod_names.append(mod)</span>
<span class="c1">#        elif mod in convert_dict.keys():</span>
<span class="c1">#            new_mod_names.append(convert_dict[mod])</span>
<span class="c1">#        else:</span>
<span class="c1">#            try:</span>
<span class="c1">#                new_mod = pose_config.modification_conversion[pose_config.modification_conversion[&#39;Modification&#39;] == mod].values[0][0]</span>
<span class="c1">#                new_mod_names.append(new_mod)</span>
<span class="c1">#            except:</span>
<span class="c1">#                failed_mod.append(mod)</span>
<span class="c1">#                new_mod_names.append(mod)</span>
<span class="c1">#    conversion_df = pd.DataFrame({&#39;PTM1&#39;:mod_list, &#39;Modification Class&#39;:new_mod_names})</span>
<span class="c1">#</span>
<span class="c1">#    #add new modification labels to data</span>
<span class="c1">#    ptmcode = ptmcode.merge(conversion_df, on = &#39;PTM1&#39;, how = &#39;left&#39;)</span>
<span class="c1">#    </span>
<span class="c1">#    #groupby by PTM1 and rename to match column names in annotation data</span>
<span class="c1">#    ptmcode = ptmcode[[&#39;UniProtKB/Swiss-Prot ID&#39;, &#39;Modification Class&#39;, &#39;Residue1&#39;, &#39;Residue2&#39;]].dropna(subset = &#39;UniProtKB/Swiss-Prot ID&#39;)</span>
<span class="c1">#    ptmcode = ptmcode.groupby([&#39;UniProtKB/Swiss-Prot ID&#39;, &#39;Modification Class&#39;, &#39;Residue1&#39;])[&#39;Residue2&#39;].agg(&#39;;&#39;.join).reset_index()</span>
<span class="c1">#    ptmcode = ptmcode.rename(columns = {&#39;UniProtKB/Swiss-Prot ID&#39;:&#39;UniProtKB Accession&#39;, &#39;Residue1&#39;:&#39;Residue&#39;, &#39;Residue2&#39;:&#39;PTMcode:Intraprotein_Interactions&#39;})</span>
<span class="c1">#    </span>
<span class="c1">#    #separate residue information into separate columns, one for amino acid and one for position</span>
<span class="c1">#    ptmcode[&#39;PTM Position in Canonical Isoform&#39;] = ptmcode[&#39;Residue&#39;].apply(lambda x: int(x[1:]))</span>
<span class="c1">#    ptmcode[&#39;Residue&#39;] = ptmcode[&#39;Residue&#39;].apply(lambda x: x[0])</span>
<span class="c1">#</span>
<span class="c1">#        #if splice data already has the annotation columns, remove them</span>
<span class="c1">#    if &#39;PTMcode:Intraprotein_Interactions&#39; in spliced_ptms.columns:</span>
<span class="c1">#        spliced_ptms = spliced_ptms.drop(columns = [&#39;PTMcode:Intraprotein_Interactions&#39;])</span>
<span class="c1">#</span>
<span class="c1">#    #explode dataframe on modifications</span>
<span class="c1">#    if spliced_ptms[&#39;Modification Class&#39;].str.contains(&#39;;&#39;).any():</span>
<span class="c1">#        spliced_ptms[&#39;Modification Class&#39;] = spliced_ptms[&#39;Modification Class&#39;].str.split(&#39;;&#39;)</span>
<span class="c1">#        spliced_ptms = spliced_ptms.explode(&#39;Modification Class&#39;).reset_index(drop = True)</span>
<span class="c1">#</span>
<span class="c1">#    #add to splice data</span>
<span class="c1">#    original_data_size = spliced_ptms.shape[0]</span>
<span class="c1">#    spliced_ptms = spliced_ptms.merge(ptmcode, how = &#39;left&#39;, on = [&#39;UniProtKB Accession&#39;, &#39;Residue&#39;, &#39;PTM Position in Canonical Isoform&#39;, &#39;Modification Class&#39;])</span>
<span class="c1">#    if spliced_ptms.shape[0] != original_data_size:</span>
<span class="c1">#        raise RuntimeError(&#39;Dataframe size has changed, check for duplicates in spliced ptms dataframe&#39;)</span>
<span class="c1">#    </span>
<span class="c1">#    #report the number of PTMs identified</span>
<span class="c1">#    if report_success:</span>
<span class="c1">#        num_ptms_with_PTMcode_data = spliced_ptms.dropna(subset = &#39;PTMcode:Intraprotein_Interactions&#39;).groupby([&#39;UniProtKB Accession&#39;, &#39;Residue&#39;]).size().shape[0]</span>
<span class="c1">#        print(f&quot;PTMcode intraprotein interactions added: {num_ptms_with_PTMcode_data} PTMs in dataset found with PTMcode intraprotein interaction information&quot;)</span>
<span class="c1">#</span>
<span class="c1">#    return spliced_ptms</span>

<span class="k">def</span> <span class="nf">extract_ids_PTMcode</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="s1">&#39;## Protein1&#39;</span><span class="p">):</span>

    <span class="c1">#add gene name to data</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">T</span>
    <span class="n">name_to_uniprot</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Gene&#39;</span><span class="p">)</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB/Swiss-Prot ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">]</span>
    <span class="n">name_to_uniprot</span> <span class="o">=</span> <span class="n">name_to_uniprot</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">,</span> <span class="n">keep</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

    <span class="c1">#protein name is provided as either ensemble gene id or gene name check for both</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">translator</span><span class="p">[[</span><span class="s1">&#39;Gene stable ID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(),</span> <span class="n">left_on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span> <span class="n">right_on</span> <span class="o">=</span> <span class="s1">&#39;Gene stable ID&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;index&#39;</span><span class="p">:</span> <span class="s1">&#39;From_ID&#39;</span><span class="p">})</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">name_to_uniprot</span><span class="p">,</span> <span class="n">left_on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span> <span class="n">right_on</span> <span class="o">=</span> <span class="s1">&#39;Gene name&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;UniProtKB/Swiss-Prot ID&#39;</span><span class="p">:</span> <span class="s1">&#39;From_Name&#39;</span><span class="p">})</span>

    <span class="c1">#grab unique id from &#39;From_ID&#39; and &#39;From_Name&#39; column, if available</span>
    <span class="n">uniprot_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;From_Name&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">combine_first</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;From_ID&#39;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">uniprot_ids</span><span class="o">.</span><span class="n">values</span>

<span class="k">def</span> <span class="nf">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">fname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;https://ptmcode.embl.de/data/PTMcode2_associations_between_proteins.txt.gz&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compression</span> <span class="o">=</span> <span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
        <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>

    <span class="c1">#grab human interactions</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Species&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Homo sapiens&#39;</span><span class="p">]</span>
    <span class="c1">#ignore intraprotein interactions</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;## Protein1&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Protein2&#39;</span><span class="p">]]</span>

    <span class="c1">#get uniprot id for primary protein and interacting protein</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">extract_ids_PTMcode</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="s1">&#39;## Protein1&#39;</span><span class="p">)</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">extract_ids_PTMcode</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="s1">&#39;Protein2&#39;</span><span class="p">)</span>

    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Protein&#39;</span><span class="p">])</span>
    <span class="c1">#remove duplicate proteins (some entries have different ids but are actually the same protein)</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]]</span>

    <span class="c1">#aggregate interactions</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Residue2&#39;</span><span class="p">]</span>


    <span class="c1">#convert modification names</span>
    <span class="n">convert_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Adp ribosylation&#39;</span><span class="p">:</span> <span class="s1">&#39;ADP Ribosylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Glutamine deamidation&#39;</span><span class="p">:</span><span class="s1">&#39;Deamidation&#39;</span><span class="p">}</span>
    <span class="n">new_mod_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">failed_mod</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">mod_list</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">mod_list</span><span class="p">:</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="n">mod</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span>
        <span class="k">if</span> <span class="s1">&#39;glycosylation&#39;</span> <span class="ow">in</span> <span class="n">mod</span><span class="p">:</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Glycosylation&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">mod</span> <span class="ow">in</span> <span class="n">convert_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
            <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">convert_dict</span><span class="p">[</span><span class="n">mod</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">new_mod</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">modification_conversion</span><span class="p">[</span><span class="s1">&#39;Modification&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">mod</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">new_mod</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">failed_mod</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
                <span class="n">new_mod_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
    <span class="n">conversion_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;PTM1&#39;</span><span class="p">:</span><span class="n">mod_list</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">:</span><span class="n">new_mod_names</span><span class="p">})</span>

    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">conversion_df</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;PTM1&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>


    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Residue1&#39;</span><span class="p">:</span><span class="s1">&#39;Residue&#39;</span><span class="p">})</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])[</span><span class="s1">&#39;Interacting Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
    <span class="n">ptmcode</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;UniProtKB/Swiss-Prot ID&#39;</span><span class="p">:</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue1&#39;</span><span class="p">:</span><span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting Residue&#39;</span><span class="p">:</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">})</span>

    <span class="c1">#separate residue information into separate columns, one for amino acid and one for position</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span>
    <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmcode</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

            <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">])</span>

        <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ptmcode</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms dataframe&#39;</span><span class="p">)</span>
    
    <span class="c1">#report the number of PTMs identified</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_PTMcode_data</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PTMcode interprotein interactions added: </span><span class="si">{</span><span class="n">num_ptms_with_PTMcode_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with PTMcode interprotein interaction information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">spliced_ptms</span>

<div class="viewcode-block" id="extract_positions_from_DEPOD"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.extract_positions_from_DEPOD">[docs]</a><span class="k">def</span> <span class="nf">extract_positions_from_DEPOD</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given string object consisting of multiple modifications in the form of &#39;Residue-Position&#39; separated by &#39;, &#39;, extract the residue and position. Ignore any excess details in the string.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)</span>
    <span class="c1">#for each residue in list, find location of &#39;Ser&#39;, &#39;Thr&#39; and &#39;Tyr&#39; in the string (should either have &#39;-&#39; or a number immediately after it)</span>
    <span class="n">new_x</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">x</span><span class="p">:</span>
        <span class="c1">#determine type of modification</span>
        <span class="k">if</span> <span class="s1">&#39;Ser&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Ser&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;S&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;Thr&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Thr&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;T&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;Tyr&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;Tyr&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span>
        <span class="k">elif</span> <span class="s1">&#39;His&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="p">[</span><span class="n">match</span><span class="o">.</span><span class="n">start</span><span class="p">()</span> <span class="k">for</span> <span class="n">match</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">finditer</span><span class="p">(</span><span class="s1">&#39;His&#39;</span><span class="p">,</span> <span class="n">item</span><span class="p">)]</span>
            <span class="n">res</span> <span class="o">=</span> <span class="s1">&#39;H&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

        <span class="c1">#check if multiple locations were found, if so grab last entry</span>
        <span class="k">if</span> <span class="n">loc</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">item</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">loc</span> <span class="o">=</span> <span class="n">loc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">make_string</span> <span class="o">=</span> <span class="kc">True</span>
        
        <span class="c1">#find integer</span>
        <span class="k">if</span> <span class="n">make_string</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">[</span><span class="n">loc</span><span class="p">:]:</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">item</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">res</span> <span class="o">+</span> <span class="n">item</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">item</span><span class="p">[</span><span class="n">loc</span><span class="o">+</span><span class="mi">3</span><span class="p">:]</span>
                <span class="n">item</span> <span class="o">=</span> <span class="n">res</span> <span class="o">+</span> <span class="n">item</span>

        <span class="n">new_x</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">new_x</span></div>

<span class="k">def</span> <span class="nf">add_DEPOD_phosphatase_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

    <span class="c1">#download data</span>
    <span class="n">depod1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;https://depod.bioss.uni-freiburg.de/download/PPase_protSubtrates_201903.xls&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;PSprots&#39;</span><span class="p">)</span>
    <span class="n">depod2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;https://depod.bioss.uni-freiburg.de/download/PPase_protSubtrates_newPairs_201903.xls&#39;</span><span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;newPSprots&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">depod1</span><span class="p">,</span> <span class="n">depod2</span><span class="p">])</span>

    <span class="c1">#remove any rows with missing sit information</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>

    <span class="c1">#remove excess annotations that make parsing difficult</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;[&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;in&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;in ref.&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>

    <span class="c1">#separate individual sites</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[(</span><span class="o">~</span><span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Isoform&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="o">~</span><span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;isoform&#39;</span><span class="p">))]</span>

    <span class="c1">#process dephosphosite strings to extract residue and position and explode so that each phosphosite is its own row</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">extract_positions_from_DEPOD</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">)</span>

    <span class="c1">#separate multiple substrate accessions into their own rows (many of these link back to the same ID, but will keep just in case)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">])</span>

    <span class="c1">#extract only needed information and add phosphorylation as modification type</span>
    <span class="c1">#extract only needed information and add phosphorylation as modification type</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Dephosphosites&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:])</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s1">&#39;Substrate accession numbers&#39;</span><span class="p">:</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Phosphatase entry names&#39;</span><span class="p">:</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="p">[[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]]</span>
    <span class="n">depod</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Phosphorylation&#39;</span>

    <span class="c1">#combine on the same PTM</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">depod</span> <span class="o">=</span> <span class="n">depod</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span>

        <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">])</span>

        <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">depod</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms dataframe&#39;</span><span class="p">)</span>
    
    <span class="c1">#report the number of PTMs identified</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_PTMcode_data</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;DEPOD Phosphatase substrates added: </span><span class="si">{</span><span class="n">num_ptms_with_PTMcode_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with Phosphatase substrate information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">spliced_ptms</span>

<span class="k">def</span> <span class="nf">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">regphos</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;http://140.138.144.141/~RegPhos/download/RegPhos_Phos_human.txt&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">dtype</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;position&#39;</span><span class="p">:</span><span class="nb">int</span><span class="p">,</span> <span class="s1">&#39;description&#39;</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span><span class="s1">&#39;catalytic kinase&#39;</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="s1">&#39;reference&#39;</span><span class="p">:</span><span class="s1">&#39;str&#39;</span><span class="p">})</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.txt&#39;</span><span class="p">)</span>
        <span class="n">regphos</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;catalytic kinase&#39;</span><span class="p">)</span>
    <span class="c1">#regphos[&#39;Residue&#39;] = regphos[&#39;code&#39;] + regphos[&#39;position&#39;].astype(str)</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;code&#39;</span><span class="p">:</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;position&#39;</span><span class="p">:</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;AC&#39;</span><span class="p">:</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;catalytic kinase&#39;</span><span class="p">:</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">})</span>
    <span class="n">regphos</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Phosphorylation&#39;</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">dropna</span><span class="p">()</span>
    <span class="n">regphos</span> <span class="o">=</span> <span class="n">regphos</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">])</span>

    <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#add to splice data</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">regphos</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataframe size has changed, check for duplicates in spliced ptms dataframe&#39;</span><span class="p">)</span>
    
    <span class="c1">#report the number of PTMs identified</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_regphos_data</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;RegPhos kinase-substrate data added: </span><span class="si">{</span><span class="n">num_ptms_with_regphos_data</span><span class="si">}</span><span class="s2"> PTMs in dataset found with kinase-substrate information&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">spliced_ptms</span>


<span class="k">def</span> <span class="nf">add_PTMsigDB_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">report_success</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="c1">#if file is None:</span>
    <span class="c1">#    ptmsigdb = pd.read_excel(&#39;https://proteomics.broadapps.org/ptmsigdb/_w_8b062d9e/appff37efd164a676afcc8e6e42e6058e01/session/a2b28c4ed29deadd6779fdd26aec33c1/download/download.xlsx?w=8b062d9e&#39;, sheet_name = &#39;human&#39;)</span>
    <span class="c1">#else:</span>
    <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.xlsx&#39;</span><span class="p">)</span>
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">sheet_name</span> <span class="o">=</span> <span class="s1">&#39;human&#39;</span><span class="p">)</span>


    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.uniprot&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">:]))</span>

    <span class="c1">#filter out excess information in some of the site.ptm column, then convert to modification class details</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;;&#39;</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">else</span> <span class="n">x</span><span class="p">)</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.ptm&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">mod_shorthand_dict</span><span class="p">)</span>

    <span class="c1">#combine signature and direction for annotation column</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;signature&#39;</span><span class="p">]</span> <span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;site.direction&#39;</span><span class="p">]</span>

    <span class="c1">#drop unneeded columns</span>
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="p">[[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="s1">&#39;category&#39;</span><span class="p">]]</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;Signature&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">),</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:&#39;</span> <span class="o">+</span> <span class="n">ptmsigdb</span><span class="p">[</span><span class="s1">&#39;category&#39;</span><span class="p">]</span> 
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

        <span class="c1">#convert to pivot table with each category being a separate column</span>
    <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">],</span> <span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;category&#39;</span><span class="p">,</span> <span class="n">values</span> <span class="o">=</span> <span class="s1">&#39;Signature&#39;</span><span class="p">,</span> <span class="n">aggfunc</span><span class="o">=</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>

    <span class="c1">#remove psp data if it is already in spliced ptms</span>
    <span class="k">if</span> <span class="s1">&#39;PSP:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:KINASE-PSP&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="s1">&#39;PSP:Disease_Association&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">ptmsigdb</span> <span class="o">=</span> <span class="n">ptmsigdb</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:DISEASE-PSP&#39;</span><span class="p">)</span>


    <span class="c1">#if splice data already has the annotation columns, remove them</span>
    <span class="k">if</span> <span class="s1">&#39;PTMsigDB:PATH-BI&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">cols_in_data</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;PTMsigDB&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">cols_in_data</span><span class="p">)</span>


    <span class="c1">#explode dataframe on modifications</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
        <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="c1">#merge with spliced_ptm info</span>
    <span class="n">original_data_size</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">ptmsigdb</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="n">original_data_size</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="s1">&#39;Dataset size changed upon merge, please make sure there are no duplicates in spliced ptms data&#39;</span><span class="p">)</span>


    <span class="c1">#report the number of ptms with motif data</span>
    <span class="k">if</span> <span class="n">report_success</span><span class="p">:</span>
        <span class="n">num_ptms_with_ikip</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:KINASE-iKiP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_path_bi</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PATH-BI&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_path_np</span><span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PATH-NP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_path_wp</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PATH-WP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_dia_pert</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PERT-P100-DIA&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_dia2_pert</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PERT-P100-DIA2&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_prm_pert</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PERT-P100-PRM&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">num_ptms_with_psp_pert</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;PTMsigDB:PERT-PSP&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">size</span><span class="p">()</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;PTMsigDB added:</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_ikip</span><span class="si">}</span><span class="s2"> PTMs associated with kinases in iKiP</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_path_wp</span><span class="si">}</span><span class="s2"> PTMs associated with molecular pathway signatures from WikiPathways</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_path_np</span><span class="si">}</span><span class="s2"> PTMs associated with molecular pathway signatures from NetPath</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_psp_pert</span><span class="si">}</span><span class="s2"> PTMs with PhosphoSitePlus perturbations</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_dia_pert</span><span class="si">}</span><span class="s2"> with perturbations in LINCS P1000 DIA dataset </span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_dia2_pert</span><span class="si">}</span><span class="s2"> with perturbations in LINCS P1000 DIA2 dataset</span><span class="se">\n\t</span><span class="s2"> -&gt;</span><span class="si">{</span><span class="n">num_ptms_with_prm_pert</span><span class="si">}</span><span class="s2"> with perturbations in LINCS P1000 PRM dataset&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span>



<span class="c1">######### Functions for combining annotations from multiple sources ########</span>

<div class="viewcode-block" id="convert_PSP_label_to_UniProt"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.convert_PSP_label_to_UniProt">[docs]</a><span class="k">def</span> <span class="nf">convert_PSP_label_to_UniProt</span><span class="p">(</span><span class="n">label</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a label for an interacting protein from PhosphoSitePlus, convert to UniProtKB accession. Required as PhosphoSitePlus interactions are recorded in various ways that aren&#39;t necessarily consistent with other databases (i.e. not always gene name)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    label: str</span>
<span class="sd">        Label for interacting protein from PhosphoSitePlus</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="c1">#using uniprot to gene name dict, construct dict to go the other direction (gene name to uniprot id)</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>


    <span class="c1">#remove isoform label if present</span>
    <span class="k">if</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span> <span class="c1">#if PSP name is gene name found in uniprot</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">label</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
    <span class="k">elif</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">:</span> <span class="c1"># if PSP name is not gene name, but is in conversion dictionary</span>
        <span class="k">return</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">[</span><span class="n">label</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span> <span class="c1">#otherwise note that gene was missed</span>
        <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span></div>
        <span class="c1">#missed_genes.append(gene)</span>

<span class="k">def</span> <span class="nf">extract_interaction_details</span><span class="p">(</span><span class="n">interaction</span><span class="p">,</span> <span class="n">column</span> <span class="o">=</span> <span class="s2">&quot;PSP:ON_PROT_INTERACT&quot;</span><span class="p">):</span>

    <span class="n">interaction_types</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;INDUCES&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;Combined:Kinase&#39;</span><span class="p">:</span><span class="s1">&#39;REGULATES&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">:</span><span class="s1">&#39;UNCLEAR&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;)&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">:</span>
        <span class="n">ptmint_type_conversion</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Inhibit&#39;</span><span class="p">:</span><span class="s1">&#39;DISRUPTS&#39;</span><span class="p">,</span> <span class="s1">&#39;Enhance&#39;</span><span class="p">:</span><span class="s2">&quot;INDUCES&quot;</span><span class="p">}</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">ptmint_type_conversion</span><span class="p">[</span><span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&gt;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">elif</span> <span class="n">column</span> <span class="o">==</span> <span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="s1">&#39;INDUCES&#39;</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">type</span> <span class="o">=</span> <span class="n">interaction_types</span><span class="p">[</span><span class="n">column</span><span class="p">]</span>
        <span class="n">protein</span> <span class="o">=</span> <span class="n">interaction</span>

    <span class="k">return</span> <span class="nb">type</span><span class="p">,</span> <span class="n">protein</span>

<div class="viewcode-block" id="unify_interaction_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.unify_interaction_data">[docs]</a><span class="k">def</span> <span class="nf">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">interaction_col</span><span class="p">,</span> <span class="n">name_dict</span> <span class="o">=</span> <span class="p">{}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm data and a column containing interaction data, extract the interacting protein, type of interaction, and convert to UniProtKB accession. This will be added as a new column labeled &#39;Interacting ID&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe containing PTM data</span>
<span class="sd">    interaction_col: str</span>
<span class="sd">        column containing interaction information from a specific database</span>
<span class="sd">    name_dict: dict</span>
<span class="sd">        dictionary to convert names within given database to UniProt IDs. For cases when name is not necessarily one of the gene names listed in UniProt</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    interact: pd.DataFrame</span>
<span class="sd">        Contains PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="c1">#using uniprot to gene name dict, construct dict to go the other direction (gene name to uniprot id)</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>

    <span class="c1">#extract PSP data from annotated PTMs, separate cases in which single PTM has multipe interactions</span>
    <span class="n">data_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Significance&#39;</span><span class="p">,</span> <span class="s1">&#39;dPSI&#39;</span><span class="p">]]</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="n">interaction_col</span><span class="p">)[[</span><span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">,</span><span class="n">interaction_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">data_cols</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">interact</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs associated with </span><span class="si">{</span><span class="n">interaction_col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">interact</span>
    
    <span class="n">interact</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">interaction_col</span><span class="p">)</span>

    <span class="c1">#extract protein and type of interaction (currently for phosphosite plus)</span>
    <span class="nb">type</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">protein</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">interact</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">processed</span> <span class="o">=</span> <span class="n">extract_interaction_details</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">interaction_col</span><span class="p">],</span> <span class="n">interaction_col</span><span class="p">)</span>
        <span class="nb">type</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">processed</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">protein</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">processed</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span>  <span class="o">=</span> <span class="nb">type</span>
    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">protein</span>
        

    <span class="c1">#convert interacting protein to uniprot id for databases that are not reported in uniprot ids</span>
    <span class="k">if</span> <span class="n">interaction_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">]:</span>
        <span class="n">interacting_id</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">missed_genes</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]:</span>
            <span class="c1">#remove isoform label if present</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span> <span class="c1">#if PSP name is gene name found in uniprot</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">gene</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()])</span>
            <span class="k">elif</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">name_dict</span><span class="p">:</span> <span class="c1"># if PSP name is not gene name, but is in conversion dictionary</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">name_dict</span><span class="p">[</span><span class="n">gene</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span> <span class="c1">#otherwise note that gene was missed</span>
                <span class="n">interacting_id</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
                <span class="n">missed_genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>

        <span class="c1">#save information</span>
        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interacting_id</span>
        <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>


        <span class="c1">#check if there multiple in one row</span>
        <span class="k">if</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
            <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
            <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Protein&#39;</span><span class="p">]</span>
    

    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="n">x</span> <span class="o">==</span> <span class="n">x</span> <span class="k">else</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">subset</span> <span class="o">=</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">)</span>
    <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="n">interaction_col</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">interact</span></div>

<div class="viewcode-block" id="add_annotation"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.add_annotation">[docs]</a><span class="k">def</span> <span class="nf">add_annotation</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">check_existing</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a desired database and annotation type, add the corresponding annotation data to the spliced ptm dataframe</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe containing PTM data</span>
<span class="sd">    database: str</span>
<span class="sd">        Database to extract annotation data from. Options include &#39;PhosphoSitePlus&#39;, &#39;PTMcode&#39;, &#39;PTMInt&#39;, &#39;RegPhos&#39;, &#39;DEPOD&#39;</span>
<span class="sd">    annotation_type: str</span>
<span class="sd">        Type of annotation to extract. Options include &#39;Function&#39;, &#39;Process&#39;, &#39;Interactions&#39;, &#39;Disease&#39;, &#39;Kinase&#39;, &#39;Phosphatase&#39;, but depend on the specific database (run analyze.get_annotation_categories())</span>
<span class="sd">    file: str</span>
<span class="sd">        File path to annotation data. If None, will download from online source, except for PhosphoSitePlus (due to licensing restrictions)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">check_existing</span><span class="p">:</span>
        <span class="n">annot_col</span> <span class="o">=</span> <span class="n">annotation_col_dict</span><span class="p">[</span><span class="n">database</span><span class="p">][</span><span class="n">annotation_type</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">annot_col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation data for </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> already present in provided dataframe, skipping. If you would like to update annotation data, set check_existing = False&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">spliced_ptms</span>

    <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s2">&quot;PhosphoSitePlus&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Function&#39;</span><span class="p">,</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">]:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_regulatory_site_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Disease&#39;</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_disease_association</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_kinase_substrate_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation type </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> not recognized for PhosphoSitePlus&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">:</span>
        <span class="c1">#if annotation_type == &#39;Intraprotein&#39;:</span>
        <span class="c1">#    if file is not None:</span>
        <span class="c1">#        check_file(file, expected_extension=&#39;.gz&#39;)</span>
        <span class="c1">#        spliced_ptms = add_PTMcode_intraprotein(spliced_ptms, file = file)</span>
        <span class="c1">#    else:</span>
        <span class="c1">#        spliced_ptms = add_PTMcode_intraprotein(spliced_ptms)</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation type </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> not recognized for PTMcode&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.csv&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMInt_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMInt_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation type </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> not recognized for PTMInt&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.txt&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation type </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> not recognized for RegPhos&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Phosphatase&#39;</span><span class="p">:</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_DEPOD_phosphatase_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Annotation type </span><span class="si">{</span><span class="n">annotation_type</span><span class="si">}</span><span class="s2"> not recognized for RegPhos&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined&#39;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;PSP:Kinase&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;PhosphoSitePlus kinase data not found in spliced PTM dataframe, please annotate with this first&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">combine_KS_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">annotation_type</span> <span class="o">==</span> <span class="s1">&#39;Interactions&#39;</span><span class="p">:</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">combine_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Database </span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s2"> not recognized&quot;</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>
    

<div class="viewcode-block" id="combine_interaction_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.combine_interaction_data">[docs]</a><span class="k">def</span> <span class="nf">combine_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">interaction_databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">,</span> <span class="s1">&#39;ELM&#39;</span><span class="p">],</span> <span class="n">include_enzyme_interactions</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given annotated spliced ptm data, extract interaction data from various databases and combine into a single dataframe. This will include the interacting protein, the type of interaction, and the source of the interaction data</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe containing PTM data and associated interaction annotations from various databases</span>
<span class="sd">    interaction_databases: list</span>
<span class="sd">        List of databases to extract interaction data from. Options include &#39;PhosphoSitePlus&#39;, &#39;PTMcode&#39;, &#39;PTMInt&#39;, &#39;RegPhos&#39;, &#39;DEPOD&#39;. These should already have annotation columns in the spliced_ptms dataframe, otherwise they will be ignored. For kinase-substrate interactions, if combined column is present, will use that instead of individual databases</span>
<span class="sd">    include_enzyme_interactions: bool</span>
<span class="sd">        If True, will include kinase-substrate and phosphatase interactions in the output dataframe</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    interact_data: list</span>
<span class="sd">        List of dataframes containing PTMs and their interacting proteins, the type of influence the PTM has on the interaction (DISRUPTS, INDUCES, or REGULATES), and the source of the interaction data</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">interact_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">combined_added</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">for</span> <span class="n">database</span> <span class="ow">in</span> <span class="n">interaction_databases</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span> <span class="ow">and</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;PhosphoSitePlus regulatory site data found and added&#39;</span><span class="p">)</span>
                <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">psp_name_dict</span><span class="p">)</span>
                <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>

  
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMcode&#39;</span> <span class="ow">and</span> <span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;PTMcode data found and added&#39;</span><span class="p">)</span>
                <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">)</span>
                <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PTMInt&#39;</span> <span class="ow">and</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;PTMInt data found and added&#39;</span><span class="p">)</span>
                <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">)</span>
                <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;ELM&#39;</span> <span class="ow">and</span> <span class="s1">&#39;ELM:Interactions&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;ELM data found and added&#39;</span><span class="p">)</span>
                <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;ELM:Interactions&#39;</span><span class="p">)</span>
                <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">include_enzyme_interactions</span><span class="p">:</span>
            <span class="c1">#dictionary to convert kinase names to gene names</span>
            <span class="n">ks_genes_to_uniprot</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABL1(ABL)&#39;</span><span class="p">:</span><span class="s1">&#39;P00519&#39;</span><span class="p">,</span> <span class="s1">&#39;ACK&#39;</span><span class="p">:</span><span class="s1">&#39;Q07912&#39;</span><span class="p">,</span> <span class="s1">&#39;AURC&#39;</span><span class="p">:</span><span class="s1">&#39;Q9UQB9&#39;</span><span class="p">,</span> <span class="s1">&#39;ERK1(MAPK3)&#39;</span><span class="p">:</span><span class="s1">&#39;P27361&#39;</span><span class="p">,</span><span class="s1">&#39;ERK2(MAPK1)&#39;</span><span class="p">:</span><span class="s1">&#39;P28482&#39;</span><span class="p">,</span>  <span class="s1">&#39;ERK5(MAPK7)&#39;</span><span class="p">:</span><span class="s1">&#39;Q13164&#39;</span><span class="p">,</span><span class="s1">&#39;JNK1(MAPK8)&#39;</span><span class="p">:</span><span class="s1">&#39;P45983&#39;</span><span class="p">,</span> <span class="s1">&#39;CK1A&#39;</span><span class="p">:</span><span class="s1">&#39;P48729&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK2(MAPK9)&#39;</span><span class="p">:</span><span class="s1">&#39;P45984&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK3(MAPK10)&#39;</span><span class="p">:</span><span class="s1">&#39;P53779&#39;</span><span class="p">,</span> <span class="s1">&#39;P38A(MAPK14)&#39;</span><span class="p">:</span><span class="s1">&#39;Q16539&#39;</span><span class="p">,</span><span class="s1">&#39;P38B(MAPK11)&#39;</span><span class="p">:</span><span class="s1">&#39;Q15759&#39;</span><span class="p">,</span> <span class="s1">&#39;P38G(MAPK12)&#39;</span><span class="p">:</span><span class="s1">&#39;P53778&#39;</span><span class="p">,</span><span class="s1">&#39;P70S6K&#39;</span> <span class="p">:</span><span class="s1">&#39;Q9UBS0&#39;</span><span class="p">,</span> <span class="s1">&#39;PAK&#39;</span><span class="p">:</span><span class="s1">&#39;Q13153&#39;</span><span class="p">,</span> <span class="s1">&#39;PKCZ&#39;</span><span class="p">:</span><span class="s1">&#39;Q05513&#39;</span><span class="p">,</span> <span class="s1">&#39;CK2A&#39;</span><span class="p">:</span><span class="s1">&#39;P19784&#39;</span><span class="p">,</span> <span class="s1">&#39;ABL2&#39;</span><span class="p">:</span><span class="s1">&#39;P42684&#39;</span><span class="p">,</span> <span class="s1">&#39;AMPKA1&#39;</span><span class="p">:</span><span class="s1">&#39;Q13131&#39;</span><span class="p">,</span> <span class="s1">&#39;AMPKA2&#39;</span><span class="p">:</span><span class="s1">&#39;Q13131&#39;</span><span class="p">,</span> <span class="s1">&#39;AURB&#39;</span><span class="p">:</span><span class="s1">&#39;Q96GD4&#39;</span><span class="p">,</span> <span class="s1">&#39;CAMK1A&#39;</span><span class="p">:</span><span class="s1">&#39;Q14012&#39;</span><span class="p">,</span> <span class="s1">&#39;CDC42BP&#39;</span><span class="p">:</span><span class="s1">&#39;Q9Y5S2&#39;</span><span class="p">,</span><span class="s1">&#39;CK1D&#39;</span><span class="p">:</span><span class="s1">&#39;P48730&#39;</span><span class="p">,</span><span class="s1">&#39;CK1E&#39;</span><span class="p">:</span><span class="s1">&#39;P49674&#39;</span><span class="p">,</span><span class="s1">&#39;CK2B&#39;</span><span class="p">:</span><span class="s1">&#39;P67870&#39;</span><span class="p">,</span><span class="s1">&#39;DMPK1&#39;</span><span class="p">:</span><span class="s1">&#39;Q09013&#39;</span><span class="p">,</span> <span class="s1">&#39;DNAPK&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span><span class="s1">&#39;DSDNA KINASE&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span> <span class="s1">&#39;EG3 KINASE&#39;</span><span class="p">:</span><span class="s1">&#39;P49840&#39;</span><span class="p">,</span><span class="s1">&#39;ERK3(MAPK6)&#39;</span><span class="p">:</span><span class="s1">&#39;Q16659&#39;</span><span class="p">,</span><span class="s1">&#39;GSK3&#39;</span><span class="p">:</span><span class="s1">&#39;P49840&#39;</span><span class="p">,</span> <span class="s1">&#39;MRCKA&#39;</span><span class="p">:</span><span class="s1">&#39;Q5VT25&#39;</span><span class="p">,</span> <span class="s1">&#39;P38D(MAPK13)&#39;</span><span class="p">:</span><span class="s1">&#39;O15264&#39;</span><span class="p">,</span><span class="s1">&#39;P70S6KB&#39;</span><span class="p">:</span><span class="s1">&#39;Q9UBS0&#39;</span><span class="p">,</span><span class="s1">&#39;PDKC&#39;</span><span class="p">:</span><span class="s1">&#39;P78527&#39;</span><span class="p">,</span><span class="s1">&#39;PKCH&#39;</span><span class="p">:</span><span class="s1">&#39;P24723&#39;</span><span class="p">,</span><span class="s1">&#39;PKCI&#39;</span><span class="p">:</span><span class="s1">&#39;P41743&#39;</span><span class="p">,</span><span class="s1">&#39;PKCT&#39;</span><span class="p">:</span><span class="s1">&#39;Q04759&#39;</span><span class="p">,</span><span class="s1">&#39;PKD3&#39;</span><span class="p">:</span><span class="s1">&#39;O94806&#39;</span><span class="p">,</span><span class="s1">&#39;PKG1&#39;</span><span class="p">:</span><span class="s1">&#39;Q13976&#39;</span><span class="p">,</span><span class="s1">&#39;PKG2&#39;</span><span class="p">:</span><span class="s1">&#39;Q13237&#39;</span><span class="p">,</span><span class="s1">&#39;SMMLCK&#39;</span><span class="p">:</span><span class="s1">&#39;Q15746&#39;</span><span class="p">}</span>
            <span class="k">if</span> <span class="s1">&#39;Combined:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">combined_added</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Combined kinase-substrate data found and added&#39;</span><span class="p">)</span>
                    <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;Combined:Kinase&#39;</span><span class="p">,</span> <span class="n">ks_genes_to_uniprot</span><span class="p">)</span>
                    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;PSP/RegPhos&#39;</span>
                    <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                    <span class="n">combined_added</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">elif</span> <span class="s1">&#39;Combined:Kinase&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">if</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;RegPhos kinase-substrate data found and added&#39;</span><span class="p">)</span>
                        <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">,</span> <span class="n">ks_genes_to_uniprot</span><span class="p">)</span>
                        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                        <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
                <span class="k">if</span> <span class="s1">&#39;PSP:Kinase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;PhosphoSitePlus kinase-substrate data found and added&#39;</span><span class="p">)</span>
                        <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">,</span> <span class="n">ks_genes_to_uniprot</span><span class="p">)</span>
                        <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                        <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;DEPOD&#39;</span> <span class="ow">and</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">():</span>
                    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;DEPOD phosphatase-substrate data found and added&#39;</span><span class="p">)</span>
                    <span class="n">interact</span> <span class="o">=</span> <span class="n">unify_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">)</span>
                    <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">database</span>
                    <span class="n">interact_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">interact</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">interact_data</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">interact_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">interact_data</span><span class="p">)</span>
        <span class="n">extra_cols</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">,</span> <span class="s1">&#39;Significance&#39;</span><span class="p">]]</span>
        <span class="n">interact_data</span> <span class="o">=</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Interacting ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Type&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">extra_cols</span><span class="p">,</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Source&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">helpers</span><span class="o">.</span><span class="n">join_unique_entries</span><span class="p">)</span>
    
        <span class="c1">#convert uniprot ids back to gene names for interpretability</span>
        <span class="n">ptm_gene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">interacting_gene</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">ptm_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">ptm_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">])</span>
            <span class="n">interacting_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span> <span class="k">else</span> <span class="n">interacting_gene</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Interacting ID&#39;</span><span class="p">])</span>
        <span class="n">interact_data</span><span class="p">[</span><span class="s1">&#39;Modified Gene&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm_gene</span>
        <span class="n">interact_data</span><span class="p">[</span><span class="s2">&quot;Interacting Gene&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interacting_gene</span>
  
  
        <span class="k">return</span> <span class="n">interact_data</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span></div>



<div class="viewcode-block" id="combine_KS_data"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.combine_KS_data">[docs]</a><span class="k">def</span> <span class="nf">combine_KS_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">ks_databases</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">],</span> <span class="n">regphos_conversion</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ERK1(MAPK3)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK3&#39;</span><span class="p">,</span> <span class="s1">&#39;ERK2(MAPK1)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">,</span> <span class="s1">&#39;JNK2(MAPK9)&#39;</span><span class="p">:</span><span class="s1">&#39;MAPK9&#39;</span><span class="p">,</span><span class="s1">&#39;CDC2&#39;</span><span class="p">:</span><span class="s1">&#39;CDK1&#39;</span><span class="p">,</span> <span class="s1">&#39;CK2A1&#39;</span><span class="p">:</span><span class="s1">&#39;CSNK2A1&#39;</span><span class="p">,</span> <span class="s1">&#39;PKACA&#39;</span><span class="p">:</span><span class="s1">&#39;PRKACA&#39;</span><span class="p">,</span> <span class="s1">&#39;ABL1(ABL)&#39;</span><span class="p">:</span><span class="s1">&#39;ABL1&#39;</span><span class="p">}):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm information, combine kinase-substrate data from multiple databases (currently support PhosphoSitePlus and RegPhos), assuming that the kinase data from these resources has already been added to the spliced ptm data. The combined kinase data will be added as a new column labeled &#39;Combined:Kinase&#39;</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Spliced PTM data from project module</span>
<span class="sd">    ks_databases: list</span>
<span class="sd">        List of databases to combine kinase data from. Currently support PhosphoSitePlus and RegPhos</span>
<span class="sd">    regphos_conversion: dict</span>
<span class="sd">        Allows conversion of RegPhos names to matching names in PhosphoSitePlus.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    splicde_ptms: pd.DataFrame</span>
<span class="sd">        Spliced PTM data with combined kinase data added</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">pose_config</span><span class="p">,</span> <span class="s1">&#39;genename_to_uniprot&#39;</span><span class="p">):</span>
        <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">flip_uniprot_dict</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">)</span>

    <span class="n">ks_data</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">db</span> <span class="ow">in</span> <span class="n">ks_databases</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">:</span>
                <span class="n">psp</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                <span class="c1">#convert PSP names to a common name (first gene name provided by uniprot)</span>
                <span class="n">psp</span> <span class="o">=</span> <span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">kin</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>  <span class="k">if</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span> <span class="k">else</span> <span class="n">kin</span> <span class="k">for</span> <span class="n">kin</span> <span class="ow">in</span> <span class="n">psp</span><span class="p">]</span>
                <span class="n">combined</span> <span class="o">+=</span> <span class="n">psp</span>
            <span class="k">elif</span> <span class="n">db</span> <span class="o">==</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">:</span>
                <span class="n">regphos</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">]</span> <span class="k">else</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">rp</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">regphos</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">rp</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                        <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">rp</span><span class="p">]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">rp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">:</span>
                        <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">uniprot_to_genename</span><span class="p">[</span><span class="n">pose_config</span><span class="o">.</span><span class="n">genename_to_uniprot</span><span class="p">[</span><span class="n">rp</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]]]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="k">elif</span> <span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">regphos_conversion</span><span class="p">:</span>
                        <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">regphos_conversion</span><span class="p">[</span><span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">regphos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">rp</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
                <span class="n">combined</span> <span class="o">+=</span> <span class="n">regphos</span>


        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">combined</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">ks_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">combined</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ks_data</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">)</span>

    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ks_data</span>
    <span class="k">return</span> <span class="n">spliced_ptms</span></div>


<div class="viewcode-block" id="check_file"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.check_file">[docs]</a><span class="k">def</span> <span class="nf">check_file</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">expected_extension</span> <span class="o">=</span> <span class="s1">&#39;.tsv&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a file name, check if the file exists and has the expected extension. If the file does not exist or has the wrong extension, raise an error.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    fname: str</span>
<span class="sd">        File name to check</span>
<span class="sd">    expected_extension: str</span>
<span class="sd">        Expected file extension. Default is &#39;.tsv&#39;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">fname</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Annotation file path must be provided&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">fname</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;File </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1"> not found&#39;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">fname</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">expected_extension</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;File </span><span class="si">{</span><span class="n">fname</span><span class="si">}</span><span class="s1"> does not have the expected extension (</span><span class="si">{</span><span class="n">expected_extension</span><span class="si">}</span><span class="s1">)&#39;</span><span class="p">)</span></div>
    




<div class="viewcode-block" id="annotate_ptms"><a class="viewcode-back" href="../../PTM_POSE.html#ptm_pose.annotate.annotate_ptms">[docs]</a><span class="k">def</span> <span class="nf">annotate_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">psp_regulatory_site_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">psp_ks_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">psp_disease_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">elm_interactions</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">elm_motifs</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">PTMint</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">PTMcode_interprotein</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">DEPOD</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">RegPhos</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">ptmsigdb_file</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">interactions_to_combine</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTMcode&#39;</span><span class="p">,</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt&#39;</span><span class="p">],</span> <span class="n">kinases_to_combine</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">],</span> <span class="n">combine_similar</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given spliced ptm data, add annotations from various databases. The annotations that can be added are the following:</span>
<span class="sd">    - PhosphoSitePlus </span>
<span class="sd">        - regulatory site data (file must be provided)</span>
<span class="sd">        - kinase-substrate data (file must be provided)</span>
<span class="sd">        - disease association data (file must be provided)</span>
<span class="sd">    - ELM </span>
<span class="sd">        - interaction data (can be downloaded automatically or provided as a file)</span>
<span class="sd">        - motif matches (elm class data can be downloaded automatically or provided as a file)</span>
<span class="sd">    - PTMInt</span>
<span class="sd">        - interaction data (will be downloaded automatically)</span>
<span class="sd">    - PTMcode</span>
<span class="sd">        - intraprotein interactions (can be downloaded automatically or provided as a file)</span>
<span class="sd">        - interprotein interactions (can be downloaded automatically or provided as a file)</span>
<span class="sd">    - DEPOD</span>
<span class="sd">        - phosphatase-substrate data (will be downloaded automatically)</span>
<span class="sd">    - RegPhos</span>
<span class="sd">        - kinase-substrate data (will be downloaded automatically)</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Spliced PTM data from project module</span>
<span class="sd">    psp_regulatory_site_file: str</span>
<span class="sd">        File path to PhosphoSitePlus regulatory site data</span>
<span class="sd">    psp_ks_file: str</span>
<span class="sd">        File path to PhosphoSitePlus kinase-substrate data</span>
<span class="sd">    psp_disease_file: str</span>
<span class="sd">        File path to PhosphoSitePlus disease association data</span>
<span class="sd">    elm_interactions: bool or str</span>
<span class="sd">        If True, download ELM interaction data automatically. If str, provide file path to ELM interaction data</span>
<span class="sd">    elm_motifs: bool or str</span>
<span class="sd">        If True, download ELM motif data automatically. If str, provide file path to ELM motif data</span>
<span class="sd">    PTMint: bool</span>
<span class="sd">        If True, download PTMInt data automatically</span>
<span class="sd">    PTMcode_intraprotein: bool or str</span>
<span class="sd">        If True, download PTMcode intraprotein data automatically. If str, provide file path to PTMcode intraprotein data</span>
<span class="sd">    PTMcode_interprotein: bool or str</span>
<span class="sd">        If True, download PTMcode interprotein data automatically. If str, provide file path to PTMcode interprotein data</span>
<span class="sd">    DEPOD: bool</span>
<span class="sd">        If True, download DEPOD data automatically</span>
<span class="sd">    RegPhos: bool</span>
<span class="sd">        If True, download RegPhos data automatically</span>
<span class="sd">    ptmsigdb_file: str</span>
<span class="sd">        File path to PTMsigDB data</span>
<span class="sd">    interactions_to_combine: list</span>
<span class="sd">        List of databases to combine interaction data from. Default is [&#39;PTMcode&#39;, &#39;PhosphoSitePlus&#39;, &#39;RegPhos&#39;, &#39;PTMInt&#39;]</span>
<span class="sd">    kinases_to_combine: list</span>
<span class="sd">        List of databases to combine kinase-substrate data from. Default is [&#39;PhosphoSitePlus&#39;, &#39;RegPhos&#39;]</span>
<span class="sd">    combine_similar: bool</span>
<span class="sd">        Whether to combine annotations of similar information (kinase, interactions, etc) from multiple databases into another column labeled as &#39;Combined&#39;. Default is True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">psp_regulatory_site_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">psp_regulatory_site_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_regulatory_site_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">psp_regulatory_site_file</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PhosphoSitePlus regulatory site data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">psp_ks_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>    
            <span class="n">check_file</span><span class="p">(</span><span class="n">psp_ks_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_kinase_substrate_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">psp_ks_file</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PhosphoSitePlus kinase-substrate data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">psp_disease_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">check_file</span><span class="p">(</span><span class="n">psp_disease_file</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PSP_disease_association</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">psp_disease_file</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PhosphoSitePlus disease association data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">elm_interactions</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">elm_interactions</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_ELM_interactions</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">elm_interactions</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">elm_interactions</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_ELM_interactions</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">elm_interactions</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;elm_interactions must be either a boolean (download elm data automatically, slower) or a string (path to elm data tsv file, faster)&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding ELM interaction data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">elm_motifs</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">elm_motifs</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">elm_motifs</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">elm_motifs</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.tsv&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_ELM_matched_motifs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">elm_motifs</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;elm_interactions must be either a boolean (download elm data automatically, slower) or a string (path to elm data tsv file, faster)&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding ELM motif matches. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">PTMint</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">PTMint</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMInt_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">PTMint</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">PTMint</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.csv&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMInt_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">PTMint</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;PTMint must be either a boolean (download PTMInt data automatically, slower) or a string (path to PTMInt data csv file, faster)&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMInt interaction data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="c1">#if PTMcode_intraprotein:</span>
    <span class="c1">#    try:</span>
    <span class="c1">#        if isinstance(PTMcode_intraprotein, bool):</span>
    <span class="c1">#            spliced_ptms = add_PTMcode_intraprotein(spliced_ptms)</span>
    <span class="c1">#        elif isinstance(PTMcode_intraprotein, str):</span>
    <span class="c1">#            check_file(PTMcode_intraprotein, expected_extension=&#39;.gz&#39;)</span>
    <span class="c1">#            spliced_ptms = add_PTMcode_intraprotein(spliced_ptms, fname = PTMcode_intraprotein)</span>
    <span class="c1">#        else:</span>
    <span class="c1">#            raise ValueError(&#39;PTMcode_intraprotein must be either a boolean (download PTMcode data automatically, slower) or a string (path to PTMcode data file, faster)&#39;)</span>
    <span class="c1">#    except Exception as e:</span>
    <span class="c1">#        print(f&#39;Error adding PTMcode intraprotein interaction data. Error message: {e}&#39;)</span>
    <span class="k">if</span> <span class="n">PTMcode_interprotein</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">PTMcode_interprotein</span><span class="p">,</span> <span class="nb">bool</span><span class="p">):</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
            <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">PTMcode_interprotein</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">PTMcode_interprotein</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMcode_interprotein</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="n">PTMcode_interprotein</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;PTMcode_interprotein must be either a boolean (download PTMcode data automatically, slower) or a string (path to PTMcode data file, faster)&#39;</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMcode interprotein interaction data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">DEPOD</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_DEPOD_phosphatase_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding DEPOD phosphatase data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">RegPhos</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">RegPhos</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="n">check_file</span><span class="p">(</span><span class="n">RegPhos</span><span class="p">,</span> <span class="n">expected_extension</span><span class="o">=</span><span class="s1">&#39;.txt&#39;</span><span class="p">)</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">RegPhos</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding RegPhos kinase substrate data data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ptmsigdb_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">add_PTMsigDB_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">file</span> <span class="o">=</span> <span class="n">ptmsigdb_file</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Error adding PTMsigDB data. Error message: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">combine_similar</span><span class="p">:</span>
        <span class="n">interaction_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PTMcode:Interprotein_Interactions&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:ON_PROT_INTERACT&#39;</span><span class="p">,</span> <span class="s1">&#39;PSP:Kinase&#39;</span><span class="p">,</span> <span class="s1">&#39;PTMInt:Interaction&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos:Kinase&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD:Phosphatase&#39;</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">set</span><span class="p">(</span><span class="n">interaction_cols</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Combining interaction data from multiple databases&#39;</span><span class="p">)</span>
            <span class="n">interact</span> <span class="o">=</span> <span class="n">combine_interaction_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">interaction_databases</span> <span class="o">=</span> <span class="n">interactions_to_combine</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">interact</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Interacting Gene&#39;</span><span class="p">]</span><span class="o">+</span><span class="s1">&#39;-&gt;&#39;</span><span class="o">+</span><span class="n">interact</span><span class="p">[</span><span class="s1">&#39;Type&#39;</span><span class="p">]</span>
                <span class="n">interact</span> <span class="o">=</span> <span class="n">interact</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">],</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>
                <span class="k">if</span> <span class="s1">&#39;Combined:Interactions&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">])</span>
    
                <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">interact</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Canonical Isoform&#39;</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Combined:Interactions&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        <span class="c1">#check for what kinase data is available</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">combine_KS_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">ks_databases</span><span class="o">=</span><span class="n">kinases_to_combine</span><span class="p">)</span> <span class="c1">#add combined kinase column</span>


    <span class="k">return</span> <span class="n">spliced_ptms</span></div>


</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>