
.. DO NOT EDIT.
.. THIS FILE WAS AUTOMATICALLY GENERATED BY SPHINX-GALLERY.
.. TO MAKE CHANGES, EDIT THE SOURCE PYTHON FILE:
.. "gallery_output/flanking_sequences/plot_sh2_domain_motifs.py"
.. LINE NUMBERS ARE GIVEN BELOW.

.. only:: html

    .. note::
        :class: sphx-glr-download-link-note

        :ref:`Go to the end <sphx_glr_download_gallery_output_flanking_sequences_plot_sh2_domain_motifs.py>`
        to download the full example code.

.. rst-class:: sphx-glr-example-title

.. _sphx_glr_gallery_output_flanking_sequences_plot_sh2_domain_motifs.py:


Identify altered SH2 domain motifs
===========================================================

A potential consequence of altered flanking sequences is that the short linear motifs important for driving PTM-specific domain interactions are disrupted, such as for the phosphotyrosine binding domains SH2 or 14-3-3 proteins (which bind to phosphoserines and threonines). Using PTM-POSE and motif data from [ELM](http://elm.eu.org/searchdb.html), we can identify and visualize the altered 14-3-3 domain motifs due to splicing events.

.. GENERATED FROM PYTHON SOURCE LINES 7-14

.. code-block:: Python


    from ptm_pose import helpers
    from ptm_pose.analyze import flank_analysis as afs

    #load example altered flanking sequence data
    altered_flanks = helpers.load_example_data(altered_flanks = True)








.. GENERATED FROM PYTHON SOURCE LINES 15-16

First, we need to identify the linear motifs present for each altered flanking sequence event. We can do this with the `compare_inclusion_motifs` function, which will identify matching any matching motifs for both the inclusion and exclusion flanking sequences. 

.. GENERATED FROM PYTHON SOURCE LINES 16-20

.. code-block:: Python


    altered_flanks = afs.compare_inclusion_motifs(altered_flanks)
    altered_flanks[['Gene', 'Residue', 'PTM Position in Isoform', 'Motif only in Inclusion', 'Motif only in Exclusion']].head()






.. raw:: html

    <div class="output_subarea output_html rendered_html output_result">
    <div>
    <style scoped>
        .dataframe tbody tr th:only-of-type {
            vertical-align: middle;
        }

        .dataframe tbody tr th {
            vertical-align: top;
        }

        .dataframe thead th {
            text-align: right;
        }
    </style>
    <table border="1" class="dataframe">
      <thead>
        <tr style="text-align: right;">
          <th></th>
          <th>Gene</th>
          <th>Residue</th>
          <th>PTM Position in Isoform</th>
          <th>Motif only in Inclusion</th>
          <th>Motif only in Exclusion</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>0</th>
          <td>ARHGAP17</td>
          <td>S</td>
          <td>497</td>
          <td>DOC_MAPK_gen_1;LIG_PDZ_Class_2;LIG_WD40_WDR5_V...</td>
          <td>LIG_SH3_3;LIG_SH3_1</td>
        </tr>
        <tr>
          <th>1</th>
          <td>ARHGAP17</td>
          <td>S</td>
          <td>497</td>
          <td>DOC_MAPK_gen_1;LIG_PDZ_Class_2;LIG_WD40_WDR5_V...</td>
          <td>LIG_SH3_3;LIG_SH3_1</td>
        </tr>
        <tr>
          <th>2</th>
          <td>KRAS</td>
          <td>K</td>
          <td>147</td>
          <td>TRG_ER_diArg_1</td>
          <td></td>
        </tr>
        <tr>
          <th>3</th>
          <td>KRAS</td>
          <td>K</td>
          <td>147</td>
          <td>TRG_ER_diArg_1</td>
          <td></td>
        </tr>
        <tr>
          <th>4</th>
          <td>KRAS</td>
          <td>T</td>
          <td>148</td>
          <td>TRG_ER_diArg_1</td>
          <td></td>
        </tr>
      </tbody>
    </table>
    </div>
    </div>
    <br />
    <br />

.. GENERATED FROM PYTHON SOURCE LINES 21-22

We can then identify the instances in which 14-3-3 motifs are altered:

.. GENERATED FROM PYTHON SOURCE LINES 22-25

.. code-block:: Python


    fourteen33_motifs = afs.identify_change_to_specific_motif(altered_flanks, elm_motif_name = '14-3-3', modification_class = 'Phosphorylation', residues = ['S','T'])
    fourteen33_motifs[['Gene', 'Residue', 'PTM Position in Isoform', 'Motif only in Inclusion', 'Motif only in Exclusion']]




.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    77 PTMs removed due to insignificant splice event (p < 0.05, dpsi >= 0.2): (51.33%)
    Final number of PTMs to be assessed: 73


.. raw:: html

    <div class="output_subarea output_html rendered_html output_result">
    <div>
    <style scoped>
        .dataframe tbody tr th:only-of-type {
            vertical-align: middle;
        }

        .dataframe tbody tr th {
            vertical-align: top;
        }

        .dataframe thead th {
            text-align: right;
        }
    </style>
    <table border="1" class="dataframe">
      <thead>
        <tr style="text-align: right;">
          <th></th>
          <th>Gene</th>
          <th>Residue</th>
          <th>PTM Position in Isoform</th>
          <th>Motif only in Inclusion</th>
          <th>Motif only in Exclusion</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <th>23</th>
          <td>MLPH</td>
          <td>S</td>
          <td>336</td>
          <td>LIG_14-3-3_CanoR_1</td>
          <td>NaN</td>
        </tr>
        <tr>
          <th>52</th>
          <td>CEACAM1</td>
          <td>T</td>
          <td>457</td>
          <td>LIG_14-3-3_CterR_2</td>
          <td>NaN</td>
        </tr>
        <tr>
          <th>68</th>
          <td>ENAH</td>
          <td>S</td>
          <td>512</td>
          <td>NaN</td>
          <td>LIG_14-3-3_CterR_2</td>
        </tr>
        <tr>
          <th>84</th>
          <td>TSC2</td>
          <td>S</td>
          <td>946</td>
          <td>NaN</td>
          <td>LIG_14-3-3_CterR_2</td>
        </tr>
      </tbody>
    </table>
    </div>
    </div>
    <br />
    <br />

.. GENERATED FROM PYTHON SOURCE LINES 26-27

And visualize the differences in sequence

.. GENERATED FROM PYTHON SOURCE LINES 27-28

.. code-block:: Python

    afs.plot_alterations_matrix(fourteen33_motifs)



.. image-sg:: /gallery_output/flanking_sequences/images/sphx_glr_plot_sh2_domain_motifs_001.png
   :alt: plot sh2 domain motifs
   :srcset: /gallery_output/flanking_sequences/images/sphx_glr_plot_sh2_domain_motifs_001.png
   :class: sphx-glr-single-img






.. rst-class:: sphx-glr-timing

   **Total running time of the script:** (0 minutes 4.535 seconds)


.. _sphx_glr_download_gallery_output_flanking_sequences_plot_sh2_domain_motifs.py:

.. only:: html

  .. container:: sphx-glr-footer sphx-glr-footer-example

    .. container:: sphx-glr-download sphx-glr-download-jupyter

      :download:`Download Jupyter notebook: plot_sh2_domain_motifs.ipynb <plot_sh2_domain_motifs.ipynb>`

    .. container:: sphx-glr-download sphx-glr-download-python

      :download:`Download Python source code: plot_sh2_domain_motifs.py <plot_sh2_domain_motifs.py>`

    .. container:: sphx-glr-download sphx-glr-download-zip

      :download:`Download zipped: plot_sh2_domain_motifs.zip <plot_sh2_domain_motifs.zip>`


.. only:: html

 .. rst-class:: sphx-glr-signature

    `Gallery generated by Sphinx-Gallery <https://sphinx-gallery.github.io>`_
