

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016) &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-binder.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-dataframe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery-rendered-html.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'Examples/ESRP1_knockdown';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Frequently Asked Questions" href="../faq.html" />
    <link rel="prev" title="Full Analysis Examples" href="Examples.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Dependencies.html">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../PTM_POSE.html">PTM-POSE Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../General_Instructions.html">Running PTM-POSE</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Examples</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../gallery_output/index.html">Types of Analysis Performed with PTM-POSE</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_protein_interactions.html">Identify protein interactions that may be impacted by splicing of PTMs</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_num_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../gallery_output/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="Examples.html">Full Analysis Examples</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/Examples/ESRP1_knockdown.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-1:-Load-the-data-and-initialize-PTM-POSE">Phase 1: Load the data and initialize PTM-POSE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-2:-Project-PTMs-onto-differentially-included-regions">Phase 2: Project PTMs onto differentially included regions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-3:-Identify-PTMs-with-altered-flanking-sequences-as-a-result-of-splice-events">Phase 3: Identify PTMs with altered flanking sequences as a result of splice events</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-4:-Annotate-PTMs-with-functional-information">Phase 4: Annotate PTMs with functional information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-5:-Analyze-Results">Phase 5: Analyze Results</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="Project-PTMs-onto-ESRP1-knockdown-data-from-MATS-(Yang-et-al,-2016)">
<h1>Project PTMs onto ESRP1 knockdown data from MATS (Yang et al, 2016)<a class="headerlink" href="#Project-PTMs-onto-ESRP1-knockdown-data-from-MATS-(Yang-et-al,-2016)" title="Permalink to this heading">#</a></h1>
<p>Here is an example of running PTM-POSE on MATS analysis of RNA sequencing data from ESRP1 knockdown experiments performed by Yang et al, 2016</p>
<p>First, let’s focus on skipped exon events.</p>
<section id="Phase-1:-Load-the-data-and-initialize-PTM-POSE">
<h2>Phase 1: Load the data and initialize PTM-POSE<a class="headerlink" href="#Phase-1:-Load-the-data-and-initialize-PTM-POSE" title="Permalink to this heading">#</a></h2>
<p>To identify differentially included PTMs as a result of ESRP1 knockdown, we need three layers of information for each splice event: 1. Chromosome 2. DNA strand 2. Start and end coordinates of the event (either hg19 or hg38)</p>
<p>Optionally, we can also provide: 1. Gene name 2. Event ID 3. Delta PSI for the event 4. Significance of the event</p>
<p>With PTM-POSE, we need to indicate where to find this information within the splice data</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">SE_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_excel</span><span class="p">(</span><span class="s1">&#39;../../ESRP1_data/Yang2016/esrp1_knockdown_data_Yang2016.xlsx&#39;</span><span class="p">,</span> <span class="n">sheet_name</span><span class="o">=</span><span class="s1">&#39;rMATS ESRP KD&#39;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">179</span><span class="p">]</span>


<span class="c1"># required column information</span>
<span class="n">chromosome_col</span> <span class="o">=</span> <span class="s1">&#39;chr&#39;</span>
<span class="n">strand_col</span> <span class="o">=</span> <span class="s1">&#39;strand&#39;</span>
<span class="n">region_start_col</span> <span class="o">=</span> <span class="s1">&#39;exonStart_0base&#39;</span>
<span class="n">region_end_col</span> <span class="o">=</span> <span class="s1">&#39;exonEnd&#39;</span>

<span class="c1"># optional column information (None if nothing is provided and will not be appended to the output)</span>
<span class="n">gene_col</span> <span class="o">=</span> <span class="s1">&#39;geneSymbol&#39;</span>
<span class="n">event_id_col</span> <span class="o">=</span> <span class="kc">None</span>   <span class="c1">#not in the data</span>
<span class="n">dPSI_col</span> <span class="o">=</span> <span class="s1">&#39;meanDeltaPSI&#39;</span>
<span class="n">sig_col</span> <span class="o">=</span> <span class="s1">&#39;FDR&#39;</span>

<span class="c1">#look at the data</span>
<span class="n">SE_data</span><span class="p">[[</span><span class="n">gene_col</span><span class="p">,</span> <span class="n">chromosome_col</span><span class="p">,</span> <span class="n">strand_col</span><span class="p">,</span> <span class="n">region_start_col</span><span class="p">,</span> <span class="n">region_end_col</span><span class="p">,</span> <span class="n">dPSI_col</span><span class="p">,</span> <span class="n">sig_col</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>geneSymbol</th>
      <th>chr</th>
      <th>strand</th>
      <th>exonStart_0base</th>
      <th>exonEnd</th>
      <th>meanDeltaPSI</th>
      <th>FDR</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SPAG9</td>
      <td>chr17</td>
      <td>-</td>
      <td>49053223</td>
      <td>49053262</td>
      <td>0.227</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ARHGAP17</td>
      <td>chr16</td>
      <td>-</td>
      <td>24950684</td>
      <td>24950918</td>
      <td>0.413</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ITGA6</td>
      <td>chr2</td>
      <td>+</td>
      <td>173366499</td>
      <td>173366629</td>
      <td>-0.361</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KRAS</td>
      <td>chr12</td>
      <td>-</td>
      <td>25368370</td>
      <td>25368494</td>
      <td>-0.068</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TCIRG1</td>
      <td>chr11</td>
      <td>+</td>
      <td>67817953</td>
      <td>67818131</td>
      <td>0.368</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>The strand can either be provided use ‘+’ and ‘-’ or using 1 and -1 to indicate the forward and reverse strand, the code will convert strand to integer format (-1 or 1) when running.</p>
<p>If this is the first time running PTM-POSE, you will need to download ptm_coordinates. If you set save = True, the coordinates will be saved for the future so you do not need to redownload them, but you can also set save = False to avoid saving the coordinates (will take ~60MB of space)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">pose_config</span>
<span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span> <span class="o">=</span> <span class="n">pose_config</span><span class="o">.</span><span class="n">download_ptm_coordinates</span><span class="p">(</span><span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Phase-2:-Project-PTMs-onto-differentially-included-regions">
<h2>Phase 2: Project PTMs onto differentially included regions<a class="headerlink" href="#Phase-2:-Project-PTMs-onto-differentially-included-regions" title="Permalink to this heading">#</a></h2>
<p>We can then use the project module of PTM-POSE to identify PTMs that can be found in these regions. This dataset uses the hg19 genome build, so we need to specify this using the ‘coordinate_type’ parameter.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">project</span>

<span class="n">splice_data</span><span class="p">,</span> <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">project</span><span class="o">.</span><span class="n">project_ptms_onto_splice_events</span><span class="p">(</span><span class="n">SE_data</span><span class="p">,</span> <span class="n">chromosome_col</span> <span class="o">=</span> <span class="n">chromosome_col</span><span class="p">,</span> <span class="n">strand_col</span> <span class="o">=</span> <span class="n">strand_col</span><span class="p">,</span> <span class="n">region_start_col</span> <span class="o">=</span> <span class="n">region_start_col</span><span class="p">,</span> <span class="n">region_end_col</span> <span class="o">=</span> <span class="n">region_end_col</span><span class="p">,</span> <span class="n">gene_col</span> <span class="o">=</span> <span class="n">gene_col</span><span class="p">,</span> <span class="n">event_id_col</span> <span class="o">=</span> <span class="n">event_id_col</span><span class="p">,</span> <span class="n">dPSI_col</span> <span class="o">=</span> <span class="n">dPSI_col</span><span class="p">,</span> <span class="n">sig_col</span> <span class="o">=</span> <span class="n">sig_col</span><span class="p">,</span> <span class="n">coordinate_type</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Translator file not found. Downloading mapping information between UniProt and Gene Names from pybiomart
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Projecting PTMs onto splice events using hg19 coordinates.: 100%|██████████| 179/179 [00:03&lt;00:00, 48.82it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PTMs projection successful (475 identified).

</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>

</pre></div></div>
</div>
<p>From this, there are two outputs: 1. The original splice dataframe with additional PTM information added</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">splice_data</span><span class="p">[[</span><span class="n">gene_col</span><span class="p">,</span> <span class="n">chromosome_col</span><span class="p">,</span> <span class="n">strand_col</span><span class="p">,</span> <span class="n">region_start_col</span><span class="p">,</span> <span class="n">region_end_col</span><span class="p">,</span> <span class="n">dPSI_col</span><span class="p">,</span> <span class="n">sig_col</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;PTMs&#39;</span><span class="p">,</span> <span class="s1">&#39;Number of PTMs Affected&#39;</span><span class="p">,</span> <span class="s1">&#39;Number of Unique PTM Sites by Position&#39;</span><span class="p">,</span> <span class="s1">&#39;Event Length&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Density (PTMs/bp)&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>geneSymbol</th>
      <th>chr</th>
      <th>strand</th>
      <th>exonStart_0base</th>
      <th>exonEnd</th>
      <th>meanDeltaPSI</th>
      <th>FDR</th>
      <th>PTMs</th>
      <th>Number of PTMs Affected</th>
      <th>Number of Unique PTM Sites by Position</th>
      <th>Event Length</th>
      <th>PTM Density (PTMs/bp)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>SPAG9</td>
      <td>17</td>
      <td>-</td>
      <td>49053223</td>
      <td>49053262</td>
      <td>0.227</td>
      <td>0</td>
      <td>NaN</td>
      <td>0</td>
      <td>0</td>
      <td>39</td>
      <td>0.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ARHGAP17</td>
      <td>16</td>
      <td>-</td>
      <td>24950684</td>
      <td>24950918</td>
      <td>0.413</td>
      <td>0</td>
      <td>Q68EM7_S575.0 (Phosphorylation)/Q68EM7_S570.0 ...</td>
      <td>6</td>
      <td>1</td>
      <td>234</td>
      <td>0.004274</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ITGA6</td>
      <td>2</td>
      <td>+</td>
      <td>173366499</td>
      <td>173366629</td>
      <td>-0.361</td>
      <td>0</td>
      <td>P23229_Ynan (Phosphorylation)/P23229_Tnan (Pho...</td>
      <td>7</td>
      <td>4</td>
      <td>130</td>
      <td>0.030769</td>
    </tr>
    <tr>
      <th>3</th>
      <td>KRAS</td>
      <td>12</td>
      <td>-</td>
      <td>25368370</td>
      <td>25368494</td>
      <td>-0.068</td>
      <td>0</td>
      <td>P01116_C186 (Methylation)/P01116_C180 (Palmito...</td>
      <td>3</td>
      <td>2</td>
      <td>124</td>
      <td>0.016129</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TCIRG1</td>
      <td>11</td>
      <td>+</td>
      <td>67817953</td>
      <td>67818131</td>
      <td>0.368</td>
      <td>0</td>
      <td>NaN</td>
      <td>0</td>
      <td>0</td>
      <td>178</td>
      <td>0.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<ol class="arabic simple" start="2">
<li><p>New dataframe that has each PTM and additional information about the PTM in its own row</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">spliced_ptms</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>dPSI</th>
      <th>Significance</th>
      <th>Gene</th>
      <th>Source of PTM</th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Gene Location (hg19)</th>
      <th>Modification</th>
      <th>Modification Class</th>
      <th>Proximity to Region Start (bp)</th>
      <th>Proximity to Region End (bp)</th>
      <th>Proximity to Splice Boundary (bp)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.413</td>
      <td>0.0</td>
      <td>ARHGAP17</td>
      <td>Q68EM7-1_S575</td>
      <td>Q68EM7</td>
      <td>S</td>
      <td>575.0</td>
      <td>24950686.0</td>
      <td>Phosphoserine</td>
      <td>Phosphorylation</td>
      <td>2.0</td>
      <td>232.0</td>
      <td>2.0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.413</td>
      <td>0.0</td>
      <td>ARHGAP17</td>
      <td>Q68EM7-1_S570</td>
      <td>Q68EM7</td>
      <td>S</td>
      <td>570.0</td>
      <td>24950701.0</td>
      <td>Phosphoserine</td>
      <td>Phosphorylation</td>
      <td>17.0</td>
      <td>217.0</td>
      <td>17.0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.413</td>
      <td>0.0</td>
      <td>ARHGAP17</td>
      <td>Q68EM7-1_S560</td>
      <td>Q68EM7</td>
      <td>S</td>
      <td>560.0</td>
      <td>24950731.0</td>
      <td>Phosphoserine</td>
      <td>Phosphorylation</td>
      <td>47.0</td>
      <td>187.0</td>
      <td>47.0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.413</td>
      <td>0.0</td>
      <td>ARHGAP17</td>
      <td>Q68EM7-1_S553</td>
      <td>Q68EM7</td>
      <td>S</td>
      <td>553.0</td>
      <td>24950752.0</td>
      <td>Phosphoserine</td>
      <td>Phosphorylation</td>
      <td>68.0</td>
      <td>166.0</td>
      <td>68.0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.413</td>
      <td>0.0</td>
      <td>ARHGAP17</td>
      <td>Q68EM7-1_S547</td>
      <td>Q68EM7</td>
      <td>S</td>
      <td>547.0</td>
      <td>24950770.0</td>
      <td>Phosphoserine</td>
      <td>Phosphorylation</td>
      <td>86.0</td>
      <td>148.0</td>
      <td>86.0</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>For MATS data, there is also a built in function for running PTM-POSE on MATS data, including all events:</p>
</section>
<section id="Phase-3:-Identify-PTMs-with-altered-flanking-sequences-as-a-result-of-splice-events">
<h2>Phase 3: Identify PTMs with altered flanking sequences as a result of splice events<a class="headerlink" href="#Phase-3:-Identify-PTMs-with-altered-flanking-sequences-as-a-result-of-splice-events" title="Permalink to this heading">#</a></h2>
<p>In addition to differential inclusion of PTMs, some PTMs may experience altered flanking sequences. We can use the project module of PTM-POSE to identify PTMs for which this happens. You will need to provide the same layers of information, plus the genomic coordinates of the regions flanking the spliced region.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">flanking_sequences</span>

<span class="n">first_flank_start_col</span> <span class="o">=</span> <span class="s1">&#39;firstFlankingES&#39;</span>
<span class="n">first_flank_end_col</span><span class="o">=</span><span class="s1">&#39;firstFlankingEE&#39;</span>
<span class="n">second_flank_start_col</span> <span class="o">=</span> <span class="s1">&#39;secondFlankingES&#39;</span>
<span class="n">second_flank_end_col</span> <span class="o">=</span> <span class="s1">&#39;secondFlankingEE&#39;</span>

<span class="n">flanks</span> <span class="o">=</span> <span class="n">flanking_sequences</span><span class="o">.</span><span class="n">get_flanking_changes_from_splice_data</span><span class="p">(</span><span class="n">SE_data</span><span class="p">,</span> <span class="n">chromosome_col</span> <span class="o">=</span> <span class="n">chromosome_col</span><span class="p">,</span> <span class="n">strand_col</span> <span class="o">=</span> <span class="n">strand_col</span><span class="p">,</span> <span class="n">first_flank_start_col</span> <span class="o">=</span> <span class="n">first_flank_start_col</span><span class="p">,</span> <span class="n">first_flank_end_col</span><span class="o">=</span><span class="n">first_flank_end_col</span><span class="p">,</span> <span class="n">second_flank_start_col</span> <span class="o">=</span> <span class="n">second_flank_start_col</span><span class="p">,</span> <span class="n">second_flank_end_col</span> <span class="o">=</span> <span class="n">second_flank_end_col</span> <span class="p">,</span> <span class="n">spliced_region_start_col</span> <span class="o">=</span> <span class="n">region_start_col</span><span class="p">,</span> <span class="n">spliced_region_end_col</span> <span class="o">=</span> <span class="n">region_end_col</span><span class="p">,</span> <span class="n">dPSI_col</span><span class="o">=</span><span class="n">dPSI_col</span><span class="p">,</span> <span class="n">sig_col</span> <span class="o">=</span> <span class="n">sig_col</span><span class="p">,</span> <span class="n">event_id_col</span> <span class="o">=</span> <span class="n">event_id_col</span><span class="p">,</span> <span class="n">coordinate_type</span> <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\Users\Sam\miniconda3\envs\testing_pose\Lib\site-packages\Bio\pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.
  warnings.warn(
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">flanks</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Event ID</th>
      <th>Source of PTM</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Inclusion Sequence</th>
      <th>Exclusion Sequence</th>
      <th>Region</th>
      <th>Translation Success</th>
      <th>Matched</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>3</td>
      <td>P01116-2_T148;P01116-1_T148</td>
      <td>T</td>
      <td>148</td>
      <td>ETSAKtRQESG</td>
      <td>ETSAKtRQGC*</td>
      <td>Second</td>
      <td>True</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>3</td>
      <td>P01116-1_K147;P01116-2_K147</td>
      <td>K</td>
      <td>147</td>
      <td>IETSAkTRQES</td>
      <td>IETSAkTRQGC</td>
      <td>Second</td>
      <td>True</td>
      <td>False</td>
    </tr>
    <tr>
      <th>0</th>
      <td>8</td>
      <td>Q9UPQ0-1_S746</td>
      <td>S</td>
      <td>746</td>
      <td>LPNLNsQGVAW</td>
      <td>LPNLNsQGGFS</td>
      <td>First</td>
      <td>True</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>8</td>
      <td>Q9UPQ0-10_S750;Q9UPQ0-6_S596;Q9UPQ0-1_S750</td>
      <td>S</td>
      <td>750</td>
      <td>PSQVDsPSSEK</td>
      <td>ILKVDsPSSEK</td>
      <td>Second</td>
      <td>True</td>
      <td>False</td>
    </tr>
    <tr>
      <th>0</th>
      <td>11</td>
      <td>P62847-1_K129</td>
      <td>K</td>
      <td>NaN</td>
      <td>NVGAGkKSVSW</td>
      <td>NVGAGkKAEGV</td>
      <td>First</td>
      <td>True</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>We can also do additional comparisons, such as comparing sequence identity and looking for matching elm motifs.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">flanks</span> <span class="o">=</span> <span class="n">flanking_sequences</span><span class="o">.</span><span class="n">compare_flanking_sequences</span><span class="p">(</span><span class="n">flanks</span><span class="p">)</span>
<span class="n">flanks</span> <span class="o">=</span> <span class="n">flanking_sequences</span><span class="o">.</span><span class="n">compare_inclusion_motifs</span><span class="p">(</span><span class="n">flanks</span><span class="p">)</span>
<span class="n">flanks</span><span class="p">[[</span><span class="s1">&#39;Source of PTM&#39;</span><span class="p">,</span><span class="s1">&#39;Sequence Identity&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Positions&#39;</span><span class="p">,</span><span class="s1">&#39;Residue Changes&#39;</span><span class="p">,</span> <span class="s1">&#39;Altered Flank Side&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif only in Inclusion&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif only in Exclusion&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
<section id="Phase-4:-Annotate-PTMs-with-functional-information">
<h2>Phase 4: Annotate PTMs with functional information<a class="headerlink" href="#Phase-4:-Annotate-PTMs-with-functional-information" title="Permalink to this heading">#</a></h2>
<p>Once we have PTMs impacted by splicing, we can also annotate them with additional information. This can be done using the annotate module of PTM-POSE, and can be used with outputs from either the project module (differentially included PTMs) or the flanking_sequence module (PTMs with altered flanking sequences).</p>
<p>Currently, there are functions for appending information from: 1. PhosphoSitePlus (function, biological process, disease association, interactions, and kinase-substrate), 2. PTMsigDB (iKiP db, perturbations) 3. RegPhos (kinase-substrate), 4. PTMcode (inter and intraprotein interactions) 5. PTMInt (interactions) 6. DEPOD (Phosphatase-substrate) 7. ELM (interactions, motifs)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">annotate</span>

<span class="c1">#where to find PhosphoSitePlus data</span>
<span class="n">psp_regulatory_file</span> <span class="o">=</span> <span class="s1">&#39;/PhosphoSitePlus/Regulatory_sites.gz&#39;</span>
<span class="n">psp_disease_file</span> <span class="o">=</span> <span class="s1">&#39;/PhosphoSitePlus/Disease-associated_sites.gz&#39;</span>
<span class="n">psp_kinase_file</span> <span class="o">=</span> <span class="s1">&#39;/Database_Information/PhosphoSitePlus/Kinase_Substrate_Dataset.gz&#39;</span>

<span class="c1">#where to find ELM data</span>

<span class="c1">#PhosphoSitePlus data (due to licencsing issues, must be downloaded manually from PhosphoSitePlus and the file path provided)</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_regulatory_site_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;/PhosphoSitePlus/Regulatory_sites.gz&#39;</span><span class="p">)</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_disease_association</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;/PhosphoSitePlus/Disease-associated_sites.gz&#39;</span><span class="p">)</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PSP_kinase_substrate_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="s1">&#39;/Database_Information/PhosphoSitePlus/Kinase_Substrate_Dataset.gz&#39;</span><span class="p">)</span>

<span class="c1">#ELM interactions (will be faster if file is downloaded manually from ELM and the file path provided)</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_ELM_interactions</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>

<span class="c1">#PTMint interactions</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_PTMint_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>

<span class="c1">#PTMcode interactions (will be faster/more reliable if file is downloaded manually from PTMcode and the file path provided)</span>
<span class="n">ptm_code_interprotein</span> <span class="o">=</span> <span class="s1">&#39;/PTMcode2_associations_between_proteins.txt.gz&#39;</span>

<span class="c1">#DEPOD phosphatase data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_DEPOD_phosphatase_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>

<span class="c1">#RegPhos data</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">add_RegPhos_data</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>

<span class="c1">#annotate ptms</span>
<span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">annotate_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
PhosphoSitePlus regulatory_site information added:
         -&gt;6 PTMs in dataset found associated with a molecular function
         -&gt;7 PTMs in dataset found associated with a biological process
         -&gt;2 PTMs in dataset found associated with a protein interaction
PhosphoSitePlus disease associations added: 1 PTM sites in dataset found associated with a disease in PhosphoSitePlus
PhosphoSitePlus kinase-substrate interactions added: 6 phosphorylation sites in dataset found associated with a kinase in PhosphoSitePlus
ELM interaction instances added: 1 PTMs in dataset found associated with at least one known ELM instance
PTMInt data added: 2 PTMs in dataset found with PTMInt interaction information
PTMcode interprotein interactions added: 27 PTMs in dataset found with PTMcode interprotein interaction information
DEPOD Phosphatase substrates added: 0 PTMs in dataset found with Phosphatase substrate information
RegPhos kinase-substrate data added: 3 PTMs in dataset found with kinase-substrate information
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
c:\Users\Sam\miniconda3\envs\testing_pose\Lib\site-packages\ptm_pose\annotate.py:558: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  regphos = pd.read_csv(&#39;http://140.138.144.141/~RegPhos/download/RegPhos_Phos_human.txt&#39;, sep = &#39;\t&#39;)
</pre></div></div>
</div>
</section>
<section id="Phase-5:-Analyze-Results">
<h2>Phase 5: Analyze Results<a class="headerlink" href="#Phase-5:-Analyze-Results" title="Permalink to this heading">#</a></h2>
<p>Once we have all of this information, we can start to assess how PTMs are impacted by splicing. Let’s first get an idea for how many PTMs have different annotations associated with them from the various sources</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">analyze</span>

<span class="n">analyze</span><span class="o">.</span><span class="n">show_available_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Examples_ESRP1_knockdown_19_0.png" src="../_images/Examples_ESRP1_knockdown_19_0.png" />
</div>
</div>
<p>There are several ptms that have previously been annotated with specific functions, let’s take a look at those:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">annotations</span><span class="p">,</span> <span class="n">annotation_counts</span> <span class="o">=</span> <span class="n">analyze</span><span class="o">.</span><span class="n">get_ptm_annotations</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">annotation_type</span> <span class="o">=</span> <span class="s1">&#39;Process&#39;</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">)</span>
<span class="n">annotations</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene</th>
      <th>UniProtKB Accession</th>
      <th>Residue</th>
      <th>PTM Position in Canonical Isoform</th>
      <th>Modification Class</th>
      <th>PSP:ON_PROCESS</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>145</th>
      <td>CEACAM1</td>
      <td>P13688</td>
      <td>S</td>
      <td>461.0</td>
      <td>Phosphorylation</td>
      <td>apoptosis, altered</td>
    </tr>
    <tr>
      <th>184</th>
      <td>YAP1</td>
      <td>P46937</td>
      <td>K</td>
      <td>342.0</td>
      <td>Ubiquitination</td>
      <td>carcinogenesis, altered</td>
    </tr>
    <tr>
      <th>217</th>
      <td>TSC2</td>
      <td>P49815</td>
      <td>S</td>
      <td>981.0</td>
      <td>Phosphorylation</td>
      <td>carcinogenesis, inhibited; cell growth, inhibi...</td>
    </tr>
    <tr>
      <th>395</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>S</td>
      <td>387.0</td>
      <td>Phosphorylation</td>
      <td>cell motility, altered</td>
    </tr>
    <tr>
      <th>407</th>
      <td>SPHK2</td>
      <td>Q9NRA0</td>
      <td>T</td>
      <td>614.0</td>
      <td>Phosphorylation</td>
      <td>cell motility, altered</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">annotation_counts</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>PSP:ON_PROCESS</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>cell motility, altered</td>
      <td>3</td>
    </tr>
    <tr>
      <th>1</th>
      <td>cell growth, induced</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>apoptosis, altered</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>carcinogenesis, altered</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>carcinogenesis, inhibited</td>
      <td>1</td>
    </tr>
    <tr>
      <th>5</th>
      <td>cell growth, inhibited</td>
      <td>1</td>
    </tr>
    <tr>
      <th>6</th>
      <td>autophagy, inhibited</td>
      <td>1</td>
    </tr>
    <tr>
      <th>7</th>
      <td>signaling pathway regulation</td>
      <td>1</td>
    </tr>
    <tr>
      <th>8</th>
      <td>cytoskeletal reorganization</td>
      <td>1</td>
    </tr>
    <tr>
      <th>9</th>
      <td>cell adhesion, inhibited</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="Examples.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Full Analysis Examples</p>
      </div>
    </a>
    <a class="right-next"
       href="../faq.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Frequently Asked Questions</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-1:-Load-the-data-and-initialize-PTM-POSE">Phase 1: Load the data and initialize PTM-POSE</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-2:-Project-PTMs-onto-differentially-included-regions">Phase 2: Project PTMs onto differentially included regions</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-3:-Identify-PTMs-with-altered-flanking-sequences-as-a-result-of-splice-events">Phase 3: Identify PTMs with altered flanking sequences as a result of splice events</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-4:-Annotate-PTMs-with-functional-information">Phase 4: Annotate PTMs with functional information</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Phase-5:-Analyze-Results">Phase 5: Analyze Results</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>