
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Kinase affinity due to PTM flanking sequence alterations &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'gallery_output/flanking_sequences/plot_kinase_library_affinity';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Frequently Asked Questions" href="../../faq.html" />
    <link rel="prev" title="Probing where and how PTM flanking sequences are altered" href="plot_location_altered_flanks.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../index.html">Gallery</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 has-children"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_ksea.html">Kinase substrate enrichment analysis (KSEA)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_protein_interactions.html">Protein interactions network</a></li>
<li class="toctree-l3"><a class="reference internal" href="../Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
</ul>
</details></li>
<li class="toctree-l2 current active has-children"><a class="reference internal" href="index.html">Analyzing altered flanking sequences</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
<li class="toctree-l3 current active"><a class="current reference internal" href="#">Kinase affinity due to PTM flanking sequence alterations</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/gallery_output/flanking_sequences/plot_kinase_library_affinity.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Kinase affinity due to PTM flanking sequence alterations</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <div class="sphx-glr-download-link-note admonition note">
<p class="admonition-title">Note</p>
<p><a class="reference internal" href="#sphx-glr-download-gallery-output-flanking-sequences-plot-kinase-library-affinity-py"><span class="std std-ref">Go to the end</span></a>
to download the full example code.</p>
</div>
<section class="sphx-glr-example-title" id="kinase-affinity-due-to-ptm-flanking-sequence-alterations">
<span id="sphx-glr-gallery-output-flanking-sequences-plot-kinase-library-affinity-py"></span><h1>Kinase affinity due to PTM flanking sequence alterations<a class="headerlink" href="#kinase-affinity-due-to-ptm-flanking-sequence-alterations" title="Link to this heading">#</a></h1>
<p>Kinases are known to recognize specific motifs surrounding Y, S, and T residues, and these motifs can be used to predict kinase interactions using a tool called Kinase Library. Here, we will use Kinase Library to identify kinases who may have a stronger preference for the flanking sequences of PTMs in one isoform vs. another. For this example, let’s focus on phosphorylation sites that have larger changes &gt;0.4 in the splicing data.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The optional dependency <cite>kinase-library</cite> is required to run this example. You can install it using <cite>pip install kinase-library</cite>.</p>
</div>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose</span><span class="w"> </span><span class="kn">import</span> <span class="n">helpers</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose.analyze</span><span class="w"> </span><span class="kn">import</span> <span class="n">enzyme</span>

<span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">load_example_data</span><span class="p">(</span><span class="n">altered_flanks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># initialize the kinase library class</span>
<span class="n">klibrary</span> <span class="o">=</span> <span class="n">enzyme</span><span class="o">.</span><span class="n">KL_flank_analysis</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">)</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>89 PTMs removed due to insignificant splice event (p &lt; 0.05, dpsi &gt;= 0.4): (86.41%)
Final number of PTMs to be assessed: 14
</pre></div>
</div>
<p>Now that we have the class initialized, we have two options. If there is a specific PTM of interest, we can run the analysis on that specific PTM. Otherwise, we can run the analysis on all PTMs in the dataframe. Let’s analyzing across all PTMs first. This will return a dataframe containing scores of each kinase for every event in the altered_flanks dataframe for both the inclusion and exclusion isoforms. In addition, it will calculate the difference and the relative difference (normalized by dPSI) between the two isoforms:
:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">klibrary</span><span class="o">.</span><span class="n">analyze_all_ptms</span><span class="p">()</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>Score sequences for inclusion isoforms
Scoring 12 ser_thr substrates
Calculating percentile for 12 ser_thr substrates

  0%|          | 0/311 [00:00&lt;?, ?it/s]
  5%|▍         | 15/311 [00:00&lt;00:02, 142.01it/s]
 10%|▉         | 30/311 [00:00&lt;00:02, 140.24it/s]
 14%|█▍        | 45/311 [00:00&lt;00:01, 133.92it/s]
 19%|█▉        | 59/311 [00:00&lt;00:01, 133.29it/s]
 23%|██▎       | 73/311 [00:00&lt;00:01, 135.06it/s]
 28%|██▊       | 87/311 [00:00&lt;00:01, 133.73it/s]
 32%|███▏      | 101/311 [00:00&lt;00:01, 133.93it/s]
 38%|███▊      | 117/311 [00:00&lt;00:01, 139.36it/s]
 42%|████▏     | 131/311 [00:00&lt;00:01, 132.01it/s]
 47%|████▋     | 146/311 [00:01&lt;00:01, 134.46it/s]
 52%|█████▏    | 161/311 [00:01&lt;00:01, 138.24it/s]
 56%|█████▋    | 175/311 [00:01&lt;00:00, 136.76it/s]
 61%|██████    | 189/311 [00:01&lt;00:00, 133.89it/s]
 65%|██████▌   | 203/311 [00:01&lt;00:00, 135.16it/s]
 70%|██████▉   | 217/311 [00:01&lt;00:00, 134.35it/s]
 74%|███████▍  | 231/311 [00:01&lt;00:00, 134.96it/s]
 79%|███████▉  | 245/311 [00:01&lt;00:00, 136.32it/s]
 83%|████████▎ | 259/311 [00:01&lt;00:00, 136.70it/s]
 88%|████████▊ | 274/311 [00:02&lt;00:00, 140.46it/s]
 93%|█████████▎| 289/311 [00:02&lt;00:00, 137.88it/s]
 97%|█████████▋| 303/311 [00:02&lt;00:00, 136.45it/s]
100%|██████████| 311/311 [00:02&lt;00:00, 134.62it/s]
Scoring 1 tyrosine substrates
Calculating percentile for 1 tyrosine substrates

  0%|          | 0/78 [00:00&lt;?, ?it/s]
100%|██████████| 78/78 [00:00&lt;00:00, 2571.41it/s]

Scoring sequences for exclusion isoforms
Scoring 12 ser_thr substrates
Calculating percentile for 12 ser_thr substrates

  0%|          | 0/311 [00:00&lt;?, ?it/s]
  5%|▍         | 14/311 [00:00&lt;00:02, 136.40it/s]
  9%|▉         | 28/311 [00:00&lt;00:02, 122.13it/s]
 14%|█▎        | 42/311 [00:00&lt;00:02, 127.18it/s]
 18%|█▊        | 55/311 [00:00&lt;00:02, 118.31it/s]
 22%|██▏       | 67/311 [00:00&lt;00:02, 117.39it/s]
 25%|██▌       | 79/311 [00:00&lt;00:02, 115.36it/s]
 29%|██▉       | 91/311 [00:00&lt;00:01, 116.23it/s]
 33%|███▎      | 104/311 [00:00&lt;00:01, 119.67it/s]
 37%|███▋      | 116/311 [00:00&lt;00:01, 118.90it/s]
 41%|████▏     | 129/311 [00:01&lt;00:01, 121.07it/s]
 46%|████▌     | 142/311 [00:01&lt;00:01, 117.83it/s]
 50%|████▉     | 155/311 [00:01&lt;00:01, 120.13it/s]
 54%|█████▍    | 168/311 [00:01&lt;00:01, 120.57it/s]
 58%|█████▊    | 181/311 [00:01&lt;00:01, 120.87it/s]
 63%|██████▎   | 195/311 [00:01&lt;00:00, 124.19it/s]
 67%|██████▋   | 208/311 [00:01&lt;00:00, 124.03it/s]
 71%|███████   | 221/311 [00:01&lt;00:00, 125.14it/s]
 75%|███████▌  | 234/311 [00:01&lt;00:00, 124.30it/s]
 80%|███████▉  | 248/311 [00:02&lt;00:00, 126.12it/s]
 84%|████████▍ | 261/311 [00:02&lt;00:00, 125.10it/s]
 88%|████████▊ | 274/311 [00:02&lt;00:00, 121.79it/s]
 92%|█████████▏| 287/311 [00:02&lt;00:00, 122.43it/s]
 96%|█████████▋| 300/311 [00:02&lt;00:00, 124.31it/s]
100%|██████████| 311/311 [00:02&lt;00:00, 121.35it/s]
Scoring 1 tyrosine substrates
Calculating percentile for 1 tyrosine substrates

  0%|          | 0/78 [00:00&lt;?, ?it/s]
100%|██████████| 78/78 [00:00&lt;00:00, 2443.92it/s]
</pre></div>
</div>
<p>With all the scores calculated, we can then identify the sites and events with the largest changes in affinity. To better represent how big of a change this can be, we can calculate a relative affinity change, which is the difference in percentile score change between inclusion and exclusion isoforms, normalized by the change in dPSI. This will give us a better idea of how much the affinity is changing relative to the change in splicing. We can then plot this data to visualize the results.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">klibrary</span><span class="o">.</span><span class="n">plot_top_changes</span><span class="p">()</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_kinase_library_affinity_001.png" srcset="../../_images/sphx_glr_plot_kinase_library_affinity_001.png" alt="Top Changes" class = "sphx-glr-single-img"/><p>Here, we can see that a site in CD44 has a shift towards interactions with CAMK2G upon the provided perturbation, and our US01 site has several different kinases which prefer the isoform expressed prior to perturbation. In addition to looking at individual sites, we can also see if there are any kinases with consistently large differences in scores/percentiles across the flanking sequence changes. This can be done using the <cite>plot_top_kinases()</cite> function, which will plot the top kinases with the largest differences in scores across all PTMs (based on median by default). This can be useful for identifying kinases that may be more broadly impacted by the splicing events.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">klibrary</span><span class="o">.</span><span class="n">plot_top_kinases</span><span class="p">(</span><span class="n">top_n</span> <span class="o">=</span> <span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_kinase_library_affinity_002.png" srcset="../../_images/sphx_glr_plot_kinase_library_affinity_002.png" alt="plot kinase library affinity" class = "sphx-glr-single-img"/><p>Shown are the top 5 kinases with the largest median difference across all assessed PTMs. You’ll see that many of the PTMs don’t result in a large change, but some kinases, like MEK5, have a large change in preference after perturbation for some sites.</p>
<p>Now let’s say we didn’t want to focus on a specific PTM, such as the USO1 S486 site. We can do this by using the <cite>analyze_single_ptm()</cite> function, which will return a dataframe with the scores of each kinase for the inclusion and exclusion isoforms, as well as the difference and relative difference between the two isoforms for the specified PTM. First, let’s take a look at what the flanking sequences look like for this PTM:</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;USO1&#39;</span>
<span class="n">loc</span> <span class="o">=</span> <span class="mi">486</span>
<span class="n">example</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[(</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">loc</span><span class="p">)]</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose.analyze</span><span class="w"> </span><span class="kn">import</span> <span class="n">flank_analysis</span>

<span class="n">flank_analysis</span><span class="o">.</span><span class="n">plot_sequence_differences</span><span class="p">(</span><span class="n">example</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">example</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">])</span>
</pre></div>
</div>
<img src="../../_images/sphx_glr_plot_kinase_library_affinity_003.png" srcset="../../_images/sphx_glr_plot_kinase_library_affinity_003.png" alt="plot kinase library affinity" class = "sphx-glr-single-img"/><p>Now we can run the kinase library analysis on this specific PTM. This will return a dataframe with the kinases that have a higher affinity for the inclusion or exclusion isoform, as well as the p-value and fold change.</p>
<div class="highlight-Python notranslate"><div class="highlight"><pre><span></span><span class="n">affinity_change</span> <span class="o">=</span> <span class="n">klibrary</span><span class="o">.</span><span class="n">analyze_single_ptm</span><span class="p">(</span><span class="s1">&#39;ARHGAP17&#39;</span><span class="p">,</span> <span class="mi">497</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">affinity_change</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
<div class="sphx-glr-script-out highlight-none notranslate"><div class="highlight"><pre><span></span>       Inclusion percentile  Exclusion percentile  Difference  Absolute Difference   dPSI  Relative Change in Preference
CDK3                   5.57                 93.56      -87.99                87.99  0.413                      -36.33987
CDK1                   7.36                 92.91      -85.55                85.55  0.413                      -35.33215
DRAK1                 97.87                 17.28       80.59                80.59  0.413                       33.28367
CDK2                  18.64                 95.50      -76.86                76.86  0.413                      -31.74318
CLK2                  17.91                 91.13      -73.22                73.22  0.413                      -30.23986
YANK3                 75.92                  3.12       72.80                72.80  0.413                       30.06640
RIPK2                 90.77                 18.16       72.61                72.61  0.413                       29.98793
RIPK3                 79.13                  6.87       72.26                72.26  0.413                       29.84338
PLK4                  77.15                  5.07       72.08                72.08  0.413                       29.76904
TTBK1                 71.18                  1.17       70.01                70.01  0.413                       28.91413
</pre></div>
</div>
<p>From this analysis, we can start to identify events that might rewire kinase interactions. When paired with analysis of differentially included PTMs, we can start to identify kinases that are more or less likely to be influenced by changes to splicing patterns.</p>
<p class="sphx-glr-timing"><strong>Total running time of the script:</strong> (0 minutes 10.063 seconds)</p>
<div class="sphx-glr-footer sphx-glr-footer-example docutils container" id="sphx-glr-download-gallery-output-flanking-sequences-plot-kinase-library-affinity-py">
<div class="sphx-glr-download sphx-glr-download-jupyter docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/27e2b3e1ad16ed1b18b30585adc81d2d/plot_kinase_library_affinity.ipynb"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Jupyter</span> <span class="pre">notebook:</span> <span class="pre">plot_kinase_library_affinity.ipynb</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-python docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/8c1e4c04f6945cce6a653ff8e03912ca/plot_kinase_library_affinity.py"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">Python</span> <span class="pre">source</span> <span class="pre">code:</span> <span class="pre">plot_kinase_library_affinity.py</span></code></a></p>
</div>
<div class="sphx-glr-download sphx-glr-download-zip docutils container">
<p><a class="reference download internal" download="" href="../../_downloads/c4651484c60daece2ba4de88425ac8b8/plot_kinase_library_affinity.zip"><code class="xref download docutils literal notranslate"><span class="pre">Download</span> <span class="pre">zipped:</span> <span class="pre">plot_kinase_library_affinity.zip</span></code></a></p>
</div>
</div>
<p class="sphx-glr-signature"><a class="reference external" href="https://sphinx-gallery.github.io">Gallery generated by Sphinx-Gallery</a></p>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="plot_location_altered_flanks.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Probing where and how PTM flanking sequences are altered</p>
      </div>
    </a>
    <a class="right-next"
       href="../../faq.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Frequently Asked Questions</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>