
.. DO NOT EDIT.
.. THIS FILE WAS AUTOMATICALLY GENERATED BY SPHINX-GALLERY.
.. TO MAKE CHANGES, EDIT THE SOURCE PYTHON FILE:
.. "gallery_output/Networks/plot_ksea.py"
.. LINE NUMBERS ARE GIVEN BELOW.

.. only:: html

    .. note::
        :class: sphx-glr-download-link-note

        :ref:`Go to the end <sphx_glr_download_gallery_output_Networks_plot_ksea.py>`
        to download the full example code.

.. rst-class:: sphx-glr-example-title

.. _sphx_glr_gallery_output_Networks_plot_ksea.py:


Kinase substrate enrichment analysis (KSEA)
============================================================================

In the realm of kinase activity analysis, Kinase Substrate Enrichment Analysis (KSEA) is a commonly used tool to identify kinases with substrates that are significantly changing between two conditions (based on phosphoproteomic data). See the publication `here <https://academic.oup.com/bioinformatics/article/33/21/3489/3892392>`_  for what the original algorithm looks like. We have adapted KSEA for use with splicing data, which uses the dPSI values to calculate a z-score to identify kinases with substrates that are either excluded or included from isoforms. This is a useful tool, but you should treat these results with caution as they will often be on only a couple kinase-substrates due to the limited kinase-substrate information available. First, let's load the example data and run KSEA on the differentially included PTMs dataframe

.. GENERATED FROM PYTHON SOURCE LINES 8-14

.. code-block:: Python


    from ptm_pose import helpers
    from ptm_pose.analyze import enzyme

    spliced_ptms = helpers.load_example_data(spliced_ptms=True)








.. GENERATED FROM PYTHON SOURCE LINES 15-16

Now we can establish the KSEA class from the enzyme module. We will need to specify the database we want to use for KSEA analysis. This can be individual databases (PhosphoSitePlus, RegPhos, DEPOD, OmniPath, or iKiP) or the combined database (Combined). Note that some of these databases are specific to modification classes (e.g. PhosphoSitePlus is specific to phosphorylation). For this example, let's start with PhosphoSitePlus, a commonly used database for kinase-substrate relationships.

.. GENERATED FROM PYTHON SOURCE LINES 16-21

.. code-block:: Python


    ksea = enzyme.KSEA(spliced_ptms, database = 'PhosphoSitePlus')
    #run KSEA for all available kinases
    ksea.runKSEA()





.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    Final number of PTMs to be assessed: 484




.. GENERATED FROM PYTHON SOURCE LINES 22-23

Now we can visualize the results. The `plot_results()` function will plot the z-scores for each kinase, as well as whether that was found to be significant (based on FDR). You can optionally include notations indicating how many substrates were identified in the splicing data for each kinase.

.. GENERATED FROM PYTHON SOURCE LINES 23-27

.. code-block:: Python


    ksea.plot_results(show_substrate_count = True)





.. image-sg:: /gallery_output/Networks/images/sphx_glr_plot_ksea_001.png
   :alt: plot ksea
   :srcset: /gallery_output/Networks/images/sphx_glr_plot_ksea_001.png
   :class: sphx-glr-single-img






.. rst-class:: sphx-glr-timing

   **Total running time of the script:** (0 minutes 0.381 seconds)


.. _sphx_glr_download_gallery_output_Networks_plot_ksea.py:

.. only:: html

  .. container:: sphx-glr-footer sphx-glr-footer-example

    .. container:: sphx-glr-download sphx-glr-download-jupyter

      :download:`Download Jupyter notebook: plot_ksea.ipynb <plot_ksea.ipynb>`

    .. container:: sphx-glr-download sphx-glr-download-python

      :download:`Download Python source code: plot_ksea.py <plot_ksea.py>`

    .. container:: sphx-glr-download sphx-glr-download-zip

      :download:`Download zipped: plot_ksea.zip <plot_ksea.zip>`


.. only:: html

 .. rst-class:: sphx-glr-signature

    `Gallery generated by Sphinx-Gallery <https://sphinx-gallery.github.io>`_
