
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ptm_pose.analyze.enzyme &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/ptm_pose/analyze/enzyme';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../gallery_output/index.html">Gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/Networks/plot_ksea.html">Kinase substrate enrichment analysis (KSEA)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/Networks/plot_protein_interactions.html">Protein interactions network</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for ptm_pose.analyze.enzyme</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pickle</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>

<span class="c1">#plotting functions</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.lines</span><span class="w"> </span><span class="kn">import</span> <span class="n">Line2D</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">matplotlib.patches</span><span class="w"> </span><span class="kn">import</span> <span class="n">Patch</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>

<span class="c1">#custom stat functions and other helper functions</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">ptm_pose</span><span class="w"> </span><span class="kn">import</span> <span class="n">stat_utils</span><span class="p">,</span> <span class="n">pose_config</span><span class="p">,</span> <span class="n">helpers</span><span class="p">,</span> <span class="n">annotate</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scipy.stats</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">stats</span>



<span class="c1">#optional analysis packages</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">kinase_library</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">kl</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="n">kl</span> <span class="o">=</span> <span class="kc">None</span>

<span class="n">package_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>


<div class="viewcode-block" id="compare_KL_for_sequence">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.enzyme.compare_KL_for_sequence">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">compare_KL_for_sequence</span><span class="p">(</span><span class="n">inclusion_seq</span><span class="p">,</span> <span class="n">exclusion_seq</span><span class="p">,</span> <span class="n">dpsi</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">comparison_type</span> <span class="o">=</span> <span class="s1">&#39;percentile&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given two sequences, compare the kinase library scores, percentiles, or ranks for each sequence. Optionally, provide a dPSI value to calculate the relative change in preference for each kinase.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    inclusion_seq : str</span>
<span class="sd">        sequence to score for inclusion preference, with modification lowercased</span>
<span class="sd">    exclusion_seq : str</span>
<span class="sd">        sequence to score for exclusion preference, with modification lowercased</span>
<span class="sd">    dpsi : float</span>
<span class="sd">        dPSI value for the PTM event, which will be used to calculate the relative change in preference for each kinase (score difference * dPSI). Default is None.</span>
<span class="sd">    comparison_type : str</span>
<span class="sd">        type of comparison to perform. Can be &#39;percentile&#39;, &#39;score&#39;, or &#39;rank&#39;. Default is &#39;percentile&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">kl</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Kinase library package not installed. To use this functionality, please install the kinase library package by running `pip install kinase-library`&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="c1">#get percentiles for inclusion sequence</span>
    <span class="n">s_inc</span> <span class="o">=</span> <span class="n">kl</span><span class="o">.</span><span class="n">Substrate</span><span class="p">(</span><span class="n">inclusion_seq</span><span class="p">)</span>
    <span class="n">inn_perc</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>

    <span class="c1">#get percentiles for exclusion sequence</span>
    <span class="n">s_exc</span> <span class="o">=</span> <span class="n">kl</span><span class="o">.</span><span class="n">Substrate</span><span class="p">(</span><span class="n">exclusion_seq</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;percentile&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">sort_by</span> <span class="o">=</span> <span class="s1">&#39;name&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;score&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">score</span><span class="p">()</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">score</span><span class="p">()</span>
    <span class="k">elif</span> <span class="n">comparison_type</span> <span class="o">==</span> <span class="s1">&#39;rank&#39;</span><span class="p">:</span>
        <span class="n">inc_results</span> <span class="o">=</span> <span class="n">s_inc</span><span class="o">.</span><span class="n">rank</span><span class="p">()</span>
        <span class="n">ex_results</span> <span class="o">=</span> <span class="n">s_exc</span><span class="o">.</span><span class="n">rank</span><span class="p">()</span>

    <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">inc_results</span><span class="p">,</span> <span class="n">ex_results</span><span class="p">],</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">results</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="sa">f</span><span class="s1">&#39;Exclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">results</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Exclusion </span><span class="si">{</span><span class="n">comparison_type</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Absolute Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
    <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">dpsi</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">dpsi</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Relative Change in Preference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">dpsi</span>

    <span class="k">return</span> <span class="n">results</span></div>


<div class="viewcode-block" id="get_all_KL_scores">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.enzyme.get_all_KL_scores">[docs]</a>
<span class="k">def</span><span class="w"> </span><span class="nf">get_all_KL_scores</span><span class="p">(</span><span class="n">seq_data</span><span class="p">,</span> <span class="n">seq_col</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ser_thr&#39;</span><span class="p">,</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">],</span> <span class="n">score_type</span> <span class="o">=</span> <span class="s1">&#39;percentiles&#39;</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given a dataset with flanking sequences, score each flanking sequence</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    seq_data : pandas dataframe</span>
<span class="sd">        processed dataframe containing flanking sequences to score</span>
<span class="sd">    seq_col : str</span>
<span class="sd">        column in seq_data containing the flanking sequences</span>
<span class="sd">    kin_type : list</span>
<span class="sd">        list of kinase types to score. Can be &#39;ser_thr&#39; or &#39;ST&#39; for serine/threonine kinases, or &#39;tyrosine&#39; or &#39;Y&#39; for tyrosine kinases. Default is [&#39;ser_thr&#39;, &#39;tyrosine&#39;].</span>
<span class="sd">    score_type : str</span>
<span class="sd">        type of score to calculate. Can be &#39;percentile&#39;, &#39;score&#39;, or &#39;rank&#39;. Default is &#39;percentile&#39;.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    merged_data : dict</span>
<span class="sd">        dictionary containing the merged dataframes for each kinase type with KinaseLibrary scores</span>
<span class="sd">    </span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">kl</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Kinase library package not installed. To use this functionality, please install the kinase library package by running `pip install kinase-library`&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    
    <span class="n">phospho_predict</span> <span class="o">=</span> <span class="n">kl</span><span class="o">.</span><span class="n">PhosphoProteomics</span><span class="p">(</span><span class="n">seq_data</span><span class="p">,</span> <span class="n">seq_col</span> <span class="o">=</span> <span class="n">seq_col</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kin_type</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">kin_type</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">phospho_predict</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">kin_type</span> <span class="o">=</span> <span class="n">kin_type</span><span class="p">)</span>
            <span class="n">merged_data</span> <span class="o">=</span> <span class="n">phospho_predict</span><span class="o">.</span><span class="n">merge_data_scores</span><span class="p">(</span><span class="n">kin_type</span> <span class="o">=</span> <span class="n">kin_type</span><span class="p">,</span> <span class="n">score_type</span> <span class="o">=</span> <span class="n">score_type</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">merged_data</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">ktype</span> <span class="ow">in</span> <span class="n">kin_type</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">ktype</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ser_thr&#39;</span><span class="p">,</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;ST&#39;</span><span class="p">]:</span>
                    <span class="k">if</span> <span class="n">ktype</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
                        <span class="n">ktype</span> <span class="o">==</span> <span class="s1">&#39;tyrosine&#39;</span>
                    <span class="k">elif</span> <span class="n">ktype</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
                        <span class="n">ktype</span> <span class="o">=</span> <span class="s1">&#39;ser_thr&#39;</span>
   
                    <span class="n">phospho_predict</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">kin_type</span> <span class="o">=</span> <span class="n">ktype</span><span class="p">)</span>
                    <span class="n">merged_data</span><span class="p">[</span><span class="n">ktype</span><span class="p">]</span> <span class="o">=</span> <span class="n">phospho_predict</span><span class="o">.</span><span class="n">merge_data_scores</span><span class="p">(</span><span class="n">kin_type</span> <span class="o">=</span> <span class="n">ktype</span><span class="p">,</span> <span class="n">score_type</span> <span class="o">=</span> <span class="n">score_type</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">ktype</span><span class="si">}</span><span class="s2"> not recognized. Must be &#39;ser_thr&#39; or &#39;ST&#39; for serine/threonine, or &#39;tyrosine&#39; or &#39;Y&#39; for tyrosine&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">kin_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>   
        <span class="n">phospho_predict</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">metric</span> <span class="o">=</span> <span class="n">score_type</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="n">kin_type</span><span class="p">)</span>
        <span class="n">merged_data</span> <span class="o">=</span> <span class="n">phospho_predict</span><span class="o">.</span><span class="n">merge_data_scores</span><span class="p">(</span><span class="n">kin_type</span> <span class="o">=</span> <span class="n">kin_type</span><span class="p">,</span> <span class="n">score_type</span> <span class="o">=</span> <span class="n">score_type</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">merged_data</span></div>




<span class="k">class</span><span class="w"> </span><span class="nc">KL_flank_analysis</span><span class="p">:</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">,</span> <span class="n">metric</span> <span class="o">=</span> <span class="s1">&#39;percentile&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class for comparing kinase preferences for inclusion and exclusion sequences from altered flanking sequences, using motif scoring from Kinase Library.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        altered_flanks : pandas dataframe</span>
<span class="sd">            dataframe containing altered flanking sequences (output during projection), including columns for &#39;Region ID&#39;, &#39;Gene&#39;, &#39;Residue&#39;, &#39;PTM Position in Isoform&#39;, &#39;Inclusion Flanking Sequence&#39;, &#39;Exclusion Flanking Sequence&#39;, &#39;Modification Class&#39;, and &#39;dPSI&#39;.</span>
<span class="sd">        metric : str</span>
<span class="sd">            metric to use for scoring. Can be &#39;percentile&#39;, &#39;score&#39;, or &#39;rank&#39;. Default is &#39;percentile&#39;.</span>
<span class="sd">        alpha : float</span>
<span class="sd">            significance threshold for p-value. Default is 0.05.</span>
<span class="sd">        min_dpsi : float</span>
<span class="sd">            effect size threshold for dPSI. Default is 0.2.</span>
<span class="sd">        **kwargs : dict</span>
<span class="sd">            additional keyword arguments to pass to the filter_ptms function</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1">#check to make sure kinase library is installed</span>
        <span class="k">if</span> <span class="n">kl</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Kinase library package not installed. To use this functionality, please install the kinase library package by running `pip install kinase-library`&#39;</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="c1">#reduce to phosphorylation sites</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="c1">#restrict to only significant PTMs</span>
        <span class="n">filter_arguments</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">extract_filter_kwargs</span><span class="p">(</span><span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">modification_class</span> <span class="o">=</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>

        <span class="c1">#restrict to only necessary columns and remove duplicate entries (from canonical and alternative isoforms)</span>
        <span class="n">optional_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span>
        <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Region ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">optional_cols</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Isoform Type&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="n">cols</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="n">keep</span> <span class="o">=</span> <span class="s1">&#39;first&#39;</span><span class="p">)</span>
        <span class="n">altered_flanks</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1">#store result</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">metric</span> <span class="o">=</span> <span class="n">metric</span>
        <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">optional_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span>


        

    <span class="k">def</span><span class="w"> </span><span class="nf">analyze_single_ptm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">loc</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Score a single PTM for its flanking sequence in the inclusion and exclusion isoforms using Kinase-Library</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">ptm_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">loc</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1">#remove duplicate entries with same splice event and .drop_duplicates(subset = [&#39;Inclusion Flanking Sequence&#39;, &#39;Exclusion Flanking Sequence&#39;, &#39;Region ID&#39;])</span>
        <span class="k">if</span> <span class="n">ptm_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No flanking sequence data found for gene </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2"> at position </span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">ptm_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># If there&#39;s only one row, squeeze it to remove the extra dimension</span>
            <span class="n">ptm_data</span> <span class="o">=</span> <span class="n">ptm_data</span><span class="o">.</span><span class="n">squeeze</span><span class="p">()</span>
            <span class="n">dpsi</span> <span class="o">=</span> <span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">ptm_data</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="n">results</span> <span class="o">=</span> <span class="n">compare_KL_for_sequence</span><span class="p">(</span><span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">dpsi</span> <span class="o">=</span> <span class="n">dpsi</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">results</span>
        <span class="k">elif</span> <span class="n">ptm_data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># if multiple entries, perform analysis on all of them</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Multiple splice events impacting the flanking sequence of </span><span class="si">{</span><span class="n">ptm_data</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}{</span><span class="n">loc</span><span class="si">}</span><span class="s2"> in </span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2"> found. Performing analysis on all of them.&quot;</span><span class="p">)</span>
            <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">ptm_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="n">dpsi</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">row</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="n">results</span><span class="p">[</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Region ID&#39;</span><span class="p">]]</span> <span class="o">=</span> <span class="n">compare_KL_for_sequence</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">],</span> <span class="n">dpsi</span> <span class="o">=</span> <span class="n">dpsi</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">score_all_ptms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Score the flanking sequences of PTMs for both the inclusion and exclusion isoforms using Kinase-Library</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">inclusion_sequences</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;Exclusion&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">exclusion_sequences</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="p">[[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;Inclusion&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Score sequences for inclusion isoforms&#39;</span><span class="p">)</span>
        <span class="n">inclusion_percentile</span> <span class="o">=</span> <span class="n">get_all_KL_scores</span><span class="p">(</span><span class="n">inclusion_sequences</span><span class="p">,</span> <span class="s1">&#39;Inclusion Flanking Sequence&#39;</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Scoring sequences for exclusion isoforms&#39;</span><span class="p">)</span>
        <span class="n">exclusion_percentile</span> <span class="o">=</span> <span class="n">get_all_KL_scores</span><span class="p">(</span><span class="n">exclusion_sequences</span><span class="p">,</span> <span class="s1">&#39;Exclusion Flanking Sequence&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">inclusion_percentile</span> <span class="o">=</span> <span class="n">inclusion_percentile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">exclusion_percentile</span> <span class="o">=</span> <span class="n">exclusion_percentile</span>
            
    <span class="k">def</span><span class="w"> </span><span class="nf">melt_score_df</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;ser_thr&#39;</span><span class="p">,</span> <span class="n">flank_type</span> <span class="o">=</span> <span class="s1">&#39;Inclusion&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Melt the percentile dataframe to long format for easier plotting</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;inclusion_percentile&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;exclusion_percentile&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">score_all_ptms</span><span class="p">()</span>

        <span class="c1">#check to make sure kin_type is valid</span>
        <span class="k">if</span> <span class="n">kin_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ser_thr&#39;</span><span class="p">,</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;kin_type must be either &#39;ser_thr&#39; or &#39;ST&#39; for serine/threonine, or &#39;tyrosine&#39;/&#39;Y&#39; for tyrosine, but got </span><span class="si">{</span><span class="n">kin_type</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">kin_type</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
            <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;tyrosine&#39;</span>
        <span class="k">elif</span> <span class="n">kin_type</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
            <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;ser_thr&#39;</span>

        <span class="c1">#grab percentile data</span>
        <span class="n">seq_col</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">flank_type</span><span class="si">}</span><span class="s1"> Flanking Sequence&#39;</span>
        <span class="k">if</span> <span class="n">flank_type</span> <span class="o">==</span> <span class="s1">&#39;Inclusion&#39;</span><span class="p">:</span>
            <span class="n">percentile_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">inclusion_percentile</span><span class="p">[</span><span class="n">kin_type</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">percentile_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">exclusion_percentile</span><span class="p">[</span><span class="n">kin_type</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1">#remove unneeded columns</span>
        <span class="n">percentile_df</span> <span class="o">=</span> <span class="n">percentile_df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;phos_res&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequence&#39;</span><span class="p">])</span>

        <span class="c1">#melt data</span>
        <span class="n">percentile_df</span> <span class="o">=</span> <span class="n">percentile_df</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Region ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="n">seq_col</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">optional_cols</span><span class="p">,</span> <span class="n">var_name</span> <span class="o">=</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="n">value_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">percentile_df</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">combine_score_dfs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Combine the inclusion and exclusion score dataframes into a single dataframe that calculates differences in predicted preference between inclusion and exclusion isoforms</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;inclusion_percentile&#39;</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;exclusion_percentile&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">score_all_ptms</span><span class="p">()</span>

        <span class="n">melted_inclusion</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">melt_score_df</span><span class="p">(</span><span class="n">flank_type</span> <span class="o">=</span> <span class="s1">&#39;Inclusion&#39;</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;ser_thr&#39;</span><span class="p">),</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">melt_score_df</span><span class="p">(</span><span class="n">flank_type</span> <span class="o">=</span> <span class="s1">&#39;Inclusion&#39;</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">melted_inclusion</span> <span class="o">=</span> <span class="n">melted_inclusion</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">():</span> <span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">})</span>

        <span class="n">melted_exclusion</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
            <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">melt_score_df</span><span class="p">(</span><span class="n">flank_type</span> <span class="o">=</span> <span class="s1">&#39;Exclusion&#39;</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;ser_thr&#39;</span><span class="p">),</span> 
            <span class="bp">self</span><span class="o">.</span><span class="n">melt_score_df</span><span class="p">(</span><span class="n">flank_type</span> <span class="o">=</span> <span class="s1">&#39;Inclusion&#39;</span><span class="p">,</span> <span class="n">kin_type</span> <span class="o">=</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">)]</span>
        <span class="p">)</span>
        <span class="n">melted_exclusion</span> <span class="o">=</span> <span class="n">melted_exclusion</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">():</span> <span class="sa">f</span><span class="s1">&#39;Exclusion </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">})</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">melted_inclusion</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">melted_exclusion</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Region ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">optional_cols</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;inner&#39;</span><span class="p">)</span>
        
        <span class="c1">#calculate the difference</span>
        <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Inclusion </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">combined</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;Exclusion </span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">metric</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">]</span>
        <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Absolute Difference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

        <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Relative Change in Preference&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Difference&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="n">combined</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span>
        
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span> <span class="o">=</span> <span class="n">combined</span>
        
    <span class="k">def</span><span class="w"> </span><span class="nf">analyze_all_ptms</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform the full analysis on all PTMs, including scoring and melting the dataframes</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">score_all_ptms</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">combine_score_dfs</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">get_kinases_with_largest_changes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">kinase_type</span> <span class="o">=</span> <span class="s1">&#39;ST&#39;</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">difference_type</span> <span class="o">=</span> <span class="s1">&#39;relative&#39;</span><span class="p">,</span> <span class="n">agg</span> <span class="o">=</span> <span class="s1">&#39;median&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the top n kinases with the largest changes in preference for inclusion or exclusion isoforms. Difference can be calculated as the normal difference in preference, the absolute difference in preference, or the relative change in preference (normalized by magnitude of splicing change, or dPSI).</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        kinase_type : str</span>
<span class="sd">            type of kinase to restrict analysis to. Can be &#39;ST&#39; or &#39;ser_thr&#39; for serine/threonine kinases, or &#39;Y&#39; or &#39;tyrosine&#39; for tyrosine kinases. Default is &#39;ST&#39;.</span>
<span class="sd">        top_n : int</span>
<span class="sd">            number of top kinases to return. Default is 5.</span>
<span class="sd">        difference_type : str</span>
<span class="sd">            type of difference to calculate. Can be &#39;normal&#39; (difference in preference), &#39;absolute&#39; (absolute difference in preference), or &#39;relative&#39; (relative change in preference normalized by magnitude of splicing change, or dPSI). Default is &#39;relative&#39;.</span>
<span class="sd">        agg : str</span>
<span class="sd">            aggregation function to use across all interactions for calculating top kinases. Default is &#39;median&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#grab kinase of interest</span>
        <span class="k">if</span> <span class="n">kinase_type</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span> <span class="ow">or</span> <span class="n">kinase_type</span> <span class="o">==</span> <span class="s1">&#39;ser_thr&#39;</span><span class="p">:</span>
            <span class="n">kinase_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;S|T&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">kinase_type</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span> <span class="ow">or</span> <span class="n">kinase_type</span> <span class="o">==</span> <span class="s1">&#39;tyrosine&#39;</span><span class="p">:</span>
            <span class="n">kinase_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Y&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;normal&#39;</span><span class="p">:</span>
            <span class="n">comp_col</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;absolute&#39;</span><span class="p">:</span>
            <span class="n">comp_col</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;relative&#39;</span><span class="p">:</span>
            <span class="n">comp_col</span> <span class="o">=</span> <span class="s1">&#39;Relative Change in Preference&#39;</span>
            <span class="n">kinase_df</span><span class="p">[</span><span class="s1">&#39;Absolute Relative&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">kinase_df</span><span class="p">[</span><span class="s1">&#39;Relative Change in Preference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">comp_col</span> <span class="o">=</span> <span class="s1">&#39;Absolute Relative&#39;</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;difference_type must be either &quot;normal&quot;, &quot;absolute&quot;, or &quot;relative&quot;&#39;</span><span class="p">)</span>

        <span class="c1">#check to make sure provided kinase type is valid</span>
        <span class="k">if</span> <span class="n">kinase_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ST&#39;</span><span class="p">,</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span><span class="s1">&#39;ser_thr&#39;</span><span class="p">,</span><span class="s1">&#39;tyrosine&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;kinase_type must be either &quot;ST&quot; or &quot;ser_thr&quot; for serine/threonine kinases or &quot;Y&quot; or &quot;tyrosine&quot; for tyrosine kinases&#39;</span><span class="p">)</span>



        <span class="n">top_kinases</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">kinase_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Kinase&#39;</span><span class="p">)[</span><span class="n">comp_col</span><span class="p">]</span><span class="o">.</span><span class="n">agg</span><span class="p">(</span><span class="n">agg</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
        <span class="n">kinase_df</span> <span class="o">=</span> <span class="n">kinase_df</span><span class="p">[</span><span class="n">kinase_df</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">top_kinases</span><span class="p">)]</span>
        <span class="k">return</span> <span class="n">kinase_df</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">plot_top_kinases</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">kinase_type</span> <span class="o">=</span> <span class="s1">&#39;ST&#39;</span><span class="p">,</span> <span class="n">difference_type</span> <span class="o">=</span> <span class="s1">&#39;relative&#39;</span><span class="p">,</span> <span class="n">agg</span> <span class="o">=</span> <span class="s1">&#39;median&#39;</span><span class="p">,</span> <span class="n">metric_type</span> <span class="o">=</span> <span class="s1">&#39;percentile&#39;</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot a swarm plot showing the top n kinases with the largest changes in preference for inclusion or exclusion isoforms&quot;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">top_kinase_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_kinases_with_largest_changes</span><span class="p">(</span><span class="n">kinase_type</span> <span class="o">=</span> <span class="n">kinase_type</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="n">top_n</span><span class="p">,</span> <span class="n">difference_type</span> <span class="o">=</span> <span class="n">difference_type</span><span class="p">,</span> <span class="n">agg</span> <span class="o">=</span> <span class="n">agg</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

        <span class="c1">#plot swarm plot</span>
        <span class="n">y_data</span> <span class="o">=</span> <span class="s1">&#39;Relative Change in Preference&#39;</span> <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;relative&#39;</span> <span class="k">else</span> <span class="s1">&#39;Absolute Difference&#39;</span> <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;absolute&#39;</span> <span class="k">else</span> <span class="s1">&#39;Difference&#39;</span>
        <span class="n">sns</span><span class="o">.</span><span class="n">swarmplot</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">top_kinase_df</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;Kinase&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="n">y_data</span><span class="p">,</span> <span class="n">dodge</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">ax</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">linestyle</span> <span class="o">=</span> <span class="s1">&#39;--&#39;</span><span class="p">,</span> <span class="n">linewidth</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;relative&#39;</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Change</span><span class="se">\n</span><span class="s1">in Preference</span><span class="se">\n</span><span class="s1">(Change*dPSI)&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;absolute&#39;</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Absolute Difference</span><span class="se">\n</span><span class="s1">in </span><span class="si">{</span><span class="n">metric_type</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;normal&#39;</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Difference</span><span class="se">\n</span><span class="s1">in </span><span class="si">{</span><span class="n">metric_type</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_top_changes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">loc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">top_n</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">difference_type</span> <span class="o">=</span> <span class="s1">&#39;relative&#39;</span><span class="p">,</span> <span class="n">metric_type</span> <span class="o">=</span> <span class="s1">&#39;percentile&#39;</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Plot the top n changes in preference after an altered flanking sequence event, either for a all events, a specific gene, or a specific PTM. Top changes can be calculated by percentile change, absolute percentile change, or relative change in preference (percentile change in affinity*dPSI)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        gene : str</span>
<span class="sd">            gene to restrict analysis to. Default is None, which will analyze all PTMs.</span>
<span class="sd">        loc : int</span>
<span class="sd">            location of PTM to restrict analysis to. Default is None, which will analyze all PTMs for the given gene or for all genes (if gene = None).</span>
<span class="sd">        top_n : int</span>
<span class="sd">            number of top changes to plot. Default is 10.</span>
<span class="sd">        difference_type : str</span>
<span class="sd">            type of difference to plot. Can be &#39;normal&#39; (difference in preference), &#39;absolute&#39; (absolute difference in preference), or &#39;relative&#39; (relative change in preference normalized by magnitude of splicing change, or dPSI). Default is &#39;relative&#39;.</span>
<span class="sd">        metric_type : str</span>
<span class="sd">            type of metric to plot. Default is &#39;percentile&#39;.</span>
<span class="sd">        ax : matplotlib axis</span>
<span class="sd">            axis to plot on. Default is None, which will create a new figure and axis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;summary_df&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">analyze_all_ptms</span><span class="p">()</span>


        <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">loc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ptm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&gt;&#39;</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">loc</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">ptm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
                <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&gt;&#39;</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ptm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;Gene&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">summary_df</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">loc</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;normal&#39;</span><span class="p">:</span>
            <span class="n">ptm</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">diff_col</span> <span class="o">=</span> <span class="s1">&#39;Difference&#39;</span>
            <span class="n">y_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Difference</span><span class="se">\n</span><span class="s1">in </span><span class="si">{</span><span class="n">metric_type</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;absolute&#39;</span><span class="p">:</span>
            <span class="n">ptm</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">diff_col</span> <span class="o">=</span> <span class="s1">&#39;Absolute Difference&#39;</span>
            <span class="n">y_label</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Absolute Difference</span><span class="se">\n</span><span class="s1">in </span><span class="si">{</span><span class="n">metric_type</span><span class="o">.</span><span class="n">capitalize</span><span class="p">()</span><span class="si">}</span><span class="s1">&#39;</span>
        <span class="k">elif</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;relative&#39;</span><span class="p">:</span>
            <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Relative Absolute&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptm</span><span class="p">[</span><span class="s1">&#39;Relative Change in Preference&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span>
            <span class="n">ptm</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;Relative Absolute&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
            <span class="n">diff_col</span> <span class="o">=</span> <span class="s1">&#39;Relative Change in Preference&#39;</span>
            <span class="n">y_label</span> <span class="o">=</span> <span class="s1">&#39;Relative Change</span><span class="se">\n</span><span class="s1">in Preference</span><span class="se">\n</span><span class="s1">(Change*dPSI)&#39;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;difference_type must be either &quot;normal&quot;, &quot;absolute&quot;, or &quot;relative&quot;&#39;</span><span class="p">)</span>

        <span class="n">plt_data</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="n">top_n</span><span class="p">)</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="n">diff_col</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1">#color bars by sign</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;coral&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;lightblue&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">plt_data</span><span class="p">[</span><span class="n">diff_col</span><span class="p">]]</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">plt_data</span><span class="p">[</span><span class="s1">&#39;Kinase&#39;</span><span class="p">],</span> <span class="n">plt_data</span><span class="p">[</span><span class="n">diff_col</span><span class="p">],</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">,</span> <span class="n">edgecolor</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">height</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">y_label</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">loc</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1"> </span><span class="si">{</span><span class="n">loc</span><span class="si">}</span><span class="s1"> - Top </span><span class="si">{</span><span class="n">top_n</span><span class="si">}</span><span class="s1"> Changes&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s1"> - Top </span><span class="si">{</span><span class="n">top_n</span><span class="si">}</span><span class="s1"> Changes&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Top Changes&#39;</span><span class="p">)</span>

        <span class="c1">#construct legend</span>
        <span class="k">if</span> <span class="n">difference_type</span> <span class="o">!=</span> <span class="s1">&#39;absolute&#39;</span><span class="p">:</span>
            <span class="n">handles</span> <span class="o">=</span> <span class="p">[</span><span class="n">plt</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;coral&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">),</span> <span class="n">plt</span><span class="o">.</span><span class="n">Rectangle</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;lightblue&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">)]</span>
            <span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Increase in Preference</span><span class="se">\n</span><span class="s1">upon Perturbation&#39;</span><span class="p">,</span> <span class="s1">&#39;Decrease in Preference </span><span class="se">\n</span><span class="s1">upon Perturbation&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">difference_type</span> <span class="o">==</span> <span class="s1">&#39;relative&#39;</span> <span class="k">else</span> <span class="p">[</span><span class="s1">&#39;Prefers Inclusion Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Prefers Exclusion Isoform&#39;</span><span class="p">]</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="p">,</span> <span class="n">labels</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="p">(</span><span class="mf">1.05</span><span class="p">,</span><span class="mf">0.5</span><span class="p">),</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>


<span class="k">class</span><span class="w"> </span><span class="nc">KSEA</span><span class="p">:</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptms</span><span class="p">,</span> <span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath Writer&#39;</span><span class="p">,</span> <span class="n">modification</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Adapted version of the Kinase Substrate Enrichment Algorithm (KSEA) from Casado et al. 2013. This version is designed to work with the PTM-POSE pipeline and uses the OmniPath database as default for enzyme-substrate information.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptms : pandas dataframe</span>
<span class="sd">            dataframe containing PTM information, including columns for &#39;PTM&#39;, &#39;Modification Class&#39;, &#39;p-value&#39;, and &#39;dPSI&#39;. This dataframe should be the output from the PTM-POSE pipeline.</span>
<span class="sd">        alpha : float</span>
<span class="sd">            significance threshold for p-value. Default is 0.05.</span>
<span class="sd">        dpsi : float</span>
<span class="sd">            effect size threshold for dPSI. Default is 0.2.</span>
<span class="sd">        database : str</span>
<span class="sd">            database to use for enzyme-substrate information. Default is &#39;OmniPath Writer&#39;. Other options include &#39;OmniPath Eraser&#39;, &#39;PhosphoSitePlus&#39;, or &#39;DEPOD&#39;.</span>
<span class="sd">        modification : str</span>
<span class="sd">            modification type to restrict analysis to. Default is None, which includes all modifications. If provided, should be one of the modification classes in the &#39;Modification Class&#39; column of the ptms dataframe.</span>
<span class="sd">        </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1">#filter ptms if desired</span>
        <span class="k">if</span> <span class="n">min_dpsi</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">min_dpsi</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">if</span> <span class="n">alpha</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">alpha</span> <span class="o">=</span> <span class="mf">1.1</span>
        <span class="c1">#filter ptms</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">modification_class</span> <span class="o">=</span> <span class="n">modification</span><span class="p">,</span> <span class="n">min_dpsi</span> <span class="o">=</span> <span class="n">min_dpsi</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1">#if omnipath database is used, make sure it was specified to either look at reader or writer enzymes</span>
        <span class="k">if</span> <span class="s1">&#39;OmniPath&#39;</span> <span class="ow">in</span> <span class="n">database</span> <span class="ow">and</span> <span class="n">database</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;OmniPath Writer&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath Eraser&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;OmniPath database must be either &quot;OmniPath Writer&quot; or &quot;OmniPath Eraser&quot;&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath Writer&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Writer Enzyme&#39;</span>  <span class="c1">#use writer enzyme annotations</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath Eraser&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="o">=</span> <span class="s1">&#39;OmniPath:Eraser Enzyme&#39;</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined Writer&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="o">=</span> <span class="s1">&#39;Combined:Writer_Enzyme&#39;</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined Eraser&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="o">=</span> <span class="s1">&#39;Combined:Eraser_Enzyme&#39;</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;PhosphoSitePlus&#39;</span><span class="p">,</span> <span class="s1">&#39;DEPOD&#39;</span><span class="p">,</span> <span class="s1">&#39;RegPhos&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath Writer&#39;</span><span class="p">,</span> <span class="s1">&#39;OmniPath Eraser&#39;</span><span class="p">,</span> <span class="s1">&#39;Combined Writer&#39;</span><span class="p">,</span> <span class="s1">&#39;Combined Eraser&#39;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;database must be either &quot;OmniPath Writer&quot;, &quot;OmniPath Eraser&quot;, &quot;PhosphoSitePlus&quot;, &quot;DEPOD&quot;, &quot;RegPhos&quot;, &quot;Combined Writer&quot;, or &quot;Combined Eraser&quot;&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">database</span><span class="si">}</span><span class="s1">:Enzyme&#39;</span>  <span class="c1">#use enzyme annotations for other databases</span>
        
        <span class="c1">#load annotations</span>
        <span class="k">if</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath Writer&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">process_database_annotations</span><span class="p">(</span><span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Writer Enzyme&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;OmniPath Eraser&#39;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">process_database_annotations</span><span class="p">(</span><span class="n">database</span> <span class="o">=</span> <span class="s1">&#39;OmniPath&#39;</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Eraser Enzyme&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined Writer&#39;</span> <span class="ow">or</span> <span class="n">database</span> <span class="o">==</span> <span class="s1">&#39;Combined Eraser&#39;</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                <span class="n">ptms</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">combine_enzyme_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">construct_annotation_dict_from_df</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">annot_col</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annot_col</span><span class="p">,</span> <span class="n">key_type</span> <span class="o">=</span> <span class="s1">&#39;annotation&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span> <span class="o">=</span> <span class="n">annotate</span><span class="o">.</span><span class="n">process_database_annotations</span><span class="p">(</span><span class="n">database</span> <span class="o">=</span> <span class="n">database</span><span class="p">,</span> <span class="n">annot_type</span> <span class="o">=</span> <span class="s1">&#39;Enzyme&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">enzymes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="c1">#add labels to the ptms dataframe</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">add_ptm_column</span><span class="p">(</span><span class="n">ptms</span><span class="p">)</span>
        <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;-&#39;</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">runKSEA_singleenzyme</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">enzyme</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a single enzyme, calculate the zscore for the enzyme based on the dPSI values of the PTMs in the dataframe.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        enzyme : str</span>
<span class="sd">            enzyme to calculate zscore for</span>

<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        z : float</span>
<span class="sd">            zscore for the enzyme</span>
<span class="sd">        m : int</span>
<span class="sd">            number of identified substrates for the enzyme in the dataset</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">annot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="n">enzyme</span><span class="p">]</span>
        <span class="c1">#check the possible residues and modifications for the current enzyme (for establishing background)</span>
        <span class="n">residues</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">annot</span><span class="p">])</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">annot</span><span class="p">])</span>

        <span class="c1">#filter the ptms dataframe to only include those that match the current enzyme</span>
        <span class="n">tmp_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptms</span><span class="p">[(</span><span class="bp">self</span><span class="o">.</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">residues</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">mod</span><span class="p">))]</span>
        <span class="k">if</span> <span class="n">tmp_ptms</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
        
        <span class="c1">#get number of sites associated with enzyme</span>
        <span class="n">m</span> <span class="o">=</span> <span class="n">tmp_ptms</span><span class="p">[</span><span class="n">tmp_ptms</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">annot</span><span class="p">)]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> 
        <span class="k">if</span> <span class="n">m</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        
        <span class="c1">#calculate the mean dPSI for all residues</span>
        <span class="n">p_mean</span> <span class="o">=</span> <span class="n">tmp_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="c1">#calculate the mean dPSI for all residues modified by the current enzyme</span>
        <span class="n">s_mean</span> <span class="o">=</span> <span class="n">tmp_ptms</span><span class="p">[</span><span class="n">tmp_ptms</span><span class="p">[</span><span class="s1">&#39;Label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">annot</span><span class="p">)][</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
        <span class="c1">#calculate the stdev of dPSI for all residues</span>
        <span class="n">p_std</span> <span class="o">=</span> <span class="n">tmp_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="c1">#calculate the zscore for the current enzyme</span>
        <span class="n">z</span> <span class="o">=</span> <span class="p">((</span><span class="n">s_mean</span> <span class="o">-</span> <span class="n">p_mean</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="n">m</span><span class="o">**</span><span class="mf">0.5</span><span class="p">))</span> <span class="o">/</span> <span class="p">(</span><span class="n">p_std</span> <span class="p">)</span>
        <span class="k">return</span> <span class="n">z</span><span class="p">,</span> <span class="n">m</span>



    <span class="k">def</span><span class="w"> </span><span class="nf">runKSEA</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Perform KSEA-style analysis for each kinase with substrates identified in experiment using dPSI values as quantitative measure.</span>

<span class="sd">            </span>
<span class="sd">        Returns</span>
<span class="sd">        -------</span>
<span class="sd">        all_results: pandas dataframe</span>
<span class="sd">            melted dataframe including all KSEA results, including zscore, number of identified substrates (m), and false discovery rate</span>
<span class="sd">        zscore: pandas dataframe</span>
<span class="sd">            dataframe which contains zscores from all data columns, updated with new zscores calculated for data_col</span>
<span class="sd">        pvals: pandas dataframe</span>
<span class="sd">            dataframe which contains p-values from all data columns, updated with new p-values calculated for data_col</span>
<span class="sd">        FDR: pandas dataframe</span>
<span class="sd">            dataframe which contains false discover rate from all data columns, updated with new false discovery rate calculated for data_col</span>
<span class="sd">        m: pandas dataframe</span>
<span class="sd">            dataframe which contains number of identified substrates for each kianse from all data columns, updated with new m calculated for data_col</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;z&#39;</span><span class="p">,</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="s1">&#39;FDR&#39;</span><span class="p">,</span> <span class="s1">&#39;m&#39;</span><span class="p">],</span> <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">enzymes</span><span class="p">)</span>
        <span class="c1">#iterate through each possible enzyme</span>
        <span class="k">for</span> <span class="n">enzyme</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotations</span><span class="p">:</span>
            <span class="c1">#get the zscore and number of substrates for the current enzyme</span>
            <span class="n">z</span><span class="p">,</span> <span class="n">m</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">runKSEA_singleenzyme</span><span class="p">(</span><span class="n">enzyme</span><span class="p">)</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enzyme</span><span class="p">,</span> <span class="s1">&#39;m&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">m</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">enzyme</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">z</span>

        <span class="c1">#calculate p-values and FDR for the results</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">cdf</span><span class="p">(</span><span class="o">-</span><span class="nb">abs</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;z&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;p&#39;</span><span class="p">,</span> <span class="n">ascending</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[</span><span class="s1">&#39;FDR&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stat_utils</span><span class="o">.</span><span class="n">adjustP</span><span class="p">(</span><span class="n">results</span><span class="p">[</span><span class="s1">&#39;p&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;BH&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results</span> <span class="o">=</span> <span class="n">results</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">plot_results</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">show_substrate_count</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a horizontal barplot of KSEA results, coloring based on significance and annotating with the number of substrates (if desired)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        show_substrate_count : bool</span>
<span class="sd">            whether to show the number of substrates on the right side of the plot. Default is True.</span>
<span class="sd">        ax : matplotlib axis</span>
<span class="sd">            axis to plot on. Default is None, which will create a new figure and axis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#create horizontal barplot of KSEA results, coloring based on significance</span>


        <span class="n">plt_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="c1">#sort results by zscore</span>
        <span class="n">plt_data</span> <span class="o">=</span>  <span class="n">plt_data</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;z&#39;</span><span class="p">)</span>

        <span class="c1">#plot results</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig_height</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">plt_data</span><span class="p">)</span><span class="o">*</span><span class="mf">0.2</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mf">3.5</span><span class="p">,</span> <span class="n">fig_height</span><span class="p">))</span>

        <span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;grey&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mf">0.05</span> <span class="k">else</span> <span class="s1">&#39;red&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">plt_data</span><span class="p">[</span><span class="s1">&#39;FDR&#39;</span><span class="p">]]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">plt_data</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">plt_data</span><span class="p">[</span><span class="s1">&#39;z&#39;</span><span class="p">],</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">,</span> <span class="n">edgecolor</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;black&#39;</span><span class="p">)</span>

        <span class="c1">#set xlim to be equal on either side of 0</span>
        <span class="n">xlim</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">plt_data</span><span class="p">[</span><span class="s1">&#39;z&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">abs</span><span class="p">()</span><span class="o">.</span><span class="n">max</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="o">-</span><span class="n">xlim</span><span class="p">,</span> <span class="n">xlim</span><span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;z-score&#39;</span><span class="p">)</span>

        <span class="c1">#add legend for significance</span>
        <span class="n">legend_elements</span> <span class="o">=</span> <span class="p">[</span><span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;grey&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;FDR &gt; 0.05&#39;</span><span class="p">),</span> <span class="n">Patch</span><span class="p">(</span><span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;red&#39;</span><span class="p">,</span> <span class="n">edgecolor</span><span class="o">=</span><span class="s1">&#39;black&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="sa">r</span><span class="s1">&#39;$FDR \leq 0.05$&#39;</span><span class="p">)]</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span> <span class="o">=</span> <span class="n">legend_elements</span><span class="p">)</span>

        <span class="c1">#annotate with number of substrates on the right side of plot</span>
        <span class="k">if</span> <span class="n">show_substrate_count</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xlim</span><span class="o">+</span><span class="mf">0.1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">plt_data</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;m&#39;</span><span class="p">,</span> <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">fontweight</span> <span class="o">=</span> <span class="s1">&#39;bold&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">plt_data</span><span class="p">[</span><span class="s1">&#39;m&#39;</span><span class="p">]):</span>
                <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">xlim</span><span class="o">+</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">)),</span> <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">,</span> <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">)</span>
        
    <span class="k">def</span><span class="w"> </span><span class="nf">save_results</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fname</span> <span class="o">=</span> <span class="s1">&#39;ksea_results&#39;</span><span class="p">,</span> <span class="n">odir</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Saves zscores, FDR, and m for all data columns in seperate .tsv files</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">results</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">odir</span><span class="si">}{</span><span class="n">fname</span><span class="si">}</span><span class="s1">.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>


<span class="k">class</span><span class="w"> </span><span class="nc">kstar_enrichment</span><span class="p">:</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">network_dir</span><span class="p">,</span> <span class="n">background_ptms</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">impact_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="s1">&#39;Included&#39;</span><span class="p">,</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">],</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given spliced ptm or PTMs with altered flanks and a single kstar network, get enrichment for each kinase in the network using a hypergeometric. Assumes the  data has already been reduced to the modification of interest (phosphotyrosine or phoshoserine/threonine)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        network_dir : dict</span>
<span class="sd">            dictionary of networks with kinase-substrate information</span>
<span class="sd">        ptms : pandas dataframe</span>
<span class="sd">            differentially included ptms</span>
<span class="sd">        background_ptms: pd.DataFrame</span>
<span class="sd">            PTMs to consider as the background for enrichment purposes, which should overlap with the spliced ptms information provided (an example might be all identified events, whether or not they are significant). If not provided, will use all ptms in the phosphoproteome.</span>
<span class="sd">        impact_type: list of str or str</span>
<span class="sd">            type of impacts to perform enrichment analysis for. Can be &#39;All&#39; (all significantly differentially included sites), &#39;Included&#39; (sites with dPSI &gt; 0), and/or &#39;Excluded&#39; (sites with dPSI &lt; 0). Default is to perform enrichment on all three groups.</span>
<span class="sd">        phospho_type : str </span>
<span class="sd">            type of phosphorylation event to extract. Can either by phosphotyrosine (&#39;Y&#39;) or phosphoserine/threonine (&#39;ST&#39;). Default is &#39;Y&#39;.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#change phospho_type to list if not already</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">phospho_type</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="k">pass</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;ST&#39;</span><span class="p">]:</span>
            <span class="n">phospho_type</span> <span class="o">=</span> <span class="p">[</span><span class="n">phospho_type</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
            <span class="n">phospho_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;ST&#39;</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;phospho_type must be either Y, ST, or All&#39;</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">impact_type</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="c1">#if impact type includes inclusion and/or exclusion, check if dPSI is present and assign impact type accordingly</span>
            <span class="k">if</span> <span class="s1">&#39;Included&#39;</span> <span class="ow">in</span> <span class="n">impact_type</span> <span class="ow">or</span> <span class="s1">&#39;Excluded&#39;</span> <span class="ow">in</span> <span class="n">impact_type</span><span class="p">:</span>
                <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Included&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;spliced_ptms must contain dPSI column to determine impact type, please either provide dPSI or restrict impact_type to &quot;All&quot;&#39;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">impact_type</span> <span class="o">=</span> <span class="n">impact_type</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">impact_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;Included&#39;</span> <span class="ow">or</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">:</span>
                <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;spliced_ptms must contain dPSI column to determine impact type, please either provide dPSI or restrict impact_type to &quot;All&quot;&#39;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Included&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">impact_type</span> <span class="o">=</span> <span class="p">[</span><span class="n">impact_type</span><span class="p">]</span>
            <span class="k">elif</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">impact_type</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;All&#39;</span><span class="p">]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;impact_type must be at least one of either &quot;All&quot;, &quot;Included&quot;, or &quot;Excluded&quot;&#39;</span><span class="p">)</span>
        
        <span class="c1">#construct background filter arguments (same as foreground but ignore significance criteria)</span>
        <span class="n">background_filter</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">extract_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">background_filter</span><span class="p">[</span><span class="s1">&#39;alpha&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1.1</span>
        <span class="n">background_filter</span><span class="p">[</span><span class="s1">&#39;min_dpsi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">networks</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="n">phospho_type</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">ptype</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Processing differentially included phosphotyrosine data&#39;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">ptype</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Processing differentially included phosphoserine/threonine data&#39;</span><span class="p">)</span>
            <span class="c1">#process ptms to only include specific phosphorylation data needed</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

            <span class="c1">#load background, either from provided data or from all phosphoproteome data</span>
            <span class="k">if</span> <span class="n">ptype</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Processing background phosphotyrosine data&#39;</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">ptype</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">Processing background phosphoserine/threonine data&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">background_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">background_ptms</span><span class="p">,</span> <span class="n">phospho_type</span><span class="o">=</span><span class="n">ptype</span><span class="p">,</span> <span class="o">**</span><span class="n">background_filter</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_ptms</span><span class="p">(</span><span class="n">pose_config</span><span class="o">.</span><span class="n">ptm_coordinates</span><span class="o">.</span><span class="n">copy</span><span class="p">(),</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">,</span> <span class="o">**</span><span class="n">background_filter</span><span class="p">)</span>

            <span class="c1">#load kstar networks</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_networks</span><span class="p">(</span><span class="n">network_dir</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="n">ptype</span><span class="p">)</span>


        <span class="c1">#save info</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span> <span class="o">=</span> <span class="n">phospho_type</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">load_networks</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">network_dir</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given network directory, load all kstar networks for specific phosphorylation type</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#check if file exists and whether a pickle has been generated: if not, load each network file individually</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">network_dir</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Network directory not found&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_dir</span><span class="si">}</span><span class="s2">/*.p&quot;</span><span class="p">):</span>
            <span class="n">networks</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_dir</span><span class="si">}</span><span class="s2">/network_</span><span class="si">{</span><span class="n">phospho_type</span><span class="si">}</span><span class="s2">.p&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span> <span class="p">)</span> <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">network_directory</span> <span class="o">=</span> <span class="n">network_dir</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;/</span><span class="si">{</span><span class="n">phospho_type</span><span class="si">}</span><span class="s1">/INDIVIDUAL_NETWORKS/&#39;</span>
            <span class="n">networks</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">network_directory</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.tsv&#39;</span><span class="p">):</span>
                    <span class="c1">#get the value of the network number</span>
                    <span class="n">file_noext</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;.tsv&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)</span>
                    <span class="n">key_name</span> <span class="o">=</span> <span class="s1">&#39;nkin&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">file_noext</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                    <span class="c1">#print(&quot;Debug: key name is %s&quot;%(key_name))</span>
                    <span class="n">networks</span><span class="p">[</span><span class="n">key_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">network_directory</span><span class="si">}{</span><span class="n">file</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">networks</span>


    <span class="k">def</span><span class="w"> </span><span class="nf">process_ptms</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptms</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given ptm information, restrict data to include only the phosphorylation type of interest and add a PTM column for matching information from KSTAR</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        ptms: pd.DataFrame</span>
<span class="sd">            ptm information containing modification type and ptm locatin information, such as the output from projection or altered flanking sequence analysis</span>
<span class="sd">        phospho_type: str</span>
<span class="sd">            type of phosphorylation event to extract. Can either by phosphotyrosine (&#39;Y&#39;) or phosphoserine/threonine (&#39;ST&#39;)</span>
<span class="sd">        </span>
<span class="sd">        Returns</span>
<span class="sd">        ptms: pd.DataFrame</span>
<span class="sd">            trimmed dataframe containing only modifications of interest and new &#39;PTM&#39; column</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1">#restrict to ptms to phosphorylation type of interest</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[(</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Phosphorylation&#39;</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Y&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">phospho_type</span> <span class="o">==</span> <span class="s1">&#39;ST&#39;</span><span class="p">:</span>
            <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">])]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;phospho_type must be either Y or ST&#39;</span><span class="p">)</span>
        
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="c1">#construct PTM column that matches KSTAR information</span>
        <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Residue&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ptms</span>

    
    <span class="k">def</span><span class="w"> </span><span class="nf">get_enrichment_single_network</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">network_key</span><span class="p">,</span> <span class="n">impact_type</span> <span class="o">=</span> <span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        in progress</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">network</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="n">network_key</span><span class="p">]</span>
        <span class="n">network</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">network</span><span class="p">[</span><span class="s1">&#39;KSTAR_ACCESSION&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">network</span><span class="p">[</span><span class="s1">&#39;KSTAR_SITE&#39;</span><span class="p">]</span>

        <span class="c1">#add network information to all significant data</span>
        <span class="k">if</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;All&#39;</span><span class="p">:</span>
            <span class="n">sig_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;Included&#39;</span><span class="p">:</span>
            <span class="n">sig_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Included&#39;</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="k">elif</span> <span class="n">impact_type</span> <span class="o">==</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">:</span>
            <span class="n">sig_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="bp">self</span><span class="o">.</span><span class="n">significant_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;impact_type must be either &quot;All&quot;, &quot;Included&quot;, or &quot;Excluded&quot;&#39;</span><span class="p">)</span>
        
        <span class="n">sig_ptms_kstar</span> <span class="o">=</span> <span class="n">sig_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">network</span><span class="p">[[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">,</span><span class="s1">&#39;PTM&#39;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;PTM&#39;</span><span class="p">)</span>

        <span class="c1">#repeat for background data</span>
        <span class="n">bg_ptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">background_ptms</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">][[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="n">bg_ptms_kstar</span> <span class="o">=</span> <span class="n">bg_ptms</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">network</span><span class="p">[[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">,</span><span class="s1">&#39;PTM&#39;</span><span class="p">]],</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;PTM&#39;</span><span class="p">)</span>

        <span class="n">results</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;k&#39;</span><span class="p">,</span><span class="s1">&#39;n&#39;</span><span class="p">,</span><span class="s1">&#39;M&#39;</span><span class="p">,</span><span class="s1">&#39;N&#39;</span><span class="p">,</span><span class="s1">&#39;p&#39;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">kinase</span> <span class="ow">in</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="c1">#get numbers for a hypergeometric test to look for enrichment of kinase substrates</span>
            <span class="n">k</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">n</span> <span class="o">=</span> <span class="n">bg_ptms_kstar</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">bg_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;KSTAR_KINASE&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">M</span> <span class="o">=</span> <span class="n">bg_ptms</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>
            <span class="n">N</span> <span class="o">=</span> <span class="n">sig_ptms_kstar</span><span class="p">[</span><span class="s1">&#39;PTM&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()</span>

            <span class="c1">#run hypergeometric test</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span><span class="s1">&#39;p&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">stat_utils</span><span class="o">.</span><span class="n">hypergeom</span><span class="p">(</span><span class="n">M</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">N</span><span class="p">,</span><span class="n">k</span><span class="p">)</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">M</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;k&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span>
            <span class="n">results</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">kinase</span><span class="p">,</span> <span class="s1">&#39;n&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n</span>

        <span class="k">return</span> <span class="n">results</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">get_enrichment_all_networks</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">impact_type</span> <span class="o">=</span> <span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="n">phospho_type</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given prostate data and a dictionary of kstar networks, get enrichment for each kinase in each network in the prostate data. Assumes the prostate data has already been reduced to the modification of interest (phosphotyrosine or phoshoserine/threonine)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        networks : dict</span>
<span class="sd">            dictionary of kstar networks</span>
<span class="sd">        prostate : pandas dataframe</span>
<span class="sd">            all PTMs identified in tCGA prostate data, regardless of significance (reduced to only include mods of interest)</span>
<span class="sd">        sig_prostate : pandas dataframe</span>
<span class="sd">            significant PTMs identified in tCGA prostate data, p &lt; 0.05 and effect size &gt; 0.25 (reduced to only include mods of interest)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">network</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">networks</span><span class="p">[</span><span class="n">phospho_type</span><span class="p">]:</span>
            <span class="n">results</span><span class="p">[</span><span class="n">network</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_enrichment_single_network</span><span class="p">(</span><span class="n">network_key</span><span class="o">=</span><span class="n">network</span><span class="p">,</span> <span class="n">impact_type</span><span class="o">=</span><span class="n">impact_type</span><span class="p">,</span> <span class="n">phospho_type</span><span class="o">=</span><span class="n">phospho_type</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">extract_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">results</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Given a dictionary of results from get_enrichment_all_networks, extract the p-values for each network and kinase, and then calculate the median p-value across all networks for each kinase</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        results : dict</span>
<span class="sd">            dictionary of results from get_enrichment_all_networks</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="s1">&#39;nkin0&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="n">results</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">for</span> <span class="n">network</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
            <span class="n">enrichment</span><span class="p">[</span><span class="n">network</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[</span><span class="n">network</span><span class="p">][</span><span class="s1">&#39;p&#39;</span><span class="p">]</span>
        <span class="n">enrichment</span><span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">enrichment</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">run_kstar_enrichment</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Run full kstar analysis to generate substrate enrichment across each of the 50 KSTAR networks and calculate the median p-value for each kinase across all networks</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">enrichment</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">median</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span><span class="p">:</span>
            <span class="n">enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">median_impacts</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">impact</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">impact_type</span><span class="p">,</span> <span class="n">desc</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Running enrichment for </span><span class="si">{</span><span class="n">ptype</span><span class="si">}</span><span class="s1"> data&#39;</span><span class="p">):</span>
                <span class="n">results</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_enrichment_all_networks</span><span class="p">(</span><span class="n">phospho_type</span><span class="o">=</span><span class="n">ptype</span><span class="p">,</span> <span class="n">impact_type</span> <span class="o">=</span> <span class="n">impact</span><span class="p">)</span>
                <span class="n">enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="n">impact</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">extract_enrichment</span><span class="p">(</span><span class="n">results</span><span class="p">)</span>
                <span class="n">median_impacts</span><span class="p">[</span><span class="n">impact</span><span class="p">]</span> <span class="o">=</span> <span class="n">enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="n">impact</span><span class="p">][</span><span class="s1">&#39;median&#39;</span><span class="p">]</span>
            
            <span class="n">median</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">median_impacts</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">enrichment_all</span> <span class="o">=</span> <span class="n">enrichment</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="o">=</span> <span class="n">median</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">return_enriched_kinases</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">impact_type</span> <span class="o">=</span> <span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return kinases with a median p-value less than the provided alpha value (substrates are enriched among the significant PTMs)</span>

<span class="sd">        Parameters</span>
<span class="sd">        ----------</span>
<span class="sd">        alpha : float</span>
<span class="sd">            significance threshold to use to subset kinases. Default is 0.05.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">run_kstar_enrichment</span><span class="p">()</span>

        <span class="n">sig_enrichment</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">ptype</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">phospho_type</span><span class="p">:</span>
            <span class="n">tmp_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="n">impact_type</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">sig_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">]</span> <span class="o">=</span> <span class="n">tmp_data</span><span class="p">[</span><span class="n">tmp_data</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
        <span class="k">return</span> <span class="n">sig_enrichment</span>
    
    <span class="k">def</span><span class="w"> </span><span class="nf">dotplot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptype</span> <span class="o">=</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="n">impact_types</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;All&#39;</span><span class="p">,</span> <span class="s1">&#39;Included&#39;</span><span class="p">,</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">],</span> <span class="n">kinase_axis</span> <span class="o">=</span> <span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">facecolor</span> <span class="o">=</span> <span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">size_legend</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">color_legend</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">max_size</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">sig_kinases_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span> <span class="n">dotsize</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> 
                 <span class="n">colormap</span><span class="o">=</span><span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;#6b838f&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">:</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s1">&#39;colorblind&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]},</span>
                 <span class="n">labelmap</span> <span class="o">=</span> <span class="p">{</span><span class="mi">0</span><span class="p">:</span> <span class="s1">&#39;FPR &gt; </span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="mf">0.05</span><span class="p">),</span> <span class="mi">1</span><span class="p">:</span><span class="s1">&#39;FPR &lt;= </span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="mf">0.05</span><span class="p">)},</span>
                 <span class="n">legend_title</span> <span class="o">=</span> <span class="s1">&#39;p-value&#39;</span><span class="p">,</span> <span class="n">size_number</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> <span class="n">size_color</span> <span class="o">=</span> <span class="s1">&#39;gray&#39;</span><span class="p">,</span> 
                 <span class="n">color_title</span> <span class="o">=</span> <span class="s1">&#39;Significant&#39;</span><span class="p">,</span> <span class="n">markersize</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> 
                 <span class="n">legend_distance</span> <span class="o">=</span> <span class="mf">1.0</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">)):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates the dotplot plot, where size is determined by values dataframe and color is determined by significant dataframe. This is a stripped down version of the code used in KSTAR to generate the dotplot for the kinase activities</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        -----------</span>
<span class="sd">        ax : matplotlib Axes instance, optional</span>
<span class="sd">            axes dotplot will be plotted on. If None then new plot generated</span>
<span class="sd">        &quot;&quot;&quot;</span> 
        <span class="n">multiplier</span> <span class="o">=</span> <span class="mi">10</span>
        <span class="n">offset</span> <span class="o">=</span> <span class="mi">5</span>
        <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="n">figsize</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">(</span><span class="n">facecolor</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>

        <span class="n">plt_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_enrichment</span><span class="p">[</span><span class="n">ptype</span><span class="p">][</span><span class="n">impact_types</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">plt_data</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">):</span>
            <span class="n">plt_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">plt_data</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">sig_kinases_only</span><span class="p">:</span>
            <span class="n">plt_data</span> <span class="o">=</span> <span class="n">plt_data</span><span class="p">[(</span><span class="n">plt_data</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">)</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)]</span>
            <span class="k">if</span> <span class="n">plt_data</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;No significant kinases found with p-value &lt; </span><span class="si">%0.2f</span><span class="s1">&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">alpha</span><span class="p">))</span>
        
        <span class="k">if</span> <span class="n">kinase_axis</span> <span class="o">==</span> <span class="s1">&#39;x&#39;</span><span class="p">:</span>
            <span class="n">plt_data</span> <span class="o">=</span> <span class="n">plt_data</span><span class="o">.</span><span class="n">T</span>
        <span class="k">elif</span> <span class="n">kinase_axis</span> <span class="o">==</span> <span class="s1">&#39;y&#39;</span><span class="p">:</span>
            <span class="n">plt_data</span> <span class="o">=</span> <span class="n">plt_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;kinase_axis must be either &quot;x&quot; or &quot;y&quot;&#39;</span><span class="p">)</span>
        
        <span class="c1"># Transform Data</span>
        <span class="n">columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">plt_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">values</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">plt_data</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">colors</span> <span class="o">=</span> <span class="p">((</span><span class="n">plt_data</span> <span class="o">&lt;=</span> <span class="n">alpha</span><span class="p">)</span> <span class="o">*</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">values</span><span class="p">[</span><span class="s1">&#39;row_index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">))</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span>
        <span class="n">colors</span><span class="p">[</span><span class="s1">&#39;row_index&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">colors</span><span class="p">))</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span>



        <span class="n">melt</span> <span class="o">=</span> <span class="n">values</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span> <span class="o">=</span> <span class="s1">&#39;row_index&#39;</span><span class="p">)</span>
        <span class="n">values</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;row_index&#39;</span><span class="p">],</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">melt</span><span class="p">[</span><span class="s1">&#39;var&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">melt</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span> <span class="p">:</span> <span class="n">columns</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        
        <span class="n">melt_color</span> <span class="o">=</span> <span class="n">colors</span><span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span> <span class="o">=</span> <span class="s1">&#39;row_index&#39;</span><span class="p">)</span>
        <span class="n">melt_color</span><span class="p">[</span><span class="s1">&#39;var&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">melt_color</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">row</span> <span class="p">:</span> <span class="n">columns</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">row</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">colors</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;row_index&#39;</span><span class="p">],</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

        <span class="c1"># Plot Data</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">melt</span><span class="p">[</span><span class="s1">&#39;var&#39;</span><span class="p">]</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">melt</span><span class="p">[</span><span class="s1">&#39;row_index&#39;</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>    <span class="c1">#needs to be done in reverse order to maintain order in the dataframe</span>
        
        
        <span class="n">s</span> <span class="o">=</span> <span class="n">melt</span><span class="o">.</span><span class="n">value</span> <span class="o">*</span> <span class="n">dotsize</span>
        
        <span class="c1">#check to see if more than 2 values are given (fprs). Otherwise get color based on binary significance</span>

        <span class="c1">#get color for each datapoint based on significance</span>
        <span class="n">melt_color</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">colormap</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">l</span><span class="p">,</span><span class="s1">&#39;black&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">melt_color</span><span class="o">.</span><span class="n">value</span><span class="p">]</span>

            
        <span class="n">c</span> <span class="o">=</span> <span class="n">melt_color</span><span class="p">[</span><span class="s1">&#39;color&#39;</span><span class="p">]</span>
        <span class="n">scatter</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">c</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">)</span>
        
        <span class="c1"># Add Color Legend</span>
        <span class="k">if</span> <span class="n">color_legend</span><span class="p">:</span>
            <span class="c1">#create the legend</span>
            <span class="n">color_legend</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">color_key</span> <span class="ow">in</span> <span class="n">colormap</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">color_legend</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                    <span class="n">Line2D</span><span class="p">([</span><span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">labelmap</span><span class="p">[</span><span class="n">color_key</span><span class="p">],</span>
                            <span class="n">markerfacecolor</span><span class="o">=</span> <span class="n">colormap</span><span class="p">[</span><span class="n">color_key</span><span class="p">],</span> <span class="n">markersize</span><span class="o">=</span><span class="n">markersize</span><span class="p">),</span>
                <span class="p">)</span>     
            <span class="n">legend1</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="n">color_legend</span><span class="p">,</span> <span class="n">loc</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;upper right&#39;</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="n">legend_distance</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span> <span class="n">title</span> <span class="o">=</span> <span class="n">color_title</span><span class="p">)</span>  

            <span class="n">legend1</span><span class="o">.</span><span class="n">set_clip_on</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">legend1</span><span class="p">)</span>
            


        <span class="c1"># Add Size Legend</span>
        <span class="k">if</span> <span class="n">size_legend</span><span class="p">:</span>
            <span class="c1">#check to see if max pval parameter was given: if so, use to create custom legend</span>
            <span class="k">if</span> <span class="n">max_size</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">s_label</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">max_size</span><span class="o">/</span><span class="n">size_number</span><span class="p">,</span><span class="n">max_size</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span><span class="n">max_size</span><span class="o">/</span><span class="n">size_number</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
                <span class="n">dsize</span> <span class="o">=</span> <span class="p">[</span><span class="n">s</span><span class="o">*</span><span class="n">dotsize</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">s_label</span><span class="p">]</span>
                <span class="n">legend_elements</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">element</span><span class="p">,</span> <span class="n">s</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">s_label</span><span class="p">,</span> <span class="n">dsize</span><span class="p">):</span>
                    <span class="n">legend_elements</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">Line2D</span><span class="p">([</span><span class="mi">0</span><span class="p">],[</span><span class="mi">0</span><span class="p">],</span> <span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s1">&#39;w&#39;</span><span class="p">,</span> <span class="n">markersize</span> <span class="o">=</span> <span class="n">s</span><span class="o">**</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">markerfacecolor</span> <span class="o">=</span> <span class="n">size_color</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="n">element</span><span class="p">))</span>
                <span class="n">legend2</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span> <span class="o">=</span> <span class="n">legend_elements</span><span class="p">,</span> <span class="n">loc</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;lower right&#39;</span><span class="p">,</span> <span class="n">title</span> <span class="o">=</span> <span class="n">legend_title</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="n">legend_distance</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span>        
            <span class="k">else</span><span class="p">:</span>
                <span class="n">kw</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span><span class="n">prop</span><span class="o">=</span><span class="s2">&quot;sizes&quot;</span><span class="p">,</span> <span class="n">num</span><span class="o">=</span><span class="n">size_number</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">size_color</span><span class="p">,</span> <span class="n">func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">s</span><span class="o">/</span><span class="n">dotsize</span><span class="p">)</span> 
                <span class="n">legend2</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="o">*</span><span class="n">scatter</span><span class="o">.</span><span class="n">legend_elements</span><span class="p">(</span><span class="o">**</span><span class="n">kw</span><span class="p">),</span>
                        <span class="n">loc</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;lower right&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">legend_title</span><span class="p">,</span> <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="n">legend_distance</span><span class="p">,</span><span class="mi">0</span><span class="p">))</span> 
            <span class="n">ax</span><span class="o">.</span><span class="n">add_artist</span><span class="p">(</span><span class="n">legend2</span><span class="p">)</span>

        
        <span class="c1"># Add Additional Plotting Information</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">tick_params</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">rotation</span> <span class="o">=</span> <span class="mi">90</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">))</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_ticks</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">))</span> <span class="o">*</span> <span class="n">multiplier</span> <span class="o">+</span> <span class="n">offset</span><span class="p">)</span>


        <span class="n">ax</span><span class="o">.</span><span class="n">set_xticklabels</span><span class="p">(</span><span class="n">plt_data</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">plt_data</span><span class="o">.</span><span class="n">index</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>  <span class="c1">#reverse order to match the dataframe</span>
        
        <span class="c1">#adjust x and y scale so that data is always equally spaced</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span><span class="o">*</span><span class="n">multiplier</span><span class="p">])</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">columns</span><span class="p">)</span><span class="o">*</span><span class="n">multiplier</span><span class="p">])</span>
        <span class="k">return</span> <span class="n">ax</span> 
    
 

        


        
      
        
</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>