
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>ptm_pose.analyze.summarize &#8212; PTM-POSE</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=362ab14a" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-binder.css?v=f4aeca0c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-dataframe.css?v=2082cf3c" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery-rendered-html.css?v=1277b6f3" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=d2a3a1a5"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = '_modules/ptm_pose/analyze/summarize';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="0.3.0" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">PTM-POSE</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Overview:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../Overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Dependencies.html">Dependencies</a></li>

<li class="toctree-l1"><a class="reference internal" href="../../../PTM_POSE.html">PTM-POSE Reference</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Instructions:</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/quickstart.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Projection_Instructions.html">Running PTM-POSE</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Filtering_PTMs.html">Filtering PTMs and Events</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../Running_PTM-POSE/Annotating_PTMs.html">Annotating PTMs with Functional Information</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Example Vignettes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../Examples/Examples.html">Examples</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../Examples/ESRP1_knockdown.html">ESRP1 Knockdown (MATS output)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../Examples/ESRP1_in_Prostate.html">ESRP1 Expression in Prostate Cancer (TCGASpliceSeq output)</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Analysis Gallery</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../gallery_output/index.html">Gallery</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/index.html">General Overview of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_NumberOfPTMs.html">Inspecting types of PTMs impacted by splicing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_PTM_annotations.html">Inspecting number of PTMs with annotation information available</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/A_Overview_of_Spliced_PTMs/plot_filter_impact.html">Impact of filtering PTMs</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/index.html">Functional Impact of Spliced PTMs</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_num_annotations.html">Identifying available PTM-specific annotation information</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_geneset_enrichment.html">Identify enriched gene sets associated with differentially spliced PTMs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/FunctionEnrichment/plot_specific_annotation.html">Assessing enriched PTM functions</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/Networks/index.html">Impact on protein interaction and regulatory networks</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/Networks/plot_kstar_enrichment.html">Identify kinases with enriched substrates in differentially included exons, using an adapted version of KSTAR</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/Networks/plot_protein_interactions.html">Protein interactions network</a></li>
</ul>
</details></li>
<li class="toctree-l2 has-children"><a class="reference internal" href="../../../gallery_output/flanking_sequences/index.html">Analyzing altered flanking sequences</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_kinase_library_affinity.html">Kinase affinity due to PTM flanking sequence alterations</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_sequence_comparison.html">Inspect difference of flanking sequence due to splice event</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_sh2_domain_motifs.html">Identify altered SH2 domain motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../gallery_output/flanking_sequences/plot_location_altered_flanks.html">Probing where and how PTM flanking sequences are altered</a></li>
</ul>
</details></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Troubleshooting</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../faq.html">Frequently Asked Questions</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/NaegleLab/PTM-POSE" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>

</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1></h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <h1>Source code for ptm_pose.analyze.summarize</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="c1">#plotting </span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>



<span class="c1">#custom stat functions</span>
<span class="kn">from</span> <span class="nn">ptm_pose</span> <span class="kn">import</span> <span class="n">helpers</span>

<div class="viewcode-block" id="combine_outputs">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.summarize.combine_outputs">[docs]</a>
<span class="k">def</span> <span class="nf">combine_outputs</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="n">altered_flanks</span><span class="p">,</span> <span class="n">report_removed_annotations</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>  <span class="n">include_stop_codon_introduction</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">remove_conflicting</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given the spliced_ptms (differentially included) and altered_flanks (altered flanking sequences) dataframes obtained from project and flanking_sequences modules, combine the two into a single dataframe that categorizes each PTM by the impact on the PTM site</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events and with annotations appended from various databases</span>
<span class="sd">    altered_flanks: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs associated with altered flanking sequences and with annotations appended from various databases</span>
<span class="sd">    include_stop_codon_introduction: bool</span>
<span class="sd">        Whether to include PTMs that introduce stop codons in the altered flanks. Default is False.</span>
<span class="sd">    remove_conflicting: bool</span>
<span class="sd">        Whether to remove PTMs that are both included and excluded across different splicing events. Default is True.</span>
<span class="sd">    kwargs: dict</span>
<span class="sd">        Additional keyword arguments to pass to the function, will be passed to `helpers.filter_ptms` if filtering is desired. Will automatically filter out insignificant events if not provided</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#filter spliced_ptms and altered_flanks dataframes to remove insignificant events or PTMs with low evidence</span>
    <span class="k">if</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">filter_arguments</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">extract_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">helpers</span><span class="o">.</span><span class="n">check_filter_kwargs</span><span class="p">(</span><span class="n">filter_arguments</span><span class="p">)</span>
        <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>
        <span class="n">filter_arguments</span><span class="p">[</span><span class="s1">&#39;remove_novel&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1">#keep novel PTMs in altered flanks, as these are not removed from isoform</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>

    <span class="c1">#extract specific direction of splicing change and add to dataframe</span>
    <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">spliced_ptms</span><span class="p">[</span><span class="s1">&#39;dPSI&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Included&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="s1">&#39;Excluded&#39;</span><span class="p">)</span>

    <span class="c1">#restrict altered flanks to those that are changed and are not disrupted by stop codons</span>
    <span class="k">if</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtypes</span> <span class="o">!=</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">include_stop_codon_introduction</span><span class="p">:</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;Stop Codon Introduced&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="k">else</span> <span class="s1">&#39;Altered Flank&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="o">~</span><span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Stop Codon Introduced&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">altered_flanks</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Altered Flank&#39;</span>

    <span class="c1">#identify annotations that are found in both datasets</span>
    <span class="n">annotation_columns_in_spliced_ptms</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
    <span class="n">annotation_columns_in_altered_flanks</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s1">&#39;:&#39;</span> <span class="ow">in</span> <span class="n">col</span><span class="p">]</span>
    <span class="n">annotation_columns</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">))</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns</span><span class="p">)</span> <span class="o">!=</span> <span class="n">annotation_columns_in_spliced_ptms</span> <span class="ow">and</span> <span class="n">report_removed_annotations</span><span class="p">:</span>
        <span class="n">annotation_columns_only_in_spliced</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">))</span>
        <span class="n">annotation_columns_only_in_altered</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_altered_flanks</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">annotation_columns_in_spliced_ptms</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns_only_in_spliced</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Warning: some annotations in spliced ptms dataframe not found in altered flanks dataframe: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_columns_only_in_spliced</span><span class="p">)</span><span class="si">}</span><span class="s1">. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">annotation_columns_only_in_altered</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Warning: some annotations in altered flanks dataframe not found in spliced ptms dataframe: </span><span class="si">{</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">annotation_columns_only_in_altered</span><span class="p">)</span><span class="si">}</span><span class="s1">. These annotations will be ignored. To avoid this, make sure to add annotations to both dataframes, or annotate the combined dataframe.&#39;</span><span class="p">)</span>

    <span class="c1">#check if dPSI or sig columns are in both dataframes</span>
    <span class="n">sig_cols</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;dPSI&#39;</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">sig_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;dPSI&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;Significance&#39;</span> <span class="ow">in</span> <span class="n">spliced_ptms</span><span class="o">.</span><span class="n">columns</span> <span class="ow">and</span> <span class="s1">&#39;Significance&#39;</span> <span class="ow">in</span> <span class="n">altered_flanks</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">sig_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Significance&#39;</span><span class="p">)</span>

    <span class="n">shared_columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">,</span> <span class="s1">&#39;Gene&#39;</span><span class="p">,</span> <span class="s1">&#39;UniProtKB Accession&#39;</span><span class="p">,</span> <span class="s1">&#39;Isoform ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Isoform Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Residue&#39;</span><span class="p">,</span> <span class="s1">&#39;PTM Position in Isoform&#39;</span><span class="p">,</span> <span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">sig_cols</span> <span class="o">+</span> <span class="n">annotation_columns</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">spliced_ptms</span><span class="p">[</span><span class="n">shared_columns</span><span class="p">],</span> <span class="n">altered_flanks</span><span class="p">[</span><span class="n">shared_columns</span><span class="p">]])</span>
    <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">combined</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="o">!=</span> <span class="s1">&#39;Impact&#39;</span><span class="p">],</span> <span class="n">as_index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="n">dropna</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s1">&#39;;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span>

    <span class="c1">#remove ptms that are both included and excluded across different events</span>
    <span class="k">if</span> <span class="n">remove_conflicting</span><span class="p">:</span>
        <span class="n">combined</span> <span class="o">=</span> <span class="n">combined</span><span class="p">[</span><span class="o">~</span><span class="p">((</span><span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Included&#39;</span><span class="p">))</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">combined</span><span class="p">[</span><span class="s1">&#39;Impact&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;Excluded&#39;</span><span class="p">)))]</span>

    <span class="k">return</span> <span class="n">combined</span></div>


<div class="viewcode-block" id="get_modification_counts">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.summarize.get_modification_counts">[docs]</a>
<span class="k">def</span> <span class="nf">get_modification_counts</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given PTM data (either spliced ptms, altered flanks, or combined data), return the counts of each modification class</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs projected onto splicing events or with altered flanking sequences</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    modification_counts: pd.Series</span>
<span class="sd">        Series with the counts of each modification class</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#filter ptms based on kwargs if provided</span>
    <span class="k">if</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">filter_arguments</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">extract_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">helpers</span><span class="o">.</span><span class="n">check_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>
        <span class="n">ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>

    <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">))</span>
    <span class="n">ptms</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span>
    <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">ptms</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">size</span><span class="p">()</span>
    <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">modification_counts</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">ascending</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">modification_counts</span></div>


<div class="viewcode-block" id="get_modification_class_data">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.summarize.get_modification_class_data">[docs]</a>
<span class="k">def</span> <span class="nf">get_modification_class_data</span><span class="p">(</span><span class="n">ptms</span><span class="p">,</span> <span class="n">mod_class</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Given ptm dataframe and a specific modification class, return a dataframe with only the PTMs of that class</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    ptms : pd.DataFrame</span>
<span class="sd">        Dataframe with ptm information, such as the spliced_ptms or altered_flanks dataframe obtained during projection</span>
<span class="sd">    mod_class : str</span>
<span class="sd">        </span>
<span class="sd">        The modification class to filter by, e.g. &#39;Phosphorylation&#39;, &#39;Acetylation&#39;, etc.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1">#check if specific modification class was provided and subset data by modification if so</span>
    <span class="n">ptms_of_interest</span> <span class="o">=</span> <span class="n">ptms</span><span class="p">[</span><span class="n">ptms</span><span class="p">[</span><span class="s1">&#39;Modification Class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">mod_class</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">ptms_of_interest</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No PTMs found for modification class &#39;</span><span class="si">{</span><span class="n">mod_class</span><span class="si">}</span><span class="s2">&#39;. Please check the input data or choose a different modification class.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ptms_of_interest</span></div>




<div class="viewcode-block" id="plot_modification_breakdown">
<a class="viewcode-back" href="../../../PTM_POSE.html#ptm_pose.analyze.summarize.plot_modification_breakdown">[docs]</a>
<span class="k">def</span> <span class="nf">plot_modification_breakdown</span><span class="p">(</span><span class="n">spliced_ptms</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">altered_flanks</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">colors</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s1">&#39;colorblind&#39;</span><span class="p">),</span> <span class="n">ax</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Plot the number of PTMs that are differentially included or have altered flanking sequences, separated by PTM type</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    spliced_ptms: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs that are differentially included</span>
<span class="sd">    altered_flanks: pd.DataFrame</span>
<span class="sd">        Dataframe with PTMs that have altered flanking sequences</span>
<span class="sd">    colors: list</span>
<span class="sd">        List of colors to use for the bar plot (first two will be used). Default is seaborn colorblind palette.</span>
<span class="sd">    ax: matplotlib.Axes</span>
<span class="sd">        Axis to plot on. If None, will create new figure. Default is None.</span>
<span class="sd">    kwargs: dict</span>
<span class="sd">        Additional keyword arguments to pass to the function, will be passed to `helpers.filter_ptms` if filtering is desired. Will automatically filter out insignificant events by min_dpsi and significance if the columns are present</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;Either spliced_ptms or altered_flanks must be provided to plot modification breakdown. Both may be provided.&#39;</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">kwargs</span><span class="p">:</span>
        <span class="n">filter_arguments</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">extract_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">helpers</span><span class="o">.</span><span class="n">check_filter_kwargs</span><span class="p">(</span><span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">spliced_ptms</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">altered_flanks</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">filter_ptms</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">,</span> <span class="o">**</span><span class="n">filter_arguments</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">ax</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>

    <span class="c1">#separate rows into unique PTM types</span>
    <span class="k">if</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">differentially_included_counts</span> <span class="o">=</span> <span class="n">get_modification_counts</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
        <span class="n">altered_flanks_counts</span> <span class="o">=</span> <span class="n">get_modification_counts</span><span class="p">(</span><span class="n">altered_flanks</span><span class="o">.</span><span class="n">copy</span><span class="p">())</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">differentially_included_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">differentially_included_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;Differentially Included PTMs&#39;</span><span class="p">)</span>
        <span class="n">altered_flanks_counts</span> <span class="o">=</span> <span class="n">altered_flanks_counts</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">differentially_included_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">fill_value</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">altered_flanks_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">altered_flanks_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">left</span> <span class="o">=</span> <span class="n">differentially_included_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;PTMs with Altered Flank&#39;</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>

        <span class="c1">#annotate with number of combined PTMs</span>
        <span class="n">total_count</span> <span class="o">=</span> <span class="n">differentially_included_counts</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">altered_flanks_counts</span><span class="p">,</span> <span class="n">fill_value</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">num_ptm</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">total_count</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">),</span> <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">)</span>  

        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="n">total_count</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">*</span><span class="mf">1.1</span><span class="p">])</span>

    <span class="k">elif</span> <span class="n">spliced_ptms</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">get_modification_counts</span><span class="p">(</span><span class="n">spliced_ptms</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">modification_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">modification_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1">#annotate with number of PTMs</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">num_ptm</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">modification_counts</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">),</span> <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">altered_flanks</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">modification_counts</span> <span class="o">=</span> <span class="n">get_modification_counts</span><span class="p">(</span><span class="n">altered_flanks</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">barh</span><span class="p">(</span><span class="n">modification_counts</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">modification_counts</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">colors</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

        <span class="c1">#annotate with number of PTMs</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">num_ptm</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">modification_counts</span><span class="o">.</span><span class="n">values</span><span class="p">):</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_ptm</span><span class="p">),</span> <span class="n">ha</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">va</span> <span class="o">=</span> <span class="s1">&#39;center&#39;</span><span class="p">)</span>

    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Number of PTMs&#39;</span><span class="p">)</span></div>


</pre></div>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Naegle Lab
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Naegle Lab.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>